{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Purpose: Plan of whole pipeline from Ref-genome/vcf-file/TSS-file/ to basenji-output"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Todo:\n",
    "ask Ziga about: vcf-gtex-gile\n",
    "run libs through with him\n",
    "\n",
    "add warning if Ref-Base is not base in VCF -> probl. due to shifted positions\n",
    "\n",
    "Classes: \n",
    "\n",
    "    [later: PipelineConfigurator: takes Gene-IDs  and sample-IDs from user, writes into output file\n",
    "            Calls all others, needs therefor all files. Makes sure that Intervall-Translator and model are compatible\n",
    "    ]\n",
    "    TODO:\n",
    "    Basenji-call with input\n",
    "    ResultWriter: puts results in Xarray with dimensions Gene/Sample/Outputvalues, stores it to disk\n",
    "    \n",
    "    Now: REPDataloader which reads in TSS, passes intervals to vcfFastaExtractor:\n",
    "    \n",
    "    TSS: not dataloader, is small...\n",
    "    but: IntervalCalculator: do...\n",
    "    \n",
    "    vcfFastaExtractor\n",
    "    random fasta access: pyfaidx\n",
    "    random vcf-access: pyVCF\n",
    "    \n",
    "    output: one-hot-encoded seq with annotations: geneID, sampleID, posiRefStart , ALLELE (0 or 1)\n",
    "    \n",
    "Input-files:\n",
    "\n",
    "   RefGenome in fasta: https://raw.githubusercontent.com/kipoi/kipoiseq/master/tests/data/hg38_chr22_32000000_32300000.fa\n",
    "   \n",
    "   (not used: intervalls-example: https://raw.githubusercontent.com/kipoi/kipoiseq/master/tests/data/intervals_51bp.tsv)\n",
    "   \n",
    "   CAGE TSS `/s/project/avsec/ExPecto/resources/geneanno.csv`\n",
    "   has:\n",
    "    lincRNA\n",
    "    lincRNA\n",
    "    protein_coding\n",
    "    rRNA\n",
    "   -> which to use?\n",
    "   \n",
    "   GTEx-vcf? maybe from /s/project/gtex-processed/vcf/GTEx_Analysis_20150112_WholeGenomeSeq_148Indiv_GATK_HaplotypeCaller.vcf.gz\n",
    "   \n",
    "   RefGenome: /s/genomes/human/hg38$ \n",
    "\n",
    "\n",
    "\n",
    "   \n",
    "   \n",
    "------\n",
    "   Use loading-param from:\n",
    "   https://github.com/kipoi/kipoi/blob/master/kipoi/data.py\n",
    "          dl = DataLoader(self, batch_size=1,\n",
    "                        collate_fn=numpy_collate_concat,    <---\n",
    "                        shuffle=False,\n",
    "                        num_workers=num_workers,\n",
    "drop_last=False)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "class IntervalCreator(object):\n",
    "    \"\"\"Converts the  file with CAGE-TSS to a .bed interval file\"\"\"\n",
    "    def __init__(tssFile, outFile):\n",
    "    def convertToBed(distUpstream,distDownstream):\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: kipoiseq in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/lib/python3.6/site-packages (0.2.2)\n",
      "Requirement already satisfied: tqdm in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/lib/python3.6/site-packages (from kipoiseq) (4.28.1)\n",
      "Requirement already satisfied: pybedtools in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/lib/python3.6/site-packages (from kipoiseq) (0.7.10)\n",
      "Requirement already satisfied: pyfaidx in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/lib/python3.6/site-packages (from kipoiseq) (0.5.4.2)\n",
      "Requirement already satisfied: h5py in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/lib/python3.6/site-packages (from kipoiseq) (2.8.0)\n",
      "Requirement already satisfied: kipoi>=0.5.5 in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/lib/python3.6/site-packages (from kipoiseq) (0.6.5)\n",
      "Requirement already satisfied: colorlog in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/lib/python3.6/site-packages (from kipoiseq) (3.1.4)\n",
      "Requirement already satisfied: related>=0.6.0 in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/lib/python3.6/site-packages (from kipoiseq) (0.7.1)\n",
      "Requirement already satisfied: pandas in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/lib/python3.6/site-packages (from kipoiseq) (0.23.4)\n",
      "Requirement already satisfied: gffutils in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/lib/python3.6/site-packages/gffutils-0.9-py3.6.egg (from kipoiseq) (0.9)\n",
      "Requirement already satisfied: numpy in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/lib/python3.6/site-packages (from kipoiseq) (1.16.0)\n",
      "Requirement already satisfied: six in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/lib/python3.6/site-packages (from pybedtools->kipoiseq) (1.12.0)\n",
      "Requirement already satisfied: setuptools>=0.7 in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/lib/python3.6/site-packages (from pyfaidx->kipoiseq) (39.1.0)\n",
      "Requirement already satisfied: future in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/lib/python3.6/site-packages (from kipoi>=0.5.5->kipoiseq) (0.17.1)\n",
      "Requirement already satisfied: enum34 in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/lib/python3.6/site-packages (from kipoi>=0.5.5->kipoiseq) (1.1.6)\n",
      "Requirement already satisfied: tinydb in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/lib/python3.6/site-packages (from kipoi>=0.5.5->kipoiseq) (3.12.1)\n",
      "Requirement already satisfied: urllib3<1.23,>=1.21.1 in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/lib/python3.6/site-packages (from kipoi>=0.5.5->kipoiseq) (1.22)\n",
      "Requirement already satisfied: pyyaml in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/lib/python3.6/site-packages (from kipoi>=0.5.5->kipoiseq) (3.13)\n",
      "Requirement already satisfied: jinja2 in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/lib/python3.6/site-packages (from kipoi>=0.5.5->kipoiseq) (2.10)\n",
      "Requirement already satisfied: cookiecutter in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/lib/python3.6/site-packages (from kipoi>=0.5.5->kipoiseq) (1.6.0)\n",
      "Requirement already satisfied: attrs in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/lib/python3.6/site-packages (from related>=0.6.0->kipoiseq) (18.2.0)\n",
      "Requirement already satisfied: python-dateutil in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/lib/python3.6/site-packages (from related>=0.6.0->kipoiseq) (2.7.5)\n",
      "Requirement already satisfied: pytz>=2011k in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/lib/python3.6/site-packages (from pandas->kipoiseq) (2018.6)\n",
      "Requirement already satisfied: argh in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/lib/python3.6/site-packages/argh-0.26.2-py3.6.egg (from gffutils->kipoiseq) (0.26.2)\n",
      "Requirement already satisfied: argcomplete in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/lib/python3.6/site-packages/argcomplete-1.9.4-py3.6.egg (from gffutils->kipoiseq) (1.9.4)\n",
      "Requirement already satisfied: simplejson in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/lib/python3.6/site-packages (from gffutils->kipoiseq) (3.15.0)\n",
      "Requirement already satisfied: MarkupSafe>=0.23 in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/lib/python3.6/site-packages (from jinja2->kipoi>=0.5.5->kipoiseq) (1.0)\n",
      "Requirement already satisfied: requests>=2.18.0 in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/lib/python3.6/site-packages (from cookiecutter->kipoi>=0.5.5->kipoiseq) (2.20.0)\n",
      "Requirement already satisfied: jinja2-time>=0.1.0 in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/lib/python3.6/site-packages (from cookiecutter->kipoi>=0.5.5->kipoiseq) (0.2.0)\n",
      "Requirement already satisfied: whichcraft>=0.4.0 in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/lib/python3.6/site-packages (from cookiecutter->kipoi>=0.5.5->kipoiseq) (0.5.2)\n",
      "Requirement already satisfied: click>=5.0 in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/lib/python3.6/site-packages (from cookiecutter->kipoi>=0.5.5->kipoiseq) (7.0)\n",
      "Requirement already satisfied: binaryornot>=0.2.0 in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/lib/python3.6/site-packages (from cookiecutter->kipoi>=0.5.5->kipoiseq) (0.4.4)\n",
      "Requirement already satisfied: poyo>=0.1.0 in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/lib/python3.6/site-packages (from cookiecutter->kipoi>=0.5.5->kipoiseq) (0.4.2)\n",
      "Requirement already satisfied: idna<2.8,>=2.5 in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/lib/python3.6/site-packages (from requests>=2.18.0->cookiecutter->kipoi>=0.5.5->kipoiseq) (2.7)\n",
      "Requirement already satisfied: chardet<3.1.0,>=3.0.2 in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/lib/python3.6/site-packages (from requests>=2.18.0->cookiecutter->kipoi>=0.5.5->kipoiseq) (3.0.4)\n",
      "Requirement already satisfied: certifi>=2017.4.17 in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/lib/python3.6/site-packages (from requests>=2.18.0->cookiecutter->kipoi>=0.5.5->kipoiseq) (2018.11.29)\n",
      "Requirement already satisfied: arrow in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/lib/python3.6/site-packages (from jinja2-time>=0.1.0->cookiecutter->kipoi>=0.5.5->kipoiseq) (0.12.1)\n",
      "\u001b[33mYou are using pip version 19.0.1, however version 19.0.2 is available.\n",
      "You should consider upgrading via the 'pip install --upgrade pip' command.\u001b[0m\n"
     ]
    }
   ],
   "source": [
    "!pip install kipoiseq\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: pyfaidx in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/lib/python3.6/site-packages (0.5.4.2)\n",
      "Requirement already satisfied: setuptools>=0.7 in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/lib/python3.6/site-packages (from pyfaidx) (39.1.0)\n",
      "Requirement already satisfied: six in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/lib/python3.6/site-packages (from pyfaidx) (1.12.0)\n",
      "\u001b[33mYou are using pip version 19.0.1, however version 19.0.2 is available.\n",
      "You should consider upgrading via the 'pip install --upgrade pip' command.\u001b[0m\n"
     ]
    }
   ],
   "source": [
    " !pip install pyfaidx"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "from pyfaidx import Fasta\n",
    "refGenome = Fasta('/s/genomes/human/hg38/hg38.fa')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "odict_keys(['chr1', 'chr10', 'chr11', 'chr11_KI270721v1_random', 'chr12', 'chr13', 'chr14', 'chr14_GL000009v2_random', 'chr14_GL000225v1_random', 'chr14_KI270722v1_random', 'chr14_GL000194v1_random', 'chr14_KI270723v1_random', 'chr14_KI270724v1_random', 'chr14_KI270725v1_random', 'chr14_KI270726v1_random', 'chr15', 'chr15_KI270727v1_random', 'chr16', 'chr16_KI270728v1_random', 'chr17', 'chr17_GL000205v2_random', 'chr17_KI270729v1_random', 'chr17_KI270730v1_random', 'chr18', 'chr19', 'chr1_KI270706v1_random', 'chr1_KI270707v1_random', 'chr1_KI270708v1_random', 'chr1_KI270709v1_random', 'chr1_KI270710v1_random', 'chr1_KI270711v1_random', 'chr1_KI270712v1_random', 'chr1_KI270713v1_random', 'chr1_KI270714v1_random', 'chr2', 'chr20', 'chr21', 'chr22', 'chr22_KI270731v1_random', 'chr22_KI270732v1_random', 'chr22_KI270733v1_random', 'chr22_KI270734v1_random', 'chr22_KI270735v1_random', 'chr22_KI270736v1_random', 'chr22_KI270737v1_random', 'chr22_KI270738v1_random', 'chr22_KI270739v1_random', 'chr2_KI270715v1_random', 'chr2_KI270716v1_random', 'chr3', 'chr3_GL000221v1_random', 'chr4', 'chr4_GL000008v2_random', 'chr5', 'chr5_GL000208v1_random', 'chr6', 'chr7', 'chr8', 'chr9', 'chr9_KI270717v1_random', 'chr9_KI270718v1_random', 'chr9_KI270719v1_random', 'chr9_KI270720v1_random', 'chr1_KI270762v1_alt', 'chr1_KI270766v1_alt', 'chr1_KI270760v1_alt', 'chr1_KI270765v1_alt', 'chr1_GL383518v1_alt', 'chr1_GL383519v1_alt', 'chr1_GL383520v2_alt', 'chr1_KI270764v1_alt', 'chr1_KI270763v1_alt', 'chr1_KI270759v1_alt', 'chr1_KI270761v1_alt', 'chr2_KI270770v1_alt', 'chr2_KI270773v1_alt', 'chr2_KI270774v1_alt', 'chr2_KI270769v1_alt', 'chr2_GL383521v1_alt', 'chr2_KI270772v1_alt', 'chr2_KI270775v1_alt', 'chr2_KI270771v1_alt', 'chr2_KI270768v1_alt', 'chr2_GL582966v2_alt', 'chr2_GL383522v1_alt', 'chr2_KI270776v1_alt', 'chr2_KI270767v1_alt', 'chr3_JH636055v2_alt', 'chr3_KI270783v1_alt', 'chr3_KI270780v1_alt', 'chr3_GL383526v1_alt', 'chr3_KI270777v1_alt', 'chr3_KI270778v1_alt', 'chr3_KI270781v1_alt', 'chr3_KI270779v1_alt', 'chr3_KI270782v1_alt', 'chr3_KI270784v1_alt', 'chr4_KI270790v1_alt', 'chr4_GL383528v1_alt', 'chr4_KI270787v1_alt', 'chr4_GL000257v2_alt', 'chr4_KI270788v1_alt', 'chr4_GL383527v1_alt', 'chr4_KI270785v1_alt', 'chr4_KI270789v1_alt', 'chr4_KI270786v1_alt', 'chr5_KI270793v1_alt', 'chr5_KI270792v1_alt', 'chr5_KI270791v1_alt', 'chr5_GL383532v1_alt', 'chr5_GL949742v1_alt', 'chr5_KI270794v1_alt', 'chr5_GL339449v2_alt', 'chr5_GL383530v1_alt', 'chr5_KI270796v1_alt', 'chr5_GL383531v1_alt', 'chr5_KI270795v1_alt', 'chr6_GL000250v2_alt', 'chr6_KI270800v1_alt', 'chr6_KI270799v1_alt', 'chr6_GL383533v1_alt', 'chr6_KI270801v1_alt', 'chr6_KI270802v1_alt', 'chr6_KB021644v2_alt', 'chr6_KI270797v1_alt', 'chr6_KI270798v1_alt', 'chr7_KI270804v1_alt', 'chr7_KI270809v1_alt', 'chr7_KI270806v1_alt', 'chr7_GL383534v2_alt', 'chr7_KI270803v1_alt', 'chr7_KI270808v1_alt', 'chr7_KI270807v1_alt', 'chr7_KI270805v1_alt', 'chr8_KI270818v1_alt', 'chr8_KI270812v1_alt', 'chr8_KI270811v1_alt', 'chr8_KI270821v1_alt', 'chr8_KI270813v1_alt', 'chr8_KI270822v1_alt', 'chr8_KI270814v1_alt', 'chr8_KI270810v1_alt', 'chr8_KI270819v1_alt', 'chr8_KI270820v1_alt', 'chr8_KI270817v1_alt', 'chr8_KI270816v1_alt', 'chr8_KI270815v1_alt', 'chr9_GL383539v1_alt', 'chr9_GL383540v1_alt', 'chr9_GL383541v1_alt', 'chr9_GL383542v1_alt', 'chr9_KI270823v1_alt', 'chr10_GL383545v1_alt', 'chr10_KI270824v1_alt', 'chr10_GL383546v1_alt', 'chr10_KI270825v1_alt', 'chr11_KI270832v1_alt', 'chr11_KI270830v1_alt', 'chr11_KI270831v1_alt', 'chr11_KI270829v1_alt', 'chr11_GL383547v1_alt', 'chr11_JH159136v1_alt', 'chr11_JH159137v1_alt', 'chr11_KI270827v1_alt', 'chr11_KI270826v1_alt', 'chr12_GL877875v1_alt', 'chr12_GL877876v1_alt', 'chr12_KI270837v1_alt', 'chr12_GL383549v1_alt', 'chr12_KI270835v1_alt', 'chr12_GL383550v2_alt', 'chr12_GL383552v1_alt', 'chr12_GL383553v2_alt', 'chr12_KI270834v1_alt', 'chr12_GL383551v1_alt', 'chr12_KI270833v1_alt', 'chr12_KI270836v1_alt', 'chr13_KI270840v1_alt', 'chr13_KI270839v1_alt', 'chr13_KI270843v1_alt', 'chr13_KI270841v1_alt', 'chr13_KI270838v1_alt', 'chr13_KI270842v1_alt', 'chr14_KI270844v1_alt', 'chr14_KI270847v1_alt', 'chr14_KI270845v1_alt', 'chr14_KI270846v1_alt', 'chr15_KI270852v1_alt', 'chr15_KI270851v1_alt', 'chr15_KI270848v1_alt', 'chr15_GL383554v1_alt', 'chr15_KI270849v1_alt', 'chr15_GL383555v2_alt', 'chr15_KI270850v1_alt', 'chr16_KI270854v1_alt', 'chr16_KI270856v1_alt', 'chr16_KI270855v1_alt', 'chr16_KI270853v1_alt', 'chr16_GL383556v1_alt', 'chr16_GL383557v1_alt', 'chr17_GL383563v3_alt', 'chr17_KI270862v1_alt', 'chr17_KI270861v1_alt', 'chr17_KI270857v1_alt', 'chr17_JH159146v1_alt', 'chr17_JH159147v1_alt', 'chr17_GL383564v2_alt', 'chr17_GL000258v2_alt', 'chr17_GL383565v1_alt', 'chr17_KI270858v1_alt', 'chr17_KI270859v1_alt', 'chr17_GL383566v1_alt', 'chr17_KI270860v1_alt', 'chr18_KI270864v1_alt', 'chr18_GL383567v1_alt', 'chr18_GL383570v1_alt', 'chr18_GL383571v1_alt', 'chr18_GL383568v1_alt', 'chr18_GL383569v1_alt', 'chr18_GL383572v1_alt', 'chr18_KI270863v1_alt', 'chr19_KI270868v1_alt', 'chr19_KI270865v1_alt', 'chr19_GL383573v1_alt', 'chr19_GL383575v2_alt', 'chr19_GL383576v1_alt', 'chr19_GL383574v1_alt', 'chr19_KI270866v1_alt', 'chr19_KI270867v1_alt', 'chr19_GL949746v1_alt', 'chr20_GL383577v2_alt', 'chr20_KI270869v1_alt', 'chr20_KI270871v1_alt', 'chr20_KI270870v1_alt', 'chr21_GL383578v2_alt', 'chr21_KI270874v1_alt', 'chr21_KI270873v1_alt', 'chr21_GL383579v2_alt', 'chr21_GL383580v2_alt', 'chr21_GL383581v2_alt', 'chr21_KI270872v1_alt', 'chr22_KI270875v1_alt', 'chr22_KI270878v1_alt', 'chr22_KI270879v1_alt', 'chr22_KI270876v1_alt', 'chr22_KI270877v1_alt', 'chr22_GL383583v2_alt', 'chr22_GL383582v2_alt', 'chrX_KI270880v1_alt', 'chrX_KI270881v1_alt', 'chr19_KI270882v1_alt', 'chr19_KI270883v1_alt', 'chr19_KI270884v1_alt', 'chr19_KI270885v1_alt', 'chr19_KI270886v1_alt', 'chr19_KI270887v1_alt', 'chr19_KI270888v1_alt', 'chr19_KI270889v1_alt', 'chr19_KI270890v1_alt', 'chr19_KI270891v1_alt', 'chr1_KI270892v1_alt', 'chr2_KI270894v1_alt', 'chr2_KI270893v1_alt', 'chr3_KI270895v1_alt', 'chr4_KI270896v1_alt', 'chr5_KI270897v1_alt', 'chr5_KI270898v1_alt', 'chr6_GL000251v2_alt', 'chr7_KI270899v1_alt', 'chr8_KI270901v1_alt', 'chr8_KI270900v1_alt', 'chr11_KI270902v1_alt', 'chr11_KI270903v1_alt', 'chr12_KI270904v1_alt', 'chr15_KI270906v1_alt', 'chr15_KI270905v1_alt', 'chr17_KI270907v1_alt', 'chr17_KI270910v1_alt', 'chr17_KI270909v1_alt', 'chr17_JH159148v1_alt', 'chr17_KI270908v1_alt', 'chr18_KI270912v1_alt', 'chr18_KI270911v1_alt', 'chr19_GL949747v2_alt', 'chr22_KB663609v1_alt', 'chrX_KI270913v1_alt', 'chr19_KI270914v1_alt', 'chr19_KI270915v1_alt', 'chr19_KI270916v1_alt', 'chr19_KI270917v1_alt', 'chr19_KI270918v1_alt', 'chr19_KI270919v1_alt', 'chr19_KI270920v1_alt', 'chr19_KI270921v1_alt', 'chr19_KI270922v1_alt', 'chr19_KI270923v1_alt', 'chr3_KI270924v1_alt', 'chr4_KI270925v1_alt', 'chr6_GL000252v2_alt', 'chr8_KI270926v1_alt', 'chr11_KI270927v1_alt', 'chr19_GL949748v2_alt', 'chr22_KI270928v1_alt', 'chr19_KI270929v1_alt', 'chr19_KI270930v1_alt', 'chr19_KI270931v1_alt', 'chr19_KI270932v1_alt', 'chr19_KI270933v1_alt', 'chr19_GL000209v2_alt', 'chr3_KI270934v1_alt', 'chr6_GL000253v2_alt', 'chr19_GL949749v2_alt', 'chr3_KI270935v1_alt', 'chr6_GL000254v2_alt', 'chr19_GL949750v2_alt', 'chr3_KI270936v1_alt', 'chr6_GL000255v2_alt', 'chr19_GL949751v2_alt', 'chr3_KI270937v1_alt', 'chr6_GL000256v2_alt', 'chr19_GL949752v1_alt', 'chr6_KI270758v1_alt', 'chr19_GL949753v2_alt', 'chr19_KI270938v1_alt', 'chrM', 'chrUn_KI270302v1', 'chrUn_KI270304v1', 'chrUn_KI270303v1', 'chrUn_KI270305v1', 'chrUn_KI270322v1', 'chrUn_KI270320v1', 'chrUn_KI270310v1', 'chrUn_KI270316v1', 'chrUn_KI270315v1', 'chrUn_KI270312v1', 'chrUn_KI270311v1', 'chrUn_KI270317v1', 'chrUn_KI270412v1', 'chrUn_KI270411v1', 'chrUn_KI270414v1', 'chrUn_KI270419v1', 'chrUn_KI270418v1', 'chrUn_KI270420v1', 'chrUn_KI270424v1', 'chrUn_KI270417v1', 'chrUn_KI270422v1', 'chrUn_KI270423v1', 'chrUn_KI270425v1', 'chrUn_KI270429v1', 'chrUn_KI270442v1', 'chrUn_KI270466v1', 'chrUn_KI270465v1', 'chrUn_KI270467v1', 'chrUn_KI270435v1', 'chrUn_KI270438v1', 'chrUn_KI270468v1', 'chrUn_KI270510v1', 'chrUn_KI270509v1', 'chrUn_KI270518v1', 'chrUn_KI270508v1', 'chrUn_KI270516v1', 'chrUn_KI270512v1', 'chrUn_KI270519v1', 'chrUn_KI270522v1', 'chrUn_KI270511v1', 'chrUn_KI270515v1', 'chrUn_KI270507v1', 'chrUn_KI270517v1', 'chrUn_KI270529v1', 'chrUn_KI270528v1', 'chrUn_KI270530v1', 'chrUn_KI270539v1', 'chrUn_KI270538v1', 'chrUn_KI270544v1', 'chrUn_KI270548v1', 'chrUn_KI270583v1', 'chrUn_KI270587v1', 'chrUn_KI270580v1', 'chrUn_KI270581v1', 'chrUn_KI270579v1', 'chrUn_KI270589v1', 'chrUn_KI270590v1', 'chrUn_KI270584v1', 'chrUn_KI270582v1', 'chrUn_KI270588v1', 'chrUn_KI270593v1', 'chrUn_KI270591v1', 'chrUn_KI270330v1', 'chrUn_KI270329v1', 'chrUn_KI270334v1', 'chrUn_KI270333v1', 'chrUn_KI270335v1', 'chrUn_KI270338v1', 'chrUn_KI270340v1', 'chrUn_KI270336v1', 'chrUn_KI270337v1', 'chrUn_KI270363v1', 'chrUn_KI270364v1', 'chrUn_KI270362v1', 'chrUn_KI270366v1', 'chrUn_KI270378v1', 'chrUn_KI270379v1', 'chrUn_KI270389v1', 'chrUn_KI270390v1', 'chrUn_KI270387v1', 'chrUn_KI270395v1', 'chrUn_KI270396v1', 'chrUn_KI270388v1', 'chrUn_KI270394v1', 'chrUn_KI270386v1', 'chrUn_KI270391v1', 'chrUn_KI270383v1', 'chrUn_KI270393v1', 'chrUn_KI270384v1', 'chrUn_KI270392v1', 'chrUn_KI270381v1', 'chrUn_KI270385v1', 'chrUn_KI270382v1', 'chrUn_KI270376v1', 'chrUn_KI270374v1', 'chrUn_KI270372v1', 'chrUn_KI270373v1', 'chrUn_KI270375v1', 'chrUn_KI270371v1', 'chrUn_KI270448v1', 'chrUn_KI270521v1', 'chrUn_GL000195v1', 'chrUn_GL000219v1', 'chrUn_GL000220v1', 'chrUn_GL000224v1', 'chrUn_KI270741v1', 'chrUn_GL000226v1', 'chrUn_GL000213v1', 'chrUn_KI270743v1', 'chrUn_KI270744v1', 'chrUn_KI270745v1', 'chrUn_KI270746v1', 'chrUn_KI270747v1', 'chrUn_KI270748v1', 'chrUn_KI270749v1', 'chrUn_KI270750v1', 'chrUn_KI270751v1', 'chrUn_KI270752v1', 'chrUn_KI270753v1', 'chrUn_KI270754v1', 'chrUn_KI270755v1', 'chrUn_KI270756v1', 'chrUn_KI270757v1', 'chrUn_GL000214v1', 'chrUn_KI270742v1', 'chrUn_GL000216v2', 'chrUn_GL000218v1', 'chrX', 'chrY', 'chrY_KI270740v1_random'])"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "refGenome.keys()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       ">chr1:40001-40100\n",
       "gcctcatggaggggatcagctgcgaggagctaagagccccctccagtcgatgctcaccaggaagctgaggtcttgtgtccagcaccctgcatagaactga"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "refGenome[\"chr1\"][40000:40100]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "import vcf\n",
    "\n",
    "vcf_reader = vcf.Reader(filename='/s/project/gtex-processed/vcf/GTEx_Analysis_20150112_WholeGenomeSeq_148Indiv_GATK_HaplotypeCaller.vcf.gz')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[Call(sample=GTEX-QV31-0003-SM-5URDH, CallData(GT=0/0, AD=[35, 0], DP=35, GQ=79, PL=[0, 79, 1169])), Call(sample=GTEX-R55E-0003-SM-5URBX, CallData(GT=0/0, AD=[27, 0], DP=27, GQ=69, PL=[0, 69, 847])), Call(sample=GTEX-X4XY-0002-SM-5URBG, CallData(GT=0/0, AD=[34, 0], DP=34, GQ=62, PL=[0, 62, 949])), Call(sample=GTEX-X62O-0001-SM-5URD5, CallData(GT=1/1, AD=[0, 55], DP=55, GQ=99, PL=[1645, 164, 0])), Call(sample=GTEX-N7MS-0009-SM-5JK3E, CallData(GT=0/0, AD=[23, 0], DP=23, GQ=60, PL=[0, 60, 900])), Call(sample=GTEX-N7MT-0009-SM-5JK4X, CallData(GT=0/1, AD=[21, 12], DP=33, GQ=99, PL=[324, 0, 619])), Call(sample=GTEX-NFK9-0004-SM-5JK4Y, CallData(GT=0/0, AD=[27, 0], DP=27, GQ=73, PL=[0, 73, 924])), Call(sample=GTEX-NL3H-0009-SM-5JK37, CallData(GT=0/0, AD=[29, 0], DP=29, GQ=67, PL=[0, 67, 1016])), Call(sample=GTEX-NL4W-0009-SM-5JK3O, CallData(GT=0/0, AD=[29, 0], DP=29, GQ=63, PL=[0, 63, 1041])), Call(sample=GTEX-NPJ7-0009-SM-5JK4M, CallData(GT=0/1, AD=[14, 18], DP=32, GQ=99, PL=[541, 0, 393])), Call(sample=GTEX-NPJ8-0004-SM-5JK3D, CallData(GT=0/0, AD=[23, 0], DP=23, GQ=60, PL=[0, 60, 900])), Call(sample=GTEX-O5YT-0010-SM-5JK31, CallData(GT=0/1, AD=[17, 20], DP=37, GQ=99, PL=[591, 0, 481])), Call(sample=GTEX-O5YV-0004-SM-5JK2N, CallData(GT=0/1, AD=[28, 24], DP=52, GQ=99, PL=[694, 0, 829])), Call(sample=GTEX-O5YW-0003-SM-5JK2Y, CallData(GT=0/0, AD=[25, 0], DP=25, GQ=60, PL=[0, 60, 879])), Call(sample=GTEX-OHPK-0003-SM-5JK4Z, CallData(GT=0/0, AD=[24, 0], DP=24, GQ=60, PL=[0, 60, 900])), Call(sample=GTEX-OHPM-0003-SM-5JK4I, CallData(GT=0/1, AD=[21, 21], DP=42, GQ=99, PL=[590, 0, 604])), Call(sample=GTEX-OHPN-0004-SM-5JK2V, CallData(GT=0/1, AD=[40, 29], DP=69, GQ=99, PL=[730, 0, 1124])), Call(sample=GTEX-OIZF-0004-SM-5JK4C, CallData(GT=0/0, AD=[24, 0], DP=24, GQ=63, PL=[0, 63, 945])), Call(sample=GTEX-OIZG-0003-SM-5JK3W, CallData(GT=0/0, AD=[25, 0], DP=25, GQ=60, PL=[0, 60, 900])), Call(sample=GTEX-OIZH-0003-SM-5JK41, CallData(GT=0/1, AD=[16, 31], DP=47, GQ=99, PL=[905, 0, 421])), Call(sample=GTEX-OIZI-1026-SM-5JK2X, CallData(GT=0/1, AD=[27, 21], DP=48, GQ=99, PL=[628, 0, 875])), Call(sample=GTEX-OOBK-0004-SM-5JK4J, CallData(GT=0/1, AD=[26, 21], DP=47, GQ=99, PL=[600, 0, 753])), Call(sample=GTEX-OXRK-0004-SM-5JK32, CallData(GT=0/0, AD=[22, 0], DP=22, GQ=60, PL=[0, 60, 805])), Call(sample=GTEX-OXRL-0003-SM-5JK3X, CallData(GT=0/0, AD=[25, 0], DP=25, GQ=63, PL=[0, 63, 806])), Call(sample=GTEX-OXRN-0003-SM-5JK51, CallData(GT=0/1, AD=[17, 13], DP=30, GQ=99, PL=[376, 0, 511])), Call(sample=GTEX-OXRO-0003-SM-5JK2R, CallData(GT=0/1, AD=[27, 25], DP=52, GQ=99, PL=[707, 0, 785])), Call(sample=GTEX-OXRP-0003-SM-5JK3P, CallData(GT=1/1, AD=[1, 38], DP=39, GQ=99, PL=[1271, 107, 0])), Call(sample=GTEX-P44H-1126-SM-5JK39, CallData(GT=0/0, AD=[28, 0], DP=28, GQ=63, PL=[0, 63, 945])), Call(sample=GTEX-P4PP-0004-SM-5JK2S, CallData(GT=0/0, AD=[27, 0], DP=27, GQ=65, PL=[0, 65, 928])), Call(sample=GTEX-P4PQ-0003-SM-5JK3Q, CallData(GT=0/0, AD=[30, 0], DP=30, GQ=65, PL=[0, 65, 947])), Call(sample=GTEX-P4QR-0004-SM-5JK2Q, CallData(GT=0/0, AD=[26, 0], DP=26, GQ=60, PL=[0, 60, 900])), Call(sample=GTEX-P4QS-0003-SM-5JK4U, CallData(GT=0/0, AD=[32, 0], DP=32, GQ=65, PL=[0, 65, 1063])), Call(sample=GTEX-P4QT-0003-SM-5JK2P, CallData(GT=0/0, AD=[26, 0], DP=26, GQ=72, PL=[0, 72, 841])), Call(sample=GTEX-P78B-0004-SM-5JK48, CallData(GT=0/0, AD=[26, 0], DP=26, GQ=63, PL=[0, 63, 938])), Call(sample=GTEX-PLZ4-0003-SM-5JK4H, CallData(GT=0/0, AD=[27, 0], DP=27, GQ=61, PL=[0, 61, 874])), Call(sample=GTEX-PLZ5-0003-SM-5JK3C, CallData(GT=0/0, AD=[25, 0], DP=25, GQ=60, PL=[0, 60, 844])), Call(sample=GTEX-PLZ6-0003-SM-5JK4W, CallData(GT=1/1, AD=[0, 24], DP=24, GQ=72, PL=[827, 72, 0])), Call(sample=GTEX-POMQ-0003-SM-5JK46, CallData(GT=0/1, AD=[20, 18], DP=38, GQ=99, PL=[516, 0, 579])), Call(sample=GTEX-POYW-1226-SM-5JK3K, CallData(GT=0/1, AD=[12, 22], DP=34, GQ=99, PL=[683, 0, 323])), Call(sample=GTEX-PSDG-0004-SM-5JK3L, CallData(GT=0/0, AD=[25, 0], DP=25, GQ=63, PL=[0, 63, 945])), Call(sample=GTEX-PVOW-1026-SM-5JK3V, CallData(GT=0/1, AD=[20, 27], DP=47, GQ=99, PL=[817, 0, 602])), Call(sample=GTEX-PW2O-0004-SM-5JK3M, CallData(GT=0/0, AD=[27, 0], DP=27, GQ=60, PL=[0, 60, 900])), Call(sample=GTEX-PWCY-0001-SM-5JK3J, CallData(GT=0/1, AD=[23, 25], DP=48, GQ=99, PL=[709, 0, 627])), Call(sample=GTEX-PWN1-0004-SM-5JK3S, CallData(GT=0/1, AD=[15, 25], DP=40, GQ=99, PL=[724, 0, 359])), Call(sample=GTEX-PWO3-0004-SM-5JK3T, CallData(GT=0/0, AD=[37, 0], DP=37, GQ=61, PL=[0, 61, 1094])), Call(sample=GTEX-PWOO-0004-SM-5JK4A, CallData(GT=0/0, AD=[25, 0], DP=25, GQ=66, PL=[0, 66, 990])), Call(sample=GTEX-Q2AG-0003-SM-5JK4S, CallData(GT=0/1, AD=[16, 15], DP=31, GQ=99, PL=[432, 0, 415])), Call(sample=GTEX-Q2AH-0004-SM-5JK52, CallData(GT=0/1, AD=[24, 19], DP=43, GQ=99, PL=[539, 0, 686])), Call(sample=GTEX-Q2AI-0003-SM-5JK3F, CallData(GT=0/1, AD=[17, 24], DP=41, GQ=99, PL=[741, 0, 450])), Call(sample=GTEX-Q734-0001-SM-5JK3N, CallData(GT=0/0, AD=[27, 0], DP=27, GQ=66, PL=[0, 66, 909])), Call(sample=GTEX-QEG4-0003-SM-5URCG, CallData(GT=0/1, AD=[22, 18], DP=40, GQ=99, PL=[424, 0, 594])), Call(sample=GTEX-QEG5-0003-SM-5URCP, CallData(GT=0/1, AD=[14, 21], DP=35, GQ=99, PL=[551, 0, 322])), Call(sample=GTEX-QESD-0001-SM-5URC9, CallData(GT=0/0, AD=[28, 0], DP=28, GQ=68, PL=[0, 68, 868])), Call(sample=GTEX-QLQW-0001-SM-5URCU, CallData(GT=0/0, AD=[23, 0], DP=23, GQ=60, PL=[0, 60, 900])), Call(sample=GTEX-QMRM-0002-SM-5URC1, CallData(GT=0/1, AD=[17, 10], DP=27, GQ=99, PL=[213, 0, 497])), Call(sample=GTEX-QV44-0003-SM-5SOEM, CallData(GT=0/0, AD=[35, 0], DP=35, GQ=84, PL=[0, 84, 1101])), Call(sample=GTEX-QVUS-0001-SM-5URDT, CallData(GT=0/1, AD=[24, 28], DP=52, GQ=99, PL=[662, 0, 641])), Call(sample=GTEX-QXCU-0004-SM-5URDS, CallData(GT=0/0, AD=[34, 0], DP=34, GQ=61, PL=[0, 61, 906])), Call(sample=GTEX-R3RS-0001-SM-5URBE, CallData(GT=0/1, AD=[20, 21], DP=41, GQ=99, PL=[552, 0, 527])), Call(sample=GTEX-R55C-0003-SM-5URCJ, CallData(GT=0/1, AD=[14, 16], DP=30, GQ=99, PL=[423, 0, 322])), Call(sample=GTEX-R55G-0003-SM-5URBQ, CallData(GT=0/0, AD=[34, 0], DP=34, GQ=71, PL=[0, 71, 961])), Call(sample=GTEX-REY6-0004-SM-5URD2, CallData(GT=0/1, AD=[22, 17], DP=39, GQ=99, PL=[428, 0, 479])), Call(sample=GTEX-RM2N-0002-SM-5SOEB, CallData(GT=1/1, AD=[0, 43], DP=43, GQ=99, PL=[1285, 128, 0])), Call(sample=GTEX-RN64-0004-SM-5URBC, CallData(GT=0/0, AD=[32, 0], DP=32, GQ=66, PL=[0, 66, 932])), Call(sample=GTEX-RNOR-0001-SM-5URCF, CallData(GT=0/1, AD=[20, 21], DP=41, GQ=99, PL=[553, 0, 528])), Call(sample=GTEX-RU1J-0002-SM-5URDF, CallData(GT=0/1, AD=[14, 23], DP=37, GQ=99, PL=[639, 0, 314])), Call(sample=GTEX-RU72-0004-SM-5SOEJ, CallData(GT=0/0, AD=[28, 0], DP=28, GQ=66, PL=[0, 66, 852])), Call(sample=GTEX-RUSQ-0001-SM-5JK34, CallData(GT=1/1, AD=[0, 52], DP=52, GQ=99, PL=[1776, 156, 0])), Call(sample=GTEX-RWS6-0004-SM-5JK3Z, CallData(GT=0/0, AD=[25, 0], DP=25, GQ=63, PL=[0, 63, 945])), Call(sample=GTEX-RWSA-0003-SM-5JK4L, CallData(GT=0/0, AD=[28, 0], DP=28, GQ=65, PL=[0, 65, 950])), Call(sample=GTEX-S32W-0001-SM-5JK4N, CallData(GT=0/0, AD=[30, 0], DP=30, GQ=64, PL=[0, 64, 1030])), Call(sample=GTEX-S341-0001-SM-5JK33, CallData(GT=0/1, AD=[21, 13], DP=34, GQ=99, PL=[354, 0, 603])), Call(sample=GTEX-S4P3-0004-SM-5JK4R, CallData(GT=0/0, AD=[27, 0], DP=27, GQ=78, PL=[0, 78, 923])), Call(sample=GTEX-S4Q7-0001-SM-5JK3A, CallData(GT=0/0, AD=[22, 0], DP=22, GQ=60, PL=[0, 60, 900])), Call(sample=GTEX-S4UY-0003-SM-5JK3Y, CallData(GT=0/1, AD=[14, 19], DP=33, GQ=99, PL=[569, 0, 374])), Call(sample=GTEX-S4Z8-0003-SM-5JK3I, CallData(GT=0/1, AD=[20, 25], DP=45, GQ=99, PL=[719, 0, 534])), Call(sample=GTEX-S7PM-0926-SM-5JK36, CallData(GT=0/0, AD=[28, 0], DP=28, GQ=66, PL=[0, 66, 990])), Call(sample=GTEX-S7SE-0004-SM-5JK4E, CallData(GT=0/0, AD=[25, 0], DP=25, GQ=60, PL=[0, 60, 843])), Call(sample=GTEX-S95S-0004-SM-5JK3U, CallData(GT=0/0, AD=[32, 0], DP=32, GQ=84, PL=[0, 84, 1067])), Call(sample=GTEX-SIU7-0002-SM-5JK4G, CallData(GT=0/1, AD=[28, 17], DP=45, GQ=99, PL=[465, 0, 825])), Call(sample=GTEX-SIU8-0002-SM-5JK47, CallData(GT=1/1, AD=[0, 44], DP=44, GQ=99, PL=[1561, 132, 0])), Call(sample=GTEX-SN8G-0002-SM-5JK4B, CallData(GT=0/0, AD=[30, 0], DP=30, GQ=84, PL=[0, 84, 1020])), Call(sample=GTEX-SNOS-0004-SM-5SOEL, CallData(GT=0/0, AD=[25, 0], DP=25, GQ=63, PL=[0, 63, 793])), Call(sample=GTEX-SSA3-0001-SM-5URCI, CallData(GT=0/0, AD=[27, 0], DP=27, GQ=63, PL=[0, 63, 829])), Call(sample=GTEX-SUCS-0001-SM-5URDQ, CallData(GT=0/1, AD=[21, 19], DP=40, GQ=99, PL=[475, 0, 515])), Call(sample=GTEX-T2IS-0002-SM-5SOEK, CallData(GT=0/0, AD=[34, 0], DP=34, GQ=61, PL=[0, 61, 975])), Call(sample=GTEX-T2YK-0003-SM-5URCD, CallData(GT=0/1, AD=[19, 32], DP=51, GQ=99, PL=[851, 0, 430])), Call(sample=GTEX-T5JC-0002-SM-5SOER, CallData(GT=0/1, AD=[19, 25], DP=44, GQ=99, PL=[658, 0, 460])), Call(sample=GTEX-T5JW-0002-SM-5URDG, CallData(GT=0/1, AD=[16, 22], DP=38, GQ=99, PL=[567, 0, 389])), Call(sample=GTEX-T6MN-0003-SM-5URB7, CallData(GT=0/0, AD=[33, 0], DP=33, GQ=81, PL=[0, 81, 1027])), Call(sample=GTEX-T6MO-0001-SM-5URCH, CallData(GT=0/1, AD=[22, 9], DP=31, GQ=99, PL=[182, 0, 573])), Call(sample=GTEX-T8EM-0001-SM-5SOEA, CallData(GT=0/0, AD=[24, 0], DP=24, GQ=62, PL=[0, 62, 862])), Call(sample=GTEX-TKQ2-0002-SM-5URCO, CallData(GT=0/0, AD=[28, 0], DP=28, GQ=68, PL=[0, 68, 841])), Call(sample=GTEX-TML8-0004-SM-5SOEP, CallData(GT=0/0, AD=[27, 0], DP=27, GQ=63, PL=[0, 63, 829])), Call(sample=GTEX-TMMY-0003-SM-5SOES, CallData(GT=1/1, AD=[0, 52], DP=52, GQ=99, PL=[1604, 156, 0])), Call(sample=GTEX-TMZS-0002-SM-5URDR, CallData(GT=0/0, AD=[28, 0], DP=28, GQ=63, PL=[0, 63, 879])), Call(sample=GTEX-TSE9-0002-SM-5URBA, CallData(GT=0/0, AD=[24, 0], DP=24, GQ=62, PL=[0, 62, 722])), Call(sample=GTEX-U3ZG-0002-SM-5SOEN, CallData(GT=1/1, AD=[0, 43], DP=43, GQ=99, PL=[1235, 129, 0])), Call(sample=GTEX-U3ZH-0003-SM-5URBK, CallData(GT=0/1, AD=[14, 27], DP=41, GQ=99, PL=[734, 0, 308])), Call(sample=GTEX-U3ZN-0001-SM-5URBO, CallData(GT=0/0, AD=[25, 0], DP=25, GQ=64, PL=[0, 64, 762])), Call(sample=GTEX-U4B1-0001-SM-5URC2, CallData(GT=0/0, AD=[32, 0], DP=32, GQ=74, PL=[0, 74, 1031])), Call(sample=GTEX-U8XE-0004-SM-5URB8, CallData(GT=0/0, AD=[27, 0], DP=27, GQ=66, PL=[0, 66, 807])), Call(sample=GTEX-UJHI-0001-SM-5SOEU, CallData(GT=0/0, AD=[30, 0], DP=30, GQ=65, PL=[0, 65, 961])), Call(sample=GTEX-UJMC-0002-SM-5SOEI, CallData(GT=0/0, AD=[33, 0], DP=33, GQ=63, PL=[0, 63, 916])), Call(sample=GTEX-UPIC-0004-SM-5SOEF, CallData(GT=0/1, AD=[16, 17], DP=33, GQ=99, PL=[414, 0, 410])), Call(sample=GTEX-UPJH-0002-SM-5SOE9, CallData(GT=0/0, AD=[30, 0], DP=30, GQ=64, PL=[0, 64, 974])), Call(sample=GTEX-UPK5-0002-SM-5SOE3, CallData(GT=0/0, AD=[26, 0], DP=26, GQ=60, PL=[0, 60, 789])), Call(sample=GTEX-UTHO-0003-SM-5SOE4, CallData(GT=1/1, AD=[0, 34], DP=34, GQ=99, PL=[1076, 102, 0])), Call(sample=GTEX-V955-0001-SM-5SOEE, CallData(GT=0/1, AD=[29, 28], DP=57, GQ=99, PL=[665, 0, 679])), Call(sample=GTEX-VJYA-0003-SM-5SOED, CallData(GT=0/1, AD=[18, 20], DP=38, GQ=99, PL=[490, 0, 408])), Call(sample=GTEX-VUSH-0002-SM-5SOEH, CallData(GT=0/1, AD=[20, 19], DP=39, GQ=99, PL=[426, 0, 508])), Call(sample=GTEX-W5WG-0003-SM-5URC3, CallData(GT=0/1, AD=[17, 20], DP=37, GQ=99, PL=[539, 0, 405])), Call(sample=GTEX-W5X1-0002-SM-5URBS, CallData(GT=0/0, AD=[25, 0], DP=25, GQ=60, PL=[0, 60, 798])), Call(sample=GTEX-WCDI-0001-SM-5SOE7, CallData(GT=0/0, AD=[26, 0], DP=26, GQ=62, PL=[0, 62, 833])), Call(sample=GTEX-WFG7-0002-SM-5SOEQ, CallData(GT=0/0, AD=[29, 0], DP=29, GQ=64, PL=[0, 64, 807])), Call(sample=GTEX-WFJO-0003-SM-5URBP, CallData(GT=0/0, AD=[22, 0], DP=22, GQ=60, PL=[0, 60, 766])), Call(sample=GTEX-WFON-0003-SM-5URD4, CallData(GT=0/0, AD=[26, 0], DP=26, GQ=63, PL=[0, 63, 916])), Call(sample=GTEX-WHPG-0002-SM-5URBH, CallData(GT=0/0, AD=[37, 0], DP=37, GQ=75, PL=[0, 75, 1071])), Call(sample=GTEX-WHSB-0004-SM-5URDE, CallData(GT=0/0, AD=[31, 0], DP=31, GQ=61, PL=[0, 61, 895])), Call(sample=GTEX-WHWD-0002-SM-5SOEC, CallData(GT=0/0, AD=[35, 0], DP=35, GQ=73, PL=[0, 73, 901])), Call(sample=GTEX-WK11-0002-SM-5SOET, CallData(GT=0/0, AD=[28, 0], DP=28, GQ=62, PL=[0, 62, 793])), Call(sample=GTEX-WL46-0002-SM-5SOE8, CallData(GT=0/1, AD=[24, 18], DP=42, GQ=99, PL=[450, 0, 632])), Call(sample=GTEX-WOFM-0003-SM-5JK3H, CallData(GT=1/1, AD=[0, 35], DP=35, GQ=99, PL=[1185, 105, 0])), Call(sample=GTEX-WVLH-0001-SM-5SOE6, CallData(GT=0/0, AD=[32, 0], DP=32, GQ=65, PL=[0, 65, 968])), Call(sample=GTEX-WWYW-0004-SM-5URDN, CallData(GT=0/1, AD=[20, 25], DP=45, GQ=99, PL=[637, 0, 466])), Call(sample=GTEX-WXYG-0001-SM-5SOE5, CallData(GT=0/1, AD=[19, 15], DP=34, GQ=99, PL=[351, 0, 461])), Call(sample=GTEX-WY7C-0002-SM-5URBW, CallData(GT=0/0, AD=[29, 0], DP=29, GQ=65, PL=[0, 65, 889])), Call(sample=GTEX-WYJK-0001-SM-5URBR, CallData(GT=0/1, AD=[13, 17], DP=30, GQ=99, PL=[441, 0, 319])), Call(sample=GTEX-WZTO-0002-SM-5URDU, CallData(GT=0/0, AD=[30, 0], DP=30, GQ=75, PL=[0, 75, 896])), Call(sample=GTEX-X261-0002-SM-5URDP, CallData(GT=0/0, AD=[27, 0], DP=27, GQ=60, PL=[0, 60, 776])), Call(sample=GTEX-X3Y1-0002-SM-5URCV, CallData(GT=0/1, AD=[20, 7], DP=27, GQ=99, PL=[134, 0, 599])), Call(sample=GTEX-X4EO-0003-SM-5URCE, CallData(GT=0/1, AD=[14, 20], DP=34, GQ=99, PL=[456, 0, 349])), Call(sample=GTEX-X4XX-0004-SM-5URCR, CallData(GT=0/0, AD=[28, 0], DP=28, GQ=76, PL=[0, 76, 866])), Call(sample=GTEX-X585-0003-SM-5URCW, CallData(GT=0/0, AD=[28, 0], DP=28, GQ=66, PL=[0, 66, 863])), Call(sample=GTEX-X8HC-0003-SM-5URDI, CallData(GT=0/1, AD=[13, 22], DP=35, GQ=99, PL=[606, 0, 292])), Call(sample=GTEX-XBEC-0002-SM-5SOEV, CallData(GT=0/0, AD=[28, 0], DP=28, GQ=63, PL=[0, 63, 832])), Call(sample=GTEX-XBED-0004-SM-5URD9, CallData(GT=0/1, AD=[13, 19], DP=32, GQ=99, PL=[518, 0, 327])), Call(sample=GTEX-XGQ4-0001-SM-5SOEG, CallData(GT=0/0, AD=[54, 0], DP=54, GQ=99, PL=[0, 109, 1532])), Call(sample=GTEX-XMK1-0003-SM-5URB9, CallData(GT=0/0, AD=[28, 0], DP=28, GQ=69, PL=[0, 69, 803])), Call(sample=GTEX-XOT4-0001-SM-5URCM, CallData(GT=0/0, AD=[28, 0], DP=28, GQ=67, PL=[0, 67, 855])), Call(sample=GTEX-XOTO-0002-SM-5URC7, CallData(GT=0/1, AD=[17, 10], DP=27, GQ=99, PL=[220, 0, 437])), Call(sample=GTEX-XPT6-0004-SM-5URDL, CallData(GT=0/0, AD=[29, 0], DP=29, GQ=68, PL=[0, 68, 878])), Call(sample=GTEX-XQ3S-0002-SM-5JK42, CallData(GT=0/0, AD=[29, 0], DP=29, GQ=75, PL=[0, 75, 893])), Call(sample=GTEX-XUJ4-0001-SM-5JK4T, CallData(GT=0/0, AD=[37, 0], DP=37, GQ=88, PL=[0, 88, 1191])), Call(sample=GTEX-XUW1-0002-SM-5JK4P, CallData(GT=0/1, AD=[26, 20], DP=46, GQ=99, PL=[572, 0, 767])), Call(sample=GTEX-XV7Q-0004-SM-5JK3G, CallData(GT=0/0, AD=[26, 0], DP=26, GQ=65, PL=[0, 65, 990])), Call(sample=GTEX-XXEK-0001-SM-5JK35, CallData(GT=0/0, AD=[29, 0], DP=29, GQ=69, PL=[0, 69, 954])), Call(sample=GTEX-XYKS-0003-SM-5JK4D, CallData(GT=0/0, AD=[31, 0], DP=31, GQ=75, PL=[0, 75, 1016]))]\n"
     ]
    }
   ],
   "source": [
    "for record in vcf_reader.fetch('20', 1110695, 1110780):  \n",
    "    print(record.samples)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "#numpy troubles: in terminal, in RIGHT env: conda install numpy==1.15\n",
    "import numpy\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "id,symbol,seqnames,strand,TSS,CAGE_representative_TSS,type\n",
      "ENSG00000000003,TSPAN6,chrX,-,99894988,99891748,protein_coding\n",
      "ENSG00000000005,TNMD,chrX,+,99839799,99839933,protein_coding\n",
      "ENSG00000000419,DPM1,chr20,-,49575092,49575069,protein_coding\n",
      "ENSG00000000457,SCYL3,chr1,-,169863408,169863037,protein_coding\n",
      "ENSG00000000460,C1orf112,chr1,+,169631245,169764186,protein_coding\n",
      "ENSG00000000938,FGR,chr1,-,27961788,27961654,protein_coding\n",
      "ENSG00000000971,CFH,chr1,+,196621008,196621174,protein_coding\n",
      "ENSG00000001036,FUCA2,chr6,-,143832827,143832857,protein_coding\n",
      "ENSG00000001084,GCLC,chr6,-,53481768,53409899,protein_coding\n"
     ]
    }
   ],
   "source": [
    "!head /s/project/avsec/ExPecto/resources/geneanno.csv\n",
    "\n",
    "import pandas as pd\n",
    "testCSV=pd.read_csv(\"/s/project/avsec/ExPecto/resources/geneanno.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 143,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>symbol</th>\n",
       "      <th>seqnames</th>\n",
       "      <th>strand</th>\n",
       "      <th>TSS</th>\n",
       "      <th>CAGE_representative_TSS</th>\n",
       "      <th>type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000000003</td>\n",
       "      <td>TSPAN6</td>\n",
       "      <td>chrX</td>\n",
       "      <td>-</td>\n",
       "      <td>99894988</td>\n",
       "      <td>99891748</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000000005</td>\n",
       "      <td>TNMD</td>\n",
       "      <td>chrX</td>\n",
       "      <td>+</td>\n",
       "      <td>99839799</td>\n",
       "      <td>99839933</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000000419</td>\n",
       "      <td>DPM1</td>\n",
       "      <td>chr20</td>\n",
       "      <td>-</td>\n",
       "      <td>49575092</td>\n",
       "      <td>49575069</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000000457</td>\n",
       "      <td>SCYL3</td>\n",
       "      <td>chr1</td>\n",
       "      <td>-</td>\n",
       "      <td>169863408</td>\n",
       "      <td>169863037</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000000460</td>\n",
       "      <td>C1orf112</td>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>169631245</td>\n",
       "      <td>169764186</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ENSG00000000938</td>\n",
       "      <td>FGR</td>\n",
       "      <td>chr1</td>\n",
       "      <td>-</td>\n",
       "      <td>27961788</td>\n",
       "      <td>27961654</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ENSG00000000971</td>\n",
       "      <td>CFH</td>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>196621008</td>\n",
       "      <td>196621174</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ENSG00000001036</td>\n",
       "      <td>FUCA2</td>\n",
       "      <td>chr6</td>\n",
       "      <td>-</td>\n",
       "      <td>143832827</td>\n",
       "      <td>143832857</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ENSG00000001084</td>\n",
       "      <td>GCLC</td>\n",
       "      <td>chr6</td>\n",
       "      <td>-</td>\n",
       "      <td>53481768</td>\n",
       "      <td>53409899</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>ENSG00000001167</td>\n",
       "      <td>NFYA</td>\n",
       "      <td>chr6</td>\n",
       "      <td>+</td>\n",
       "      <td>41040684</td>\n",
       "      <td>41040714</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>ENSG00000001460</td>\n",
       "      <td>STPG1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>-</td>\n",
       "      <td>24743424</td>\n",
       "      <td>24740215</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>ENSG00000001461</td>\n",
       "      <td>NIPAL3</td>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>24742284</td>\n",
       "      <td>24742292</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>ENSG00000001497</td>\n",
       "      <td>LAS1L</td>\n",
       "      <td>chrX</td>\n",
       "      <td>-</td>\n",
       "      <td>64754655</td>\n",
       "      <td>64754667</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>ENSG00000001561</td>\n",
       "      <td>ENPP4</td>\n",
       "      <td>chr6</td>\n",
       "      <td>+</td>\n",
       "      <td>46097730</td>\n",
       "      <td>46097792</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>ENSG00000001617</td>\n",
       "      <td>SEMA3F</td>\n",
       "      <td>chr3</td>\n",
       "      <td>+</td>\n",
       "      <td>50192478</td>\n",
       "      <td>50192515</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>ENSG00000001626</td>\n",
       "      <td>CFTR</td>\n",
       "      <td>chr7</td>\n",
       "      <td>+</td>\n",
       "      <td>117105838</td>\n",
       "      <td>117120079</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>ENSG00000001629</td>\n",
       "      <td>ANKIB1</td>\n",
       "      <td>chr7</td>\n",
       "      <td>+</td>\n",
       "      <td>91875548</td>\n",
       "      <td>91875338</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>ENSG00000001630</td>\n",
       "      <td>CYP51A1</td>\n",
       "      <td>chr7</td>\n",
       "      <td>-</td>\n",
       "      <td>91772266</td>\n",
       "      <td>91763788</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>ENSG00000001631</td>\n",
       "      <td>KRIT1</td>\n",
       "      <td>chr7</td>\n",
       "      <td>-</td>\n",
       "      <td>91875480</td>\n",
       "      <td>91875228</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>ENSG00000002016</td>\n",
       "      <td>RAD52</td>\n",
       "      <td>chr12</td>\n",
       "      <td>-</td>\n",
       "      <td>1099219</td>\n",
       "      <td>1058808</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>ENSG00000002330</td>\n",
       "      <td>BAD</td>\n",
       "      <td>chr11</td>\n",
       "      <td>-</td>\n",
       "      <td>64052176</td>\n",
       "      <td>64052126</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>ENSG00000002549</td>\n",
       "      <td>LAP3</td>\n",
       "      <td>chr4</td>\n",
       "      <td>+</td>\n",
       "      <td>17578815</td>\n",
       "      <td>17579120</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>ENSG00000002586</td>\n",
       "      <td>CD99</td>\n",
       "      <td>chrX</td>\n",
       "      <td>+</td>\n",
       "      <td>2609220</td>\n",
       "      <td>2609367</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>ENSG00000002587</td>\n",
       "      <td>HS3ST1</td>\n",
       "      <td>chr4</td>\n",
       "      <td>-</td>\n",
       "      <td>11431389</td>\n",
       "      <td>11430518</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>ENSG00000002726</td>\n",
       "      <td>AOC1</td>\n",
       "      <td>chr7</td>\n",
       "      <td>+</td>\n",
       "      <td>150521715</td>\n",
       "      <td>150549606</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>ENSG00000002745</td>\n",
       "      <td>WNT16</td>\n",
       "      <td>chr7</td>\n",
       "      <td>+</td>\n",
       "      <td>120965421</td>\n",
       "      <td>120969054</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>ENSG00000002746</td>\n",
       "      <td>HECW1</td>\n",
       "      <td>chr7</td>\n",
       "      <td>+</td>\n",
       "      <td>43152198</td>\n",
       "      <td>43152246</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>ENSG00000002822</td>\n",
       "      <td>MAD1L1</td>\n",
       "      <td>chr7</td>\n",
       "      <td>-</td>\n",
       "      <td>2272878</td>\n",
       "      <td>2272547</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>ENSG00000002834</td>\n",
       "      <td>LASP1</td>\n",
       "      <td>chr17</td>\n",
       "      <td>+</td>\n",
       "      <td>37026112</td>\n",
       "      <td>37026311</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>ENSG00000002919</td>\n",
       "      <td>SNX11</td>\n",
       "      <td>chr17</td>\n",
       "      <td>+</td>\n",
       "      <td>46180719</td>\n",
       "      <td>46185139</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24309</th>\n",
       "      <td>ENSG00000259675</td>\n",
       "      <td>RP11-507B12.1</td>\n",
       "      <td>chr15</td>\n",
       "      <td>-</td>\n",
       "      <td>62007370</td>\n",
       "      <td>61945444</td>\n",
       "      <td>lincRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24310</th>\n",
       "      <td>ENSG00000259676</td>\n",
       "      <td>RP11-343B18.2</td>\n",
       "      <td>chr15</td>\n",
       "      <td>-</td>\n",
       "      <td>89341965</td>\n",
       "      <td>89341965</td>\n",
       "      <td>lincRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24311</th>\n",
       "      <td>ENSG00000259680</td>\n",
       "      <td>RP11-812E19.9</td>\n",
       "      <td>chr16</td>\n",
       "      <td>-</td>\n",
       "      <td>33647696</td>\n",
       "      <td>33647775</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24312</th>\n",
       "      <td>ENSG00000259681</td>\n",
       "      <td>RP11-96O20.2</td>\n",
       "      <td>chr15</td>\n",
       "      <td>+</td>\n",
       "      <td>45850796</td>\n",
       "      <td>45850796</td>\n",
       "      <td>lincRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24313</th>\n",
       "      <td>ENSG00000259685</td>\n",
       "      <td>CTD-2315E11.1</td>\n",
       "      <td>chr15</td>\n",
       "      <td>+</td>\n",
       "      <td>90645881</td>\n",
       "      <td>90645881</td>\n",
       "      <td>lincRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24314</th>\n",
       "      <td>ENSG00000259687</td>\n",
       "      <td>LINC01220</td>\n",
       "      <td>chr14</td>\n",
       "      <td>+</td>\n",
       "      <td>75761107</td>\n",
       "      <td>75761144</td>\n",
       "      <td>lincRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24315</th>\n",
       "      <td>ENSG00000259692</td>\n",
       "      <td>RP11-499F3.2</td>\n",
       "      <td>chr15</td>\n",
       "      <td>-</td>\n",
       "      <td>82163466</td>\n",
       "      <td>82163466</td>\n",
       "      <td>lincRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24316</th>\n",
       "      <td>ENSG00000259694</td>\n",
       "      <td>RP13-262C2.3</td>\n",
       "      <td>chr15</td>\n",
       "      <td>-</td>\n",
       "      <td>84841946</td>\n",
       "      <td>84841946</td>\n",
       "      <td>lincRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24317</th>\n",
       "      <td>ENSG00000259697</td>\n",
       "      <td>RP11-299H22.1</td>\n",
       "      <td>chr15</td>\n",
       "      <td>-</td>\n",
       "      <td>62495517</td>\n",
       "      <td>62495517</td>\n",
       "      <td>lincRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24318</th>\n",
       "      <td>ENSG00000259702</td>\n",
       "      <td>RP11-236L14.1</td>\n",
       "      <td>chr15</td>\n",
       "      <td>-</td>\n",
       "      <td>96607655</td>\n",
       "      <td>96607586</td>\n",
       "      <td>lincRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24319</th>\n",
       "      <td>ENSG00000259703</td>\n",
       "      <td>LINC00593</td>\n",
       "      <td>chr15</td>\n",
       "      <td>+</td>\n",
       "      <td>70127573</td>\n",
       "      <td>70127573</td>\n",
       "      <td>lincRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24320</th>\n",
       "      <td>ENSG00000259705</td>\n",
       "      <td>RP11-227D13.1</td>\n",
       "      <td>chr15</td>\n",
       "      <td>+</td>\n",
       "      <td>48938148</td>\n",
       "      <td>48938148</td>\n",
       "      <td>lincRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24321</th>\n",
       "      <td>ENSG00000259711</td>\n",
       "      <td>CTD-3032H12.2</td>\n",
       "      <td>chr16</td>\n",
       "      <td>+</td>\n",
       "      <td>54968825</td>\n",
       "      <td>54968825</td>\n",
       "      <td>lincRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24322</th>\n",
       "      <td>ENSG00000259715</td>\n",
       "      <td>CTD-3110H11.1</td>\n",
       "      <td>chr15</td>\n",
       "      <td>-</td>\n",
       "      <td>50652391</td>\n",
       "      <td>50652391</td>\n",
       "      <td>lincRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24323</th>\n",
       "      <td>ENSG00000259717</td>\n",
       "      <td>LINC00677</td>\n",
       "      <td>chr14</td>\n",
       "      <td>-</td>\n",
       "      <td>103589344</td>\n",
       "      <td>103589392</td>\n",
       "      <td>lincRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24324</th>\n",
       "      <td>ENSG00000259719</td>\n",
       "      <td>RP11-930O11.1</td>\n",
       "      <td>chr14</td>\n",
       "      <td>+</td>\n",
       "      <td>56777598</td>\n",
       "      <td>56777598</td>\n",
       "      <td>lincRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24325</th>\n",
       "      <td>ENSG00000259724</td>\n",
       "      <td>LINC01581</td>\n",
       "      <td>chr15</td>\n",
       "      <td>-</td>\n",
       "      <td>94651167</td>\n",
       "      <td>94651167</td>\n",
       "      <td>lincRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24326</th>\n",
       "      <td>ENSG00000259725</td>\n",
       "      <td>CTD-3032H12.1</td>\n",
       "      <td>chr16</td>\n",
       "      <td>-</td>\n",
       "      <td>54972583</td>\n",
       "      <td>54972583</td>\n",
       "      <td>lincRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24327</th>\n",
       "      <td>ENSG00000259727</td>\n",
       "      <td>RP11-1069G10.2</td>\n",
       "      <td>chr15</td>\n",
       "      <td>+</td>\n",
       "      <td>63188011</td>\n",
       "      <td>63187993</td>\n",
       "      <td>lincRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24328</th>\n",
       "      <td>ENSG00000259730</td>\n",
       "      <td>CTD-2007H18.1</td>\n",
       "      <td>chr15</td>\n",
       "      <td>+</td>\n",
       "      <td>48236241</td>\n",
       "      <td>48236241</td>\n",
       "      <td>lincRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24329</th>\n",
       "      <td>ENSG00000259731</td>\n",
       "      <td>RP11-326N17.1</td>\n",
       "      <td>chr15</td>\n",
       "      <td>-</td>\n",
       "      <td>39051614</td>\n",
       "      <td>39051614</td>\n",
       "      <td>lincRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24330</th>\n",
       "      <td>ENSG00000259737</td>\n",
       "      <td>RP11-475A13.1</td>\n",
       "      <td>chr15</td>\n",
       "      <td>-</td>\n",
       "      <td>36763217</td>\n",
       "      <td>36763217</td>\n",
       "      <td>lincRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24331</th>\n",
       "      <td>ENSG00000259747</td>\n",
       "      <td>RP11-275I4.2</td>\n",
       "      <td>chr15</td>\n",
       "      <td>+</td>\n",
       "      <td>38964048</td>\n",
       "      <td>38964096</td>\n",
       "      <td>lincRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24332</th>\n",
       "      <td>ENSG00000259754</td>\n",
       "      <td>RP11-208K4.1</td>\n",
       "      <td>chr15</td>\n",
       "      <td>+</td>\n",
       "      <td>48176507</td>\n",
       "      <td>48176507</td>\n",
       "      <td>lincRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24333</th>\n",
       "      <td>ENSG00000259757</td>\n",
       "      <td>RP11-129I19.2</td>\n",
       "      <td>chr5</td>\n",
       "      <td>-</td>\n",
       "      <td>2737873</td>\n",
       "      <td>2737873</td>\n",
       "      <td>lincRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24334</th>\n",
       "      <td>ENSG00000259759</td>\n",
       "      <td>RP11-324D17.2</td>\n",
       "      <td>chr16</td>\n",
       "      <td>+</td>\n",
       "      <td>54273605</td>\n",
       "      <td>54273605</td>\n",
       "      <td>lincRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24335</th>\n",
       "      <td>ENSG00000259761</td>\n",
       "      <td>RP11-138H10.2</td>\n",
       "      <td>chr15</td>\n",
       "      <td>+</td>\n",
       "      <td>87585285</td>\n",
       "      <td>87585285</td>\n",
       "      <td>lincRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24336</th>\n",
       "      <td>ENSG00000259763</td>\n",
       "      <td>RP11-327J17.1</td>\n",
       "      <td>chr15</td>\n",
       "      <td>+</td>\n",
       "      <td>96779014</td>\n",
       "      <td>96779014</td>\n",
       "      <td>lincRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24337</th>\n",
       "      <td>ENSG00000259772</td>\n",
       "      <td>RP11-16E12.2</td>\n",
       "      <td>chr15</td>\n",
       "      <td>-</td>\n",
       "      <td>31523041</td>\n",
       "      <td>31521585</td>\n",
       "      <td>lincRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24338</th>\n",
       "      <td>ENSG00000259774</td>\n",
       "      <td>RP11-182J1.13</td>\n",
       "      <td>chr15</td>\n",
       "      <td>+</td>\n",
       "      <td>84977316</td>\n",
       "      <td>84977316</td>\n",
       "      <td>lincRNA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>24339 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    id          symbol seqnames strand        TSS  \\\n",
       "0      ENSG00000000003          TSPAN6     chrX      -   99894988   \n",
       "1      ENSG00000000005            TNMD     chrX      +   99839799   \n",
       "2      ENSG00000000419            DPM1    chr20      -   49575092   \n",
       "3      ENSG00000000457           SCYL3     chr1      -  169863408   \n",
       "4      ENSG00000000460        C1orf112     chr1      +  169631245   \n",
       "5      ENSG00000000938             FGR     chr1      -   27961788   \n",
       "6      ENSG00000000971             CFH     chr1      +  196621008   \n",
       "7      ENSG00000001036           FUCA2     chr6      -  143832827   \n",
       "8      ENSG00000001084            GCLC     chr6      -   53481768   \n",
       "9      ENSG00000001167            NFYA     chr6      +   41040684   \n",
       "10     ENSG00000001460           STPG1     chr1      -   24743424   \n",
       "11     ENSG00000001461          NIPAL3     chr1      +   24742284   \n",
       "12     ENSG00000001497           LAS1L     chrX      -   64754655   \n",
       "13     ENSG00000001561           ENPP4     chr6      +   46097730   \n",
       "14     ENSG00000001617          SEMA3F     chr3      +   50192478   \n",
       "15     ENSG00000001626            CFTR     chr7      +  117105838   \n",
       "16     ENSG00000001629          ANKIB1     chr7      +   91875548   \n",
       "17     ENSG00000001630         CYP51A1     chr7      -   91772266   \n",
       "18     ENSG00000001631           KRIT1     chr7      -   91875480   \n",
       "19     ENSG00000002016           RAD52    chr12      -    1099219   \n",
       "20     ENSG00000002330             BAD    chr11      -   64052176   \n",
       "21     ENSG00000002549            LAP3     chr4      +   17578815   \n",
       "22     ENSG00000002586            CD99     chrX      +    2609220   \n",
       "23     ENSG00000002587          HS3ST1     chr4      -   11431389   \n",
       "24     ENSG00000002726            AOC1     chr7      +  150521715   \n",
       "25     ENSG00000002745           WNT16     chr7      +  120965421   \n",
       "26     ENSG00000002746           HECW1     chr7      +   43152198   \n",
       "27     ENSG00000002822          MAD1L1     chr7      -    2272878   \n",
       "28     ENSG00000002834           LASP1    chr17      +   37026112   \n",
       "29     ENSG00000002919           SNX11    chr17      +   46180719   \n",
       "...                ...             ...      ...    ...        ...   \n",
       "24309  ENSG00000259675   RP11-507B12.1    chr15      -   62007370   \n",
       "24310  ENSG00000259676   RP11-343B18.2    chr15      -   89341965   \n",
       "24311  ENSG00000259680   RP11-812E19.9    chr16      -   33647696   \n",
       "24312  ENSG00000259681    RP11-96O20.2    chr15      +   45850796   \n",
       "24313  ENSG00000259685   CTD-2315E11.1    chr15      +   90645881   \n",
       "24314  ENSG00000259687       LINC01220    chr14      +   75761107   \n",
       "24315  ENSG00000259692    RP11-499F3.2    chr15      -   82163466   \n",
       "24316  ENSG00000259694    RP13-262C2.3    chr15      -   84841946   \n",
       "24317  ENSG00000259697   RP11-299H22.1    chr15      -   62495517   \n",
       "24318  ENSG00000259702   RP11-236L14.1    chr15      -   96607655   \n",
       "24319  ENSG00000259703       LINC00593    chr15      +   70127573   \n",
       "24320  ENSG00000259705   RP11-227D13.1    chr15      +   48938148   \n",
       "24321  ENSG00000259711   CTD-3032H12.2    chr16      +   54968825   \n",
       "24322  ENSG00000259715   CTD-3110H11.1    chr15      -   50652391   \n",
       "24323  ENSG00000259717       LINC00677    chr14      -  103589344   \n",
       "24324  ENSG00000259719   RP11-930O11.1    chr14      +   56777598   \n",
       "24325  ENSG00000259724       LINC01581    chr15      -   94651167   \n",
       "24326  ENSG00000259725   CTD-3032H12.1    chr16      -   54972583   \n",
       "24327  ENSG00000259727  RP11-1069G10.2    chr15      +   63188011   \n",
       "24328  ENSG00000259730   CTD-2007H18.1    chr15      +   48236241   \n",
       "24329  ENSG00000259731   RP11-326N17.1    chr15      -   39051614   \n",
       "24330  ENSG00000259737   RP11-475A13.1    chr15      -   36763217   \n",
       "24331  ENSG00000259747    RP11-275I4.2    chr15      +   38964048   \n",
       "24332  ENSG00000259754    RP11-208K4.1    chr15      +   48176507   \n",
       "24333  ENSG00000259757   RP11-129I19.2     chr5      -    2737873   \n",
       "24334  ENSG00000259759   RP11-324D17.2    chr16      +   54273605   \n",
       "24335  ENSG00000259761   RP11-138H10.2    chr15      +   87585285   \n",
       "24336  ENSG00000259763   RP11-327J17.1    chr15      +   96779014   \n",
       "24337  ENSG00000259772    RP11-16E12.2    chr15      -   31523041   \n",
       "24338  ENSG00000259774   RP11-182J1.13    chr15      +   84977316   \n",
       "\n",
       "       CAGE_representative_TSS            type  \n",
       "0                     99891748  protein_coding  \n",
       "1                     99839933  protein_coding  \n",
       "2                     49575069  protein_coding  \n",
       "3                    169863037  protein_coding  \n",
       "4                    169764186  protein_coding  \n",
       "5                     27961654  protein_coding  \n",
       "6                    196621174  protein_coding  \n",
       "7                    143832857  protein_coding  \n",
       "8                     53409899  protein_coding  \n",
       "9                     41040714  protein_coding  \n",
       "10                    24740215  protein_coding  \n",
       "11                    24742292  protein_coding  \n",
       "12                    64754667  protein_coding  \n",
       "13                    46097792  protein_coding  \n",
       "14                    50192515  protein_coding  \n",
       "15                   117120079  protein_coding  \n",
       "16                    91875338  protein_coding  \n",
       "17                    91763788  protein_coding  \n",
       "18                    91875228  protein_coding  \n",
       "19                     1058808  protein_coding  \n",
       "20                    64052126  protein_coding  \n",
       "21                    17579120  protein_coding  \n",
       "22                     2609367  protein_coding  \n",
       "23                    11430518  protein_coding  \n",
       "24                   150549606  protein_coding  \n",
       "25                   120969054  protein_coding  \n",
       "26                    43152246  protein_coding  \n",
       "27                     2272547  protein_coding  \n",
       "28                    37026311  protein_coding  \n",
       "29                    46185139  protein_coding  \n",
       "...                        ...             ...  \n",
       "24309                 61945444         lincRNA  \n",
       "24310                 89341965         lincRNA  \n",
       "24311                 33647775  protein_coding  \n",
       "24312                 45850796         lincRNA  \n",
       "24313                 90645881         lincRNA  \n",
       "24314                 75761144         lincRNA  \n",
       "24315                 82163466         lincRNA  \n",
       "24316                 84841946         lincRNA  \n",
       "24317                 62495517         lincRNA  \n",
       "24318                 96607586         lincRNA  \n",
       "24319                 70127573         lincRNA  \n",
       "24320                 48938148         lincRNA  \n",
       "24321                 54968825         lincRNA  \n",
       "24322                 50652391         lincRNA  \n",
       "24323                103589392         lincRNA  \n",
       "24324                 56777598         lincRNA  \n",
       "24325                 94651167         lincRNA  \n",
       "24326                 54972583         lincRNA  \n",
       "24327                 63187993         lincRNA  \n",
       "24328                 48236241         lincRNA  \n",
       "24329                 39051614         lincRNA  \n",
       "24330                 36763217         lincRNA  \n",
       "24331                 38964096         lincRNA  \n",
       "24332                 48176507         lincRNA  \n",
       "24333                  2737873         lincRNA  \n",
       "24334                 54273605         lincRNA  \n",
       "24335                 87585285         lincRNA  \n",
       "24336                 96779014         lincRNA  \n",
       "24337                 31521585         lincRNA  \n",
       "24338                 84977316         lincRNA  \n",
       "\n",
       "[24339 rows x 7 columns]"
      ]
     },
     "execution_count": 143,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "testCSV"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from torch.utils.data import Dataset, DataLoader\n",
    "class vcfFastaExtractor(Data):\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/modules/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/importlib/_bootstrap.py:219: RuntimeWarning: numpy.dtype size changed, may indicate binary incompatibility. Expected 96, got 88\n",
      "  return f(*args, **kwds)\n",
      "/opt/modules/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/importlib/_bootstrap.py:219: RuntimeWarning: numpy.dtype size changed, may indicate binary incompatibility. Expected 96, got 88\n",
      "  return f(*args, **kwds)\n"
     ]
    }
   ],
   "source": [
    "from kipoiseq.dataloaders import SeqIntervalDl\n",
    "\n",
    "class VariantTransformer(object):\n",
    "    \"\"\"Includes the required variants into an reference-genome DNA-sequence\"\"\"\n",
    "    def __init__(self,vcfFile): \n",
    "    def __call__(seq,genomePosi,sample_idx):\n",
    "\n",
    "class VariantDataloader(SeqIntervalDl):\n",
    "    \n",
    "    \n",
    "#adapted from super class (kipoiseq)\n",
    "    def __init__(self,\n",
    "                    intervals_file,\n",
    "                    fasta_file,\n",
    "                    vcfFile):\n",
    "        super().__init__(intervals_file,fasta_file)\n",
    "        self.var_transform=VariantTransformer(vcfFile)\n",
    "        \n",
    "    def __len__(self):\n",
    "        return len(self.seq_dl)*len(self.var_transform)\n",
    "\n",
    "    def __getitem__(self, idx):\n",
    "        idInterval=idx/len(self.var_transform)\n",
    "        idSample=idx%len(self.var_transform)\n",
    "        ret = self.seq_dl[idInterval]\n",
    "        \n",
    "        ret['inputs'] = self.var_transform(str(ret[\"inputs\"]),ret[\"metadata\"],idSample)\n",
    "        ret['inputs'] = self.input_transform(str(ret[\"inputs\"]))\n",
    "        return ret"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[4,5,6)\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[-1,0,1)"
      ]
     },
     "execution_count": 67,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "class Interval(object):\n",
    "    \"\"\"\n",
    "    stores window-range-info\n",
    "    as usual in python, an [ ) interval\"\"\"\n",
    "    def __init__(self, n_up,n_down,anchor=None):\n",
    "        if n_up==n_down:\n",
    "            raise AttributeError(\"An interval without elements is not supported. Your upper and lower bound(exclusive) are the same.\")\n",
    "            \n",
    "        \n",
    "        self.lower=n_up\n",
    "        self.upper=n_down\n",
    "        self.anchor=anchor\n",
    "        if anchor is None:\n",
    "            self.anchor=int((self.lower+self.upper)/2)\n",
    "    def setAnchor(self,newAnchor):\n",
    "        self.lower=newAnchor-(self.anchor-self.lower)\n",
    "        self.upper=newAnchor+(self.upper-self.anchor)\n",
    "        self.anchor=newAnchor\n",
    "    def __str__(self):\n",
    "        return \"[{},{},{})\".format(self.lower,self.anchor,self.upper)\n",
    "    def __repr__(self):\n",
    "        return str(self)\n",
    "    def __len__(self):\n",
    "        return self.upper-self.lower\n",
    "    def create_genomic_location(self, tssPosi, chrom, strand):\n",
    "        \n",
    "test=Interval(4,6)\n",
    "print(test)\n",
    "test.setAnchor(0)\n",
    "test"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['AA', 'CC']"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "class VcfFastaDummy(object):\n",
    "    def __init__(self,a,b):\n",
    "        pass\n",
    "    def extract(self,interval,sample): \n",
    "        allele1=\"A\"*len(interval)\n",
    "        allele2=\"C\"*len(interval)\n",
    "        return [allele1,allele2]\n",
    "    \n",
    "test=VcfFastaDummy(\"a\",\"b\")\n",
    "test.extract(Interval(3,5),\"testSample\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Solving environment: done\n",
      "\n",
      "\n",
      "==> WARNING: A newer version of conda exists. <==\n",
      "  current version: 4.5.12\n",
      "  latest version: 4.6.7\n",
      "\n",
      "Please update conda by running\n",
      "\n",
      "    $ conda update -n base -c defaults conda\n",
      "\n",
      "\n",
      "\n",
      "## Package Plan ##\n",
      "\n",
      "  environment location: /opt/modules/i12g/anaconda/3-5.0.1/envs/rep\n",
      "\n",
      "  added / updated specs: \n",
      "    - pytorch\n",
      "\n",
      "\n",
      "The following packages will be downloaded:\n",
      "\n",
      "    package                    |            build\n",
      "    ---------------------------|-----------------\n",
      "    cudnn-7.3.1                |        cuda9.0_0       331.3 MB\n",
      "\n",
      "The following NEW packages will be INSTALLED:\n",
      "\n",
      "    cffi:        1.11.5-py36he75722e_1\n",
      "    cudatoolkit: 9.0-h13b8566_0       \n",
      "    cudnn:       7.3.1-cuda9.0_0      \n",
      "    nccl:        1.3.5-cuda9.0_0      \n",
      "    ninja:       1.8.2-py36h6bb024c_1 \n",
      "    pycparser:   2.19-py36_0          \n",
      "    pytorch:     0.4.1-py36ha74772b_0 \n",
      "\n",
      "\n",
      "Downloading and Extracting Packages\n",
      "cudnn-7.3.1          | 331.3 MB  | ##################################### | 100% \n",
      "Preparing transaction: done\n",
      "Verifying transaction: done\n",
      "Executing transaction: done\n"
     ]
    }
   ],
   "source": [
    "! conda install pytorch -y"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[1., 0., 0., 0.],\n",
       "       [0., 1., 0., 0.],\n",
       "       [0., 1., 0., 0.],\n",
       "       [0., 1., 0., 0.]])"
      ]
     },
     "execution_count": 61,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "        from kipoiseq.transforms import ReorderedOneHot\n",
    "        testTransform=ReorderedOneHot()\n",
    "        testTransform(\"ACCC\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Working with new kipoiseq.extractor modules, install and so on.."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 144,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "gzip: stdout: Broken pipe\n"
     ]
    }
   ],
   "source": [
    "vcfFile=\"/s/project/gtex-processed/vcf/GTEx_Analysis_20150112_WholeGenomeSeq_148Indiv_GATK_HaplotypeCaller.vcf.gz\"\n",
    "vcfTest=\"vcfTest\"\n",
    "! zcat /s/project/gtex-processed/vcf/GTEx_Analysis_20150112_WholeGenomeSeq_148Indiv_GATK_HaplotypeCaller.vcf.gz | head -n 300 >vcfTest"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 160,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting kipoiseq\n",
      "  Downloading https://files.pythonhosted.org/packages/e6/6c/fedbbcb4e5525c87cc1973c025c64e44e8fe3a540a588d648936be0b7bae/kipoiseq-0.2.3.tar.gz\n",
      "Requirement already satisfied, skipping upgrade: kipoi>=0.5.5 in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (from kipoiseq) (0.6.5)\n",
      "Requirement already satisfied, skipping upgrade: pybedtools in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (from kipoiseq) (0.8.0)\n",
      "Requirement already satisfied, skipping upgrade: pyfaidx in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (from kipoiseq) (0.5.5.2)\n",
      "Requirement already satisfied, skipping upgrade: numpy in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (from kipoiseq) (1.16.0)\n",
      "Requirement already satisfied, skipping upgrade: pandas in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (from kipoiseq) (0.23.4)\n",
      "Requirement already satisfied, skipping upgrade: tqdm in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (from kipoiseq) (4.28.1)\n",
      "Requirement already satisfied, skipping upgrade: gffutils in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (from kipoiseq) (0.9)\n",
      "Requirement already satisfied, skipping upgrade: kipoi-utils>=0.1.1 in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (from kipoiseq) (0.1.1)\n",
      "Requirement already satisfied, skipping upgrade: kipoi-conda>=0.1.0 in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (from kipoiseq) (0.1.0)\n",
      "Requirement already satisfied, skipping upgrade: enum34 in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (from kipoi>=0.5.5->kipoiseq) (1.1.6)\n",
      "Requirement already satisfied, skipping upgrade: colorlog in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (from kipoi>=0.5.5->kipoiseq) (4.0.2)\n",
      "Requirement already satisfied, skipping upgrade: tinydb in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (from kipoi>=0.5.5->kipoiseq) (3.12.2)\n",
      "Requirement already satisfied, skipping upgrade: pyyaml in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (from kipoi>=0.5.5->kipoiseq) (3.13)\n",
      "Requirement already satisfied, skipping upgrade: future in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (from kipoi>=0.5.5->kipoiseq) (0.17.1)\n",
      "Requirement already satisfied, skipping upgrade: h5py in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (from kipoi>=0.5.5->kipoiseq) (2.8.0)\n",
      "Requirement already satisfied, skipping upgrade: related>=0.6.0 in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (from kipoi>=0.5.5->kipoiseq) (0.7.2)\n",
      "Requirement already satisfied, skipping upgrade: urllib3<1.23,>=1.21.1 in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (from kipoi>=0.5.5->kipoiseq) (1.22)\n",
      "Requirement already satisfied, skipping upgrade: cookiecutter in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (from kipoi>=0.5.5->kipoiseq) (1.6.0)\n",
      "Requirement already satisfied, skipping upgrade: jinja2 in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (from kipoi>=0.5.5->kipoiseq) (2.10)\n",
      "Requirement already satisfied, skipping upgrade: pysam in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (from pybedtools->kipoiseq) (0.15.1)\n",
      "Requirement already satisfied, skipping upgrade: six in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (from pybedtools->kipoiseq) (1.12.0)\n",
      "Requirement already satisfied, skipping upgrade: setuptools>=0.7 in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (from pyfaidx->kipoiseq) (40.6.3)\n",
      "Requirement already satisfied, skipping upgrade: python-dateutil>=2.5.0 in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (from pandas->kipoiseq) (2.7.5)\n",
      "Requirement already satisfied, skipping upgrade: pytz>=2011k in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (from pandas->kipoiseq) (2018.7)\n",
      "Requirement already satisfied, skipping upgrade: argcomplete in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (from gffutils->kipoiseq) (1.9.4)\n",
      "Requirement already satisfied, skipping upgrade: argh in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (from gffutils->kipoiseq) (0.26.2)\n",
      "Requirement already satisfied, skipping upgrade: simplejson in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (from gffutils->kipoiseq) (3.16.0)\n",
      "Requirement already satisfied, skipping upgrade: attrs in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (from kipoi-utils>=0.1.1->kipoiseq) (18.2.0)\n",
      "Requirement already satisfied, skipping upgrade: click>=5.0 in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (from cookiecutter->kipoi>=0.5.5->kipoiseq) (7.0)\n",
      "Requirement already satisfied, skipping upgrade: poyo>=0.1.0 in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (from cookiecutter->kipoi>=0.5.5->kipoiseq) (0.4.2)\n",
      "Requirement already satisfied, skipping upgrade: requests>=2.18.0 in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (from cookiecutter->kipoi>=0.5.5->kipoiseq) (2.21.0)\n",
      "Requirement already satisfied, skipping upgrade: whichcraft>=0.4.0 in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (from cookiecutter->kipoi>=0.5.5->kipoiseq) (0.5.2)\n",
      "Requirement already satisfied, skipping upgrade: binaryornot>=0.2.0 in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (from cookiecutter->kipoi>=0.5.5->kipoiseq) (0.4.4)\n",
      "Requirement already satisfied, skipping upgrade: jinja2-time>=0.1.0 in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (from cookiecutter->kipoi>=0.5.5->kipoiseq) (0.2.0)\n",
      "Requirement already satisfied, skipping upgrade: MarkupSafe>=0.23 in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (from jinja2->kipoi>=0.5.5->kipoiseq) (1.1.0)\n",
      "Requirement already satisfied, skipping upgrade: chardet<3.1.0,>=3.0.2 in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (from requests>=2.18.0->cookiecutter->kipoi>=0.5.5->kipoiseq) (3.0.4)\n",
      "Requirement already satisfied, skipping upgrade: certifi>=2017.4.17 in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (from requests>=2.18.0->cookiecutter->kipoi>=0.5.5->kipoiseq) (2018.11.29)\n",
      "Requirement already satisfied, skipping upgrade: idna<2.9,>=2.5 in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (from requests>=2.18.0->cookiecutter->kipoi>=0.5.5->kipoiseq) (2.8)\n",
      "Requirement already satisfied, skipping upgrade: arrow in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (from jinja2-time>=0.1.0->cookiecutter->kipoi>=0.5.5->kipoiseq) (0.12.1)\n",
      "Building wheels for collected packages: kipoiseq\n",
      "  Running setup.py bdist_wheel for kipoiseq ... \u001b[?25ldone\n",
      "\u001b[?25h  Stored in directory: /data/ouga/home/ag_gagneur/reinharj/.cache/pip/wheels/dd/59/ba/5133d66b274ad58a3e0971752a0b2451543a1982710ed2a5c8\n",
      "Successfully built kipoiseq\n",
      "Installing collected packages: kipoiseq\n",
      "  Found existing installation: kipoiseq 0.2.2\n",
      "    Uninstalling kipoiseq-0.2.2:\n",
      "      Successfully uninstalled kipoiseq-0.2.2\n",
      "Successfully installed kipoiseq-0.2.3\n"
     ]
    }
   ],
   "source": [
    "!pip install -U kipoiseq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 186,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<module 'kipoiseq.extractors' from '/opt/modules/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages/kipoiseq/extractors/__init__.py'>"
      ]
     },
     "execution_count": 186,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from importlib import reload\n",
    "reload(kipoiseq)\n",
    "reload(kipoiseq.extractors)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting cyvcf2\n",
      "  Using cached https://files.pythonhosted.org/packages/63/21/2f7546d7121c001f73b73bc7941c511a4848771afdb4de300f50301ce458/cyvcf2-0.10.8.tar.gz\n",
      "Requirement already satisfied, skipping upgrade: numpy in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/lib/python3.6/site-packages (from cyvcf2) (1.15.0)\n",
      "Building wheels for collected packages: cyvcf2\n",
      "  Building wheel for cyvcf2 (setup.py) ... \u001b[?25lerror\n",
      "  Complete output from command /opt/modules/i12g/anaconda/3-5.0.1/bin/python -u -c \"import setuptools, tokenize;__file__='/scratch/tmp/reinharj/pip-install-z3bdr5su/cyvcf2/setup.py';f=getattr(tokenize, 'open', open)(__file__);code=f.read().replace('\\r\\n', '\\n');f.close();exec(compile(code, __file__, 'exec'))\" bdist_wheel -d /scratch/tmp/reinharj/pip-wheel-cqoylj8z --python-tag cp36:\n",
      "  running bdist_wheel\n",
      "  running build\n",
      "  running build_py\n",
      "  creating build\n",
      "  creating build/lib.linux-x86_64-3.6\n",
      "  creating build/lib.linux-x86_64-3.6/cyvcf2\n",
      "  copying cyvcf2/__main__.py -> build/lib.linux-x86_64-3.6/cyvcf2\n",
      "  copying cyvcf2/cli.py -> build/lib.linux-x86_64-3.6/cyvcf2\n",
      "  copying cyvcf2/__init__.py -> build/lib.linux-x86_64-3.6/cyvcf2\n",
      "  creating build/lib.linux-x86_64-3.6/cyvcf2/tests\n",
      "  copying cyvcf2/tests/test_cli.py -> build/lib.linux-x86_64-3.6/cyvcf2/tests\n",
      "  copying cyvcf2/tests/__init__.py -> build/lib.linux-x86_64-3.6/cyvcf2/tests\n",
      "  copying cyvcf2/tests/test_hemi.py -> build/lib.linux-x86_64-3.6/cyvcf2/tests\n",
      "  copying cyvcf2/tests/test_reader.py -> build/lib.linux-x86_64-3.6/cyvcf2/tests\n",
      "  running egg_info\n",
      "  writing cyvcf2.egg-info/PKG-INFO\n",
      "  writing dependency_links to cyvcf2.egg-info/dependency_links.txt\n",
      "  writing entry points to cyvcf2.egg-info/entry_points.txt\n",
      "  writing requirements to cyvcf2.egg-info/requires.txt\n",
      "  writing top-level names to cyvcf2.egg-info/top_level.txt\n",
      "  reading manifest file 'cyvcf2.egg-info/SOURCES.txt'\n",
      "  reading manifest template 'MANIFEST.in'\n",
      "  writing manifest file 'cyvcf2.egg-info/SOURCES.txt'\n",
      "  copying cyvcf2/cyvcf2.c -> build/lib.linux-x86_64-3.6/cyvcf2\n",
      "  copying cyvcf2/cyvcf2.pxd -> build/lib.linux-x86_64-3.6/cyvcf2\n",
      "  copying cyvcf2/cyvcf2.pyx -> build/lib.linux-x86_64-3.6/cyvcf2\n",
      "  copying cyvcf2/helpers.c -> build/lib.linux-x86_64-3.6/cyvcf2\n",
      "  copying cyvcf2/helpers.h -> build/lib.linux-x86_64-3.6/cyvcf2\n",
      "  copying cyvcf2/relatedness.h -> build/lib.linux-x86_64-3.6/cyvcf2\n",
      "  copying cyvcf2/tests/__init__.pyc -> build/lib.linux-x86_64-3.6/cyvcf2/tests\n",
      "  copying cyvcf2/tests/bug.vcf.gz -> build/lib.linux-x86_64-3.6/cyvcf2/tests\n",
      "  copying cyvcf2/tests/decomposed.vcf -> build/lib.linux-x86_64-3.6/cyvcf2/tests\n",
      "  copying cyvcf2/tests/empty.vcf -> build/lib.linux-x86_64-3.6/cyvcf2/tests\n",
      "  copying cyvcf2/tests/issue_44.vcf -> build/lib.linux-x86_64-3.6/cyvcf2/tests\n",
      "  copying cyvcf2/tests/o.vcf.gz -> build/lib.linux-x86_64-3.6/cyvcf2/tests\n",
      "  copying cyvcf2/tests/seg.vcf.gz -> build/lib.linux-x86_64-3.6/cyvcf2/tests\n",
      "  copying cyvcf2/tests/test-alt-repr.vcf -> build/lib.linux-x86_64-3.6/cyvcf2/tests\n",
      "  copying cyvcf2/tests/test-format-string.vcf -> build/lib.linux-x86_64-3.6/cyvcf2/tests\n",
      "  copying cyvcf2/tests/test-haploidX.vcf -> build/lib.linux-x86_64-3.6/cyvcf2/tests\n",
      "  copying cyvcf2/tests/test-hemi.vcf -> build/lib.linux-x86_64-3.6/cyvcf2/tests\n",
      "  copying cyvcf2/tests/test-multiallelic-homozygous-alt.vcf.gz -> build/lib.linux-x86_64-3.6/cyvcf2/tests\n",
      "  copying cyvcf2/tests/test-multiallelic-homozygous-alt.vcf.gz.tbi -> build/lib.linux-x86_64-3.6/cyvcf2/tests\n",
      "  copying cyvcf2/tests/test-strict-gt-option-flag.vcf.gz -> build/lib.linux-x86_64-3.6/cyvcf2/tests\n",
      "  copying cyvcf2/tests/test-strict-gt-option-flag.vcf.gz.tbi -> build/lib.linux-x86_64-3.6/cyvcf2/tests\n",
      "  copying cyvcf2/tests/test.comp_het.3.vcf -> build/lib.linux-x86_64-3.6/cyvcf2/tests\n",
      "  copying cyvcf2/tests/test.snpeff.bcf -> build/lib.linux-x86_64-3.6/cyvcf2/tests\n",
      "  copying cyvcf2/tests/test.snpeff.bcf.csi -> build/lib.linux-x86_64-3.6/cyvcf2/tests\n",
      "  copying cyvcf2/tests/test.snpeff.vcf -> build/lib.linux-x86_64-3.6/cyvcf2/tests\n",
      "  copying cyvcf2/tests/test.vcf.gz -> build/lib.linux-x86_64-3.6/cyvcf2/tests\n",
      "  copying cyvcf2/tests/test.vcf.gz.tbi -> build/lib.linux-x86_64-3.6/cyvcf2/tests\n",
      "  copying cyvcf2/tests/test_cli.pyc -> build/lib.linux-x86_64-3.6/cyvcf2/tests\n",
      "  copying cyvcf2/tests/test_gt_alt_freqs.vcf -> build/lib.linux-x86_64-3.6/cyvcf2/tests\n",
      "  copying cyvcf2/tests/test_gt_bases.vcf.gz -> build/lib.linux-x86_64-3.6/cyvcf2/tests\n",
      "  copying cyvcf2/tests/test_hemi.pyc -> build/lib.linux-x86_64-3.6/cyvcf2/tests\n",
      "  copying cyvcf2/tests/test_reader.pyc -> build/lib.linux-x86_64-3.6/cyvcf2/tests\n",
      "  running build_ext\n",
      "  cythoning cyvcf2/cyvcf2.pyx to cyvcf2/cyvcf2.c\n",
      "  warning: cyvcf2/cyvcf2.pxd:26:16: Declarations should not be declared inline.\n",
      "  warning: cyvcf2/cyvcf2.pxd:279:30: Declarations should not be declared inline.\n",
      "  building 'cyvcf2.cyvcf2' extension\n",
      "  creating build/temp.linux-x86_64-3.6\n",
      "  creating build/temp.linux-x86_64-3.6/cyvcf2\n",
      "  creating build/temp.linux-x86_64-3.6/htslib\n",
      "  creating build/temp.linux-x86_64-3.6/htslib/cram\n",
      "  /opt/modules/i12g/anaconda/3-5.0.1/bin/x86_64-conda_cos6-linux-gnu-cc -DNDEBUG -fwrapv -O2 -Wall -Wstrict-prototypes -I/opt/modules/i12g/samtools/src/1.3.1/samtools-1.3.1 -DNDEBUG -D_FORTIFY_SOURCE=2 -O2 -fPIC -Ihtslib -Icyvcf2 -I/opt/modules/i12g/anaconda/3-5.0.1/lib/python3.6/site-packages/numpy/core/include -I/opt/modules/i12g/anaconda/3-5.0.1/include/python3.6m -c cyvcf2/cyvcf2.c -o build/temp.linux-x86_64-3.6/cyvcf2/cyvcf2.o\n",
      "  In file included from /opt/modules/i12g/anaconda/3-5.0.1/lib/python3.6/site-packages/numpy/core/include/numpy/ndarraytypes.h:1821:0,\n",
      "                   from /opt/modules/i12g/anaconda/3-5.0.1/lib/python3.6/site-packages/numpy/core/include/numpy/ndarrayobject.h:18,\n",
      "                   from /opt/modules/i12g/anaconda/3-5.0.1/lib/python3.6/site-packages/numpy/core/include/numpy/arrayobject.h:4,\n",
      "                   from cyvcf2/cyvcf2.c:526:\n",
      "  /opt/modules/i12g/anaconda/3-5.0.1/lib/python3.6/site-packages/numpy/core/include/numpy/npy_1_7_deprecated_api.h:15:2: warning: #warning \"Using deprecated NumPy API, disable it by \" \"#defining NPY_NO_DEPRECATED_API NPY_1_7_API_VERSION\" [-Wcpp]\n",
      "   #warning \"Using deprecated NumPy API, disable it by \" \\\n",
      "    ^~~~~~~\n",
      "  In file included from htslib/htslib/hts.h:35:0,\n",
      "                   from htslib/htslib/vcf.h:39,\n",
      "                   from cyvcf2/helpers.h:1,\n",
      "                   from cyvcf2/cyvcf2.c:529:\n",
      "  htslib/htslib/hts_log.h:52:6: warning: function declaration isn't a prototype [-Wstrict-prototypes]\n",
      "   enum htsLogLevel hts_get_log_level();\n",
      "        ^~~~~~~~~~~\n",
      "  cyvcf2/cyvcf2.c: In function '__pyx_pf_6cyvcf2_6cyvcf2_3VCF_4add_to_header':\n",
      "  cyvcf2/cyvcf2.c:8123:32: warning: passing argument 1 of 'bcf_hdr_append' discards 'const' qualifier from pointer target type [-Wdiscarded-qualifiers]\n",
      "     __pyx_v_ret = bcf_hdr_append(__pyx_v_self->hdr, __pyx_t_2);\n",
      "                                  ^~~~~~~~~~~~\n",
      "  In file included from cyvcf2/helpers.h:1:0,\n",
      "                   from cyvcf2/cyvcf2.c:529:\n",
      "  htslib/htslib/vcf.h:443:9: note: expected 'bcf_hdr_t * {aka struct <anonymous> *}' but argument is of type 'const bcf_hdr_t * {aka const struct <anonymous> *}'\n",
      "       int bcf_hdr_append(bcf_hdr_t *h, const char *line);\n",
      "           ^~~~~~~~~~~~~~\n",
      "  cyvcf2/cyvcf2.c:8165:30: warning: passing argument 1 of 'bcf_hdr_sync' discards 'const' qualifier from pointer target type [-Wdiscarded-qualifiers]\n",
      "     __pyx_v_ret = bcf_hdr_sync(__pyx_v_self->hdr);\n",
      "                                ^~~~~~~~~~~~\n",
      "  In file included from cyvcf2/helpers.h:1:0,\n",
      "                   from cyvcf2/cyvcf2.c:529:\n",
      "  htslib/htslib/vcf.h:479:9: note: expected 'bcf_hdr_t * {aka struct <anonymous> *}' but argument is of type 'const bcf_hdr_t * {aka const struct <anonymous> *}'\n",
      "       int bcf_hdr_sync(bcf_hdr_t *h);\n",
      "           ^~~~~~~~~~~~\n",
      "  cyvcf2/cyvcf2.c: In function '__pyx_pf_6cyvcf2_6cyvcf2_3VCF_12set_samples':\n",
      "  cyvcf2/cyvcf2.c:8774:37: warning: passing argument 1 of 'bcf_hdr_set_samples' discards 'const' qualifier from pointer target type [-Wdiscarded-qualifiers]\n",
      "     __pyx_v_ret = bcf_hdr_set_samples(__pyx_v_self->hdr, ((char const *)__pyx_t_6), 0);\n",
      "                                       ^~~~~~~~~~~~\n",
      "  In file included from cyvcf2/helpers.h:1:0,\n",
      "                   from cyvcf2/cyvcf2.c:529:\n",
      "  htslib/htslib/vcf.h:315:9: note: expected 'bcf_hdr_t * {aka struct <anonymous> *}' but argument is of type 'const bcf_hdr_t * {aka const struct <anonymous> *}'\n",
      "       int bcf_hdr_set_samples(bcf_hdr_t *hdr, const char *samples, int is_file);\n",
      "           ^~~~~~~~~~~~~~~~~~~\n",
      "  cyvcf2/cyvcf2.c: In function '__pyx_pf_6cyvcf2_6cyvcf2_3VCF_14update':\n",
      "  cyvcf2/cyvcf2.c:9048:32: warning: passing argument 1 of 'bcf_hdr_append' discards 'const' qualifier from pointer target type [-Wdiscarded-qualifiers]\n",
      "     __pyx_v_ret = bcf_hdr_append(__pyx_v_self->hdr, __pyx_t_4);\n",
      "                                  ^~~~~~~~~~~~\n",
      "  In file included from cyvcf2/helpers.h:1:0,\n",
      "                   from cyvcf2/cyvcf2.c:529:\n",
      "  htslib/htslib/vcf.h:443:9: note: expected 'bcf_hdr_t * {aka struct <anonymous> *}' but argument is of type 'const bcf_hdr_t * {aka const struct <anonymous> *}'\n",
      "       int bcf_hdr_append(bcf_hdr_t *h, const char *line);\n",
      "           ^~~~~~~~~~~~~~\n",
      "  cyvcf2/cyvcf2.c:9101:30: warning: passing argument 1 of 'bcf_hdr_sync' discards 'const' qualifier from pointer target type [-Wdiscarded-qualifiers]\n",
      "     __pyx_v_ret = bcf_hdr_sync(__pyx_v_self->hdr);\n",
      "                                ^~~~~~~~~~~~\n",
      "  In file included from cyvcf2/helpers.h:1:0,\n",
      "                   from cyvcf2/cyvcf2.c:529:\n",
      "  htslib/htslib/vcf.h:479:9: note: expected 'bcf_hdr_t * {aka struct <anonymous> *}' but argument is of type 'const bcf_hdr_t * {aka const struct <anonymous> *}'\n",
      "       int bcf_hdr_sync(bcf_hdr_t *h);\n",
      "           ^~~~~~~~~~~~\n",
      "  cyvcf2/cyvcf2.c: In function '__pyx_gb_6cyvcf2_6cyvcf2_3VCF_18generator':\n",
      "  htslib/htslib/vcf.h:796:102: warning: passing argument 4 of 'hts_itr_querys' discards 'const' qualifier from pointer target type [-Wdiscarded-qualifiers]\n",
      "       #define bcf_itr_querys(idx, hdr, s) hts_itr_querys((idx), (s), (hts_name2id_f)(bcf_hdr_name2id), (hdr), hts_itr_query, bcf_readrec)\n",
      "                                                                                                        ^\n",
      "  cyvcf2/cyvcf2.c:9311:34: note: in expansion of macro 'bcf_itr_querys'\n",
      "     __pyx_cur_scope->__pyx_v_itr = bcf_itr_querys(__pyx_cur_scope->__pyx_v_self->hidx, __pyx_cur_scope->__pyx_v_self->hdr, __pyx_t_2);\n",
      "                                    ^~~~~~~~~~~~~~\n",
      "  In file included from htslib/htslib/vcf.h:39:0,\n",
      "                   from cyvcf2/helpers.h:1,\n",
      "                   from cyvcf2/cyvcf2.c:529:\n",
      "  htslib/htslib/hts.h:675:16: note: expected 'void *' but argument is of type 'const bcf_hdr_t * {aka const struct <anonymous> *}'\n",
      "       hts_itr_t *hts_itr_querys(const hts_idx_t *idx, const char *reg, hts_name2id_f getid, void *hdr, hts_itr_query_func *itr_query, hts_readrec_func *readrec);\n",
      "                  ^~~~~~~~~~~~~~\n",
      "  cyvcf2/cyvcf2.c: In function '__pyx_gb_6cyvcf2_6cyvcf2_3VCF_24generator2':\n",
      "  cyvcf2/cyvcf2.c:10686:134: warning: passing argument 2 of '__pyx_f_6cyvcf2_6cyvcf2_newHREC' discards 'const' qualifier from pointer target type [-Wdiscarded-qualifiers]\n",
      "       __pyx_t_3 = ((PyObject *)__pyx_f_6cyvcf2_6cyvcf2_newHREC((__pyx_cur_scope->__pyx_v_self->hdr->hrec[__pyx_cur_scope->__pyx_v_i]), __pyx_cur_scope->__pyx_v_self->hdr)); if (unlikely(!__pyx_t_3)) __PYX_ERR(0, 408, __pyx_L1_error)\n",
      "                                                                                                                                        ^~~~~~~~~~~~~~~\n",
      "  cyvcf2/cyvcf2.c:2730:61: note: expected 'bcf_hdr_t * {aka struct <anonymous> *}' but argument is of type 'const bcf_hdr_t * {aka const struct <anonymous> *}'\n",
      "   static CYTHON_INLINE struct __pyx_obj_6cyvcf2_6cyvcf2_HREC *__pyx_f_6cyvcf2_6cyvcf2_newHREC(bcf_hrec_t *, bcf_hdr_t *); /*proto*/\n",
      "                                                               ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "  cyvcf2/cyvcf2.c: In function '__pyx_pf_6cyvcf2_6cyvcf2_3VCF_35__dealloc__':\n",
      "  cyvcf2/cyvcf2.c:12360:21: warning: passing argument 1 of 'bcf_hdr_destroy' discards 'const' qualifier from pointer target type [-Wdiscarded-qualifiers]\n",
      "       bcf_hdr_destroy(__pyx_v_self->hdr);\n",
      "                       ^~~~~~~~~~~~\n",
      "  In file included from cyvcf2/helpers.h:1:0,\n",
      "                   from cyvcf2/cyvcf2.c:529:\n",
      "  htslib/htslib/vcf.h:261:10: note: expected 'bcf_hdr_t * {aka struct <anonymous> *}' but argument is of type 'const bcf_hdr_t * {aka const struct <anonymous> *}'\n",
      "       void bcf_hdr_destroy(bcf_hdr_t *h);\n",
      "            ^~~~~~~~~~~~~~~\n",
      "  cyvcf2/cyvcf2.c: In function '__pyx_pf_6cyvcf2_6cyvcf2_3VCF_10raw_header___get__':\n",
      "  cyvcf2/cyvcf2.c:12848:3: warning: 'bcf_hdr_fmt_text' is deprecated: use bcf_hdr_format() instead [-Wdeprecated-declarations]\n",
      "     __pyx_v_s = bcf_hdr_fmt_text(__pyx_v_self->hdr, 0, (&__pyx_v_hlen));\n",
      "     ^~~~~~~~~\n",
      "  In file included from cyvcf2/helpers.h:1:0,\n",
      "                   from cyvcf2/cyvcf2.c:529:\n",
      "  htslib/htslib/vcf.h:439:11: note: declared here\n",
      "       char *bcf_hdr_fmt_text(const bcf_hdr_t *hdr, int is_bcf, int *len)\n",
      "             ^~~~~~~~~~~~~~~~\n",
      "  cyvcf2/cyvcf2.c: In function '__pyx_pf_6cyvcf2_6cyvcf2_3VCF_8seqnames___get__':\n",
      "  cyvcf2/cyvcf2.c:13098:18: warning: assignment from incompatible pointer type [-Wincompatible-pointer-types]\n",
      "     __pyx_v_cnames = bcf_hdr_seqnames(__pyx_v_self->hdr, (&__pyx_v_n));\n",
      "                    ^\n",
      "  In file included from cyvcf2/helpers.h:1:0,\n",
      "                   from cyvcf2/cyvcf2.c:529:\n",
      "  htslib/htslib/vcf.h:799:111: warning: passing argument 4 of 'hts_idx_seqnames' discards 'const' qualifier from pointer target type [-Wdiscarded-qualifiers]\n",
      "       #define bcf_index_seqnames(idx, hdr, nptr) hts_idx_seqnames((idx),(nptr),(hts_id2name_f)(bcf_hdr_id2name),(hdr))\n",
      "                                                                                                                 ^\n",
      "  cyvcf2/cyvcf2.c:13225:24: note: in expansion of macro 'bcf_index_seqnames'\n",
      "         __pyx_v_cnames = bcf_index_seqnames(__pyx_v_self->hidx, __pyx_v_self->hdr, (&__pyx_v_n));\n",
      "                          ^~~~~~~~~~~~~~~~~~\n",
      "  In file included from htslib/htslib/vcf.h:39:0,\n",
      "                   from cyvcf2/helpers.h:1,\n",
      "                   from cyvcf2/cyvcf2.c:529:\n",
      "  htslib/htslib/hts.h:677:18: note: expected 'void *' but argument is of type 'const bcf_hdr_t * {aka const struct <anonymous> *}'\n",
      "       const char **hts_idx_seqnames(const hts_idx_t *idx, int *n, hts_id2name_f getid, void *hdr); // free only the array, not the values\n",
      "                    ^~~~~~~~~~~~~~~~\n",
      "  cyvcf2/cyvcf2.c:13225:22: warning: assignment from incompatible pointer type [-Wincompatible-pointer-types]\n",
      "         __pyx_v_cnames = bcf_index_seqnames(__pyx_v_self->hidx, __pyx_v_self->hdr, (&__pyx_v_n));\n",
      "                        ^\n",
      "  cyvcf2/cyvcf2.c:13308:22: warning: assignment from incompatible pointer type [-Wincompatible-pointer-types]\n",
      "         __pyx_v_cnames = tbx_seqnames(__pyx_v_self->idx, (&__pyx_v_n));\n",
      "                        ^\n",
      "  cyvcf2/cyvcf2.c: In function '__pyx_pf_6cyvcf2_6cyvcf2_7Variant_6FORMAT___get__':\n",
      "  cyvcf2/cyvcf2.c:26554:17: warning: assignment discards 'const' qualifier from pointer target type [-Wdiscarded-qualifiers]\n",
      "       __pyx_v_key = bcf_hdr_int2id(__pyx_v_self->vcf->hdr, BCF_DT_ID, __pyx_v_fmt.id);\n",
      "                   ^\n",
      "  cyvcf2/cyvcf2.c: In function '__pyx_pf_6cyvcf2_6cyvcf2_7Variant_6FILTER_2__set__':\n",
      "  cyvcf2/cyvcf2.c:37178:13: warning: assignment discards 'const' qualifier from pointer target type [-Wdiscarded-qualifiers]\n",
      "     __pyx_v_h = __pyx_t_5;\n",
      "               ^\n",
      "  cyvcf2/cyvcf2.c: In function '__pyx_pf_6cyvcf2_6cyvcf2_4INFO_12__next__':\n",
      "  cyvcf2/cyvcf2.c:40278:16: warning: assignment discards 'const' qualifier from pointer target type [-Wdiscarded-qualifiers]\n",
      "     __pyx_v_name = bcf_hdr_int2id(__pyx_v_self->hdr, BCF_DT_ID, __pyx_v_info->key);\n",
      "                  ^\n",
      "  cyvcf2/cyvcf2.c: In function '__pyx_f_6cyvcf2_6cyvcf2_newVariant':\n",
      "  cyvcf2/cyvcf2.c:41259:18: warning: assignment discards 'const' qualifier from pointer target type [-Wdiscarded-qualifiers]\n",
      "     __pyx_v_i->hdr = __pyx_t_4;\n",
      "                    ^\n",
      "  cyvcf2/cyvcf2.c: In function '__pyx_pf_6cyvcf2_6cyvcf2_6Writer___init__':\n",
      "  cyvcf2/cyvcf2.c:41954:16: warning: passing argument 1 of 'bcf_hdr_sync' discards 'const' qualifier from pointer target type [-Wdiscarded-qualifiers]\n",
      "     bcf_hdr_sync(__pyx_v_tmpl->hdr);\n",
      "                  ^~~~~~~~~~~~\n",
      "  In file included from cyvcf2/helpers.h:1:0,\n",
      "                   from cyvcf2/cyvcf2.c:529:\n",
      "  htslib/htslib/vcf.h:479:9: note: expected 'bcf_hdr_t * {aka struct <anonymous> *}' but argument is of type 'const bcf_hdr_t * {aka const struct <anonymous> *}'\n",
      "       int bcf_hdr_sync(bcf_hdr_t *h);\n",
      "           ^~~~~~~~~~~~\n",
      "  cyvcf2/cyvcf2.c:41982:16: warning: passing argument 1 of 'bcf_hdr_sync' discards 'const' qualifier from pointer target type [-Wdiscarded-qualifiers]\n",
      "     bcf_hdr_sync(__pyx_v_self->__pyx_base.hdr);\n",
      "                  ^~~~~~~~~~~~\n",
      "  In file included from cyvcf2/helpers.h:1:0,\n",
      "                   from cyvcf2/cyvcf2.c:529:\n",
      "  htslib/htslib/vcf.h:479:9: note: expected 'bcf_hdr_t * {aka struct <anonymous> *}' but argument is of type 'const bcf_hdr_t * {aka const struct <anonymous> *}'\n",
      "       int bcf_hdr_sync(bcf_hdr_t *h);\n",
      "           ^~~~~~~~~~~~\n",
      "  cyvcf2/cyvcf2.c: In function '__pyx_pf_6cyvcf2_6cyvcf2_6Writer_2write_header':\n",
      "  cyvcf2/cyvcf2.c:42049:47: warning: passing argument 2 of 'bcf_hdr_write' discards 'const' qualifier from pointer target type [-Wdiscarded-qualifiers]\n",
      "     bcf_hdr_write(__pyx_v_self->__pyx_base.hts, __pyx_v_self->__pyx_base.hdr);\n",
      "                                                 ^~~~~~~~~~~~\n",
      "  In file included from cyvcf2/helpers.h:1:0,\n",
      "                   from cyvcf2/cyvcf2.c:529:\n",
      "  htslib/htslib/vcf.h:320:9: note: expected 'bcf_hdr_t * {aka struct <anonymous> *}' but argument is of type 'const bcf_hdr_t * {aka const struct <anonymous> *}'\n",
      "       int bcf_hdr_write(htsFile *fp, bcf_hdr_t *h);\n",
      "           ^~~~~~~~~~~~~\n",
      "  cyvcf2/cyvcf2.c: In function '__pyx_pf_6cyvcf2_6cyvcf2_6Writer_4write_record':\n",
      "  cyvcf2/cyvcf2.c:42229:36: warning: passing argument 1 of 'bcf_hdr_add_hrec' discards 'const' qualifier from pointer target type [-Wdiscarded-qualifiers]\n",
      "       __pyx_t_1 = ((bcf_hdr_add_hrec(__pyx_v_self->__pyx_base.hdr, __pyx_v_h) < 0) != 0);\n",
      "                                      ^~~~~~~~~~~~\n",
      "  In file included from cyvcf2/helpers.h:1:0,\n",
      "                   from cyvcf2/cyvcf2.c:529:\n",
      "  htslib/htslib/vcf.h:482:9: note: expected 'bcf_hdr_t * {aka struct <anonymous> *}' but argument is of type 'const bcf_hdr_t * {aka const struct <anonymous> *}'\n",
      "       int bcf_hdr_add_hrec(bcf_hdr_t *hdr, bcf_hrec_t *hrec);\n",
      "           ^~~~~~~~~~~~~~~~\n",
      "  cyvcf2/cyvcf2.c:42272:18: warning: passing argument 1 of 'bcf_hdr_sync' discards 'const' qualifier from pointer target type [-Wdiscarded-qualifiers]\n",
      "       bcf_hdr_sync(__pyx_v_self->__pyx_base.hdr);\n",
      "                    ^~~~~~~~~~~~\n",
      "  In file included from cyvcf2/helpers.h:1:0,\n",
      "                   from cyvcf2/cyvcf2.c:529:\n",
      "  htslib/htslib/vcf.h:479:9: note: expected 'bcf_hdr_t * {aka struct <anonymous> *}' but argument is of type 'const bcf_hdr_t * {aka const struct <anonymous> *}'\n",
      "       int bcf_hdr_sync(bcf_hdr_t *h);\n",
      "           ^~~~~~~~~~~~\n",
      "  cyvcf2/cyvcf2.c:42336:76: warning: passing argument 2 of 'bcf_write' discards 'const' qualifier from pointer target type [-Wdiscarded-qualifiers]\n",
      "     __pyx_t_3 = __Pyx_PyInt_From_int(bcf_write(__pyx_v_self->__pyx_base.hts, __pyx_v_self->__pyx_base.hdr, __pyx_v_var->b)); if (unlikely(!__pyx_t_3)) __PYX_ERR(0, 2168, __pyx_L1_error)\n",
      "                                                                              ^~~~~~~~~~~~\n",
      "  In file included from cyvcf2/helpers.h:1:0,\n",
      "                   from cyvcf2/cyvcf2.c:529:\n",
      "  htslib/htslib/vcf.h:370:9: note: expected 'bcf_hdr_t * {aka struct <anonymous> *}' but argument is of type 'const bcf_hdr_t * {aka const struct <anonymous> *}'\n",
      "       int bcf_write(htsFile *fp, bcf_hdr_t *h, bcf1_t *v);\n",
      "           ^~~~~~~~~\n",
      "  At top level:\n",
      "  cyvcf2/cyvcf2.c:6725:18: warning: '__pyx_f_6cyvcf2_6cyvcf2_xstr' defined but not used [-Wunused-function]\n",
      "   static PyObject *__pyx_f_6cyvcf2_6cyvcf2_xstr(PyObject *__pyx_v_s) {\n",
      "                    ^~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "  /opt/modules/i12g/anaconda/3-5.0.1/bin/x86_64-conda_cos6-linux-gnu-cc -DNDEBUG -fwrapv -O2 -Wall -Wstrict-prototypes -I/opt/modules/i12g/samtools/src/1.3.1/samtools-1.3.1 -DNDEBUG -D_FORTIFY_SOURCE=2 -O2 -fPIC -Ihtslib -Icyvcf2 -I/opt/modules/i12g/anaconda/3-5.0.1/lib/python3.6/site-packages/numpy/core/include -I/opt/modules/i12g/anaconda/3-5.0.1/include/python3.6m -c htslib/errmod.c -o build/temp.linux-x86_64-3.6/htslib/errmod.o\n",
      "  In file included from htslib/htslib/hts.h:35:0,\n",
      "                   from htslib/errmod.c:29:\n",
      "  htslib/htslib/hts_log.h:52:6: warning: function declaration isn't a prototype [-Wstrict-prototypes]\n",
      "   enum htsLogLevel hts_get_log_level();\n",
      "        ^~~~~~~~~~~\n",
      "  /opt/modules/i12g/anaconda/3-5.0.1/bin/x86_64-conda_cos6-linux-gnu-cc -DNDEBUG -fwrapv -O2 -Wall -Wstrict-prototypes -I/opt/modules/i12g/samtools/src/1.3.1/samtools-1.3.1 -DNDEBUG -D_FORTIFY_SOURCE=2 -O2 -fPIC -Ihtslib -Icyvcf2 -I/opt/modules/i12g/anaconda/3-5.0.1/lib/python3.6/site-packages/numpy/core/include -I/opt/modules/i12g/anaconda/3-5.0.1/include/python3.6m -c htslib/vcf_sweep.c -o build/temp.linux-x86_64-3.6/htslib/vcf_sweep.o\n",
      "  In file included from htslib/htslib/hts.h:35:0,\n",
      "                   from htslib/htslib/vcf_sweep.h:29,\n",
      "                   from htslib/vcf_sweep.c:27:\n",
      "  htslib/htslib/hts_log.h:52:6: warning: function declaration isn't a prototype [-Wstrict-prototypes]\n",
      "   enum htsLogLevel hts_get_log_level();\n",
      "        ^~~~~~~~~~~\n",
      "  In file included from htslib/vcf_sweep.c:28:0:\n",
      "  htslib/htslib/bgzf.h:35:10: fatal error: zlib.h: No such file or directory\n",
      "   #include <zlib.h>\n",
      "            ^~~~~~~~\n",
      "  compilation terminated.\n",
      "  error: command '/opt/modules/i12g/anaconda/3-5.0.1/bin/x86_64-conda_cos6-linux-gnu-cc' failed with exit status 1\n",
      "  \n",
      "  ----------------------------------------\n",
      "\u001b[31m  Failed building wheel for cyvcf2\u001b[0m\n",
      "\u001b[?25h  Running setup.py clean for cyvcf2\n",
      "Failed to build cyvcf2\n",
      "\u001b[31mmmsplice 0.2.7 has requirement cyvcf2==0.9.0, but you'll have cyvcf2 0.10.8 which is incompatible.\u001b[0m\n",
      "Installing collected packages: cyvcf2\n",
      "  Found existing installation: cyvcf2 0.9.0\n",
      "    Uninstalling cyvcf2-0.9.0:\n",
      "      Successfully uninstalled cyvcf2-0.9.0\n",
      "  Running setup.py install for cyvcf2 ... \u001b[?25lerror\n",
      "    Complete output from command /opt/modules/i12g/anaconda/3-5.0.1/bin/python -u -c \"import setuptools, tokenize;__file__='/scratch/tmp/reinharj/pip-install-z3bdr5su/cyvcf2/setup.py';f=getattr(tokenize, 'open', open)(__file__);code=f.read().replace('\\r\\n', '\\n');f.close();exec(compile(code, __file__, 'exec'))\" install --record /scratch/tmp/reinharj/pip-record-m3n4wo_g/install-record.txt --single-version-externally-managed --compile:\n",
      "    running install\n",
      "    running build\n",
      "    running build_py\n",
      "    creating build\n",
      "    creating build/lib.linux-x86_64-3.6\n",
      "    creating build/lib.linux-x86_64-3.6/cyvcf2\n",
      "    copying cyvcf2/__main__.py -> build/lib.linux-x86_64-3.6/cyvcf2\n",
      "    copying cyvcf2/cli.py -> build/lib.linux-x86_64-3.6/cyvcf2\n",
      "    copying cyvcf2/__init__.py -> build/lib.linux-x86_64-3.6/cyvcf2\n",
      "    creating build/lib.linux-x86_64-3.6/cyvcf2/tests\n",
      "    copying cyvcf2/tests/test_cli.py -> build/lib.linux-x86_64-3.6/cyvcf2/tests\n",
      "    copying cyvcf2/tests/__init__.py -> build/lib.linux-x86_64-3.6/cyvcf2/tests\n",
      "    copying cyvcf2/tests/test_hemi.py -> build/lib.linux-x86_64-3.6/cyvcf2/tests\n",
      "    copying cyvcf2/tests/test_reader.py -> build/lib.linux-x86_64-3.6/cyvcf2/tests\n",
      "    running egg_info\n",
      "    writing cyvcf2.egg-info/PKG-INFO\n",
      "    writing dependency_links to cyvcf2.egg-info/dependency_links.txt\n",
      "    writing entry points to cyvcf2.egg-info/entry_points.txt\n",
      "    writing requirements to cyvcf2.egg-info/requires.txt\n",
      "    writing top-level names to cyvcf2.egg-info/top_level.txt\n",
      "    reading manifest file 'cyvcf2.egg-info/SOURCES.txt'\n",
      "    reading manifest template 'MANIFEST.in'\n",
      "    writing manifest file 'cyvcf2.egg-info/SOURCES.txt'\n",
      "    copying cyvcf2/cyvcf2.c -> build/lib.linux-x86_64-3.6/cyvcf2\n",
      "    copying cyvcf2/cyvcf2.pxd -> build/lib.linux-x86_64-3.6/cyvcf2\n",
      "    copying cyvcf2/cyvcf2.pyx -> build/lib.linux-x86_64-3.6/cyvcf2\n",
      "    copying cyvcf2/helpers.c -> build/lib.linux-x86_64-3.6/cyvcf2\n",
      "    copying cyvcf2/helpers.h -> build/lib.linux-x86_64-3.6/cyvcf2\n",
      "    copying cyvcf2/relatedness.h -> build/lib.linux-x86_64-3.6/cyvcf2\n",
      "    copying cyvcf2/tests/__init__.pyc -> build/lib.linux-x86_64-3.6/cyvcf2/tests\n",
      "    copying cyvcf2/tests/bug.vcf.gz -> build/lib.linux-x86_64-3.6/cyvcf2/tests\n",
      "    copying cyvcf2/tests/decomposed.vcf -> build/lib.linux-x86_64-3.6/cyvcf2/tests\n",
      "    copying cyvcf2/tests/empty.vcf -> build/lib.linux-x86_64-3.6/cyvcf2/tests\n",
      "    copying cyvcf2/tests/issue_44.vcf -> build/lib.linux-x86_64-3.6/cyvcf2/tests\n",
      "    copying cyvcf2/tests/o.vcf.gz -> build/lib.linux-x86_64-3.6/cyvcf2/tests\n",
      "    copying cyvcf2/tests/seg.vcf.gz -> build/lib.linux-x86_64-3.6/cyvcf2/tests\n",
      "    copying cyvcf2/tests/test-alt-repr.vcf -> build/lib.linux-x86_64-3.6/cyvcf2/tests\n",
      "    copying cyvcf2/tests/test-format-string.vcf -> build/lib.linux-x86_64-3.6/cyvcf2/tests\n",
      "    copying cyvcf2/tests/test-haploidX.vcf -> build/lib.linux-x86_64-3.6/cyvcf2/tests\n",
      "    copying cyvcf2/tests/test-hemi.vcf -> build/lib.linux-x86_64-3.6/cyvcf2/tests\n",
      "    copying cyvcf2/tests/test-multiallelic-homozygous-alt.vcf.gz -> build/lib.linux-x86_64-3.6/cyvcf2/tests\n",
      "    copying cyvcf2/tests/test-multiallelic-homozygous-alt.vcf.gz.tbi -> build/lib.linux-x86_64-3.6/cyvcf2/tests\n",
      "    copying cyvcf2/tests/test-strict-gt-option-flag.vcf.gz -> build/lib.linux-x86_64-3.6/cyvcf2/tests\n",
      "    copying cyvcf2/tests/test-strict-gt-option-flag.vcf.gz.tbi -> build/lib.linux-x86_64-3.6/cyvcf2/tests\n",
      "    copying cyvcf2/tests/test.comp_het.3.vcf -> build/lib.linux-x86_64-3.6/cyvcf2/tests\n",
      "    copying cyvcf2/tests/test.snpeff.bcf -> build/lib.linux-x86_64-3.6/cyvcf2/tests\n",
      "    copying cyvcf2/tests/test.snpeff.bcf.csi -> build/lib.linux-x86_64-3.6/cyvcf2/tests\n",
      "    copying cyvcf2/tests/test.snpeff.vcf -> build/lib.linux-x86_64-3.6/cyvcf2/tests\n",
      "    copying cyvcf2/tests/test.vcf.gz -> build/lib.linux-x86_64-3.6/cyvcf2/tests\n",
      "    copying cyvcf2/tests/test.vcf.gz.tbi -> build/lib.linux-x86_64-3.6/cyvcf2/tests\n",
      "    copying cyvcf2/tests/test_cli.pyc -> build/lib.linux-x86_64-3.6/cyvcf2/tests\n",
      "    copying cyvcf2/tests/test_gt_alt_freqs.vcf -> build/lib.linux-x86_64-3.6/cyvcf2/tests\n",
      "    copying cyvcf2/tests/test_gt_bases.vcf.gz -> build/lib.linux-x86_64-3.6/cyvcf2/tests\n",
      "    copying cyvcf2/tests/test_hemi.pyc -> build/lib.linux-x86_64-3.6/cyvcf2/tests\n",
      "    copying cyvcf2/tests/test_reader.pyc -> build/lib.linux-x86_64-3.6/cyvcf2/tests\n",
      "    running build_ext\n",
      "    skipping 'cyvcf2/cyvcf2.c' Cython extension (up-to-date)\n",
      "    building 'cyvcf2.cyvcf2' extension\n",
      "    creating build/temp.linux-x86_64-3.6\n",
      "    creating build/temp.linux-x86_64-3.6/cyvcf2\n",
      "    creating build/temp.linux-x86_64-3.6/htslib\n",
      "    creating build/temp.linux-x86_64-3.6/htslib/cram\n",
      "    /opt/modules/i12g/anaconda/3-5.0.1/bin/x86_64-conda_cos6-linux-gnu-cc -DNDEBUG -fwrapv -O2 -Wall -Wstrict-prototypes -I/opt/modules/i12g/samtools/src/1.3.1/samtools-1.3.1 -DNDEBUG -D_FORTIFY_SOURCE=2 -O2 -fPIC -Ihtslib -Icyvcf2 -I/opt/modules/i12g/anaconda/3-5.0.1/lib/python3.6/site-packages/numpy/core/include -I/opt/modules/i12g/anaconda/3-5.0.1/include/python3.6m -c cyvcf2/cyvcf2.c -o build/temp.linux-x86_64-3.6/cyvcf2/cyvcf2.o\n",
      "    In file included from /opt/modules/i12g/anaconda/3-5.0.1/lib/python3.6/site-packages/numpy/core/include/numpy/ndarraytypes.h:1821:0,\n",
      "                     from /opt/modules/i12g/anaconda/3-5.0.1/lib/python3.6/site-packages/numpy/core/include/numpy/ndarrayobject.h:18,\n",
      "                     from /opt/modules/i12g/anaconda/3-5.0.1/lib/python3.6/site-packages/numpy/core/include/numpy/arrayobject.h:4,\n",
      "                     from cyvcf2/cyvcf2.c:526:\n",
      "    /opt/modules/i12g/anaconda/3-5.0.1/lib/python3.6/site-packages/numpy/core/include/numpy/npy_1_7_deprecated_api.h:15:2: warning: #warning \"Using deprecated NumPy API, disable it by \" \"#defining NPY_NO_DEPRECATED_API NPY_1_7_API_VERSION\" [-Wcpp]\n",
      "     #warning \"Using deprecated NumPy API, disable it by \" \\\n",
      "      ^~~~~~~\n",
      "    In file included from htslib/htslib/hts.h:35:0,\n",
      "                     from htslib/htslib/vcf.h:39,\n",
      "                     from cyvcf2/helpers.h:1,\n",
      "                     from cyvcf2/cyvcf2.c:529:\n",
      "    htslib/htslib/hts_log.h:52:6: warning: function declaration isn't a prototype [-Wstrict-prototypes]\n",
      "     enum htsLogLevel hts_get_log_level();\n",
      "          ^~~~~~~~~~~\n",
      "    cyvcf2/cyvcf2.c: In function '__pyx_pf_6cyvcf2_6cyvcf2_3VCF_4add_to_header':\n",
      "    cyvcf2/cyvcf2.c:8123:32: warning: passing argument 1 of 'bcf_hdr_append' discards 'const' qualifier from pointer target type [-Wdiscarded-qualifiers]\n",
      "       __pyx_v_ret = bcf_hdr_append(__pyx_v_self->hdr, __pyx_t_2);\n",
      "                                    ^~~~~~~~~~~~\n",
      "    In file included from cyvcf2/helpers.h:1:0,\n",
      "                     from cyvcf2/cyvcf2.c:529:\n",
      "    htslib/htslib/vcf.h:443:9: note: expected 'bcf_hdr_t * {aka struct <anonymous> *}' but argument is of type 'const bcf_hdr_t * {aka const struct <anonymous> *}'\n",
      "         int bcf_hdr_append(bcf_hdr_t *h, const char *line);\n",
      "             ^~~~~~~~~~~~~~\n",
      "    cyvcf2/cyvcf2.c:8165:30: warning: passing argument 1 of 'bcf_hdr_sync' discards 'const' qualifier from pointer target type [-Wdiscarded-qualifiers]\n",
      "       __pyx_v_ret = bcf_hdr_sync(__pyx_v_self->hdr);\n",
      "                                  ^~~~~~~~~~~~\n",
      "    In file included from cyvcf2/helpers.h:1:0,\n",
      "                     from cyvcf2/cyvcf2.c:529:\n",
      "    htslib/htslib/vcf.h:479:9: note: expected 'bcf_hdr_t * {aka struct <anonymous> *}' but argument is of type 'const bcf_hdr_t * {aka const struct <anonymous> *}'\n",
      "         int bcf_hdr_sync(bcf_hdr_t *h);\n",
      "             ^~~~~~~~~~~~\n",
      "    cyvcf2/cyvcf2.c: In function '__pyx_pf_6cyvcf2_6cyvcf2_3VCF_12set_samples':\n",
      "    cyvcf2/cyvcf2.c:8774:37: warning: passing argument 1 of 'bcf_hdr_set_samples' discards 'const' qualifier from pointer target type [-Wdiscarded-qualifiers]\n",
      "       __pyx_v_ret = bcf_hdr_set_samples(__pyx_v_self->hdr, ((char const *)__pyx_t_6), 0);\n",
      "                                         ^~~~~~~~~~~~\n",
      "    In file included from cyvcf2/helpers.h:1:0,\n",
      "                     from cyvcf2/cyvcf2.c:529:\n",
      "    htslib/htslib/vcf.h:315:9: note: expected 'bcf_hdr_t * {aka struct <anonymous> *}' but argument is of type 'const bcf_hdr_t * {aka const struct <anonymous> *}'\n",
      "         int bcf_hdr_set_samples(bcf_hdr_t *hdr, const char *samples, int is_file);\n",
      "             ^~~~~~~~~~~~~~~~~~~\n",
      "    cyvcf2/cyvcf2.c: In function '__pyx_pf_6cyvcf2_6cyvcf2_3VCF_14update':\n",
      "    cyvcf2/cyvcf2.c:9048:32: warning: passing argument 1 of 'bcf_hdr_append' discards 'const' qualifier from pointer target type [-Wdiscarded-qualifiers]\n",
      "       __pyx_v_ret = bcf_hdr_append(__pyx_v_self->hdr, __pyx_t_4);\n",
      "                                    ^~~~~~~~~~~~\n",
      "    In file included from cyvcf2/helpers.h:1:0,\n",
      "                     from cyvcf2/cyvcf2.c:529:\n",
      "    htslib/htslib/vcf.h:443:9: note: expected 'bcf_hdr_t * {aka struct <anonymous> *}' but argument is of type 'const bcf_hdr_t * {aka const struct <anonymous> *}'\n",
      "         int bcf_hdr_append(bcf_hdr_t *h, const char *line);\n",
      "             ^~~~~~~~~~~~~~\n",
      "    cyvcf2/cyvcf2.c:9101:30: warning: passing argument 1 of 'bcf_hdr_sync' discards 'const' qualifier from pointer target type [-Wdiscarded-qualifiers]\n",
      "       __pyx_v_ret = bcf_hdr_sync(__pyx_v_self->hdr);\n",
      "                                  ^~~~~~~~~~~~\n",
      "    In file included from cyvcf2/helpers.h:1:0,\n",
      "                     from cyvcf2/cyvcf2.c:529:\n",
      "    htslib/htslib/vcf.h:479:9: note: expected 'bcf_hdr_t * {aka struct <anonymous> *}' but argument is of type 'const bcf_hdr_t * {aka const struct <anonymous> *}'\n",
      "         int bcf_hdr_sync(bcf_hdr_t *h);\n",
      "             ^~~~~~~~~~~~\n",
      "    cyvcf2/cyvcf2.c: In function '__pyx_gb_6cyvcf2_6cyvcf2_3VCF_18generator':\n",
      "    htslib/htslib/vcf.h:796:102: warning: passing argument 4 of 'hts_itr_querys' discards 'const' qualifier from pointer target type [-Wdiscarded-qualifiers]\n",
      "         #define bcf_itr_querys(idx, hdr, s) hts_itr_querys((idx), (s), (hts_name2id_f)(bcf_hdr_name2id), (hdr), hts_itr_query, bcf_readrec)\n",
      "                                                                                                          ^\n",
      "    cyvcf2/cyvcf2.c:9311:34: note: in expansion of macro 'bcf_itr_querys'\n",
      "       __pyx_cur_scope->__pyx_v_itr = bcf_itr_querys(__pyx_cur_scope->__pyx_v_self->hidx, __pyx_cur_scope->__pyx_v_self->hdr, __pyx_t_2);\n",
      "                                      ^~~~~~~~~~~~~~\n",
      "    In file included from htslib/htslib/vcf.h:39:0,\n",
      "                     from cyvcf2/helpers.h:1,\n",
      "                     from cyvcf2/cyvcf2.c:529:\n",
      "    htslib/htslib/hts.h:675:16: note: expected 'void *' but argument is of type 'const bcf_hdr_t * {aka const struct <anonymous> *}'\n",
      "         hts_itr_t *hts_itr_querys(const hts_idx_t *idx, const char *reg, hts_name2id_f getid, void *hdr, hts_itr_query_func *itr_query, hts_readrec_func *readrec);\n",
      "                    ^~~~~~~~~~~~~~\n",
      "    cyvcf2/cyvcf2.c: In function '__pyx_gb_6cyvcf2_6cyvcf2_3VCF_24generator2':\n",
      "    cyvcf2/cyvcf2.c:10686:134: warning: passing argument 2 of '__pyx_f_6cyvcf2_6cyvcf2_newHREC' discards 'const' qualifier from pointer target type [-Wdiscarded-qualifiers]\n",
      "         __pyx_t_3 = ((PyObject *)__pyx_f_6cyvcf2_6cyvcf2_newHREC((__pyx_cur_scope->__pyx_v_self->hdr->hrec[__pyx_cur_scope->__pyx_v_i]), __pyx_cur_scope->__pyx_v_self->hdr)); if (unlikely(!__pyx_t_3)) __PYX_ERR(0, 408, __pyx_L1_error)\n",
      "                                                                                                                                          ^~~~~~~~~~~~~~~\n",
      "    cyvcf2/cyvcf2.c:2730:61: note: expected 'bcf_hdr_t * {aka struct <anonymous> *}' but argument is of type 'const bcf_hdr_t * {aka const struct <anonymous> *}'\n",
      "     static CYTHON_INLINE struct __pyx_obj_6cyvcf2_6cyvcf2_HREC *__pyx_f_6cyvcf2_6cyvcf2_newHREC(bcf_hrec_t *, bcf_hdr_t *); /*proto*/\n",
      "                                                                 ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "    cyvcf2/cyvcf2.c: In function '__pyx_pf_6cyvcf2_6cyvcf2_3VCF_35__dealloc__':\n",
      "    cyvcf2/cyvcf2.c:12360:21: warning: passing argument 1 of 'bcf_hdr_destroy' discards 'const' qualifier from pointer target type [-Wdiscarded-qualifiers]\n",
      "         bcf_hdr_destroy(__pyx_v_self->hdr);\n",
      "                         ^~~~~~~~~~~~\n",
      "    In file included from cyvcf2/helpers.h:1:0,\n",
      "                     from cyvcf2/cyvcf2.c:529:\n",
      "    htslib/htslib/vcf.h:261:10: note: expected 'bcf_hdr_t * {aka struct <anonymous> *}' but argument is of type 'const bcf_hdr_t * {aka const struct <anonymous> *}'\n",
      "         void bcf_hdr_destroy(bcf_hdr_t *h);\n",
      "              ^~~~~~~~~~~~~~~\n",
      "    cyvcf2/cyvcf2.c: In function '__pyx_pf_6cyvcf2_6cyvcf2_3VCF_10raw_header___get__':\n",
      "    cyvcf2/cyvcf2.c:12848:3: warning: 'bcf_hdr_fmt_text' is deprecated: use bcf_hdr_format() instead [-Wdeprecated-declarations]\n",
      "       __pyx_v_s = bcf_hdr_fmt_text(__pyx_v_self->hdr, 0, (&__pyx_v_hlen));\n",
      "       ^~~~~~~~~\n",
      "    In file included from cyvcf2/helpers.h:1:0,\n",
      "                     from cyvcf2/cyvcf2.c:529:\n",
      "    htslib/htslib/vcf.h:439:11: note: declared here\n",
      "         char *bcf_hdr_fmt_text(const bcf_hdr_t *hdr, int is_bcf, int *len)\n",
      "               ^~~~~~~~~~~~~~~~\n",
      "    cyvcf2/cyvcf2.c: In function '__pyx_pf_6cyvcf2_6cyvcf2_3VCF_8seqnames___get__':\n",
      "    cyvcf2/cyvcf2.c:13098:18: warning: assignment from incompatible pointer type [-Wincompatible-pointer-types]\n",
      "       __pyx_v_cnames = bcf_hdr_seqnames(__pyx_v_self->hdr, (&__pyx_v_n));\n",
      "                      ^\n",
      "    In file included from cyvcf2/helpers.h:1:0,\n",
      "                     from cyvcf2/cyvcf2.c:529:\n",
      "    htslib/htslib/vcf.h:799:111: warning: passing argument 4 of 'hts_idx_seqnames' discards 'const' qualifier from pointer target type [-Wdiscarded-qualifiers]\n",
      "         #define bcf_index_seqnames(idx, hdr, nptr) hts_idx_seqnames((idx),(nptr),(hts_id2name_f)(bcf_hdr_id2name),(hdr))\n",
      "                                                                                                                   ^\n",
      "    cyvcf2/cyvcf2.c:13225:24: note: in expansion of macro 'bcf_index_seqnames'\n",
      "           __pyx_v_cnames = bcf_index_seqnames(__pyx_v_self->hidx, __pyx_v_self->hdr, (&__pyx_v_n));\n",
      "                            ^~~~~~~~~~~~~~~~~~\n",
      "    In file included from htslib/htslib/vcf.h:39:0,\n",
      "                     from cyvcf2/helpers.h:1,\n",
      "                     from cyvcf2/cyvcf2.c:529:\n",
      "    htslib/htslib/hts.h:677:18: note: expected 'void *' but argument is of type 'const bcf_hdr_t * {aka const struct <anonymous> *}'\n",
      "         const char **hts_idx_seqnames(const hts_idx_t *idx, int *n, hts_id2name_f getid, void *hdr); // free only the array, not the values\n",
      "                      ^~~~~~~~~~~~~~~~\n",
      "    cyvcf2/cyvcf2.c:13225:22: warning: assignment from incompatible pointer type [-Wincompatible-pointer-types]\n",
      "           __pyx_v_cnames = bcf_index_seqnames(__pyx_v_self->hidx, __pyx_v_self->hdr, (&__pyx_v_n));\n",
      "                          ^\n",
      "    cyvcf2/cyvcf2.c:13308:22: warning: assignment from incompatible pointer type [-Wincompatible-pointer-types]\n",
      "           __pyx_v_cnames = tbx_seqnames(__pyx_v_self->idx, (&__pyx_v_n));\n",
      "                          ^\n",
      "    cyvcf2/cyvcf2.c: In function '__pyx_pf_6cyvcf2_6cyvcf2_7Variant_6FORMAT___get__':\n",
      "    cyvcf2/cyvcf2.c:26554:17: warning: assignment discards 'const' qualifier from pointer target type [-Wdiscarded-qualifiers]\n",
      "         __pyx_v_key = bcf_hdr_int2id(__pyx_v_self->vcf->hdr, BCF_DT_ID, __pyx_v_fmt.id);\n",
      "                     ^\n",
      "    cyvcf2/cyvcf2.c: In function '__pyx_pf_6cyvcf2_6cyvcf2_7Variant_6FILTER_2__set__':\n",
      "    cyvcf2/cyvcf2.c:37178:13: warning: assignment discards 'const' qualifier from pointer target type [-Wdiscarded-qualifiers]\n",
      "       __pyx_v_h = __pyx_t_5;\n",
      "                 ^\n",
      "    cyvcf2/cyvcf2.c: In function '__pyx_pf_6cyvcf2_6cyvcf2_4INFO_12__next__':\n",
      "    cyvcf2/cyvcf2.c:40278:16: warning: assignment discards 'const' qualifier from pointer target type [-Wdiscarded-qualifiers]\n",
      "       __pyx_v_name = bcf_hdr_int2id(__pyx_v_self->hdr, BCF_DT_ID, __pyx_v_info->key);\n",
      "                    ^\n",
      "    cyvcf2/cyvcf2.c: In function '__pyx_f_6cyvcf2_6cyvcf2_newVariant':\n",
      "    cyvcf2/cyvcf2.c:41259:18: warning: assignment discards 'const' qualifier from pointer target type [-Wdiscarded-qualifiers]\n",
      "       __pyx_v_i->hdr = __pyx_t_4;\n",
      "                      ^\n",
      "    cyvcf2/cyvcf2.c: In function '__pyx_pf_6cyvcf2_6cyvcf2_6Writer___init__':\n",
      "    cyvcf2/cyvcf2.c:41954:16: warning: passing argument 1 of 'bcf_hdr_sync' discards 'const' qualifier from pointer target type [-Wdiscarded-qualifiers]\n",
      "       bcf_hdr_sync(__pyx_v_tmpl->hdr);\n",
      "                    ^~~~~~~~~~~~\n",
      "    In file included from cyvcf2/helpers.h:1:0,\n",
      "                     from cyvcf2/cyvcf2.c:529:\n",
      "    htslib/htslib/vcf.h:479:9: note: expected 'bcf_hdr_t * {aka struct <anonymous> *}' but argument is of type 'const bcf_hdr_t * {aka const struct <anonymous> *}'\n",
      "         int bcf_hdr_sync(bcf_hdr_t *h);\n",
      "             ^~~~~~~~~~~~\n",
      "    cyvcf2/cyvcf2.c:41982:16: warning: passing argument 1 of 'bcf_hdr_sync' discards 'const' qualifier from pointer target type [-Wdiscarded-qualifiers]\n",
      "       bcf_hdr_sync(__pyx_v_self->__pyx_base.hdr);\n",
      "                    ^~~~~~~~~~~~\n",
      "    In file included from cyvcf2/helpers.h:1:0,\n",
      "                     from cyvcf2/cyvcf2.c:529:\n",
      "    htslib/htslib/vcf.h:479:9: note: expected 'bcf_hdr_t * {aka struct <anonymous> *}' but argument is of type 'const bcf_hdr_t * {aka const struct <anonymous> *}'\n",
      "         int bcf_hdr_sync(bcf_hdr_t *h);\n",
      "             ^~~~~~~~~~~~\n",
      "    cyvcf2/cyvcf2.c: In function '__pyx_pf_6cyvcf2_6cyvcf2_6Writer_2write_header':\n",
      "    cyvcf2/cyvcf2.c:42049:47: warning: passing argument 2 of 'bcf_hdr_write' discards 'const' qualifier from pointer target type [-Wdiscarded-qualifiers]\n",
      "       bcf_hdr_write(__pyx_v_self->__pyx_base.hts, __pyx_v_self->__pyx_base.hdr);\n",
      "                                                   ^~~~~~~~~~~~\n",
      "    In file included from cyvcf2/helpers.h:1:0,\n",
      "                     from cyvcf2/cyvcf2.c:529:\n",
      "    htslib/htslib/vcf.h:320:9: note: expected 'bcf_hdr_t * {aka struct <anonymous> *}' but argument is of type 'const bcf_hdr_t * {aka const struct <anonymous> *}'\n",
      "         int bcf_hdr_write(htsFile *fp, bcf_hdr_t *h);\n",
      "             ^~~~~~~~~~~~~\n",
      "    cyvcf2/cyvcf2.c: In function '__pyx_pf_6cyvcf2_6cyvcf2_6Writer_4write_record':\n",
      "    cyvcf2/cyvcf2.c:42229:36: warning: passing argument 1 of 'bcf_hdr_add_hrec' discards 'const' qualifier from pointer target type [-Wdiscarded-qualifiers]\n",
      "         __pyx_t_1 = ((bcf_hdr_add_hrec(__pyx_v_self->__pyx_base.hdr, __pyx_v_h) < 0) != 0);\n",
      "                                        ^~~~~~~~~~~~\n",
      "    In file included from cyvcf2/helpers.h:1:0,\n",
      "                     from cyvcf2/cyvcf2.c:529:\n",
      "    htslib/htslib/vcf.h:482:9: note: expected 'bcf_hdr_t * {aka struct <anonymous> *}' but argument is of type 'const bcf_hdr_t * {aka const struct <anonymous> *}'\n",
      "         int bcf_hdr_add_hrec(bcf_hdr_t *hdr, bcf_hrec_t *hrec);\n",
      "             ^~~~~~~~~~~~~~~~\n",
      "    cyvcf2/cyvcf2.c:42272:18: warning: passing argument 1 of 'bcf_hdr_sync' discards 'const' qualifier from pointer target type [-Wdiscarded-qualifiers]\n",
      "         bcf_hdr_sync(__pyx_v_self->__pyx_base.hdr);\n",
      "                      ^~~~~~~~~~~~\n",
      "    In file included from cyvcf2/helpers.h:1:0,\n",
      "                     from cyvcf2/cyvcf2.c:529:\n",
      "    htslib/htslib/vcf.h:479:9: note: expected 'bcf_hdr_t * {aka struct <anonymous> *}' but argument is of type 'const bcf_hdr_t * {aka const struct <anonymous> *}'\n",
      "         int bcf_hdr_sync(bcf_hdr_t *h);\n",
      "             ^~~~~~~~~~~~\n",
      "    cyvcf2/cyvcf2.c:42336:76: warning: passing argument 2 of 'bcf_write' discards 'const' qualifier from pointer target type [-Wdiscarded-qualifiers]\n",
      "       __pyx_t_3 = __Pyx_PyInt_From_int(bcf_write(__pyx_v_self->__pyx_base.hts, __pyx_v_self->__pyx_base.hdr, __pyx_v_var->b)); if (unlikely(!__pyx_t_3)) __PYX_ERR(0, 2168, __pyx_L1_error)\n",
      "                                                                                ^~~~~~~~~~~~\n",
      "    In file included from cyvcf2/helpers.h:1:0,\n",
      "                     from cyvcf2/cyvcf2.c:529:\n",
      "    htslib/htslib/vcf.h:370:9: note: expected 'bcf_hdr_t * {aka struct <anonymous> *}' but argument is of type 'const bcf_hdr_t * {aka const struct <anonymous> *}'\n",
      "         int bcf_write(htsFile *fp, bcf_hdr_t *h, bcf1_t *v);\n",
      "             ^~~~~~~~~\n",
      "    At top level:\n",
      "    cyvcf2/cyvcf2.c:6725:18: warning: '__pyx_f_6cyvcf2_6cyvcf2_xstr' defined but not used [-Wunused-function]\n",
      "     static PyObject *__pyx_f_6cyvcf2_6cyvcf2_xstr(PyObject *__pyx_v_s) {\n",
      "                      ^~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "    /opt/modules/i12g/anaconda/3-5.0.1/bin/x86_64-conda_cos6-linux-gnu-cc -DNDEBUG -fwrapv -O2 -Wall -Wstrict-prototypes -I/opt/modules/i12g/samtools/src/1.3.1/samtools-1.3.1 -DNDEBUG -D_FORTIFY_SOURCE=2 -O2 -fPIC -Ihtslib -Icyvcf2 -I/opt/modules/i12g/anaconda/3-5.0.1/lib/python3.6/site-packages/numpy/core/include -I/opt/modules/i12g/anaconda/3-5.0.1/include/python3.6m -c htslib/errmod.c -o build/temp.linux-x86_64-3.6/htslib/errmod.o\n",
      "    In file included from htslib/htslib/hts.h:35:0,\n",
      "                     from htslib/errmod.c:29:\n",
      "    htslib/htslib/hts_log.h:52:6: warning: function declaration isn't a prototype [-Wstrict-prototypes]\n",
      "     enum htsLogLevel hts_get_log_level();\n",
      "          ^~~~~~~~~~~\n",
      "    /opt/modules/i12g/anaconda/3-5.0.1/bin/x86_64-conda_cos6-linux-gnu-cc -DNDEBUG -fwrapv -O2 -Wall -Wstrict-prototypes -I/opt/modules/i12g/samtools/src/1.3.1/samtools-1.3.1 -DNDEBUG -D_FORTIFY_SOURCE=2 -O2 -fPIC -Ihtslib -Icyvcf2 -I/opt/modules/i12g/anaconda/3-5.0.1/lib/python3.6/site-packages/numpy/core/include -I/opt/modules/i12g/anaconda/3-5.0.1/include/python3.6m -c htslib/vcf_sweep.c -o build/temp.linux-x86_64-3.6/htslib/vcf_sweep.o\n",
      "    In file included from htslib/htslib/hts.h:35:0,\n",
      "                     from htslib/htslib/vcf_sweep.h:29,\n",
      "                     from htslib/vcf_sweep.c:27:\n",
      "    htslib/htslib/hts_log.h:52:6: warning: function declaration isn't a prototype [-Wstrict-prototypes]\n",
      "     enum htsLogLevel hts_get_log_level();\n",
      "          ^~~~~~~~~~~\n",
      "    In file included from htslib/vcf_sweep.c:28:0:\n",
      "    htslib/htslib/bgzf.h:35:10: fatal error: zlib.h: No such file or directory\n",
      "     #include <zlib.h>\n",
      "              ^~~~~~~~\n",
      "    compilation terminated.\n",
      "    error: command '/opt/modules/i12g/anaconda/3-5.0.1/bin/x86_64-conda_cos6-linux-gnu-cc' failed with exit status 1\n",
      "    \n",
      "    ----------------------------------------\n",
      "\u001b[?25h  Rolling back uninstall of cyvcf2\n",
      "\u001b[31mCould not install packages due to an EnvironmentError: [Errno 22] Invalid argument: '/data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/lib/python3.6/site-packages/-yvcf2-0.9.0.dist-info/.'\n",
      "\u001b[0m\n",
      "\u001b[33mYou are using pip version 19.0.1, however version 19.0.3 is available.\n",
      "You should consider upgrading via the 'pip install --upgrade pip' command.\u001b[0m\n"
     ]
    }
   ],
   "source": [
    "!pip3 install -U cyvcf2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "ename": "ModuleNotFoundError",
     "evalue": "No module named 'cyvcf2'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mModuleNotFoundError\u001b[0m                       Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-4-817846f577e6>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0;32mimport\u001b[0m \u001b[0mcyvcf2\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mModuleNotFoundError\u001b[0m: No module named 'cyvcf2'"
     ]
    }
   ],
   "source": [
    "import cyvcf2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "ename": "ModuleNotFoundError",
     "evalue": "No module named 'kipoiseq'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mModuleNotFoundError\u001b[0m                       Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-1-362d7ad5302a>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0;32mfrom\u001b[0m \u001b[0mkipoiseq\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mextractors\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mvcf_seq\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0mMultiSampleVCF\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      2\u001b[0m \u001b[0mvcfReader\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mMultiSampleVCF\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mvcfTest\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mModuleNotFoundError\u001b[0m: No module named 'kipoiseq'"
     ]
    }
   ],
   "source": [
    "from kipoiseq.extractors.vcf_seq import MultiSampleVCF\n",
    "vcfReader=MultiSampleVCF(vcfTest)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 177,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'0.2.3'"
      ]
     },
     "execution_count": 177,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "kipoiseq.__version__"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from kipoiseq.extractors import BaseVCFSeqExtractor\n",
    "class AllelicVCFSeqExtractor(BaseVCFSeqExtractor):\n",
    "    \n",
    "    def splitVariants(self,variants):\n",
    "        \"\"\"splits variants into homo-allelic and hetero-allelic\"\"\"\n",
    "        \n",
    "        \n",
    "    def get_allelic_variants(self,allVars):\n",
    "        \"\"\"\n",
    "        hetero-allelic variants are only used on first allele, homo- on both\n",
    "        \"\"\"\n",
    "        homoVars,heteroVars=splitVariants(self,allVars)\n",
    "        return [homoVars+heteroVars,homoVars]\n",
    "        \n",
    "    def extract(self, interval, anchor=None, sample_id=None, fixed_len=True):\n",
    "        all_variants=self.vcf.fetch_variants(interval, sample_id)\n",
    "        \n",
    "        vars1,vars2=get_allelic_variants()\n",
    "        \n",
    "        allele1=self.variant_extractor.extract(\n",
    "                interval, variants=vars2,\n",
    "                anchor=anchor, fixed_len=fixed_len)        \n",
    "        allele2=self.variant_extractor.extract(\n",
    "                interval, variants=vars1,\n",
    "                anchor=anchor, fixed_len=fixed_len)\n",
    "        \n",
    "        return [allele1,allele2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 129,
   "metadata": {},
   "outputs": [],
   "source": [
    "from torch.utils.data import Dataset\n",
    "\n",
    "\n",
    "\n",
    "#formerly \"repDataloader\"\n",
    "class repDataset(Dataset):\n",
    "    from pybedtools import Interval\n",
    "    \n",
    "    def __init__(self,vcfFile,refGenFasta,tssFile,n_upstream,n_downstream,sample_ids):\n",
    "        \"\"\"set instance-vars, read TSS completely, load (dummy) VCFfastaExtractor\"\"\"\n",
    "\n",
    "        import pandas as pd\n",
    "\n",
    "        from kipoiseq.transforms import ReorderedOneHot\n",
    "        self.transform=ReorderedOneHot()\n",
    "        \n",
    "        \n",
    "        self.vcfFile=vcfFile\n",
    "        self.refGenFasta=refGenFasta\n",
    "        self.varExtractor=VcfFastaDummy(vcfFile,refGenFasta)\n",
    "        \n",
    "        self.sampleIDs=sample_ids\n",
    "        self.interval=Interval(n_upstream,n_downstream)\n",
    "        \n",
    "        #TODO use only prot-coding? Separate file or bound to schema of example file?\n",
    "        self.tssCollection=pd.read_csv(tssFile) # needs field \"TSS\" with RegGen position and \"strand\" with \"+\" or \"-\"\n",
    "        \n",
    "\n",
    "    def __len__(self):\n",
    "        return len(self.sampleIDs)*len(self.tssCollection)\n",
    "    def __getitem__(self,idx):\n",
    "        import numpy as np\n",
    "        \"\"\"transform idx to sample- and tss-id, then get both alleles from extractor\"\"\"\n",
    "        sampleIdx=int(idx/len(self.tssCollection))\n",
    "        tssIdx=int(idx%len(self.tssCollection))\n",
    "        \n",
    "        \n",
    "        chromo=self.tssCollection.loc[tssIdx,\"seqname\"]\n",
    "        strand=self.tssCollection.loc[tssIdx,\"strand\"]\n",
    "        tssStart=self.tssCollection.loc[tssIdx,\"TSS\"]\n",
    "        \n",
    "        genomic_interval=self.interval.getGenomicInterval(chromo,tssStart,strand)  \n",
    "        \n",
    "        alleles= self.varExtractor.extract(genomic_interval,self.sampleIDs[sampleIdx]) #strand-info where?\n",
    "        \n",
    "        #formatting output\n",
    "        gene_id=self.tssCollection.loc[tssIdx,\"id\"]\n",
    "        sample_id=self.sampleIDs[sampleIdx]\n",
    "        returnBatch={}\n",
    "        returnBatch[\"inputs\"]=[self.transform(alleles[0]),self.transform(alleles[1])]\n",
    "        #need np.array-conversion, or it will concat on other dimension for batchsize>1\n",
    "        returnBatch[\"metadata\"]={\"gene_id\":np.array([gene_id]*2),\"sample_id\":np.array([sample_id]*2),\"allele_id\":np.array([0,1])} \n",
    "        return returnBatch\n",
    "    \n",
    "    def getGenomicInterval(chromosome,chromoAnchor,strand):\n",
    "        \"\"\"converts internal interval to bedInterval as used in extractor\"\"\"\n",
    "        from pybedtools import Interval\n",
    "        self.interval.setAnchor(chromoAnchor)\n",
    "        return Interval(chromosome, start=self.interval.lower, end=self.interval.upper, strand=strand, otherfields=None)\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "#make test-tss file\n",
    "! head -n 20 /s/project/avsec/ExPecto/resources/geneanno.csv> testTSS.csv\n",
    "tssFile=\"testTSS.csv\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 133,
   "metadata": {},
   "outputs": [],
   "source": [
    "#data-loader tut here: https://stanford.edu/~shervine/blog/pytorch-how-to-generate-data-parallel\n",
    "basenjiSeqSize=131072 \n",
    "testData = repDataset(\"dummy\",\"dummy\",tssFile,-int(basenjiSeqSize/2),int(basenjiSeqSize/2),[\"dummySampleID\"])\n",
    "\n",
    "testBatch=testData.__getitem__(18)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 123,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'input': [array([[1., 0., 0., 0.],\n",
      "       [1., 0., 0., 0.],\n",
      "       [1., 0., 0., 0.],\n",
      "       [1., 0., 0., 0.],\n",
      "       [1., 0., 0., 0.],\n",
      "       [1., 0., 0., 0.]]), array([[0., 1., 0., 0.],\n",
      "       [0., 1., 0., 0.],\n",
      "       [0., 1., 0., 0.],\n",
      "       [0., 1., 0., 0.],\n",
      "       [0., 1., 0., 0.],\n",
      "       [0., 1., 0., 0.]])], 'metadata': {'gene_id': array(['ENSG00000000003', 'ENSG00000000003', 'ENSG00000000005',\n",
      "       'ENSG00000000005', 'ENSG00000000419', 'ENSG00000000419'],\n",
      "      dtype='<U15'), 'sample_id': array(['dummySampleID', 'dummySampleID', 'dummySampleID', 'dummySampleID',\n",
      "       'dummySampleID', 'dummySampleID'], dtype='<U13'), 'allele_id': array([0, 1, 0, 1, 0, 1])}}\n",
      "{'input': [array([[1., 0., 0., 0.],\n",
      "       [1., 0., 0., 0.],\n",
      "       [1., 0., 0., 0.],\n",
      "       [1., 0., 0., 0.],\n",
      "       [1., 0., 0., 0.],\n",
      "       [1., 0., 0., 0.]]), array([[0., 1., 0., 0.],\n",
      "       [0., 1., 0., 0.],\n",
      "       [0., 1., 0., 0.],\n",
      "       [0., 1., 0., 0.],\n",
      "       [0., 1., 0., 0.],\n",
      "       [0., 1., 0., 0.]])], 'metadata': {'gene_id': array(['ENSG00000000457', 'ENSG00000000457', 'ENSG00000000460',\n",
      "       'ENSG00000000460', 'ENSG00000000938', 'ENSG00000000938'],\n",
      "      dtype='<U15'), 'sample_id': array(['dummySampleID', 'dummySampleID', 'dummySampleID', 'dummySampleID',\n",
      "       'dummySampleID', 'dummySampleID'], dtype='<U13'), 'allele_id': array([0, 1, 0, 1, 0, 1])}}\n",
      "{'input': [array([[1., 0., 0., 0.],\n",
      "       [1., 0., 0., 0.],\n",
      "       [1., 0., 0., 0.],\n",
      "       [1., 0., 0., 0.],\n",
      "       [1., 0., 0., 0.],\n",
      "       [1., 0., 0., 0.]]), array([[0., 1., 0., 0.],\n",
      "       [0., 1., 0., 0.],\n",
      "       [0., 1., 0., 0.],\n",
      "       [0., 1., 0., 0.],\n",
      "       [0., 1., 0., 0.],\n",
      "       [0., 1., 0., 0.]])], 'metadata': {'gene_id': array(['ENSG00000000971', 'ENSG00000000971', 'ENSG00000001036',\n",
      "       'ENSG00000001036', 'ENSG00000001084', 'ENSG00000001084'],\n",
      "      dtype='<U15'), 'sample_id': array(['dummySampleID', 'dummySampleID', 'dummySampleID', 'dummySampleID',\n",
      "       'dummySampleID', 'dummySampleID'], dtype='<U13'), 'allele_id': array([0, 1, 0, 1, 0, 1])}}\n",
      "{'input': [array([[1., 0., 0., 0.],\n",
      "       [1., 0., 0., 0.],\n",
      "       [1., 0., 0., 0.],\n",
      "       [1., 0., 0., 0.],\n",
      "       [1., 0., 0., 0.],\n",
      "       [1., 0., 0., 0.]]), array([[0., 1., 0., 0.],\n",
      "       [0., 1., 0., 0.],\n",
      "       [0., 1., 0., 0.],\n",
      "       [0., 1., 0., 0.],\n",
      "       [0., 1., 0., 0.],\n",
      "       [0., 1., 0., 0.]])], 'metadata': {'gene_id': array(['ENSG00000001167', 'ENSG00000001167', 'ENSG00000001460',\n",
      "       'ENSG00000001460', 'ENSG00000001461', 'ENSG00000001461'],\n",
      "      dtype='<U15'), 'sample_id': array(['dummySampleID', 'dummySampleID', 'dummySampleID', 'dummySampleID',\n",
      "       'dummySampleID', 'dummySampleID'], dtype='<U13'), 'allele_id': array([0, 1, 0, 1, 0, 1])}}\n",
      "{'input': [array([[1., 0., 0., 0.],\n",
      "       [1., 0., 0., 0.],\n",
      "       [1., 0., 0., 0.],\n",
      "       [1., 0., 0., 0.],\n",
      "       [1., 0., 0., 0.],\n",
      "       [1., 0., 0., 0.]]), array([[0., 1., 0., 0.],\n",
      "       [0., 1., 0., 0.],\n",
      "       [0., 1., 0., 0.],\n",
      "       [0., 1., 0., 0.],\n",
      "       [0., 1., 0., 0.],\n",
      "       [0., 1., 0., 0.]])], 'metadata': {'gene_id': array(['ENSG00000001497', 'ENSG00000001497', 'ENSG00000001561',\n",
      "       'ENSG00000001561', 'ENSG00000001617', 'ENSG00000001617'],\n",
      "      dtype='<U15'), 'sample_id': array(['dummySampleID', 'dummySampleID', 'dummySampleID', 'dummySampleID',\n",
      "       'dummySampleID', 'dummySampleID'], dtype='<U13'), 'allele_id': array([0, 1, 0, 1, 0, 1])}}\n",
      "{'input': [array([[1., 0., 0., 0.],\n",
      "       [1., 0., 0., 0.],\n",
      "       [1., 0., 0., 0.],\n",
      "       [1., 0., 0., 0.],\n",
      "       [1., 0., 0., 0.],\n",
      "       [1., 0., 0., 0.]]), array([[0., 1., 0., 0.],\n",
      "       [0., 1., 0., 0.],\n",
      "       [0., 1., 0., 0.],\n",
      "       [0., 1., 0., 0.],\n",
      "       [0., 1., 0., 0.],\n",
      "       [0., 1., 0., 0.]])], 'metadata': {'gene_id': array(['ENSG00000001626', 'ENSG00000001626', 'ENSG00000001629',\n",
      "       'ENSG00000001629', 'ENSG00000001630', 'ENSG00000001630'],\n",
      "      dtype='<U15'), 'sample_id': array(['dummySampleID', 'dummySampleID', 'dummySampleID', 'dummySampleID',\n",
      "       'dummySampleID', 'dummySampleID'], dtype='<U13'), 'allele_id': array([0, 1, 0, 1, 0, 1])}}\n",
      "{'input': [array([[1., 0., 0., 0.],\n",
      "       [1., 0., 0., 0.]]), array([[0., 1., 0., 0.],\n",
      "       [0., 1., 0., 0.]])], 'metadata': {'gene_id': array(['ENSG00000001631', 'ENSG00000001631'], dtype='<U15'), 'sample_id': array(['dummySampleID', 'dummySampleID'], dtype='<U13'), 'allele_id': array([0, 1])}}\n"
     ]
    }
   ],
   "source": [
    "##MAIN routine!\n",
    "\n",
    "from torch.utils import data\n",
    "from kipoi.data_utils import numpy_collate_concat\n",
    "generator = data.DataLoader(testData,batch_size=3, collate_fn=numpy_collate_concat)\n",
    "\n",
    "import kipoi\n",
    "model = kipoi.get_model('Basenji')\n",
    "\n",
    "writer=ResultWriter()\n",
    "\n",
    "\n",
    "for batch in generator:\n",
    "    preds=model.predict_on_batch(batch['inputs'])\n",
    "    #TODO choose/compute right bin AFTER TSS, not over. For now, just use middle\n",
    "    #see https://github.com/kipoi/models/issues/87 for info n dimensions\n",
    "    posidx = int(960/2)\n",
    "    #remember: id_inputvec(batchsize,here allele1 and 2), id_genomicInterval, id_features/\"tissues\"\n",
    "    tssExpr=preds[:,pos_idx:(pos_idx+1),:]\n",
    "    batch[\"preds\"]=tssExpr\n",
    "    writer.write_results(batch)\n",
    "    \n",
    "writer.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import kipoi\n",
    "model = kipoi.get_model('Basenji')\n",
    "\n",
    "\n",
    "#Standard prediction process in kipoi-basenji\n",
    "# Download example dataloader kwargs\n",
    "dl_kwargs = model.default_dataloader.download_example('example')\n",
    "# Get the dataloader and instantiate it\n",
    "dl = model.default_dataloader(**dl_kwargs)\n",
    "# get a batch iterator\n",
    "it = dl.batch_iter(batch_size=2) #does not work with anything else\n",
    "# predict for a batch\n",
    "batch = next(it)\n",
    "preds=model.predict_on_batch(batch['input'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 131,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Using downloaded and verified file: /data/nasif12/home_if12/reinharj/.kipoi/models/Basenji/downloaded/model_files/ckp/model.meta\n",
      "Using downloaded and verified file: /data/nasif12/home_if12/reinharj/.kipoi/models/Basenji/downloaded/model_files/ckp/model.index\n",
      "Using downloaded and verified file: /data/nasif12/home_if12/reinharj/.kipoi/models/Basenji/downloaded/model_files/ckp/model.data-00000-of-00001\n",
      "Using downloaded and verified file: /data/nasif12/home_if12/reinharj/.kipoi/models/Basenji/downloaded/model_files/const_feed_dict_pkl/3a76c37eb9ad255680ba774b110de1be\n",
      "INFO:tensorflow:Restoring parameters from /data/nasif12/home_if12/reinharj/.kipoi/models/Basenji/downloaded/model_files/ckp/model\n"
     ]
    },
    {
     "ename": "ValueError",
     "evalue": "Cannot feed value of shape (2, 2, 4) for Tensor 'inputs:0', which has shape '(2, 131072, 4)'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-131-8f7354449ea8>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0mkipoi\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      2\u001b[0m \u001b[0mmodel\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mkipoi\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget_model\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'Basenji'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 3\u001b[0;31m \u001b[0mpreds\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mmodel\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mpredict_on_batch\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mtestBatch\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'inputs'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m/opt/modules/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages/kipoi/model.py\u001b[0m in \u001b[0;36mpredict_on_batch\u001b[0;34m(self, x)\u001b[0m\n\u001b[1;32m   1512\u001b[0m         \u001b[0mfeed_dict\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_build_feed_dict\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mx\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1513\u001b[0m         return self.sess.run(self.target_ops,\n\u001b[0;32m-> 1514\u001b[0;31m                              feed_dict=merge_dicts(feed_dict, self.const_feed_dict))\n\u001b[0m\u001b[1;32m   1515\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1516\u001b[0m     \u001b[0;34m@\u001b[0m\u001b[0mclassmethod\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/opt/modules/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages/tensorflow/python/client/session.py\u001b[0m in \u001b[0;36mrun\u001b[0;34m(self, fetches, feed_dict, options, run_metadata)\u001b[0m\n\u001b[1;32m    927\u001b[0m     \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    928\u001b[0m       result = self._run(None, fetches, feed_dict, options_ptr,\n\u001b[0;32m--> 929\u001b[0;31m                          run_metadata_ptr)\n\u001b[0m\u001b[1;32m    930\u001b[0m       \u001b[0;32mif\u001b[0m \u001b[0mrun_metadata\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    931\u001b[0m         \u001b[0mproto_data\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mtf_session\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mTF_GetBuffer\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mrun_metadata_ptr\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/opt/modules/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages/tensorflow/python/client/session.py\u001b[0m in \u001b[0;36m_run\u001b[0;34m(self, handle, fetches, feed_dict, options, run_metadata)\u001b[0m\n\u001b[1;32m   1126\u001b[0m                              \u001b[0;34m'which has shape %r'\u001b[0m \u001b[0;34m%\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1127\u001b[0m                              (np_val.shape, subfeed_t.name,\n\u001b[0;32m-> 1128\u001b[0;31m                               str(subfeed_t.get_shape())))\n\u001b[0m\u001b[1;32m   1129\u001b[0m           \u001b[0;32mif\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mgraph\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mis_feedable\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msubfeed_t\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1130\u001b[0m             \u001b[0;32mraise\u001b[0m \u001b[0mValueError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'Tensor %s may not be fed.'\u001b[0m \u001b[0;34m%\u001b[0m \u001b[0msubfeed_t\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mValueError\u001b[0m: Cannot feed value of shape (2, 2, 4) for Tensor 'inputs:0', which has shape '(2, 131072, 4)'"
     ]
    }
   ],
   "source": [
    "import kipoi\n",
    "model = kipoi.get_model('Basenji')\n",
    "preds=model.predict_on_batch(testBatch['inputs'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 134,
   "metadata": {},
   "outputs": [],
   "source": [
    "preds=model.predict_on_batch(testBatch['inputs'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 141,
   "metadata": {},
   "outputs": [],
   "source": [
    "preds.shape\n",
    "#TODO choose/compute right bin AFTER TSS, not over. For now, just use middle\n",
    "#see https://github.com/kipoi/models/issues/87 for info n dimensions\n",
    "tssExpr=preds[:,int(960/2),:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 142,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[ 1.8517308 ,  3.6696124 ,  1.8414515 , ...,  7.1143007 ,\n",
       "         4.61417   , 10.097965  ],\n",
       "       [ 0.07730692, 14.220292  ,  0.02221763, ...,  2.2907472 ,\n",
       "         0.02211935,  1.0927494 ]], dtype=float32)"
      ]
     },
     "execution_count": 142,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tssExpr"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Output-writer with Zarr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting zarr\n",
      "\u001b[?25l  Downloading https://files.pythonhosted.org/packages/30/5d/95f88adf4f50947949942364e9501f0bb06d3e3159018e20c23c1d462ae8/zarr-2.2.0.tar.gz (3.3MB)\n",
      "\u001b[K    100% |████████████████████████████████| 3.3MB 1.3MB/s \n",
      "\u001b[?25hCollecting asciitree (from zarr)\n",
      "  Downloading https://files.pythonhosted.org/packages/2d/6a/885bc91484e1aa8f618f6f0228d76d0e67000b0fdd6090673b777e311913/asciitree-0.3.3.tar.gz\n",
      "Requirement already satisfied: numpy>=1.7 in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (from zarr) (1.16.0)\n",
      "Collecting fasteners (from zarr)\n",
      "  Downloading https://files.pythonhosted.org/packages/14/3a/096c7ad18e102d4f219f5dd15951f9728ca5092a3385d2e8f79a7c1e1017/fasteners-0.14.1-py2.py3-none-any.whl\n",
      "Collecting numcodecs>=0.5.3 (from zarr)\n",
      "\u001b[?25l  Downloading https://files.pythonhosted.org/packages/6f/1c/b1537659c5900312f77454f3eca25c728bf1b0ab24ccbcca5e1d712a661b/numcodecs-0.6.3.tar.gz (3.8MB)\n",
      "\u001b[K    100% |████████████████████████████████| 3.8MB 1.0MB/s \n",
      "\u001b[?25hRequirement already satisfied: six in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (from fasteners->zarr) (1.12.0)\n",
      "Collecting monotonic>=0.1 (from fasteners->zarr)\n",
      "  Downloading https://files.pythonhosted.org/packages/ac/aa/063eca6a416f397bd99552c534c6d11d57f58f2e94c14780f3bbf818c4cf/monotonic-1.5-py2.py3-none-any.whl\n",
      "Building wheels for collected packages: zarr, asciitree, numcodecs\n",
      "  Running setup.py bdist_wheel for zarr ... \u001b[?25ldone\n",
      "\u001b[?25h  Stored in directory: /data/ouga/home/ag_gagneur/reinharj/.cache/pip/wheels/60/68/03/ed754a88e8af5cc9d77d6b043ac7913a69a0c4bccf036b56b2\n",
      "  Running setup.py bdist_wheel for asciitree ... \u001b[?25ldone\n",
      "\u001b[?25h  Stored in directory: /data/ouga/home/ag_gagneur/reinharj/.cache/pip/wheels/1d/d9/58/9808b306744df0208fccc640d3d9952a5bc7468502d42897d5\n",
      "  Running setup.py bdist_wheel for numcodecs ... \u001b[?25ldone\n",
      "\u001b[?25h  Stored in directory: /data/ouga/home/ag_gagneur/reinharj/.cache/pip/wheels/75/ea/37/8a1c2b6e937606f90dd0a9659783790d922efd1cf98d7a82a9\n",
      "Successfully built zarr asciitree numcodecs\n",
      "Installing collected packages: asciitree, monotonic, fasteners, numcodecs, zarr\n",
      "Successfully installed asciitree-0.3.3 fasteners-0.14.1 monotonic-1.5 numcodecs-0.6.3 zarr-2.2.0\n"
     ]
    }
   ],
   "source": [
    "! pip install zarr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(1000, 1000)"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "(1000,)*2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "class ZarrBatchWriter(object):\n",
    "    def __init__(self,outputFile,genes,samples,genePositions,predSize):\n",
    "        \"\"\"set up data array, annotate with attributes\"\"\" #TODO maybe use dicts(name-idx) for O(1) lookup??\n",
    "        import zarr\n",
    "        self.filename=outputFile\n",
    "        self.cGenes=len(genes)\n",
    "        self.cSamples=len(samples)\n",
    "        self.cGenePositions=len(genePositions)\n",
    "        self.cPredSize=predSize\n",
    "        #Allele got canceled, just append allele 2 to allele1\n",
    "        self.data=zarr.open(self.filename, mode='w', shape=(self.cGenes,self.cSamples,self.cGenePositions,self.cPredSize),\n",
    "                 dtype='f8') #TODO: precision right? chunksize left out for now, are there default values? \"chunks=(1000, 1000),\"\n",
    "        \n",
    "        self.data.attrs['genes'] = self.as_idx_dict(genes)\n",
    "        self.data.attrs[\"samples\"]=self.as_idx_dict(samples)\n",
    "        self.data.attrs[\"genePositions\"]=self.as_idx_dict(genePositions)\n",
    "    @staticmethod    \n",
    "    def as_idx_dict(lis):\n",
    "        \"\"\"for O1 lookup\"\"\"\n",
    "        dic={}\n",
    "        for idx,elem in enumerate(lis):\n",
    "            dic[elem]=idx\n",
    "        return dic\n",
    "        \n",
    "    def batch_write(predData):\n",
    "        \"\"\"plug in predictions in right position, given by metadata and annotation-array\n",
    "        Exprects whole range of geneLocations at once!\n",
    "        \"\"\"\n",
    "        genes=set(predData[\"metadata\"][\"gene_id\"])\n",
    "        samples=set(predData[\"metadata\"][\"sample_id\"])\n",
    "        \n",
    "        if len(genes)!=1 or len(samples)!=1:\n",
    "            raise AttributeError(\"Wrong number of genes and/or samples for one write operation. Only accepting exactly one of each per write-call\")\n",
    "        geneId=self.data.attrs[\"genes\"][genes[0]]\n",
    "        sampleId=self.data.attrs[\"samples\"][samples[0]]\n",
    "        \n",
    "        \n",
    "        cSeq=predData[\"preds\"].shape[0]\n",
    "        concatPreds=np.reshape(np.stack(predData[\"preds\"],axis=1),(self.cGenePositions,self.cPredSize) #Todo own function, test...\n",
    "        \n",
    "        self.data[geneId,sampleId]=concatPreds\n",
    "        \n",
    "        return\n",
    "        \n",
    "        \n",
    "test=ZarrBatchWriter(\"test\",[\"gene1\",\"gene2\",\"gene4\"],[\"samp1\",\"samp2\"],[0],10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "import zarr\n",
    "test=zarr.zeros((10,10))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [],
   "source": [
    "#testing array reshaping\n",
    "import numpy as np\n",
    "test=np.array([range(24)])\n",
    "test=np.reshape(test,(2,3,4))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[[ 0,  1,  2,  3],\n",
       "        [ 4,  5,  6,  7],\n",
       "        [ 8,  9, 10, 11]],\n",
       "\n",
       "       [[12, 13, 14, 15],\n",
       "        [16, 17, 18, 19],\n",
       "        [20, 21, 22, 23]]])"
      ]
     },
     "execution_count": 67,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test[:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[[ 0,  1,  2,  3],\n",
       "        [ 4,  5,  6,  7],\n",
       "        [ 8,  9, 10, 11]],\n",
       "\n",
       "       [[12, 13, 14, 15],\n",
       "        [16, 17, 18, 19],\n",
       "        [20, 21, 22, 23]]])"
      ]
     },
     "execution_count": 94,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test2=np.stack(test,axis=1)\n",
    "test"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[ 0,  1,  2,  3, 12, 13, 14, 15],\n",
       "       [ 4,  5,  6,  7, 16, 17, 18, 19],\n",
       "       [ 8,  9, 10, 11, 20, 21, 22, 23]])"
      ]
     },
     "execution_count": 95,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.reshape(test2,(3,8))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
