{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Goal\n",
    "Preprocessing rnaseq data\n",
    "\n",
    "This is a tutorial with the main workflow for preparing a training dataset for a basic linear regression model. \n",
    "The output (X,Y) are all pair combination between tissues per each sample, where X is input, Y target."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# code autoreload\n",
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: numpy in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (1.15.4)\n",
      "Requirement already satisfied: pandas in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (0.23.4)\n",
      "Requirement already satisfied: numpy>=1.9.0 in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (from pandas) (1.15.4)\n",
      "Requirement already satisfied: python-dateutil>=2.5.0 in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (from pandas) (2.7.5)\n",
      "Requirement already satisfied: pytz>=2011k in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (from pandas) (2018.7)\n",
      "Requirement already satisfied: six>=1.5 in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (from python-dateutil>=2.5.0->pandas) (1.12.0)\n",
      "Requirement already satisfied: anndata in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (0.6.16)\n",
      "Requirement already satisfied: pandas>=0.21.0 in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (from anndata) (0.23.4)\n",
      "Requirement already satisfied: scipy~=1.0 in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (from anndata) (1.2.0)\n",
      "Requirement already satisfied: natsort in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (from anndata) (5.5.0)\n",
      "Requirement already satisfied: numpy~=1.14 in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (from anndata) (1.15.4)\n",
      "Requirement already satisfied: h5py in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (from anndata) (2.9.0)\n",
      "Requirement already satisfied: python-dateutil>=2.5.0 in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (from pandas>=0.21.0->anndata) (2.7.5)\n",
      "Requirement already satisfied: pytz>=2011k in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (from pandas>=0.21.0->anndata) (2018.7)\n",
      "Requirement already satisfied: six in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (from h5py->anndata) (1.12.0)\n",
      "Requirement already satisfied: scanpy in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (1.3.6)\n",
      "Requirement already satisfied: natsort in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (from scanpy) (5.5.0)\n",
      "Requirement already satisfied: networkx in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (from scanpy) (2.2)\n",
      "Requirement already satisfied: seaborn in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (from scanpy) (0.9.0)\n",
      "Requirement already satisfied: joblib in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (from scanpy) (0.13.0)\n",
      "Requirement already satisfied: numba>=0.40.0 in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (from scanpy) (0.41.0)\n",
      "Requirement already satisfied: scipy in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (from scanpy) (1.2.0)\n",
      "Requirement already satisfied: scikit-learn>=0.19.1 in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (from scanpy) (0.20.2)\n",
      "Requirement already satisfied: pandas>=0.21 in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (from scanpy) (0.23.4)\n",
      "Requirement already satisfied: h5py in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (from scanpy) (2.9.0)\n",
      "Requirement already satisfied: matplotlib>=2.2 in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (from scanpy) (3.0.2)\n",
      "Requirement already satisfied: statsmodels in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (from scanpy) (0.9.0)\n",
      "Requirement already satisfied: anndata>=0.6.13 in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (from scanpy) (0.6.16)\n",
      "Requirement already satisfied: tables in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (from scanpy) (3.4.4)\n",
      "Requirement already satisfied: decorator>=4.3.0 in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (from networkx->scanpy) (4.3.0)\n",
      "Requirement already satisfied: numpy>=1.9.3 in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (from seaborn->scanpy) (1.15.4)\n",
      "Requirement already satisfied: llvmlite>=0.26.0dev0 in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (from numba>=0.40.0->scanpy) (0.26.0)\n",
      "Requirement already satisfied: pytz>=2011k in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (from pandas>=0.21->scanpy) (2018.7)\n",
      "Requirement already satisfied: python-dateutil>=2.5.0 in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (from pandas>=0.21->scanpy) (2.7.5)\n",
      "Requirement already satisfied: six in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (from h5py->scanpy) (1.12.0)\n",
      "Requirement already satisfied: cycler>=0.10 in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (from matplotlib>=2.2->scanpy) (0.10.0)\n",
      "Requirement already satisfied: pyparsing!=2.0.4,!=2.1.2,!=2.1.6,>=2.0.1 in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (from matplotlib>=2.2->scanpy) (2.3.0)\n",
      "Requirement already satisfied: kiwisolver>=1.0.1 in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (from matplotlib>=2.2->scanpy) (1.0.1)\n",
      "Requirement already satisfied: patsy in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (from statsmodels->scanpy) (0.5.1)\n",
      "Requirement already satisfied: numexpr>=2.5.2 in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (from tables->scanpy) (2.6.8)\n",
      "Requirement already satisfied: setuptools in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (from kiwisolver>=1.0.1->matplotlib>=2.2->scanpy) (40.6.3)\n",
      "Requirement already satisfied: argh in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (0.26.2)\n"
     ]
    }
   ],
   "source": [
    "!pip install numpy\n",
    "!pip install pandas\n",
    "!pip install anndata\n",
    "\n",
    "!pip install scanpy\n",
    "!pip install argh"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Modules to reload:\n",
      "all-except-skipped\n",
      "\n",
      "Modules to skip:\n",
      "\n"
     ]
    }
   ],
   "source": [
    "import rep.preprocessing as p\n",
    "from rep.constants import ANNDATA_CST as a\n",
    "\n",
    "%aimport"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 1. Load raw count matrix + annotation and put it into an annodata object"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting data.csv\n"
     ]
    }
   ],
   "source": [
    "%%writefile data.csv\n",
    "T1_s1,T2_s2,T3_s1,T4_s2,T5_s1,T6_s1\n",
    "G1,10,20,30,40,5,6\n",
    "G2,5,10,20,30,5,6\n",
    "G3,6,7,8,9,1,1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting anno_var.csv\n"
     ]
    }
   ],
   "source": [
    "%%writefile anno_var.csv\n",
    "Sample,Gender,Tissue,Individual,Technology\n",
    "T1_s1,F,Tissue1,Sample1,rnaseq\n",
    "T2_s2,M,Tissue2,Sample2,rnaseq\n",
    "T3_s1,F,Tissue3,Sample1,rnaseq\n",
    "T4_s2,M,Tissue4,Sample2,rnaseq\n",
    "T5_s1,M,Tissue5,Sample1,rnaseq\n",
    "T6_s1,M,Tissue6,Sample1,rnaseq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Writing anno_obs.csv\n"
     ]
    }
   ],
   "source": [
    "%%writefile anno_obs.csv\n",
    "G1,hg19,T1,chr1,1111,-\n",
    "G2,hg19,T2,chr2,2222,-\n",
    "G3,hg19,T3,chr3,3333,-"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "anndata.X\n",
      "[[10. 20. 30. 40.  5.  6.]\n",
      " [ 5. 10. 20. 30.  5.  6.]\n",
      " [ 6.  7.  8.  9.  1.  1.]]\n",
      "anndata.var\n",
      "Index(['T1_s1', 'T2_s2', 'T3_s1', 'T4_s2', 'T5_s1', 'T6_s1'], dtype='object')\n",
      "      Gender   Tissue Individual Technology\n",
      "T1_s1      F  Tissue1    Sample1     rnaseq\n",
      "T2_s2      M  Tissue2    Sample2     rnaseq\n",
      "T3_s1      F  Tissue3    Sample1     rnaseq\n",
      "T4_s2      M  Tissue4    Sample2     rnaseq\n",
      "T5_s1      M  Tissue5    Sample1     rnaseq\n",
      "T6_s1      M  Tissue6    Sample1     rnaseq\n",
      "anndata.obs\n",
      "Index(['G1', 'G2', 'G3'], dtype='object')\n",
      "       1   2     3     4  5\n",
      "G1  hg19  T1  chr1  1111  -\n",
      "G2  hg19  T2  chr2  2222  -\n",
      "G3  hg19  T3  chr3  3333  -\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# annobj is the equivalent of the summarized experiments in R. \n",
    "# It contains \n",
    "# - annobj.X - counts\n",
    "# - annobj.var - data.frame annotation object for the columns of the count matrix (in our case Tissue_Sample)\n",
    "# - annobj.obs - data.frame annotation object for the rows of the count matrix (= features that we measure, in our case Genes)\n",
    "# \n",
    "annobj = p.load_count_matrix(\"data.csv\", sep=\",\", varanno=\"anno_var.csv\", obsanno=\"anno_obs.csv\")\n",
    "p.print_anndata(annobj)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Save data to h2ad format"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "... storing '1' as categorical\n",
      "... storing '5' as categorical\n",
      "... storing '1' as categorical\n",
      "... storing '3' as categorical\n",
      "... storing '4' as categorical\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "'/data/nasif12/home_if12/giurgiu/rep_gagneur/rep/notebooks/preprocessing/tmp1545316680.h5ad'"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "name = p.save(annobj)\n",
    "name"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 2. Prepare training sample:\n",
    "- select genes and tissue_samples for training\n",
    "- slice the anndata object\n",
    "- compute the cross tissue matrix pair (X,Y)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "      Gender   Tissue Individual Technology\n",
      "T1_s1      F  Tissue1    Sample1     rnaseq\n",
      "T2_s2      M  Tissue2    Sample2     rnaseq\n",
      "T3_s1      F  Tissue3    Sample1     rnaseq\n",
      "T4_s2      M  Tissue4    Sample2     rnaseq\n",
      "T5_s1      M  Tissue5    Sample1     rnaseq\n",
      "T6_s1      M  Tissue6    Sample1     rnaseq\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "(['T2_s2', 'T6_s1', 'T5_s1', 'T4_s2'], ['G1', 'G2'])"
      ]
     },
     "execution_count": 65,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# filtering by different values in anndata.var and anndata.obs\n",
    "# stored in a json format by specifying the {column over we would like to filter : array of admitted values}\n",
    "# in this example we filter the anndata.var by gender (M) and anndata.obs should contain only G1 and G2\n",
    "(var, obs) = p.filter_anndata_by_value(annobj, {a.VAR: {'Gender': ['M']}, a.OBS: {0: ['G1', 'G2']}})\n",
    "\n",
    "# these are the rows and columns selected for the training set\n",
    "(var,obs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'Sample1': ['T6_s1', 'T5_s1'], 'Sample2': ['T2_s2', 'T4_s2']}"
      ]
     },
     "execution_count": 67,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# since we want to impute gene expression across different tissues, we\n",
    "# would like to compute the cross tissue pairs per sample and not across samples\n",
    "\n",
    "# for this purpose we can use the group_by(df, column, index_subset) \n",
    "# function to group tissues by sample\n",
    "\n",
    "dict_samples_var = p.group_by(annobj.var, 'Individual', var)\n",
    "dict_samples_var"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "4.3 Slice anndata\n",
      "\n",
      "anndata.X\n",
      "[[20.  5. 30.  6. 10. 40.]\n",
      " [10.  5. 20.  6.  5. 30.]]\n",
      "anndata.var\n",
      "Index(['T2_s2', 'T5_s1', 'T3_s1', 'T6_s1', 'T1_s1', 'T4_s2'], dtype='object')\n",
      "      Gender   Tissue Individual Technology\n",
      "T2_s2      M  Tissue2    Sample2     rnaseq\n",
      "T5_s1      M  Tissue5    Sample1     rnaseq\n",
      "T3_s1      F  Tissue3    Sample1     rnaseq\n",
      "T6_s1      M  Tissue6    Sample1     rnaseq\n",
      "T1_s1      F  Tissue1    Sample1     rnaseq\n",
      "T4_s2      M  Tissue4    Sample2     rnaseq\n",
      "anndata.obs\n",
      "Index(['G1', 'G2'], dtype='object')\n",
      "       1   2     3     4  5\n",
      "G1  hg19  T1  chr1  1111  -\n",
      "G2  hg19  T2  chr2  2222  -\n",
      "\n",
      "4.4 Build the two matrices X and Y\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# compute cross tissue matrix pair (X, Y)\n",
    "(X, Y) = p.rnaseq_cross_tissue(annobj, individuals=['Sample1','Sample2'], gene_ids=obs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>G1</th>\n",
       "      <th>G2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>T1_s1_T6_s1</th>\n",
       "      <td>10</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T1_s1_T3_s1</th>\n",
       "      <td>10</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T1_s1_T5_s1</th>\n",
       "      <td>10</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T6_s1_T1_s1</th>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T6_s1_T3_s1</th>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T6_s1_T5_s1</th>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T3_s1_T1_s1</th>\n",
       "      <td>30</td>\n",
       "      <td>20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T3_s1_T6_s1</th>\n",
       "      <td>30</td>\n",
       "      <td>20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T3_s1_T5_s1</th>\n",
       "      <td>30</td>\n",
       "      <td>20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T5_s1_T1_s1</th>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T5_s1_T6_s1</th>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T5_s1_T3_s1</th>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T2_s2_T4_s2</th>\n",
       "      <td>20</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T4_s2_T2_s2</th>\n",
       "      <td>40</td>\n",
       "      <td>30</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             G1  G2\n",
       "T1_s1_T6_s1  10   5\n",
       "T1_s1_T3_s1  10   5\n",
       "T1_s1_T5_s1  10   5\n",
       "T6_s1_T1_s1   6   6\n",
       "T6_s1_T3_s1   6   6\n",
       "T6_s1_T5_s1   6   6\n",
       "T3_s1_T1_s1  30  20\n",
       "T3_s1_T6_s1  30  20\n",
       "T3_s1_T5_s1  30  20\n",
       "T5_s1_T1_s1   5   5\n",
       "T5_s1_T6_s1   5   5\n",
       "T5_s1_T3_s1   5   5\n",
       "T2_s2_T4_s2  20  10\n",
       "T4_s2_T2_s2  40  30"
      ]
     },
     "execution_count": 73,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# train matrix\n",
    "X"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>G1</th>\n",
       "      <th>G2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>T1_s1_T6_s1</th>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T1_s1_T3_s1</th>\n",
       "      <td>30</td>\n",
       "      <td>20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T1_s1_T5_s1</th>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T6_s1_T1_s1</th>\n",
       "      <td>10</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T6_s1_T3_s1</th>\n",
       "      <td>30</td>\n",
       "      <td>20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T6_s1_T5_s1</th>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T3_s1_T1_s1</th>\n",
       "      <td>10</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T3_s1_T6_s1</th>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T3_s1_T5_s1</th>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T5_s1_T1_s1</th>\n",
       "      <td>10</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T5_s1_T6_s1</th>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T5_s1_T3_s1</th>\n",
       "      <td>30</td>\n",
       "      <td>20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T2_s2_T4_s2</th>\n",
       "      <td>40</td>\n",
       "      <td>30</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T4_s2_T2_s2</th>\n",
       "      <td>20</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             G1  G2\n",
       "T1_s1_T6_s1   6   6\n",
       "T1_s1_T3_s1  30  20\n",
       "T1_s1_T5_s1   5   5\n",
       "T6_s1_T1_s1  10   5\n",
       "T6_s1_T3_s1  30  20\n",
       "T6_s1_T5_s1   5   5\n",
       "T3_s1_T1_s1  10   5\n",
       "T3_s1_T6_s1   6   6\n",
       "T3_s1_T5_s1   5   5\n",
       "T5_s1_T1_s1  10   5\n",
       "T5_s1_T6_s1   6   6\n",
       "T5_s1_T3_s1  30  20\n",
       "T2_s2_T4_s2  40  30\n",
       "T4_s2_T2_s2  20  10"
      ]
     },
     "execution_count": 74,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# label matrix\n",
    "Y"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3. Create toy large dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: numpy in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (1.15.4)\n",
      "Requirement already satisfied: matplotlib in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (3.0.2)\n",
      "Requirement already satisfied: cycler>=0.10 in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (from matplotlib) (0.10.0)\n",
      "Requirement already satisfied: numpy>=1.10.0 in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (from matplotlib) (1.15.4)\n",
      "Requirement already satisfied: pyparsing!=2.0.4,!=2.1.2,!=2.1.6,>=2.0.1 in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (from matplotlib) (2.3.0)\n",
      "Requirement already satisfied: kiwisolver>=1.0.1 in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (from matplotlib) (1.0.1)\n",
      "Requirement already satisfied: python-dateutil>=2.1 in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (from matplotlib) (2.7.5)\n",
      "Requirement already satisfied: six in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (from cycler>=0.10->matplotlib) (1.12.0)\n",
      "Requirement already satisfied: setuptools in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (from kiwisolver>=1.0.1->matplotlib) (40.6.3)\n"
     ]
    }
   ],
   "source": [
    "!pip install numpy\n",
    "!pip install matplotlib\n",
    "import random\n",
    "import math\n",
    "import numpy.random as nr\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "from scipy import stats\n",
    "\n",
    "import anndata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Gene_19991',\n",
       " 'Gene_19992',\n",
       " 'Gene_19993',\n",
       " 'Gene_19994',\n",
       " 'Gene_19995',\n",
       " 'Gene_19996',\n",
       " 'Gene_19997',\n",
       " 'Gene_19998',\n",
       " 'Gene_19999',\n",
       " 'Gene_20000']"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# genes\n",
    "genes = list(map(lambda x: \"Gene_\" + str(x) , list(range(1, 20001))))\n",
    "genes[19990:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([4223, 1803, 3079, 1555, 3025, 2965, 2215, 3119, 8627, 1596])"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "raw_counts = nr.negative_binomial(3,0.001,size=20000)\n",
    "# sns.distplot(raw_counts)\n",
    "raw_counts[:10]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "# tissues\n",
    "tissues = ['Muscle', 'Blood', 'Skin', 'Lung', 'Heart', 'Stomatch', 'Colon', 'Testis', 'Brain', 'Liver']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([1.43544583, 2.17944366, 1.84529393, 2.3820589 , 6.49733261,\n",
       "       3.90651559, 5.98537936, 8.64013829, 2.92092658, 2.4304256 ])"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# simulate how many tissues are collected per patient\n",
    "distribution_tissues = nr.uniform(1,len(tissues),1000)\n",
    "distribution_tissues[:10]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "# generate samples - raw counts per gene per tissue\n",
    "frames = []\n",
    "for i in range(1,10): # patients\n",
    "  \n",
    "  # choose tissues\n",
    "  how_many_tissues = math.floor(distribution_tissues[random.randint(1,len(distribution_tissues))])\n",
    "  random.shuffle(tissues)\n",
    "  t_set = tissues[:how_many_tissues]  \n",
    "  \n",
    "  # initialize count matrix\n",
    "  counts = pd.DataFrame(columns=genes,index=map(lambda x: \"Patient_\" + str(i) + \"_\" + str(x), t_set))\n",
    "\n",
    "  for j in range(0,len(t_set)):\n",
    "    \n",
    "    col_name = \"Patient_\" + str(i) + \"_\" + t_set[j]\n",
    "    \n",
    "    # generate raw counts per tissue per set of genes\n",
    "    raw_counts = nr.negative_binomial(3,0.001,size=20000)\n",
    "    counts.loc[col_name] = raw_counts.reshape(1,20000)\n",
    "    \n",
    "  # append expression per tissue for the patient \n",
    "  frames.append(counts)\n",
    "      \n",
    "\n",
    "df_big = pd.concat(frames)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(20000, 42)"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# transpose the matrix to be genes x patient_tissue\n",
    "count_matrix = df_big.transpose()\n",
    "count_matrix.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Individual</th>\n",
       "      <th>Tissue</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Patient_1_Liver</th>\n",
       "      <td>Patient_1</td>\n",
       "      <td>Liver</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Patient_1_Blood</th>\n",
       "      <td>Patient_1</td>\n",
       "      <td>Blood</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Patient_2_Testis</th>\n",
       "      <td>Patient_2</td>\n",
       "      <td>Testis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Patient_2_Brain</th>\n",
       "      <td>Patient_2</td>\n",
       "      <td>Brain</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Patient_2_Liver</th>\n",
       "      <td>Patient_2</td>\n",
       "      <td>Liver</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Patient_3_Colon</th>\n",
       "      <td>Patient_3</td>\n",
       "      <td>Colon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Patient_3_Heart</th>\n",
       "      <td>Patient_3</td>\n",
       "      <td>Heart</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Patient_3_Blood</th>\n",
       "      <td>Patient_3</td>\n",
       "      <td>Blood</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Patient_3_Muscle</th>\n",
       "      <td>Patient_3</td>\n",
       "      <td>Muscle</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Patient_3_Skin</th>\n",
       "      <td>Patient_3</td>\n",
       "      <td>Skin</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 Individual  Tissue\n",
       "Patient_1_Liver   Patient_1   Liver\n",
       "Patient_1_Blood   Patient_1   Blood\n",
       "Patient_2_Testis  Patient_2  Testis\n",
       "Patient_2_Brain   Patient_2   Brain\n",
       "Patient_2_Liver   Patient_2   Liver\n",
       "Patient_3_Colon   Patient_3   Colon\n",
       "Patient_3_Heart   Patient_3   Heart\n",
       "Patient_3_Blood   Patient_3   Blood\n",
       "Patient_3_Muscle  Patient_3  Muscle\n",
       "Patient_3_Skin    Patient_3    Skin"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# create annotation dataframes\n",
    "# column = 0 its the index! be careful\n",
    "samples_anno = pd.DataFrame(index=count_matrix.columns,columns=['Individual','Tissue'])\n",
    "\n",
    "for i in range(0,samples_anno.shape[0]):\n",
    "  sample = samples_anno.index[i]\n",
    "  [pat,number,tissue] = sample.split(\"_\")\n",
    "  \n",
    "  samples_anno.at[sample, 'Individual'] = pat + \"_\" + str(number)\n",
    "  samples_anno.at[sample, 'Tissue'] = tissue\n",
    "\n",
    "samples_anno.drop_duplicates().shape\n",
    "samples_anno[:10]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(20000, 0)"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes_anno = pd.DataFrame(index=count_matrix.index)\n",
    "genes_anno.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "anndata.X\n",
      "[[5954. 3762. 2766. ... 2754. 2417. 3411.]\n",
      " [2390.  966. 6234. ... 1869. 1225. 2182.]\n",
      " [4451.  974. 3649. ...  571. 6226. 1150.]\n",
      " ...\n",
      " [2644. 2462. 5296. ... 5027. 3532. 2245.]\n",
      " [1821. 3628.  380. ... 3414. 1150. 4381.]\n",
      " [1416. 1910. 1993. ... 1053. 1768. 4921.]]\n",
      "anndata.var\n",
      "Index(['Patient_1_Liver', 'Patient_1_Blood', 'Patient_2_Testis',\n",
      "       'Patient_2_Brain', 'Patient_2_Liver', 'Patient_3_Colon',\n",
      "       'Patient_3_Heart', 'Patient_3_Blood', 'Patient_3_Muscle',\n",
      "       'Patient_3_Skin', 'Patient_4_Heart', 'Patient_4_Colon',\n",
      "       'Patient_4_Liver', 'Patient_4_Stomatch', 'Patient_4_Blood',\n",
      "       'Patient_5_Muscle', 'Patient_5_Colon', 'Patient_5_Skin',\n",
      "       'Patient_5_Brain', 'Patient_5_Lung', 'Patient_5_Testis',\n",
      "       'Patient_5_Blood', 'Patient_5_Stomatch', 'Patient_5_Heart',\n",
      "       'Patient_6_Brain', 'Patient_6_Blood', 'Patient_6_Muscle',\n",
      "       'Patient_6_Stomatch', 'Patient_7_Skin', 'Patient_7_Heart',\n",
      "       'Patient_7_Blood', 'Patient_7_Muscle', 'Patient_7_Liver',\n",
      "       'Patient_7_Testis', 'Patient_7_Brain', 'Patient_7_Lung',\n",
      "       'Patient_8_Testis', 'Patient_8_Muscle', 'Patient_8_Blood',\n",
      "       'Patient_9_Lung', 'Patient_9_Muscle', 'Patient_9_Stomatch'],\n",
      "      dtype='object')\n",
      "                   Individual    Tissue\n",
      "Patient_1_Liver     Patient_1     Liver\n",
      "Patient_1_Blood     Patient_1     Blood\n",
      "Patient_2_Testis    Patient_2    Testis\n",
      "Patient_2_Brain     Patient_2     Brain\n",
      "Patient_2_Liver     Patient_2     Liver\n",
      "Patient_3_Colon     Patient_3     Colon\n",
      "Patient_3_Heart     Patient_3     Heart\n",
      "Patient_3_Blood     Patient_3     Blood\n",
      "Patient_3_Muscle    Patient_3    Muscle\n",
      "Patient_3_Skin      Patient_3      Skin\n",
      "Patient_4_Heart     Patient_4     Heart\n",
      "Patient_4_Colon     Patient_4     Colon\n",
      "Patient_4_Liver     Patient_4     Liver\n",
      "Patient_4_Stomatch  Patient_4  Stomatch\n",
      "Patient_4_Blood     Patient_4     Blood\n",
      "Patient_5_Muscle    Patient_5    Muscle\n",
      "Patient_5_Colon     Patient_5     Colon\n",
      "Patient_5_Skin      Patient_5      Skin\n",
      "Patient_5_Brain     Patient_5     Brain\n",
      "Patient_5_Lung      Patient_5      Lung\n",
      "Patient_5_Testis    Patient_5    Testis\n",
      "Patient_5_Blood     Patient_5     Blood\n",
      "Patient_5_Stomatch  Patient_5  Stomatch\n",
      "Patient_5_Heart     Patient_5     Heart\n",
      "Patient_6_Brain     Patient_6     Brain\n",
      "Patient_6_Blood     Patient_6     Blood\n",
      "Patient_6_Muscle    Patient_6    Muscle\n",
      "Patient_6_Stomatch  Patient_6  Stomatch\n",
      "Patient_7_Skin      Patient_7      Skin\n",
      "Patient_7_Heart     Patient_7     Heart\n",
      "Patient_7_Blood     Patient_7     Blood\n",
      "Patient_7_Muscle    Patient_7    Muscle\n",
      "Patient_7_Liver     Patient_7     Liver\n",
      "Patient_7_Testis    Patient_7    Testis\n",
      "Patient_7_Brain     Patient_7     Brain\n",
      "Patient_7_Lung      Patient_7      Lung\n",
      "Patient_8_Testis    Patient_8    Testis\n",
      "Patient_8_Muscle    Patient_8    Muscle\n",
      "Patient_8_Blood     Patient_8     Blood\n",
      "Patient_9_Lung      Patient_9      Lung\n",
      "Patient_9_Muscle    Patient_9    Muscle\n",
      "Patient_9_Stomatch  Patient_9  Stomatch\n",
      "anndata.obs\n",
      "Index(['Gene_1', 'Gene_2', 'Gene_3', 'Gene_4', 'Gene_5', 'Gene_6', 'Gene_7',\n",
      "       'Gene_8', 'Gene_9', 'Gene_10',\n",
      "       ...\n",
      "       'Gene_19991', 'Gene_19992', 'Gene_19993', 'Gene_19994', 'Gene_19995',\n",
      "       'Gene_19996', 'Gene_19997', 'Gene_19998', 'Gene_19999', 'Gene_20000'],\n",
      "      dtype='object', length=20000)\n",
      "Empty DataFrame\n",
      "Columns: []\n",
      "Index: [Gene_1, Gene_2, Gene_3, Gene_4, Gene_5, Gene_6, Gene_7, Gene_8, Gene_9, Gene_10, Gene_11, Gene_12, Gene_13, Gene_14, Gene_15, Gene_16, Gene_17, Gene_18, Gene_19, Gene_20, Gene_21, Gene_22, Gene_23, Gene_24, Gene_25, Gene_26, Gene_27, Gene_28, Gene_29, Gene_30, Gene_31, Gene_32, Gene_33, Gene_34, Gene_35, Gene_36, Gene_37, Gene_38, Gene_39, Gene_40, Gene_41, Gene_42, Gene_43, Gene_44, Gene_45, Gene_46, Gene_47, Gene_48, Gene_49, Gene_50, Gene_51, Gene_52, Gene_53, Gene_54, Gene_55, Gene_56, Gene_57, Gene_58, Gene_59, Gene_60, Gene_61, Gene_62, Gene_63, Gene_64, Gene_65, Gene_66, Gene_67, Gene_68, Gene_69, Gene_70, Gene_71, Gene_72, Gene_73, Gene_74, Gene_75, Gene_76, Gene_77, Gene_78, Gene_79, Gene_80, Gene_81, Gene_82, Gene_83, Gene_84, Gene_85, Gene_86, Gene_87, Gene_88, Gene_89, Gene_90, Gene_91, Gene_92, Gene_93, Gene_94, Gene_95, Gene_96, Gene_97, Gene_98, Gene_99, Gene_100, ...]\n",
      "\n",
      "[20000 rows x 0 columns]\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# create annobj\n",
    "annobj = anndata.AnnData(count_matrix)\n",
    "annobj.var = samples_anno\n",
    "annobj.var_names = list(samples_anno.index)\n",
    "annobj.obs = genes_anno\n",
    "annobj.obs_names = list(genes_anno.index)\n",
    "\n",
    "p.print_anndata(annobj)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Blood', 'Brain', 'Heart', 'Muscle', 'Testis']"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# filter using a subset of the tissues\n",
    "random.shuffle(tissues)\n",
    "subset_tissue = tissues[:5]\n",
    "subset_tissue"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'var': {'Tissue': ['Blood', 'Brain', 'Heart', 'Muscle', 'Testis']}}"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# create filtering json\n",
    "json_filter = {a.VAR: {'Tissue':subset_tissue}}\n",
    "# json_filter[a.OBS] = {0:genes[:10]}\n",
    "json_filter"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                   Individual    Tissue\n",
      "Patient_1_Liver     Patient_1     Liver\n",
      "Patient_1_Blood     Patient_1     Blood\n",
      "Patient_2_Testis    Patient_2    Testis\n",
      "Patient_2_Brain     Patient_2     Brain\n",
      "Patient_2_Liver     Patient_2     Liver\n",
      "Patient_3_Colon     Patient_3     Colon\n",
      "Patient_3_Heart     Patient_3     Heart\n",
      "Patient_3_Blood     Patient_3     Blood\n",
      "Patient_3_Muscle    Patient_3    Muscle\n",
      "Patient_3_Skin      Patient_3      Skin\n",
      "Patient_4_Heart     Patient_4     Heart\n",
      "Patient_4_Colon     Patient_4     Colon\n",
      "Patient_4_Liver     Patient_4     Liver\n",
      "Patient_4_Stomatch  Patient_4  Stomatch\n",
      "Patient_4_Blood     Patient_4     Blood\n",
      "Patient_5_Muscle    Patient_5    Muscle\n",
      "Patient_5_Colon     Patient_5     Colon\n",
      "Patient_5_Skin      Patient_5      Skin\n",
      "Patient_5_Brain     Patient_5     Brain\n",
      "Patient_5_Lung      Patient_5      Lung\n",
      "Patient_5_Testis    Patient_5    Testis\n",
      "Patient_5_Blood     Patient_5     Blood\n",
      "Patient_5_Stomatch  Patient_5  Stomatch\n",
      "Patient_5_Heart     Patient_5     Heart\n",
      "Patient_6_Brain     Patient_6     Brain\n",
      "Patient_6_Blood     Patient_6     Blood\n",
      "Patient_6_Muscle    Patient_6    Muscle\n",
      "Patient_6_Stomatch  Patient_6  Stomatch\n",
      "Patient_7_Skin      Patient_7      Skin\n",
      "Patient_7_Heart     Patient_7     Heart\n",
      "Patient_7_Blood     Patient_7     Blood\n",
      "Patient_7_Muscle    Patient_7    Muscle\n",
      "Patient_7_Liver     Patient_7     Liver\n",
      "Patient_7_Testis    Patient_7    Testis\n",
      "Patient_7_Brain     Patient_7     Brain\n",
      "Patient_7_Lung      Patient_7      Lung\n",
      "Patient_8_Testis    Patient_8    Testis\n",
      "Patient_8_Muscle    Patient_8    Muscle\n",
      "Patient_8_Blood     Patient_8     Blood\n",
      "Patient_9_Lung      Patient_9      Lung\n",
      "Patient_9_Muscle    Patient_9    Muscle\n",
      "Patient_9_Stomatch  Patient_9  Stomatch\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "['Patient_6_Blood',\n",
       " 'Patient_7_Brain',\n",
       " 'Patient_7_Heart',\n",
       " 'Patient_4_Heart',\n",
       " 'Patient_2_Testis',\n",
       " 'Patient_6_Brain',\n",
       " 'Patient_5_Muscle',\n",
       " 'Patient_4_Blood',\n",
       " 'Patient_3_Blood',\n",
       " 'Patient_5_Brain']"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "(var, obs) = p.filter_anndata_by_value(annobj, json_filter)\n",
    "var[:10]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "4.3 Slice anndata\n",
      "\n",
      "anndata.X\n",
      "[[3762. 2766. 5954. 4545. 5159.]\n",
      " [ 966. 6234. 2390. 3354. 1559.]\n",
      " [ 974. 3649. 4451. 4068. 2982.]\n",
      " ...\n",
      " [2462. 5296. 2644. 1339. 1233.]\n",
      " [3628.  380. 1821. 3780. 1911.]\n",
      " [1910. 1993. 1416. 2386. 1513.]]\n",
      "anndata.var\n",
      "Index(['Patient_1_Blood', 'Patient_2_Testis', 'Patient_1_Liver',\n",
      "       'Patient_2_Brain', 'Patient_2_Liver'],\n",
      "      dtype='object')\n",
      "                 Individual  Tissue\n",
      "Patient_1_Blood   Patient_1   Blood\n",
      "Patient_2_Testis  Patient_2  Testis\n",
      "Patient_1_Liver   Patient_1   Liver\n",
      "Patient_2_Brain   Patient_2   Brain\n",
      "Patient_2_Liver   Patient_2   Liver\n",
      "anndata.obs\n",
      "Index(['Gene_1', 'Gene_2', 'Gene_3', 'Gene_4', 'Gene_5', 'Gene_6', 'Gene_7',\n",
      "       'Gene_8', 'Gene_9', 'Gene_10',\n",
      "       ...\n",
      "       'Gene_19991', 'Gene_19992', 'Gene_19993', 'Gene_19994', 'Gene_19995',\n",
      "       'Gene_19996', 'Gene_19997', 'Gene_19998', 'Gene_19999', 'Gene_20000'],\n",
      "      dtype='object', length=20000)\n",
      "Empty DataFrameView\n",
      "Columns: []\n",
      "Index: [Gene_1, Gene_2, Gene_3, Gene_4, Gene_5, Gene_6, Gene_7, Gene_8, Gene_9, Gene_10, Gene_11, Gene_12, Gene_13, Gene_14, Gene_15, Gene_16, Gene_17, Gene_18, Gene_19, Gene_20, Gene_21, Gene_22, Gene_23, Gene_24, Gene_25, Gene_26, Gene_27, Gene_28, Gene_29, Gene_30, Gene_31, Gene_32, Gene_33, Gene_34, Gene_35, Gene_36, Gene_37, Gene_38, Gene_39, Gene_40, Gene_41, Gene_42, Gene_43, Gene_44, Gene_45, Gene_46, Gene_47, Gene_48, Gene_49, Gene_50, Gene_51, Gene_52, Gene_53, Gene_54, Gene_55, Gene_56, Gene_57, Gene_58, Gene_59, Gene_60, Gene_61, Gene_62, Gene_63, Gene_64, Gene_65, Gene_66, Gene_67, Gene_68, Gene_69, Gene_70, Gene_71, Gene_72, Gene_73, Gene_74, Gene_75, Gene_76, Gene_77, Gene_78, Gene_79, Gene_80, Gene_81, Gene_82, Gene_83, Gene_84, Gene_85, Gene_86, Gene_87, Gene_88, Gene_89, Gene_90, Gene_91, Gene_92, Gene_93, Gene_94, Gene_95, Gene_96, Gene_97, Gene_98, Gene_99, Gene_100, ...]\n",
      "\n",
      "[20000 rows x 0 columns]\n",
      "\n",
      "4.4 Build the two matrices X and Y\n",
      "\n",
      "(8, 20000)\n",
      "(8, 20000)\n"
     ]
    }
   ],
   "source": [
    "# slice  + compute all pairs\n",
    "(X, Y) = p.rnaseq_cross_tissue(annobj, individuals=['Patient_1','Patient_2'], gene_ids=obs)\n",
    "print(X.shape)\n",
    "print(Y.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene_1</th>\n",
       "      <th>Gene_2</th>\n",
       "      <th>Gene_3</th>\n",
       "      <th>Gene_4</th>\n",
       "      <th>Gene_5</th>\n",
       "      <th>Gene_6</th>\n",
       "      <th>Gene_7</th>\n",
       "      <th>Gene_8</th>\n",
       "      <th>Gene_9</th>\n",
       "      <th>Gene_10</th>\n",
       "      <th>...</th>\n",
       "      <th>Gene_19991</th>\n",
       "      <th>Gene_19992</th>\n",
       "      <th>Gene_19993</th>\n",
       "      <th>Gene_19994</th>\n",
       "      <th>Gene_19995</th>\n",
       "      <th>Gene_19996</th>\n",
       "      <th>Gene_19997</th>\n",
       "      <th>Gene_19998</th>\n",
       "      <th>Gene_19999</th>\n",
       "      <th>Gene_20000</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Patient_1_Blood_Patient_1_Liver</th>\n",
       "      <td>3762</td>\n",
       "      <td>966</td>\n",
       "      <td>974</td>\n",
       "      <td>2223</td>\n",
       "      <td>4763</td>\n",
       "      <td>1724</td>\n",
       "      <td>5866</td>\n",
       "      <td>2025</td>\n",
       "      <td>4844</td>\n",
       "      <td>5496</td>\n",
       "      <td>...</td>\n",
       "      <td>4696</td>\n",
       "      <td>1132</td>\n",
       "      <td>468</td>\n",
       "      <td>775</td>\n",
       "      <td>6098</td>\n",
       "      <td>1150</td>\n",
       "      <td>7856</td>\n",
       "      <td>2462</td>\n",
       "      <td>3628</td>\n",
       "      <td>1910</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Patient_1_Liver_Patient_1_Blood</th>\n",
       "      <td>5954</td>\n",
       "      <td>2390</td>\n",
       "      <td>4451</td>\n",
       "      <td>2333</td>\n",
       "      <td>3371</td>\n",
       "      <td>3715</td>\n",
       "      <td>2096</td>\n",
       "      <td>2483</td>\n",
       "      <td>2641</td>\n",
       "      <td>7564</td>\n",
       "      <td>...</td>\n",
       "      <td>3967</td>\n",
       "      <td>1500</td>\n",
       "      <td>1003</td>\n",
       "      <td>6824</td>\n",
       "      <td>2320</td>\n",
       "      <td>7065</td>\n",
       "      <td>2660</td>\n",
       "      <td>2644</td>\n",
       "      <td>1821</td>\n",
       "      <td>1416</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Patient_2_Testis_Patient_2_Brain</th>\n",
       "      <td>2766</td>\n",
       "      <td>6234</td>\n",
       "      <td>3649</td>\n",
       "      <td>1282</td>\n",
       "      <td>8075</td>\n",
       "      <td>5825</td>\n",
       "      <td>4944</td>\n",
       "      <td>108</td>\n",
       "      <td>1688</td>\n",
       "      <td>1189</td>\n",
       "      <td>...</td>\n",
       "      <td>1150</td>\n",
       "      <td>6609</td>\n",
       "      <td>1465</td>\n",
       "      <td>570</td>\n",
       "      <td>1969</td>\n",
       "      <td>4336</td>\n",
       "      <td>2644</td>\n",
       "      <td>5296</td>\n",
       "      <td>380</td>\n",
       "      <td>1993</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Patient_2_Testis_Patient_2_Liver</th>\n",
       "      <td>2766</td>\n",
       "      <td>6234</td>\n",
       "      <td>3649</td>\n",
       "      <td>1282</td>\n",
       "      <td>8075</td>\n",
       "      <td>5825</td>\n",
       "      <td>4944</td>\n",
       "      <td>108</td>\n",
       "      <td>1688</td>\n",
       "      <td>1189</td>\n",
       "      <td>...</td>\n",
       "      <td>1150</td>\n",
       "      <td>6609</td>\n",
       "      <td>1465</td>\n",
       "      <td>570</td>\n",
       "      <td>1969</td>\n",
       "      <td>4336</td>\n",
       "      <td>2644</td>\n",
       "      <td>5296</td>\n",
       "      <td>380</td>\n",
       "      <td>1993</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Patient_2_Brain_Patient_2_Testis</th>\n",
       "      <td>4545</td>\n",
       "      <td>3354</td>\n",
       "      <td>4068</td>\n",
       "      <td>4765</td>\n",
       "      <td>2463</td>\n",
       "      <td>4179</td>\n",
       "      <td>2773</td>\n",
       "      <td>3208</td>\n",
       "      <td>2042</td>\n",
       "      <td>2168</td>\n",
       "      <td>...</td>\n",
       "      <td>1524</td>\n",
       "      <td>6672</td>\n",
       "      <td>2991</td>\n",
       "      <td>2425</td>\n",
       "      <td>2158</td>\n",
       "      <td>4529</td>\n",
       "      <td>1961</td>\n",
       "      <td>1339</td>\n",
       "      <td>3780</td>\n",
       "      <td>2386</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Patient_2_Brain_Patient_2_Liver</th>\n",
       "      <td>4545</td>\n",
       "      <td>3354</td>\n",
       "      <td>4068</td>\n",
       "      <td>4765</td>\n",
       "      <td>2463</td>\n",
       "      <td>4179</td>\n",
       "      <td>2773</td>\n",
       "      <td>3208</td>\n",
       "      <td>2042</td>\n",
       "      <td>2168</td>\n",
       "      <td>...</td>\n",
       "      <td>1524</td>\n",
       "      <td>6672</td>\n",
       "      <td>2991</td>\n",
       "      <td>2425</td>\n",
       "      <td>2158</td>\n",
       "      <td>4529</td>\n",
       "      <td>1961</td>\n",
       "      <td>1339</td>\n",
       "      <td>3780</td>\n",
       "      <td>2386</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Patient_2_Liver_Patient_2_Testis</th>\n",
       "      <td>5159</td>\n",
       "      <td>1559</td>\n",
       "      <td>2982</td>\n",
       "      <td>6594</td>\n",
       "      <td>2101</td>\n",
       "      <td>3903</td>\n",
       "      <td>1786</td>\n",
       "      <td>3455</td>\n",
       "      <td>1685</td>\n",
       "      <td>1449</td>\n",
       "      <td>...</td>\n",
       "      <td>6190</td>\n",
       "      <td>2781</td>\n",
       "      <td>8495</td>\n",
       "      <td>1543</td>\n",
       "      <td>2569</td>\n",
       "      <td>2710</td>\n",
       "      <td>3536</td>\n",
       "      <td>1233</td>\n",
       "      <td>1911</td>\n",
       "      <td>1513</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Patient_2_Liver_Patient_2_Brain</th>\n",
       "      <td>5159</td>\n",
       "      <td>1559</td>\n",
       "      <td>2982</td>\n",
       "      <td>6594</td>\n",
       "      <td>2101</td>\n",
       "      <td>3903</td>\n",
       "      <td>1786</td>\n",
       "      <td>3455</td>\n",
       "      <td>1685</td>\n",
       "      <td>1449</td>\n",
       "      <td>...</td>\n",
       "      <td>6190</td>\n",
       "      <td>2781</td>\n",
       "      <td>8495</td>\n",
       "      <td>1543</td>\n",
       "      <td>2569</td>\n",
       "      <td>2710</td>\n",
       "      <td>3536</td>\n",
       "      <td>1233</td>\n",
       "      <td>1911</td>\n",
       "      <td>1513</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Patient_3_Heart_Patient_3_Blood</th>\n",
       "      <td>1816</td>\n",
       "      <td>2330</td>\n",
       "      <td>490</td>\n",
       "      <td>2100</td>\n",
       "      <td>4136</td>\n",
       "      <td>3471</td>\n",
       "      <td>2373</td>\n",
       "      <td>1224</td>\n",
       "      <td>2062</td>\n",
       "      <td>6767</td>\n",
       "      <td>...</td>\n",
       "      <td>2462</td>\n",
       "      <td>2730</td>\n",
       "      <td>1484</td>\n",
       "      <td>3102</td>\n",
       "      <td>3307</td>\n",
       "      <td>1028</td>\n",
       "      <td>1685</td>\n",
       "      <td>2926</td>\n",
       "      <td>345</td>\n",
       "      <td>1881</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Patient_3_Heart_Patient_3_Muscle</th>\n",
       "      <td>1816</td>\n",
       "      <td>2330</td>\n",
       "      <td>490</td>\n",
       "      <td>2100</td>\n",
       "      <td>4136</td>\n",
       "      <td>3471</td>\n",
       "      <td>2373</td>\n",
       "      <td>1224</td>\n",
       "      <td>2062</td>\n",
       "      <td>6767</td>\n",
       "      <td>...</td>\n",
       "      <td>2462</td>\n",
       "      <td>2730</td>\n",
       "      <td>1484</td>\n",
       "      <td>3102</td>\n",
       "      <td>3307</td>\n",
       "      <td>1028</td>\n",
       "      <td>1685</td>\n",
       "      <td>2926</td>\n",
       "      <td>345</td>\n",
       "      <td>1881</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10 rows  20000 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                  Gene_1  Gene_2  Gene_3  Gene_4  Gene_5  \\\n",
       "Patient_1_Blood_Patient_1_Liver     3762     966     974    2223    4763   \n",
       "Patient_1_Liver_Patient_1_Blood     5954    2390    4451    2333    3371   \n",
       "Patient_2_Testis_Patient_2_Brain    2766    6234    3649    1282    8075   \n",
       "Patient_2_Testis_Patient_2_Liver    2766    6234    3649    1282    8075   \n",
       "Patient_2_Brain_Patient_2_Testis    4545    3354    4068    4765    2463   \n",
       "Patient_2_Brain_Patient_2_Liver     4545    3354    4068    4765    2463   \n",
       "Patient_2_Liver_Patient_2_Testis    5159    1559    2982    6594    2101   \n",
       "Patient_2_Liver_Patient_2_Brain     5159    1559    2982    6594    2101   \n",
       "Patient_3_Heart_Patient_3_Blood     1816    2330     490    2100    4136   \n",
       "Patient_3_Heart_Patient_3_Muscle    1816    2330     490    2100    4136   \n",
       "\n",
       "                                  Gene_6  Gene_7  Gene_8  Gene_9  Gene_10  \\\n",
       "Patient_1_Blood_Patient_1_Liver     1724    5866    2025    4844     5496   \n",
       "Patient_1_Liver_Patient_1_Blood     3715    2096    2483    2641     7564   \n",
       "Patient_2_Testis_Patient_2_Brain    5825    4944     108    1688     1189   \n",
       "Patient_2_Testis_Patient_2_Liver    5825    4944     108    1688     1189   \n",
       "Patient_2_Brain_Patient_2_Testis    4179    2773    3208    2042     2168   \n",
       "Patient_2_Brain_Patient_2_Liver     4179    2773    3208    2042     2168   \n",
       "Patient_2_Liver_Patient_2_Testis    3903    1786    3455    1685     1449   \n",
       "Patient_2_Liver_Patient_2_Brain     3903    1786    3455    1685     1449   \n",
       "Patient_3_Heart_Patient_3_Blood     3471    2373    1224    2062     6767   \n",
       "Patient_3_Heart_Patient_3_Muscle    3471    2373    1224    2062     6767   \n",
       "\n",
       "                                     ...      Gene_19991  Gene_19992  \\\n",
       "Patient_1_Blood_Patient_1_Liver      ...            4696        1132   \n",
       "Patient_1_Liver_Patient_1_Blood      ...            3967        1500   \n",
       "Patient_2_Testis_Patient_2_Brain     ...            1150        6609   \n",
       "Patient_2_Testis_Patient_2_Liver     ...            1150        6609   \n",
       "Patient_2_Brain_Patient_2_Testis     ...            1524        6672   \n",
       "Patient_2_Brain_Patient_2_Liver      ...            1524        6672   \n",
       "Patient_2_Liver_Patient_2_Testis     ...            6190        2781   \n",
       "Patient_2_Liver_Patient_2_Brain      ...            6190        2781   \n",
       "Patient_3_Heart_Patient_3_Blood      ...            2462        2730   \n",
       "Patient_3_Heart_Patient_3_Muscle     ...            2462        2730   \n",
       "\n",
       "                                  Gene_19993  Gene_19994  Gene_19995  \\\n",
       "Patient_1_Blood_Patient_1_Liver          468         775        6098   \n",
       "Patient_1_Liver_Patient_1_Blood         1003        6824        2320   \n",
       "Patient_2_Testis_Patient_2_Brain        1465         570        1969   \n",
       "Patient_2_Testis_Patient_2_Liver        1465         570        1969   \n",
       "Patient_2_Brain_Patient_2_Testis        2991        2425        2158   \n",
       "Patient_2_Brain_Patient_2_Liver         2991        2425        2158   \n",
       "Patient_2_Liver_Patient_2_Testis        8495        1543        2569   \n",
       "Patient_2_Liver_Patient_2_Brain         8495        1543        2569   \n",
       "Patient_3_Heart_Patient_3_Blood         1484        3102        3307   \n",
       "Patient_3_Heart_Patient_3_Muscle        1484        3102        3307   \n",
       "\n",
       "                                  Gene_19996  Gene_19997  Gene_19998  \\\n",
       "Patient_1_Blood_Patient_1_Liver         1150        7856        2462   \n",
       "Patient_1_Liver_Patient_1_Blood         7065        2660        2644   \n",
       "Patient_2_Testis_Patient_2_Brain        4336        2644        5296   \n",
       "Patient_2_Testis_Patient_2_Liver        4336        2644        5296   \n",
       "Patient_2_Brain_Patient_2_Testis        4529        1961        1339   \n",
       "Patient_2_Brain_Patient_2_Liver         4529        1961        1339   \n",
       "Patient_2_Liver_Patient_2_Testis        2710        3536        1233   \n",
       "Patient_2_Liver_Patient_2_Brain         2710        3536        1233   \n",
       "Patient_3_Heart_Patient_3_Blood         1028        1685        2926   \n",
       "Patient_3_Heart_Patient_3_Muscle        1028        1685        2926   \n",
       "\n",
       "                                  Gene_19999  Gene_20000  \n",
       "Patient_1_Blood_Patient_1_Liver         3628        1910  \n",
       "Patient_1_Liver_Patient_1_Blood         1821        1416  \n",
       "Patient_2_Testis_Patient_2_Brain         380        1993  \n",
       "Patient_2_Testis_Patient_2_Liver         380        1993  \n",
       "Patient_2_Brain_Patient_2_Testis        3780        2386  \n",
       "Patient_2_Brain_Patient_2_Liver         3780        2386  \n",
       "Patient_2_Liver_Patient_2_Testis        1911        1513  \n",
       "Patient_2_Liver_Patient_2_Brain         1911        1513  \n",
       "Patient_3_Heart_Patient_3_Blood          345        1881  \n",
       "Patient_3_Heart_Patient_3_Muscle         345        1881  \n",
       "\n",
       "[10 rows x 20000 columns]"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# train data\n",
    "X[:10]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene_1</th>\n",
       "      <th>Gene_2</th>\n",
       "      <th>Gene_3</th>\n",
       "      <th>Gene_4</th>\n",
       "      <th>Gene_5</th>\n",
       "      <th>Gene_6</th>\n",
       "      <th>Gene_7</th>\n",
       "      <th>Gene_8</th>\n",
       "      <th>Gene_9</th>\n",
       "      <th>Gene_10</th>\n",
       "      <th>...</th>\n",
       "      <th>Gene_19991</th>\n",
       "      <th>Gene_19992</th>\n",
       "      <th>Gene_19993</th>\n",
       "      <th>Gene_19994</th>\n",
       "      <th>Gene_19995</th>\n",
       "      <th>Gene_19996</th>\n",
       "      <th>Gene_19997</th>\n",
       "      <th>Gene_19998</th>\n",
       "      <th>Gene_19999</th>\n",
       "      <th>Gene_20000</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Patient_1_Blood_Patient_1_Liver</th>\n",
       "      <td>5954</td>\n",
       "      <td>2390</td>\n",
       "      <td>4451</td>\n",
       "      <td>2333</td>\n",
       "      <td>3371</td>\n",
       "      <td>3715</td>\n",
       "      <td>2096</td>\n",
       "      <td>2483</td>\n",
       "      <td>2641</td>\n",
       "      <td>7564</td>\n",
       "      <td>...</td>\n",
       "      <td>3967</td>\n",
       "      <td>1500</td>\n",
       "      <td>1003</td>\n",
       "      <td>6824</td>\n",
       "      <td>2320</td>\n",
       "      <td>7065</td>\n",
       "      <td>2660</td>\n",
       "      <td>2644</td>\n",
       "      <td>1821</td>\n",
       "      <td>1416</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Patient_1_Liver_Patient_1_Blood</th>\n",
       "      <td>3762</td>\n",
       "      <td>966</td>\n",
       "      <td>974</td>\n",
       "      <td>2223</td>\n",
       "      <td>4763</td>\n",
       "      <td>1724</td>\n",
       "      <td>5866</td>\n",
       "      <td>2025</td>\n",
       "      <td>4844</td>\n",
       "      <td>5496</td>\n",
       "      <td>...</td>\n",
       "      <td>4696</td>\n",
       "      <td>1132</td>\n",
       "      <td>468</td>\n",
       "      <td>775</td>\n",
       "      <td>6098</td>\n",
       "      <td>1150</td>\n",
       "      <td>7856</td>\n",
       "      <td>2462</td>\n",
       "      <td>3628</td>\n",
       "      <td>1910</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Patient_2_Testis_Patient_2_Brain</th>\n",
       "      <td>4545</td>\n",
       "      <td>3354</td>\n",
       "      <td>4068</td>\n",
       "      <td>4765</td>\n",
       "      <td>2463</td>\n",
       "      <td>4179</td>\n",
       "      <td>2773</td>\n",
       "      <td>3208</td>\n",
       "      <td>2042</td>\n",
       "      <td>2168</td>\n",
       "      <td>...</td>\n",
       "      <td>1524</td>\n",
       "      <td>6672</td>\n",
       "      <td>2991</td>\n",
       "      <td>2425</td>\n",
       "      <td>2158</td>\n",
       "      <td>4529</td>\n",
       "      <td>1961</td>\n",
       "      <td>1339</td>\n",
       "      <td>3780</td>\n",
       "      <td>2386</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Patient_2_Testis_Patient_2_Liver</th>\n",
       "      <td>5159</td>\n",
       "      <td>1559</td>\n",
       "      <td>2982</td>\n",
       "      <td>6594</td>\n",
       "      <td>2101</td>\n",
       "      <td>3903</td>\n",
       "      <td>1786</td>\n",
       "      <td>3455</td>\n",
       "      <td>1685</td>\n",
       "      <td>1449</td>\n",
       "      <td>...</td>\n",
       "      <td>6190</td>\n",
       "      <td>2781</td>\n",
       "      <td>8495</td>\n",
       "      <td>1543</td>\n",
       "      <td>2569</td>\n",
       "      <td>2710</td>\n",
       "      <td>3536</td>\n",
       "      <td>1233</td>\n",
       "      <td>1911</td>\n",
       "      <td>1513</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Patient_2_Brain_Patient_2_Testis</th>\n",
       "      <td>2766</td>\n",
       "      <td>6234</td>\n",
       "      <td>3649</td>\n",
       "      <td>1282</td>\n",
       "      <td>8075</td>\n",
       "      <td>5825</td>\n",
       "      <td>4944</td>\n",
       "      <td>108</td>\n",
       "      <td>1688</td>\n",
       "      <td>1189</td>\n",
       "      <td>...</td>\n",
       "      <td>1150</td>\n",
       "      <td>6609</td>\n",
       "      <td>1465</td>\n",
       "      <td>570</td>\n",
       "      <td>1969</td>\n",
       "      <td>4336</td>\n",
       "      <td>2644</td>\n",
       "      <td>5296</td>\n",
       "      <td>380</td>\n",
       "      <td>1993</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Patient_2_Brain_Patient_2_Liver</th>\n",
       "      <td>5159</td>\n",
       "      <td>1559</td>\n",
       "      <td>2982</td>\n",
       "      <td>6594</td>\n",
       "      <td>2101</td>\n",
       "      <td>3903</td>\n",
       "      <td>1786</td>\n",
       "      <td>3455</td>\n",
       "      <td>1685</td>\n",
       "      <td>1449</td>\n",
       "      <td>...</td>\n",
       "      <td>6190</td>\n",
       "      <td>2781</td>\n",
       "      <td>8495</td>\n",
       "      <td>1543</td>\n",
       "      <td>2569</td>\n",
       "      <td>2710</td>\n",
       "      <td>3536</td>\n",
       "      <td>1233</td>\n",
       "      <td>1911</td>\n",
       "      <td>1513</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Patient_2_Liver_Patient_2_Testis</th>\n",
       "      <td>2766</td>\n",
       "      <td>6234</td>\n",
       "      <td>3649</td>\n",
       "      <td>1282</td>\n",
       "      <td>8075</td>\n",
       "      <td>5825</td>\n",
       "      <td>4944</td>\n",
       "      <td>108</td>\n",
       "      <td>1688</td>\n",
       "      <td>1189</td>\n",
       "      <td>...</td>\n",
       "      <td>1150</td>\n",
       "      <td>6609</td>\n",
       "      <td>1465</td>\n",
       "      <td>570</td>\n",
       "      <td>1969</td>\n",
       "      <td>4336</td>\n",
       "      <td>2644</td>\n",
       "      <td>5296</td>\n",
       "      <td>380</td>\n",
       "      <td>1993</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Patient_2_Liver_Patient_2_Brain</th>\n",
       "      <td>4545</td>\n",
       "      <td>3354</td>\n",
       "      <td>4068</td>\n",
       "      <td>4765</td>\n",
       "      <td>2463</td>\n",
       "      <td>4179</td>\n",
       "      <td>2773</td>\n",
       "      <td>3208</td>\n",
       "      <td>2042</td>\n",
       "      <td>2168</td>\n",
       "      <td>...</td>\n",
       "      <td>1524</td>\n",
       "      <td>6672</td>\n",
       "      <td>2991</td>\n",
       "      <td>2425</td>\n",
       "      <td>2158</td>\n",
       "      <td>4529</td>\n",
       "      <td>1961</td>\n",
       "      <td>1339</td>\n",
       "      <td>3780</td>\n",
       "      <td>2386</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>8 rows  20000 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                  Gene_1  Gene_2  Gene_3  Gene_4  Gene_5  \\\n",
       "Patient_1_Blood_Patient_1_Liver     5954    2390    4451    2333    3371   \n",
       "Patient_1_Liver_Patient_1_Blood     3762     966     974    2223    4763   \n",
       "Patient_2_Testis_Patient_2_Brain    4545    3354    4068    4765    2463   \n",
       "Patient_2_Testis_Patient_2_Liver    5159    1559    2982    6594    2101   \n",
       "Patient_2_Brain_Patient_2_Testis    2766    6234    3649    1282    8075   \n",
       "Patient_2_Brain_Patient_2_Liver     5159    1559    2982    6594    2101   \n",
       "Patient_2_Liver_Patient_2_Testis    2766    6234    3649    1282    8075   \n",
       "Patient_2_Liver_Patient_2_Brain     4545    3354    4068    4765    2463   \n",
       "\n",
       "                                  Gene_6  Gene_7  Gene_8  Gene_9  Gene_10  \\\n",
       "Patient_1_Blood_Patient_1_Liver     3715    2096    2483    2641     7564   \n",
       "Patient_1_Liver_Patient_1_Blood     1724    5866    2025    4844     5496   \n",
       "Patient_2_Testis_Patient_2_Brain    4179    2773    3208    2042     2168   \n",
       "Patient_2_Testis_Patient_2_Liver    3903    1786    3455    1685     1449   \n",
       "Patient_2_Brain_Patient_2_Testis    5825    4944     108    1688     1189   \n",
       "Patient_2_Brain_Patient_2_Liver     3903    1786    3455    1685     1449   \n",
       "Patient_2_Liver_Patient_2_Testis    5825    4944     108    1688     1189   \n",
       "Patient_2_Liver_Patient_2_Brain     4179    2773    3208    2042     2168   \n",
       "\n",
       "                                     ...      Gene_19991  Gene_19992  \\\n",
       "Patient_1_Blood_Patient_1_Liver      ...            3967        1500   \n",
       "Patient_1_Liver_Patient_1_Blood      ...            4696        1132   \n",
       "Patient_2_Testis_Patient_2_Brain     ...            1524        6672   \n",
       "Patient_2_Testis_Patient_2_Liver     ...            6190        2781   \n",
       "Patient_2_Brain_Patient_2_Testis     ...            1150        6609   \n",
       "Patient_2_Brain_Patient_2_Liver      ...            6190        2781   \n",
       "Patient_2_Liver_Patient_2_Testis     ...            1150        6609   \n",
       "Patient_2_Liver_Patient_2_Brain      ...            1524        6672   \n",
       "\n",
       "                                  Gene_19993  Gene_19994  Gene_19995  \\\n",
       "Patient_1_Blood_Patient_1_Liver         1003        6824        2320   \n",
       "Patient_1_Liver_Patient_1_Blood          468         775        6098   \n",
       "Patient_2_Testis_Patient_2_Brain        2991        2425        2158   \n",
       "Patient_2_Testis_Patient_2_Liver        8495        1543        2569   \n",
       "Patient_2_Brain_Patient_2_Testis        1465         570        1969   \n",
       "Patient_2_Brain_Patient_2_Liver         8495        1543        2569   \n",
       "Patient_2_Liver_Patient_2_Testis        1465         570        1969   \n",
       "Patient_2_Liver_Patient_2_Brain         2991        2425        2158   \n",
       "\n",
       "                                  Gene_19996  Gene_19997  Gene_19998  \\\n",
       "Patient_1_Blood_Patient_1_Liver         7065        2660        2644   \n",
       "Patient_1_Liver_Patient_1_Blood         1150        7856        2462   \n",
       "Patient_2_Testis_Patient_2_Brain        4529        1961        1339   \n",
       "Patient_2_Testis_Patient_2_Liver        2710        3536        1233   \n",
       "Patient_2_Brain_Patient_2_Testis        4336        2644        5296   \n",
       "Patient_2_Brain_Patient_2_Liver         2710        3536        1233   \n",
       "Patient_2_Liver_Patient_2_Testis        4336        2644        5296   \n",
       "Patient_2_Liver_Patient_2_Brain         4529        1961        1339   \n",
       "\n",
       "                                  Gene_19999  Gene_20000  \n",
       "Patient_1_Blood_Patient_1_Liver         1821        1416  \n",
       "Patient_1_Liver_Patient_1_Blood         3628        1910  \n",
       "Patient_2_Testis_Patient_2_Brain        3780        2386  \n",
       "Patient_2_Testis_Patient_2_Liver        1911        1513  \n",
       "Patient_2_Brain_Patient_2_Testis         380        1993  \n",
       "Patient_2_Brain_Patient_2_Liver         1911        1513  \n",
       "Patient_2_Liver_Patient_2_Testis         380        1993  \n",
       "Patient_2_Liver_Patient_2_Brain         3780        2386  \n",
       "\n",
       "[8 rows x 20000 columns]"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Y[:10]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python (rep)",
   "language": "python",
   "name": "rep"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
