{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Goal\n",
    "Preprocessing rnaseq data\n",
    "\n",
    "This is a tutorial with the main workflow for preparing a training dataset for a basic linear regression model. \n",
    "The output (X,Y) are all pair combination between tissues per each sample, where X is input, Y target."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The autoreload extension is already loaded. To reload it, use:\n",
      "  %reload_ext autoreload\n"
     ]
    }
   ],
   "source": [
    "# code autoreload\n",
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: numpy in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (1.15.4)\n",
      "Requirement already satisfied: pandas in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (0.23.4)\n",
      "Requirement already satisfied: numpy>=1.9.0 in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (from pandas) (1.15.4)\n",
      "Requirement already satisfied: python-dateutil>=2.5.0 in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (from pandas) (2.7.5)\n",
      "Requirement already satisfied: pytz>=2011k in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (from pandas) (2018.7)\n",
      "Requirement already satisfied: six>=1.5 in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (from python-dateutil>=2.5.0->pandas) (1.12.0)\n",
      "Requirement already satisfied: anndata in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (0.6.16)\n",
      "Requirement already satisfied: pandas>=0.21.0 in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (from anndata) (0.23.4)\n",
      "Requirement already satisfied: scipy~=1.0 in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (from anndata) (1.2.0)\n",
      "Requirement already satisfied: natsort in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (from anndata) (5.5.0)\n",
      "Requirement already satisfied: numpy~=1.14 in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (from anndata) (1.15.4)\n",
      "Requirement already satisfied: h5py in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (from anndata) (2.9.0)\n",
      "Requirement already satisfied: python-dateutil>=2.5.0 in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (from pandas>=0.21.0->anndata) (2.7.5)\n",
      "Requirement already satisfied: pytz>=2011k in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (from pandas>=0.21.0->anndata) (2018.7)\n",
      "Requirement already satisfied: six in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (from h5py->anndata) (1.12.0)\n",
      "Requirement already satisfied: scanpy in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (1.3.6)\n",
      "Requirement already satisfied: natsort in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (from scanpy) (5.5.0)\n",
      "Requirement already satisfied: networkx in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (from scanpy) (2.2)\n",
      "Requirement already satisfied: seaborn in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (from scanpy) (0.9.0)\n",
      "Requirement already satisfied: joblib in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (from scanpy) (0.13.0)\n",
      "Requirement already satisfied: numba>=0.40.0 in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (from scanpy) (0.41.0)\n",
      "Requirement already satisfied: scipy in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (from scanpy) (1.2.0)\n",
      "Requirement already satisfied: scikit-learn>=0.19.1 in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (from scanpy) (0.20.2)\n",
      "Requirement already satisfied: pandas>=0.21 in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (from scanpy) (0.23.4)\n",
      "Requirement already satisfied: h5py in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (from scanpy) (2.9.0)\n",
      "Requirement already satisfied: matplotlib>=2.2 in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (from scanpy) (3.0.2)\n",
      "Requirement already satisfied: statsmodels in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (from scanpy) (0.9.0)\n",
      "Requirement already satisfied: anndata>=0.6.13 in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (from scanpy) (0.6.16)\n",
      "Requirement already satisfied: tables in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (from scanpy) (3.4.4)\n",
      "Requirement already satisfied: decorator>=4.3.0 in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (from networkx->scanpy) (4.3.0)\n",
      "Requirement already satisfied: numpy>=1.9.3 in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (from seaborn->scanpy) (1.15.4)\n",
      "Requirement already satisfied: llvmlite>=0.26.0dev0 in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (from numba>=0.40.0->scanpy) (0.26.0)\n",
      "Requirement already satisfied: pytz>=2011k in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (from pandas>=0.21->scanpy) (2018.7)\n",
      "Requirement already satisfied: python-dateutil>=2.5.0 in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (from pandas>=0.21->scanpy) (2.7.5)\n",
      "Requirement already satisfied: six in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (from h5py->scanpy) (1.12.0)\n",
      "Requirement already satisfied: cycler>=0.10 in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (from matplotlib>=2.2->scanpy) (0.10.0)\n",
      "Requirement already satisfied: pyparsing!=2.0.4,!=2.1.2,!=2.1.6,>=2.0.1 in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (from matplotlib>=2.2->scanpy) (2.3.0)\n",
      "Requirement already satisfied: kiwisolver>=1.0.1 in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (from matplotlib>=2.2->scanpy) (1.0.1)\n",
      "Requirement already satisfied: patsy in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (from statsmodels->scanpy) (0.5.1)\n",
      "Requirement already satisfied: numexpr>=2.5.2 in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (from tables->scanpy) (2.6.8)\n",
      "Requirement already satisfied: setuptools in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (from kiwisolver>=1.0.1->matplotlib>=2.2->scanpy) (40.6.3)\n",
      "Requirement already satisfied: argh in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (0.26.2)\n"
     ]
    }
   ],
   "source": [
    "!pip install numpy\n",
    "!pip install pandas\n",
    "!pip install anndata\n",
    "\n",
    "!pip install scanpy\n",
    "!pip install argh"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Modules to reload:\n",
      "all-except-skipped\n",
      "\n",
      "Modules to skip:\n",
      "\n"
     ]
    }
   ],
   "source": [
    "import rep.preprocessing as p\n",
    "from rep.constants import ANNDATA_CST as a\n",
    "\n",
    "%aimport"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 1. Load raw count matrix + annotation and put it into an annodata object"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting data.csv\n"
     ]
    }
   ],
   "source": [
    "%%writefile data.csv\n",
    "T1_s1,T2_s2,T3_s1,T4_s2,T5_s1,T6_s1\n",
    "G1,10,20,30,40,5,6\n",
    "G2,5,10,20,30,5,6\n",
    "G3,6,7,8,9,1,1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting anno_var.csv\n"
     ]
    }
   ],
   "source": [
    "%%writefile anno_var.csv\n",
    "Sample,Gender,Tissue,Individual,Technology\n",
    "T1_s1,F,Tissue1,Sample1,rnaseq\n",
    "T2_s2,M,Tissue2,Sample2,rnaseq\n",
    "T3_s1,F,Tissue3,Sample1,rnaseq\n",
    "T4_s2,M,Tissue4,Sample2,rnaseq\n",
    "T5_s1,M,Tissue5,Sample1,rnaseq\n",
    "T6_s1,M,Tissue6,Sample1,rnaseq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Writing anno_obs.csv\n"
     ]
    }
   ],
   "source": [
    "%%writefile anno_obs.csv\n",
    "G1,hg19,T1,chr1,1111,-\n",
    "G2,hg19,T2,chr2,2222,-\n",
    "G3,hg19,T3,chr3,3333,-"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "anndata.X\n",
      "[[10. 20. 30. 40.  5.  6.]\n",
      " [ 5. 10. 20. 30.  5.  6.]\n",
      " [ 6.  7.  8.  9.  1.  1.]]\n",
      "anndata.var\n",
      "Index(['T1_s1', 'T2_s2', 'T3_s1', 'T4_s2', 'T5_s1', 'T6_s1'], dtype='object')\n",
      "      Gender   Tissue Individual Technology\n",
      "T1_s1      F  Tissue1    Sample1     rnaseq\n",
      "T2_s2      M  Tissue2    Sample2     rnaseq\n",
      "T3_s1      F  Tissue3    Sample1     rnaseq\n",
      "T4_s2      M  Tissue4    Sample2     rnaseq\n",
      "T5_s1      M  Tissue5    Sample1     rnaseq\n",
      "T6_s1      M  Tissue6    Sample1     rnaseq\n",
      "anndata.obs\n",
      "Index(['G1', 'G2', 'G3'], dtype='object')\n",
      "       1   2     3     4  5\n",
      "G1  hg19  T1  chr1  1111  -\n",
      "G2  hg19  T2  chr2  2222  -\n",
      "G3  hg19  T3  chr3  3333  -\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# annobj is the equivalent of the summarized experiments in R. \n",
    "# It contains \n",
    "# - annobj.X - counts\n",
    "# - annobj.var - data.frame annotation object for the columns of the count matrix (in our case Tissue_Sample)\n",
    "# - annobj.obs - data.frame annotation object for the rows of the count matrix (= features that we measure, in our case Genes)\n",
    "# \n",
    "annobj = p.load_count_matrix(\"data.csv\", sep=\",\", varanno=\"anno_var.csv\", obsanno=\"anno_obs.csv\")\n",
    "p.print_anndata(annobj)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Save data to h2ad format"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "... storing '1' as categorical\n",
      "... storing '5' as categorical\n",
      "... storing '1' as categorical\n",
      "... storing '3' as categorical\n",
      "... storing '4' as categorical\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "'/data/nasif12/home_if12/giurgiu/rep_gagneur/rep/notebooks/preprocessing/tmp1545316680.h5ad'"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "name = p.save(annobj)\n",
    "name"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 2. Prepare training sample:\n",
    "- select genes and tissue_samples for training\n",
    "- slice the anndata object\n",
    "- compute the cross tissue matrix pair (X,Y)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "      Gender   Tissue Individual Technology\n",
      "T1_s1      F  Tissue1    Sample1     rnaseq\n",
      "T2_s2      M  Tissue2    Sample2     rnaseq\n",
      "T3_s1      F  Tissue3    Sample1     rnaseq\n",
      "T4_s2      M  Tissue4    Sample2     rnaseq\n",
      "T5_s1      M  Tissue5    Sample1     rnaseq\n",
      "T6_s1      M  Tissue6    Sample1     rnaseq\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "(['T2_s2', 'T6_s1', 'T5_s1', 'T4_s2'], ['G1', 'G2'])"
      ]
     },
     "execution_count": 65,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# filtering by different values in anndata.var and anndata.obs\n",
    "# stored in a json format by specifying the {column over we would like to filter : array of admitted values}\n",
    "# in this example we filter the anndata.var by gender (M) and anndata.obs should contain only G1 and G2\n",
    "(var, obs) = p.filter_anndata_by_value(annobj, {a.VAR: {'Gender': ['M']}, a.OBS: {0: ['G1', 'G2']}})\n",
    "\n",
    "# these are the rows and columns selected for the training set\n",
    "(var,obs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'Sample1': ['T6_s1', 'T5_s1'], 'Sample2': ['T2_s2', 'T4_s2']}"
      ]
     },
     "execution_count": 67,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# since we want to impute gene expression across different tissues, we\n",
    "# would like to compute the cross tissue pairs per sample and not across samples\n",
    "\n",
    "# for this purpose we can use the group_by(df, column, index_subset) \n",
    "# function to group tissues by sample\n",
    "\n",
    "dict_samples_var = p.group_by(annobj.var, 'Individual', var)\n",
    "dict_samples_var"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "4.3 Slice anndata\n",
      "\n",
      "anndata.X\n",
      "[[20.  5. 30.  6. 10. 40.]\n",
      " [10.  5. 20.  6.  5. 30.]]\n",
      "anndata.var\n",
      "Index(['T2_s2', 'T5_s1', 'T3_s1', 'T6_s1', 'T1_s1', 'T4_s2'], dtype='object')\n",
      "      Gender   Tissue Individual Technology\n",
      "T2_s2      M  Tissue2    Sample2     rnaseq\n",
      "T5_s1      M  Tissue5    Sample1     rnaseq\n",
      "T3_s1      F  Tissue3    Sample1     rnaseq\n",
      "T6_s1      M  Tissue6    Sample1     rnaseq\n",
      "T1_s1      F  Tissue1    Sample1     rnaseq\n",
      "T4_s2      M  Tissue4    Sample2     rnaseq\n",
      "anndata.obs\n",
      "Index(['G1', 'G2'], dtype='object')\n",
      "       1   2     3     4  5\n",
      "G1  hg19  T1  chr1  1111  -\n",
      "G2  hg19  T2  chr2  2222  -\n",
      "\n",
      "4.4 Build the two matrices X and Y\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# compute cross tissue matrix pair (X, Y)\n",
    "(X, Y) = p.rnaseq_cross_tissue(annobj, individuals=['Sample1','Sample2'], gene_ids=obs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>G1</th>\n",
       "      <th>G2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>T1_s1_T6_s1</th>\n",
       "      <td>10</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T1_s1_T3_s1</th>\n",
       "      <td>10</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T1_s1_T5_s1</th>\n",
       "      <td>10</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T6_s1_T1_s1</th>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T6_s1_T3_s1</th>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T6_s1_T5_s1</th>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T3_s1_T1_s1</th>\n",
       "      <td>30</td>\n",
       "      <td>20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T3_s1_T6_s1</th>\n",
       "      <td>30</td>\n",
       "      <td>20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T3_s1_T5_s1</th>\n",
       "      <td>30</td>\n",
       "      <td>20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T5_s1_T1_s1</th>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T5_s1_T6_s1</th>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T5_s1_T3_s1</th>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T2_s2_T4_s2</th>\n",
       "      <td>20</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T4_s2_T2_s2</th>\n",
       "      <td>40</td>\n",
       "      <td>30</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             G1  G2\n",
       "T1_s1_T6_s1  10   5\n",
       "T1_s1_T3_s1  10   5\n",
       "T1_s1_T5_s1  10   5\n",
       "T6_s1_T1_s1   6   6\n",
       "T6_s1_T3_s1   6   6\n",
       "T6_s1_T5_s1   6   6\n",
       "T3_s1_T1_s1  30  20\n",
       "T3_s1_T6_s1  30  20\n",
       "T3_s1_T5_s1  30  20\n",
       "T5_s1_T1_s1   5   5\n",
       "T5_s1_T6_s1   5   5\n",
       "T5_s1_T3_s1   5   5\n",
       "T2_s2_T4_s2  20  10\n",
       "T4_s2_T2_s2  40  30"
      ]
     },
     "execution_count": 73,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# train matrix\n",
    "X"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>G1</th>\n",
       "      <th>G2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>T1_s1_T6_s1</th>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T1_s1_T3_s1</th>\n",
       "      <td>30</td>\n",
       "      <td>20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T1_s1_T5_s1</th>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T6_s1_T1_s1</th>\n",
       "      <td>10</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T6_s1_T3_s1</th>\n",
       "      <td>30</td>\n",
       "      <td>20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T6_s1_T5_s1</th>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T3_s1_T1_s1</th>\n",
       "      <td>10</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T3_s1_T6_s1</th>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T3_s1_T5_s1</th>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T5_s1_T1_s1</th>\n",
       "      <td>10</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T5_s1_T6_s1</th>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T5_s1_T3_s1</th>\n",
       "      <td>30</td>\n",
       "      <td>20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T2_s2_T4_s2</th>\n",
       "      <td>40</td>\n",
       "      <td>30</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T4_s2_T2_s2</th>\n",
       "      <td>20</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             G1  G2\n",
       "T1_s1_T6_s1   6   6\n",
       "T1_s1_T3_s1  30  20\n",
       "T1_s1_T5_s1   5   5\n",
       "T6_s1_T1_s1  10   5\n",
       "T6_s1_T3_s1  30  20\n",
       "T6_s1_T5_s1   5   5\n",
       "T3_s1_T1_s1  10   5\n",
       "T3_s1_T6_s1   6   6\n",
       "T3_s1_T5_s1   5   5\n",
       "T5_s1_T1_s1  10   5\n",
       "T5_s1_T6_s1   6   6\n",
       "T5_s1_T3_s1  30  20\n",
       "T2_s2_T4_s2  40  30\n",
       "T4_s2_T2_s2  20  10"
      ]
     },
     "execution_count": 74,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# label matrix\n",
    "Y"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3. Create toy large dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: numpy in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (1.15.4)\n",
      "Requirement already satisfied: matplotlib in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (3.0.2)\n",
      "Requirement already satisfied: kiwisolver>=1.0.1 in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (from matplotlib) (1.0.1)\n",
      "Requirement already satisfied: pyparsing!=2.0.4,!=2.1.2,!=2.1.6,>=2.0.1 in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (from matplotlib) (2.3.0)\n",
      "Requirement already satisfied: numpy>=1.10.0 in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (from matplotlib) (1.15.4)\n",
      "Requirement already satisfied: cycler>=0.10 in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (from matplotlib) (0.10.0)\n",
      "Requirement already satisfied: python-dateutil>=2.1 in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (from matplotlib) (2.7.5)\n",
      "Requirement already satisfied: setuptools in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (from kiwisolver>=1.0.1->matplotlib) (40.6.3)\n",
      "Requirement already satisfied: six in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (from cycler>=0.10->matplotlib) (1.12.0)\n"
     ]
    }
   ],
   "source": [
    "!pip install numpy\n",
    "!pip install matplotlib\n",
    "import random\n",
    "import math\n",
    "import numpy.random as nr\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "from scipy import stats\n",
    "\n",
    "import anndata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Gene_19991',\n",
       " 'Gene_19992',\n",
       " 'Gene_19993',\n",
       " 'Gene_19994',\n",
       " 'Gene_19995',\n",
       " 'Gene_19996',\n",
       " 'Gene_19997',\n",
       " 'Gene_19998',\n",
       " 'Gene_19999',\n",
       " 'Gene_20000']"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# genes\n",
    "genes = list(map(lambda x: \"Gene_\" + str(x) , list(range(1, 20001))))\n",
    "genes[19990:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([2748, 7615, 1717, 2546, 2959, 3118, 1490, 5260,  630, 1818])"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "raw_counts = nr.negative_binomial(3,0.001,size=20000)\n",
    "# sns.distplot(raw_counts)\n",
    "raw_counts[:10]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# tissues\n",
    "tissues = ['Muscle', 'Blood', 'Skin', 'Lung', 'Heart', 'Stomatch', 'Colon', 'Testis', 'Brain', 'Liver']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([7.92409257, 1.85795368, 7.1120728 , 3.13443305, 6.07305385,\n",
       "       6.03105284, 1.96623577, 8.85059353, 5.00515778, 7.85631287])"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# simulate how many tissues are collected per patient\n",
    "distribution_tissues = nr.uniform(1,len(tissues),1000)\n",
    "distribution_tissues[:10]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# generate samples - raw counts per gene per tissue\n",
    "frames = []\n",
    "for i in range(1,100): # patients\n",
    "  \n",
    "  # choose tissues\n",
    "  how_many_tissues = math.floor(distribution_tissues[random.randint(1,len(distribution_tissues))])\n",
    "  random.shuffle(tissues)\n",
    "  t_set = tissues[:how_many_tissues]  \n",
    "  \n",
    "  # initialize count matrix\n",
    "  counts = pd.DataFrame(columns=genes,index=map(lambda x: \"Patient_\" + str(i) + \"_\" + str(x), t_set))\n",
    "\n",
    "  for j in range(0,len(t_set)):\n",
    "    \n",
    "    col_name = \"Patient_\" + str(i) + \"_\" + t_set[j]\n",
    "    \n",
    "    # generate raw counts per tissue per set of genes\n",
    "    raw_counts = nr.negative_binomial(3,0.001,size=20000)\n",
    "    counts.loc[col_name] = raw_counts.reshape(1,20000)\n",
    "    \n",
    "  # append expression per tissue for the patient \n",
    "  frames.append(counts)\n",
    "      \n",
    "\n",
    "df_big = pd.concat(frames)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(20000, 515)"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# transpose the matrix to be genes x patient_tissue\n",
    "count_matrix = df_big.transpose()\n",
    "count_matrix.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Individual</th>\n",
       "      <th>Tissue</th>\n",
       "      <th>Gender</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Patient_1_Colon</th>\n",
       "      <td>Patient_1</td>\n",
       "      <td>Colon</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Patient_1_Testis</th>\n",
       "      <td>Patient_1</td>\n",
       "      <td>Testis</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Patient_2_Muscle</th>\n",
       "      <td>Patient_2</td>\n",
       "      <td>Muscle</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Patient_2_Testis</th>\n",
       "      <td>Patient_2</td>\n",
       "      <td>Testis</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Patient_3_Stomatch</th>\n",
       "      <td>Patient_3</td>\n",
       "      <td>Stomatch</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Patient_4_Muscle</th>\n",
       "      <td>Patient_4</td>\n",
       "      <td>Muscle</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Patient_4_Blood</th>\n",
       "      <td>Patient_4</td>\n",
       "      <td>Blood</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Patient_4_Colon</th>\n",
       "      <td>Patient_4</td>\n",
       "      <td>Colon</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Patient_4_Stomatch</th>\n",
       "      <td>Patient_4</td>\n",
       "      <td>Stomatch</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Patient_4_Heart</th>\n",
       "      <td>Patient_4</td>\n",
       "      <td>Heart</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   Individual    Tissue  Gender\n",
       "Patient_1_Colon     Patient_1     Colon     2.0\n",
       "Patient_1_Testis    Patient_1    Testis     2.0\n",
       "Patient_2_Muscle    Patient_2    Muscle     1.0\n",
       "Patient_2_Testis    Patient_2    Testis     1.0\n",
       "Patient_3_Stomatch  Patient_3  Stomatch     2.0\n",
       "Patient_4_Muscle    Patient_4    Muscle     1.0\n",
       "Patient_4_Blood     Patient_4     Blood     1.0\n",
       "Patient_4_Colon     Patient_4     Colon     1.0\n",
       "Patient_4_Stomatch  Patient_4  Stomatch     1.0\n",
       "Patient_4_Heart     Patient_4     Heart     1.0"
      ]
     },
     "execution_count": 71,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# create annotation dataframes\n",
    "# column = 0 its the index! be careful\n",
    "samples_anno = pd.DataFrame(index=count_matrix.columns,columns=['Individual','Tissue'])\n",
    "gender=[1,2]\n",
    "\n",
    "for i in range(0,samples_anno.shape[0]):\n",
    "  sample = samples_anno.index[i]\n",
    "  [pat,number,tissue] = sample.split(\"_\")\n",
    "  \n",
    "  samples_anno.at[sample, 'Individual'] = pat + \"_\" + str(number)\n",
    "  samples_anno.at[sample, 'Tissue'] = tissue\n",
    "  samples_anno.at[sample, 'Gender'] = gender[int(number)%2]\n",
    "\n",
    "samples_anno.drop_duplicates().shape\n",
    "samples_anno[:10]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(100, 0)"
      ]
     },
     "execution_count": 72,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes_anno = pd.DataFrame(index=count_matrix.index)\n",
    "genes_anno.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "anndata.X ----\n",
      "[[2691. 2133. 4818. ... 1695. 5303. 8355.]\n",
      " [2242. 5293. 1021. ... 2374. 5922. 1623.]\n",
      " [4557. 2922. 2317. ...  938. 5487. 1925.]\n",
      " ...\n",
      " [1168. 1726. 2143. ... 2327. 4561. 2079.]\n",
      " [1784. 2701. 1817. ...  766. 4523. 2383.]\n",
      " [2142. 1778. 4253. ... 6785. 3840.  819.]]\n",
      "anndata.var ----\n",
      "                       Individual    Tissue  Gender\n",
      "Patient_1_Colon         Patient_1     Colon     2.0\n",
      "Patient_1_Testis        Patient_1    Testis     2.0\n",
      "Patient_2_Muscle        Patient_2    Muscle     1.0\n",
      "Patient_2_Testis        Patient_2    Testis     1.0\n",
      "Patient_3_Stomatch      Patient_3  Stomatch     2.0\n",
      "Patient_4_Muscle        Patient_4    Muscle     1.0\n",
      "Patient_4_Blood         Patient_4     Blood     1.0\n",
      "Patient_4_Colon         Patient_4     Colon     1.0\n",
      "Patient_4_Stomatch      Patient_4  Stomatch     1.0\n",
      "Patient_4_Heart         Patient_4     Heart     1.0\n",
      "Patient_5_Colon         Patient_5     Colon     2.0\n",
      "Patient_5_Testis        Patient_5    Testis     2.0\n",
      "Patient_5_Brain         Patient_5     Brain     2.0\n",
      "Patient_5_Blood         Patient_5     Blood     2.0\n",
      "Patient_5_Liver         Patient_5     Liver     2.0\n",
      "Patient_6_Heart         Patient_6     Heart     1.0\n",
      "Patient_6_Skin          Patient_6      Skin     1.0\n",
      "Patient_6_Colon         Patient_6     Colon     1.0\n",
      "Patient_6_Testis        Patient_6    Testis     1.0\n",
      "Patient_6_Muscle        Patient_6    Muscle     1.0\n",
      "Patient_6_Brain         Patient_6     Brain     1.0\n",
      "Patient_6_Blood         Patient_6     Blood     1.0\n",
      "Patient_6_Lung          Patient_6      Lung     1.0\n",
      "Patient_6_Liver         Patient_6     Liver     1.0\n",
      "Patient_7_Brain         Patient_7     Brain     2.0\n",
      "Patient_7_Testis        Patient_7    Testis     2.0\n",
      "Patient_7_Stomatch      Patient_7  Stomatch     2.0\n",
      "Patient_7_Colon         Patient_7     Colon     2.0\n",
      "Patient_7_Lung          Patient_7      Lung     2.0\n",
      "Patient_7_Liver         Patient_7     Liver     2.0\n",
      "...                           ...       ...     ...\n",
      "Patient_592_Muscle    Patient_592    Muscle     1.0\n",
      "Patient_592_Colon     Patient_592     Colon     1.0\n",
      "Patient_593_Brain     Patient_593     Brain     2.0\n",
      "Patient_593_Stomatch  Patient_593  Stomatch     2.0\n",
      "Patient_593_Heart     Patient_593     Heart     2.0\n",
      "Patient_593_Lung      Patient_593      Lung     2.0\n",
      "Patient_594_Testis    Patient_594    Testis     1.0\n",
      "Patient_594_Muscle    Patient_594    Muscle     1.0\n",
      "Patient_594_Heart     Patient_594     Heart     1.0\n",
      "Patient_594_Blood     Patient_594     Blood     1.0\n",
      "Patient_595_Lung      Patient_595      Lung     2.0\n",
      "Patient_595_Heart     Patient_595     Heart     2.0\n",
      "Patient_596_Blood     Patient_596     Blood     1.0\n",
      "Patient_596_Testis    Patient_596    Testis     1.0\n",
      "Patient_596_Liver     Patient_596     Liver     1.0\n",
      "Patient_596_Brain     Patient_596     Brain     1.0\n",
      "Patient_597_Brain     Patient_597     Brain     2.0\n",
      "Patient_597_Testis    Patient_597    Testis     2.0\n",
      "Patient_597_Blood     Patient_597     Blood     2.0\n",
      "Patient_597_Heart     Patient_597     Heart     2.0\n",
      "Patient_597_Skin      Patient_597      Skin     2.0\n",
      "Patient_597_Muscle    Patient_597    Muscle     2.0\n",
      "Patient_597_Stomatch  Patient_597  Stomatch     2.0\n",
      "Patient_597_Lung      Patient_597      Lung     2.0\n",
      "Patient_598_Skin      Patient_598      Skin     1.0\n",
      "Patient_598_Heart     Patient_598     Heart     1.0\n",
      "Patient_598_Brain     Patient_598     Brain     1.0\n",
      "Patient_598_Testis    Patient_598    Testis     1.0\n",
      "Patient_598_Lung      Patient_598      Lung     1.0\n",
      "Patient_599_Stomatch  Patient_599  Stomatch     2.0\n",
      "\n",
      "[3041 rows x 3 columns]\n",
      "anndata.obs ----\n",
      "Empty DataFrame\n",
      "Columns: []\n",
      "Index: [Gene_1, Gene_2, Gene_3, Gene_4, Gene_5, Gene_6, Gene_7, Gene_8, Gene_9, Gene_10, Gene_11, Gene_12, Gene_13, Gene_14, Gene_15, Gene_16, Gene_17, Gene_18, Gene_19, Gene_20, Gene_21, Gene_22, Gene_23, Gene_24, Gene_25, Gene_26, Gene_27, Gene_28, Gene_29, Gene_30, Gene_31, Gene_32, Gene_33, Gene_34, Gene_35, Gene_36, Gene_37, Gene_38, Gene_39, Gene_40, Gene_41, Gene_42, Gene_43, Gene_44, Gene_45, Gene_46, Gene_47, Gene_48, Gene_49, Gene_50, Gene_51, Gene_52, Gene_53, Gene_54, Gene_55, Gene_56, Gene_57, Gene_58, Gene_59, Gene_60, Gene_61, Gene_62, Gene_63, Gene_64, Gene_65, Gene_66, Gene_67, Gene_68, Gene_69, Gene_70, Gene_71, Gene_72, Gene_73, Gene_74, Gene_75, Gene_76, Gene_77, Gene_78, Gene_79, Gene_80, Gene_81, Gene_82, Gene_83, Gene_84, Gene_85, Gene_86, Gene_87, Gene_88, Gene_89, Gene_90, Gene_91, Gene_92, Gene_93, Gene_94, Gene_95, Gene_96, Gene_97, Gene_98, Gene_99, Gene_100]\n",
      "\n",
      "[100 rows x 0 columns]\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# create annobj\n",
    "annobj = anndata.AnnData(count_matrix)\n",
    "annobj.var = samples_anno\n",
    "annobj.var_names = list(samples_anno.index)\n",
    "annobj.obs = genes_anno\n",
    "annobj.obs_names = list(genes_anno.index)\n",
    "\n",
    "p.print_anndata(annobj)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Liver', 'Stomatch', 'Testis', 'Colon', 'Blood']"
      ]
     },
     "execution_count": 74,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# filter using a subset of the tissues\n",
    "random.shuffle(tissues)\n",
    "subset_tissue = tissues[:5]\n",
    "subset_tissue"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(599, 2)\n",
      "    Individual  Gender\n",
      "0    Patient_1     2.0\n",
      "2    Patient_2     1.0\n",
      "4    Patient_3     2.0\n",
      "5    Patient_4     1.0\n",
      "10   Patient_5     2.0\n",
      "15   Patient_6     1.0\n",
      "24   Patient_7     2.0\n",
      "33   Patient_8     1.0\n",
      "40   Patient_9     2.0\n",
      "45  Patient_10     1.0\n",
      "<pandas.core.groupby.groupby.DataFrameGroupBy object at 0x7f2114ed6978>\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "(360, 120, 119)"
      ]
     },
     "execution_count": 75,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "(train,valid,test) = p.split_by_individuals(annobj,groupby=['Gender'])\n",
    "len(train),len(valid),len(test)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total pairs: 9578\n",
      "4.4 Build the two matrices X and Y\n",
      "CPU times: user 4min 44s, sys: 330 ms, total: 4min 45s\n",
      "Wall time: 2min 43s\n",
      "Total pairs: 3568\n",
      "4.4 Build the two matrices X and Y\n",
      "CPU times: user 26.4 s, sys: 3.2 s, total: 29.6 s\n",
      "Wall time: 1min 13s\n",
      "Total pairs: 3340\n",
      "4.4 Build the two matrices X and Y\n",
      "CPU times: user 19.8 s, sys: 131 ms, total: 19.9 s\n",
      "Wall time: 20.9 s\n"
     ]
    }
   ],
   "source": [
    "# slice  + compute all pairs\n",
    "%time (X_train, Y_train) = p.rnaseq_cross_tissue(annobj, individuals=train,gene_ids=annobj.obs_names)\n",
    "%time (X_valid, Y_valid) = p.rnaseq_cross_tissue(annobj, individuals=valid,gene_ids=annobj.obs_names)\n",
    "%time (X_test, Y_test) = p.rnaseq_cross_tissue(annobj, individuals=test,gene_ids=annobj.obs_names)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene_1</th>\n",
       "      <th>Gene_2</th>\n",
       "      <th>Gene_3</th>\n",
       "      <th>Gene_4</th>\n",
       "      <th>Gene_5</th>\n",
       "      <th>Gene_6</th>\n",
       "      <th>Gene_7</th>\n",
       "      <th>Gene_8</th>\n",
       "      <th>Gene_9</th>\n",
       "      <th>Gene_10</th>\n",
       "      <th>...</th>\n",
       "      <th>Gene_19991</th>\n",
       "      <th>Gene_19992</th>\n",
       "      <th>Gene_19993</th>\n",
       "      <th>Gene_19994</th>\n",
       "      <th>Gene_19995</th>\n",
       "      <th>Gene_19996</th>\n",
       "      <th>Gene_19997</th>\n",
       "      <th>Gene_19998</th>\n",
       "      <th>Gene_19999</th>\n",
       "      <th>Gene_20000</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Patient_1_Blood_Patient_1_Liver</th>\n",
       "      <td>3762</td>\n",
       "      <td>966</td>\n",
       "      <td>974</td>\n",
       "      <td>2223</td>\n",
       "      <td>4763</td>\n",
       "      <td>1724</td>\n",
       "      <td>5866</td>\n",
       "      <td>2025</td>\n",
       "      <td>4844</td>\n",
       "      <td>5496</td>\n",
       "      <td>...</td>\n",
       "      <td>4696</td>\n",
       "      <td>1132</td>\n",
       "      <td>468</td>\n",
       "      <td>775</td>\n",
       "      <td>6098</td>\n",
       "      <td>1150</td>\n",
       "      <td>7856</td>\n",
       "      <td>2462</td>\n",
       "      <td>3628</td>\n",
       "      <td>1910</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Patient_1_Liver_Patient_1_Blood</th>\n",
       "      <td>5954</td>\n",
       "      <td>2390</td>\n",
       "      <td>4451</td>\n",
       "      <td>2333</td>\n",
       "      <td>3371</td>\n",
       "      <td>3715</td>\n",
       "      <td>2096</td>\n",
       "      <td>2483</td>\n",
       "      <td>2641</td>\n",
       "      <td>7564</td>\n",
       "      <td>...</td>\n",
       "      <td>3967</td>\n",
       "      <td>1500</td>\n",
       "      <td>1003</td>\n",
       "      <td>6824</td>\n",
       "      <td>2320</td>\n",
       "      <td>7065</td>\n",
       "      <td>2660</td>\n",
       "      <td>2644</td>\n",
       "      <td>1821</td>\n",
       "      <td>1416</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Patient_2_Testis_Patient_2_Brain</th>\n",
       "      <td>2766</td>\n",
       "      <td>6234</td>\n",
       "      <td>3649</td>\n",
       "      <td>1282</td>\n",
       "      <td>8075</td>\n",
       "      <td>5825</td>\n",
       "      <td>4944</td>\n",
       "      <td>108</td>\n",
       "      <td>1688</td>\n",
       "      <td>1189</td>\n",
       "      <td>...</td>\n",
       "      <td>1150</td>\n",
       "      <td>6609</td>\n",
       "      <td>1465</td>\n",
       "      <td>570</td>\n",
       "      <td>1969</td>\n",
       "      <td>4336</td>\n",
       "      <td>2644</td>\n",
       "      <td>5296</td>\n",
       "      <td>380</td>\n",
       "      <td>1993</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Patient_2_Testis_Patient_2_Liver</th>\n",
       "      <td>2766</td>\n",
       "      <td>6234</td>\n",
       "      <td>3649</td>\n",
       "      <td>1282</td>\n",
       "      <td>8075</td>\n",
       "      <td>5825</td>\n",
       "      <td>4944</td>\n",
       "      <td>108</td>\n",
       "      <td>1688</td>\n",
       "      <td>1189</td>\n",
       "      <td>...</td>\n",
       "      <td>1150</td>\n",
       "      <td>6609</td>\n",
       "      <td>1465</td>\n",
       "      <td>570</td>\n",
       "      <td>1969</td>\n",
       "      <td>4336</td>\n",
       "      <td>2644</td>\n",
       "      <td>5296</td>\n",
       "      <td>380</td>\n",
       "      <td>1993</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Patient_2_Brain_Patient_2_Testis</th>\n",
       "      <td>4545</td>\n",
       "      <td>3354</td>\n",
       "      <td>4068</td>\n",
       "      <td>4765</td>\n",
       "      <td>2463</td>\n",
       "      <td>4179</td>\n",
       "      <td>2773</td>\n",
       "      <td>3208</td>\n",
       "      <td>2042</td>\n",
       "      <td>2168</td>\n",
       "      <td>...</td>\n",
       "      <td>1524</td>\n",
       "      <td>6672</td>\n",
       "      <td>2991</td>\n",
       "      <td>2425</td>\n",
       "      <td>2158</td>\n",
       "      <td>4529</td>\n",
       "      <td>1961</td>\n",
       "      <td>1339</td>\n",
       "      <td>3780</td>\n",
       "      <td>2386</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Patient_2_Brain_Patient_2_Liver</th>\n",
       "      <td>4545</td>\n",
       "      <td>3354</td>\n",
       "      <td>4068</td>\n",
       "      <td>4765</td>\n",
       "      <td>2463</td>\n",
       "      <td>4179</td>\n",
       "      <td>2773</td>\n",
       "      <td>3208</td>\n",
       "      <td>2042</td>\n",
       "      <td>2168</td>\n",
       "      <td>...</td>\n",
       "      <td>1524</td>\n",
       "      <td>6672</td>\n",
       "      <td>2991</td>\n",
       "      <td>2425</td>\n",
       "      <td>2158</td>\n",
       "      <td>4529</td>\n",
       "      <td>1961</td>\n",
       "      <td>1339</td>\n",
       "      <td>3780</td>\n",
       "      <td>2386</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Patient_2_Liver_Patient_2_Testis</th>\n",
       "      <td>5159</td>\n",
       "      <td>1559</td>\n",
       "      <td>2982</td>\n",
       "      <td>6594</td>\n",
       "      <td>2101</td>\n",
       "      <td>3903</td>\n",
       "      <td>1786</td>\n",
       "      <td>3455</td>\n",
       "      <td>1685</td>\n",
       "      <td>1449</td>\n",
       "      <td>...</td>\n",
       "      <td>6190</td>\n",
       "      <td>2781</td>\n",
       "      <td>8495</td>\n",
       "      <td>1543</td>\n",
       "      <td>2569</td>\n",
       "      <td>2710</td>\n",
       "      <td>3536</td>\n",
       "      <td>1233</td>\n",
       "      <td>1911</td>\n",
       "      <td>1513</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Patient_2_Liver_Patient_2_Brain</th>\n",
       "      <td>5159</td>\n",
       "      <td>1559</td>\n",
       "      <td>2982</td>\n",
       "      <td>6594</td>\n",
       "      <td>2101</td>\n",
       "      <td>3903</td>\n",
       "      <td>1786</td>\n",
       "      <td>3455</td>\n",
       "      <td>1685</td>\n",
       "      <td>1449</td>\n",
       "      <td>...</td>\n",
       "      <td>6190</td>\n",
       "      <td>2781</td>\n",
       "      <td>8495</td>\n",
       "      <td>1543</td>\n",
       "      <td>2569</td>\n",
       "      <td>2710</td>\n",
       "      <td>3536</td>\n",
       "      <td>1233</td>\n",
       "      <td>1911</td>\n",
       "      <td>1513</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Patient_3_Heart_Patient_3_Blood</th>\n",
       "      <td>1816</td>\n",
       "      <td>2330</td>\n",
       "      <td>490</td>\n",
       "      <td>2100</td>\n",
       "      <td>4136</td>\n",
       "      <td>3471</td>\n",
       "      <td>2373</td>\n",
       "      <td>1224</td>\n",
       "      <td>2062</td>\n",
       "      <td>6767</td>\n",
       "      <td>...</td>\n",
       "      <td>2462</td>\n",
       "      <td>2730</td>\n",
       "      <td>1484</td>\n",
       "      <td>3102</td>\n",
       "      <td>3307</td>\n",
       "      <td>1028</td>\n",
       "      <td>1685</td>\n",
       "      <td>2926</td>\n",
       "      <td>345</td>\n",
       "      <td>1881</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Patient_3_Heart_Patient_3_Muscle</th>\n",
       "      <td>1816</td>\n",
       "      <td>2330</td>\n",
       "      <td>490</td>\n",
       "      <td>2100</td>\n",
       "      <td>4136</td>\n",
       "      <td>3471</td>\n",
       "      <td>2373</td>\n",
       "      <td>1224</td>\n",
       "      <td>2062</td>\n",
       "      <td>6767</td>\n",
       "      <td>...</td>\n",
       "      <td>2462</td>\n",
       "      <td>2730</td>\n",
       "      <td>1484</td>\n",
       "      <td>3102</td>\n",
       "      <td>3307</td>\n",
       "      <td>1028</td>\n",
       "      <td>1685</td>\n",
       "      <td>2926</td>\n",
       "      <td>345</td>\n",
       "      <td>1881</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10 rows  20000 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                  Gene_1  Gene_2  Gene_3  Gene_4  Gene_5  \\\n",
       "Patient_1_Blood_Patient_1_Liver     3762     966     974    2223    4763   \n",
       "Patient_1_Liver_Patient_1_Blood     5954    2390    4451    2333    3371   \n",
       "Patient_2_Testis_Patient_2_Brain    2766    6234    3649    1282    8075   \n",
       "Patient_2_Testis_Patient_2_Liver    2766    6234    3649    1282    8075   \n",
       "Patient_2_Brain_Patient_2_Testis    4545    3354    4068    4765    2463   \n",
       "Patient_2_Brain_Patient_2_Liver     4545    3354    4068    4765    2463   \n",
       "Patient_2_Liver_Patient_2_Testis    5159    1559    2982    6594    2101   \n",
       "Patient_2_Liver_Patient_2_Brain     5159    1559    2982    6594    2101   \n",
       "Patient_3_Heart_Patient_3_Blood     1816    2330     490    2100    4136   \n",
       "Patient_3_Heart_Patient_3_Muscle    1816    2330     490    2100    4136   \n",
       "\n",
       "                                  Gene_6  Gene_7  Gene_8  Gene_9  Gene_10  \\\n",
       "Patient_1_Blood_Patient_1_Liver     1724    5866    2025    4844     5496   \n",
       "Patient_1_Liver_Patient_1_Blood     3715    2096    2483    2641     7564   \n",
       "Patient_2_Testis_Patient_2_Brain    5825    4944     108    1688     1189   \n",
       "Patient_2_Testis_Patient_2_Liver    5825    4944     108    1688     1189   \n",
       "Patient_2_Brain_Patient_2_Testis    4179    2773    3208    2042     2168   \n",
       "Patient_2_Brain_Patient_2_Liver     4179    2773    3208    2042     2168   \n",
       "Patient_2_Liver_Patient_2_Testis    3903    1786    3455    1685     1449   \n",
       "Patient_2_Liver_Patient_2_Brain     3903    1786    3455    1685     1449   \n",
       "Patient_3_Heart_Patient_3_Blood     3471    2373    1224    2062     6767   \n",
       "Patient_3_Heart_Patient_3_Muscle    3471    2373    1224    2062     6767   \n",
       "\n",
       "                                     ...      Gene_19991  Gene_19992  \\\n",
       "Patient_1_Blood_Patient_1_Liver      ...            4696        1132   \n",
       "Patient_1_Liver_Patient_1_Blood      ...            3967        1500   \n",
       "Patient_2_Testis_Patient_2_Brain     ...            1150        6609   \n",
       "Patient_2_Testis_Patient_2_Liver     ...            1150        6609   \n",
       "Patient_2_Brain_Patient_2_Testis     ...            1524        6672   \n",
       "Patient_2_Brain_Patient_2_Liver      ...            1524        6672   \n",
       "Patient_2_Liver_Patient_2_Testis     ...            6190        2781   \n",
       "Patient_2_Liver_Patient_2_Brain      ...            6190        2781   \n",
       "Patient_3_Heart_Patient_3_Blood      ...            2462        2730   \n",
       "Patient_3_Heart_Patient_3_Muscle     ...            2462        2730   \n",
       "\n",
       "                                  Gene_19993  Gene_19994  Gene_19995  \\\n",
       "Patient_1_Blood_Patient_1_Liver          468         775        6098   \n",
       "Patient_1_Liver_Patient_1_Blood         1003        6824        2320   \n",
       "Patient_2_Testis_Patient_2_Brain        1465         570        1969   \n",
       "Patient_2_Testis_Patient_2_Liver        1465         570        1969   \n",
       "Patient_2_Brain_Patient_2_Testis        2991        2425        2158   \n",
       "Patient_2_Brain_Patient_2_Liver         2991        2425        2158   \n",
       "Patient_2_Liver_Patient_2_Testis        8495        1543        2569   \n",
       "Patient_2_Liver_Patient_2_Brain         8495        1543        2569   \n",
       "Patient_3_Heart_Patient_3_Blood         1484        3102        3307   \n",
       "Patient_3_Heart_Patient_3_Muscle        1484        3102        3307   \n",
       "\n",
       "                                  Gene_19996  Gene_19997  Gene_19998  \\\n",
       "Patient_1_Blood_Patient_1_Liver         1150        7856        2462   \n",
       "Patient_1_Liver_Patient_1_Blood         7065        2660        2644   \n",
       "Patient_2_Testis_Patient_2_Brain        4336        2644        5296   \n",
       "Patient_2_Testis_Patient_2_Liver        4336        2644        5296   \n",
       "Patient_2_Brain_Patient_2_Testis        4529        1961        1339   \n",
       "Patient_2_Brain_Patient_2_Liver         4529        1961        1339   \n",
       "Patient_2_Liver_Patient_2_Testis        2710        3536        1233   \n",
       "Patient_2_Liver_Patient_2_Brain         2710        3536        1233   \n",
       "Patient_3_Heart_Patient_3_Blood         1028        1685        2926   \n",
       "Patient_3_Heart_Patient_3_Muscle        1028        1685        2926   \n",
       "\n",
       "                                  Gene_19999  Gene_20000  \n",
       "Patient_1_Blood_Patient_1_Liver         3628        1910  \n",
       "Patient_1_Liver_Patient_1_Blood         1821        1416  \n",
       "Patient_2_Testis_Patient_2_Brain         380        1993  \n",
       "Patient_2_Testis_Patient_2_Liver         380        1993  \n",
       "Patient_2_Brain_Patient_2_Testis        3780        2386  \n",
       "Patient_2_Brain_Patient_2_Liver         3780        2386  \n",
       "Patient_2_Liver_Patient_2_Testis        1911        1513  \n",
       "Patient_2_Liver_Patient_2_Brain         1911        1513  \n",
       "Patient_3_Heart_Patient_3_Blood          345        1881  \n",
       "Patient_3_Heart_Patient_3_Muscle         345        1881  \n",
       "\n",
       "[10 rows x 20000 columns]"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# train data\n",
    "X_train[:10]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene_1</th>\n",
       "      <th>Gene_2</th>\n",
       "      <th>Gene_3</th>\n",
       "      <th>Gene_4</th>\n",
       "      <th>Gene_5</th>\n",
       "      <th>Gene_6</th>\n",
       "      <th>Gene_7</th>\n",
       "      <th>Gene_8</th>\n",
       "      <th>Gene_9</th>\n",
       "      <th>Gene_10</th>\n",
       "      <th>...</th>\n",
       "      <th>Gene_19991</th>\n",
       "      <th>Gene_19992</th>\n",
       "      <th>Gene_19993</th>\n",
       "      <th>Gene_19994</th>\n",
       "      <th>Gene_19995</th>\n",
       "      <th>Gene_19996</th>\n",
       "      <th>Gene_19997</th>\n",
       "      <th>Gene_19998</th>\n",
       "      <th>Gene_19999</th>\n",
       "      <th>Gene_20000</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Patient_1_Blood_Patient_1_Liver</th>\n",
       "      <td>5954</td>\n",
       "      <td>2390</td>\n",
       "      <td>4451</td>\n",
       "      <td>2333</td>\n",
       "      <td>3371</td>\n",
       "      <td>3715</td>\n",
       "      <td>2096</td>\n",
       "      <td>2483</td>\n",
       "      <td>2641</td>\n",
       "      <td>7564</td>\n",
       "      <td>...</td>\n",
       "      <td>3967</td>\n",
       "      <td>1500</td>\n",
       "      <td>1003</td>\n",
       "      <td>6824</td>\n",
       "      <td>2320</td>\n",
       "      <td>7065</td>\n",
       "      <td>2660</td>\n",
       "      <td>2644</td>\n",
       "      <td>1821</td>\n",
       "      <td>1416</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Patient_1_Liver_Patient_1_Blood</th>\n",
       "      <td>3762</td>\n",
       "      <td>966</td>\n",
       "      <td>974</td>\n",
       "      <td>2223</td>\n",
       "      <td>4763</td>\n",
       "      <td>1724</td>\n",
       "      <td>5866</td>\n",
       "      <td>2025</td>\n",
       "      <td>4844</td>\n",
       "      <td>5496</td>\n",
       "      <td>...</td>\n",
       "      <td>4696</td>\n",
       "      <td>1132</td>\n",
       "      <td>468</td>\n",
       "      <td>775</td>\n",
       "      <td>6098</td>\n",
       "      <td>1150</td>\n",
       "      <td>7856</td>\n",
       "      <td>2462</td>\n",
       "      <td>3628</td>\n",
       "      <td>1910</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Patient_2_Testis_Patient_2_Brain</th>\n",
       "      <td>4545</td>\n",
       "      <td>3354</td>\n",
       "      <td>4068</td>\n",
       "      <td>4765</td>\n",
       "      <td>2463</td>\n",
       "      <td>4179</td>\n",
       "      <td>2773</td>\n",
       "      <td>3208</td>\n",
       "      <td>2042</td>\n",
       "      <td>2168</td>\n",
       "      <td>...</td>\n",
       "      <td>1524</td>\n",
       "      <td>6672</td>\n",
       "      <td>2991</td>\n",
       "      <td>2425</td>\n",
       "      <td>2158</td>\n",
       "      <td>4529</td>\n",
       "      <td>1961</td>\n",
       "      <td>1339</td>\n",
       "      <td>3780</td>\n",
       "      <td>2386</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Patient_2_Testis_Patient_2_Liver</th>\n",
       "      <td>5159</td>\n",
       "      <td>1559</td>\n",
       "      <td>2982</td>\n",
       "      <td>6594</td>\n",
       "      <td>2101</td>\n",
       "      <td>3903</td>\n",
       "      <td>1786</td>\n",
       "      <td>3455</td>\n",
       "      <td>1685</td>\n",
       "      <td>1449</td>\n",
       "      <td>...</td>\n",
       "      <td>6190</td>\n",
       "      <td>2781</td>\n",
       "      <td>8495</td>\n",
       "      <td>1543</td>\n",
       "      <td>2569</td>\n",
       "      <td>2710</td>\n",
       "      <td>3536</td>\n",
       "      <td>1233</td>\n",
       "      <td>1911</td>\n",
       "      <td>1513</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Patient_2_Brain_Patient_2_Testis</th>\n",
       "      <td>2766</td>\n",
       "      <td>6234</td>\n",
       "      <td>3649</td>\n",
       "      <td>1282</td>\n",
       "      <td>8075</td>\n",
       "      <td>5825</td>\n",
       "      <td>4944</td>\n",
       "      <td>108</td>\n",
       "      <td>1688</td>\n",
       "      <td>1189</td>\n",
       "      <td>...</td>\n",
       "      <td>1150</td>\n",
       "      <td>6609</td>\n",
       "      <td>1465</td>\n",
       "      <td>570</td>\n",
       "      <td>1969</td>\n",
       "      <td>4336</td>\n",
       "      <td>2644</td>\n",
       "      <td>5296</td>\n",
       "      <td>380</td>\n",
       "      <td>1993</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Patient_2_Brain_Patient_2_Liver</th>\n",
       "      <td>5159</td>\n",
       "      <td>1559</td>\n",
       "      <td>2982</td>\n",
       "      <td>6594</td>\n",
       "      <td>2101</td>\n",
       "      <td>3903</td>\n",
       "      <td>1786</td>\n",
       "      <td>3455</td>\n",
       "      <td>1685</td>\n",
       "      <td>1449</td>\n",
       "      <td>...</td>\n",
       "      <td>6190</td>\n",
       "      <td>2781</td>\n",
       "      <td>8495</td>\n",
       "      <td>1543</td>\n",
       "      <td>2569</td>\n",
       "      <td>2710</td>\n",
       "      <td>3536</td>\n",
       "      <td>1233</td>\n",
       "      <td>1911</td>\n",
       "      <td>1513</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Patient_2_Liver_Patient_2_Testis</th>\n",
       "      <td>2766</td>\n",
       "      <td>6234</td>\n",
       "      <td>3649</td>\n",
       "      <td>1282</td>\n",
       "      <td>8075</td>\n",
       "      <td>5825</td>\n",
       "      <td>4944</td>\n",
       "      <td>108</td>\n",
       "      <td>1688</td>\n",
       "      <td>1189</td>\n",
       "      <td>...</td>\n",
       "      <td>1150</td>\n",
       "      <td>6609</td>\n",
       "      <td>1465</td>\n",
       "      <td>570</td>\n",
       "      <td>1969</td>\n",
       "      <td>4336</td>\n",
       "      <td>2644</td>\n",
       "      <td>5296</td>\n",
       "      <td>380</td>\n",
       "      <td>1993</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Patient_2_Liver_Patient_2_Brain</th>\n",
       "      <td>4545</td>\n",
       "      <td>3354</td>\n",
       "      <td>4068</td>\n",
       "      <td>4765</td>\n",
       "      <td>2463</td>\n",
       "      <td>4179</td>\n",
       "      <td>2773</td>\n",
       "      <td>3208</td>\n",
       "      <td>2042</td>\n",
       "      <td>2168</td>\n",
       "      <td>...</td>\n",
       "      <td>1524</td>\n",
       "      <td>6672</td>\n",
       "      <td>2991</td>\n",
       "      <td>2425</td>\n",
       "      <td>2158</td>\n",
       "      <td>4529</td>\n",
       "      <td>1961</td>\n",
       "      <td>1339</td>\n",
       "      <td>3780</td>\n",
       "      <td>2386</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>8 rows  20000 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                  Gene_1  Gene_2  Gene_3  Gene_4  Gene_5  \\\n",
       "Patient_1_Blood_Patient_1_Liver     5954    2390    4451    2333    3371   \n",
       "Patient_1_Liver_Patient_1_Blood     3762     966     974    2223    4763   \n",
       "Patient_2_Testis_Patient_2_Brain    4545    3354    4068    4765    2463   \n",
       "Patient_2_Testis_Patient_2_Liver    5159    1559    2982    6594    2101   \n",
       "Patient_2_Brain_Patient_2_Testis    2766    6234    3649    1282    8075   \n",
       "Patient_2_Brain_Patient_2_Liver     5159    1559    2982    6594    2101   \n",
       "Patient_2_Liver_Patient_2_Testis    2766    6234    3649    1282    8075   \n",
       "Patient_2_Liver_Patient_2_Brain     4545    3354    4068    4765    2463   \n",
       "\n",
       "                                  Gene_6  Gene_7  Gene_8  Gene_9  Gene_10  \\\n",
       "Patient_1_Blood_Patient_1_Liver     3715    2096    2483    2641     7564   \n",
       "Patient_1_Liver_Patient_1_Blood     1724    5866    2025    4844     5496   \n",
       "Patient_2_Testis_Patient_2_Brain    4179    2773    3208    2042     2168   \n",
       "Patient_2_Testis_Patient_2_Liver    3903    1786    3455    1685     1449   \n",
       "Patient_2_Brain_Patient_2_Testis    5825    4944     108    1688     1189   \n",
       "Patient_2_Brain_Patient_2_Liver     3903    1786    3455    1685     1449   \n",
       "Patient_2_Liver_Patient_2_Testis    5825    4944     108    1688     1189   \n",
       "Patient_2_Liver_Patient_2_Brain     4179    2773    3208    2042     2168   \n",
       "\n",
       "                                     ...      Gene_19991  Gene_19992  \\\n",
       "Patient_1_Blood_Patient_1_Liver      ...            3967        1500   \n",
       "Patient_1_Liver_Patient_1_Blood      ...            4696        1132   \n",
       "Patient_2_Testis_Patient_2_Brain     ...            1524        6672   \n",
       "Patient_2_Testis_Patient_2_Liver     ...            6190        2781   \n",
       "Patient_2_Brain_Patient_2_Testis     ...            1150        6609   \n",
       "Patient_2_Brain_Patient_2_Liver      ...            6190        2781   \n",
       "Patient_2_Liver_Patient_2_Testis     ...            1150        6609   \n",
       "Patient_2_Liver_Patient_2_Brain      ...            1524        6672   \n",
       "\n",
       "                                  Gene_19993  Gene_19994  Gene_19995  \\\n",
       "Patient_1_Blood_Patient_1_Liver         1003        6824        2320   \n",
       "Patient_1_Liver_Patient_1_Blood          468         775        6098   \n",
       "Patient_2_Testis_Patient_2_Brain        2991        2425        2158   \n",
       "Patient_2_Testis_Patient_2_Liver        8495        1543        2569   \n",
       "Patient_2_Brain_Patient_2_Testis        1465         570        1969   \n",
       "Patient_2_Brain_Patient_2_Liver         8495        1543        2569   \n",
       "Patient_2_Liver_Patient_2_Testis        1465         570        1969   \n",
       "Patient_2_Liver_Patient_2_Brain         2991        2425        2158   \n",
       "\n",
       "                                  Gene_19996  Gene_19997  Gene_19998  \\\n",
       "Patient_1_Blood_Patient_1_Liver         7065        2660        2644   \n",
       "Patient_1_Liver_Patient_1_Blood         1150        7856        2462   \n",
       "Patient_2_Testis_Patient_2_Brain        4529        1961        1339   \n",
       "Patient_2_Testis_Patient_2_Liver        2710        3536        1233   \n",
       "Patient_2_Brain_Patient_2_Testis        4336        2644        5296   \n",
       "Patient_2_Brain_Patient_2_Liver         2710        3536        1233   \n",
       "Patient_2_Liver_Patient_2_Testis        4336        2644        5296   \n",
       "Patient_2_Liver_Patient_2_Brain         4529        1961        1339   \n",
       "\n",
       "                                  Gene_19999  Gene_20000  \n",
       "Patient_1_Blood_Patient_1_Liver         1821        1416  \n",
       "Patient_1_Liver_Patient_1_Blood         3628        1910  \n",
       "Patient_2_Testis_Patient_2_Brain        3780        2386  \n",
       "Patient_2_Testis_Patient_2_Liver        1911        1513  \n",
       "Patient_2_Brain_Patient_2_Testis         380        1993  \n",
       "Patient_2_Brain_Patient_2_Liver         1911        1513  \n",
       "Patient_2_Liver_Patient_2_Testis         380        1993  \n",
       "Patient_2_Liver_Patient_2_Brain         3780        2386  \n",
       "\n",
       "[8 rows x 20000 columns]"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Y_train[:10]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python (rep)",
   "language": "python",
   "name": "rep"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
