{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Goal\n",
    "Preprocessing rnaseq data\n",
    "\n",
    "This is a tutorial with the main workflow for preparing a training dataset for a basic linear regression model. \n",
    "The output (X,Y) are all pair combination between tissues per each sample, where X is input, Y target."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The autoreload extension is already loaded. To reload it, use:\n",
      "  %reload_ext autoreload\n"
     ]
    }
   ],
   "source": [
    "# code autoreload\n",
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: numpy in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (1.15.4)\n",
      "Requirement already satisfied: pandas in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (0.23.4)\n",
      "Requirement already satisfied: numpy>=1.9.0 in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (from pandas) (1.15.4)\n",
      "Requirement already satisfied: python-dateutil>=2.5.0 in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (from pandas) (2.7.5)\n",
      "Requirement already satisfied: pytz>=2011k in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (from pandas) (2018.7)\n",
      "Requirement already satisfied: six>=1.5 in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (from python-dateutil>=2.5.0->pandas) (1.12.0)\n",
      "Requirement already satisfied: anndata in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (0.6.16)\n",
      "Requirement already satisfied: pandas>=0.21.0 in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (from anndata) (0.23.4)\n",
      "Requirement already satisfied: scipy~=1.0 in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (from anndata) (1.2.0)\n",
      "Requirement already satisfied: natsort in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (from anndata) (5.5.0)\n",
      "Requirement already satisfied: numpy~=1.14 in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (from anndata) (1.15.4)\n",
      "Requirement already satisfied: h5py in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (from anndata) (2.9.0)\n",
      "Requirement already satisfied: python-dateutil>=2.5.0 in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (from pandas>=0.21.0->anndata) (2.7.5)\n",
      "Requirement already satisfied: pytz>=2011k in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (from pandas>=0.21.0->anndata) (2018.7)\n",
      "Requirement already satisfied: six in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (from h5py->anndata) (1.12.0)\n",
      "Requirement already satisfied: scanpy in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (1.3.6)\n",
      "Requirement already satisfied: natsort in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (from scanpy) (5.5.0)\n",
      "Requirement already satisfied: networkx in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (from scanpy) (2.2)\n",
      "Requirement already satisfied: seaborn in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (from scanpy) (0.9.0)\n",
      "Requirement already satisfied: joblib in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (from scanpy) (0.13.0)\n",
      "Requirement already satisfied: numba>=0.40.0 in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (from scanpy) (0.41.0)\n",
      "Requirement already satisfied: scipy in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (from scanpy) (1.2.0)\n",
      "Requirement already satisfied: scikit-learn>=0.19.1 in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (from scanpy) (0.20.2)\n",
      "Requirement already satisfied: pandas>=0.21 in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (from scanpy) (0.23.4)\n",
      "Requirement already satisfied: h5py in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (from scanpy) (2.9.0)\n",
      "Requirement already satisfied: matplotlib>=2.2 in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (from scanpy) (3.0.2)\n",
      "Requirement already satisfied: statsmodels in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (from scanpy) (0.9.0)\n",
      "Requirement already satisfied: anndata>=0.6.13 in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (from scanpy) (0.6.16)\n",
      "Requirement already satisfied: tables in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (from scanpy) (3.4.4)\n",
      "Requirement already satisfied: decorator>=4.3.0 in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (from networkx->scanpy) (4.3.0)\n",
      "Requirement already satisfied: numpy>=1.9.3 in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (from seaborn->scanpy) (1.15.4)\n",
      "Requirement already satisfied: llvmlite>=0.26.0dev0 in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (from numba>=0.40.0->scanpy) (0.26.0)\n",
      "Requirement already satisfied: pytz>=2011k in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (from pandas>=0.21->scanpy) (2018.7)\n",
      "Requirement already satisfied: python-dateutil>=2.5.0 in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (from pandas>=0.21->scanpy) (2.7.5)\n",
      "Requirement already satisfied: six in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (from h5py->scanpy) (1.12.0)\n",
      "Requirement already satisfied: cycler>=0.10 in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (from matplotlib>=2.2->scanpy) (0.10.0)\n",
      "Requirement already satisfied: pyparsing!=2.0.4,!=2.1.2,!=2.1.6,>=2.0.1 in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (from matplotlib>=2.2->scanpy) (2.3.0)\n",
      "Requirement already satisfied: kiwisolver>=1.0.1 in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (from matplotlib>=2.2->scanpy) (1.0.1)\n",
      "Requirement already satisfied: patsy in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (from statsmodels->scanpy) (0.5.1)\n",
      "Requirement already satisfied: numexpr>=2.5.2 in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (from tables->scanpy) (2.6.8)\n",
      "Requirement already satisfied: setuptools in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (from kiwisolver>=1.0.1->matplotlib>=2.2->scanpy) (40.6.3)\n",
      "Requirement already satisfied: argh in /data/nasif12/modules_if12/SL7/i12g/anaconda/3-5.0.1/envs/rep/lib/python3.6/site-packages (0.26.2)\n"
     ]
    }
   ],
   "source": [
    "!pip install numpy\n",
    "!pip install pandas\n",
    "!pip install anndata\n",
    "\n",
    "!pip install scanpy\n",
    "!pip install argh"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Modules to reload:\n",
      "all-except-skipped\n",
      "\n",
      "Modules to skip:\n",
      "\n"
     ]
    }
   ],
   "source": [
    "import rep.preprocessing as p\n",
    "from rep.constants import ANNDATA_CST as a\n",
    "\n",
    "%aimport"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 1. Load raw count matrix + annotation and put it into an annodata object"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting data.csv\n"
     ]
    }
   ],
   "source": [
    "%%writefile data.csv\n",
    "T1_s1,T2_s2,T3_s1,T4_s2,T5_s1,T6_s1,T1_s3\n",
    "G1,10,20,30,40,5,6,23\n",
    "G2,5,10,20,30,5,6,24\n",
    "G3,6,7,8,9,1,1,25"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting anno_var.csv\n"
     ]
    }
   ],
   "source": [
    "%%writefile anno_var.csv\n",
    "Sample,Gender,Tissue,Individual,Seq\n",
    "T1_s1,F,Tissue1,Sample1,WGS\n",
    "T2_s2,M,Tissue2,Sample2,WXS\n",
    "T3_s1,F,Tissue3,Sample1,WGS\n",
    "T4_s2,M,Tissue4,Sample2,WXS\n",
    "T5_s1,M,Tissue5,Sample1,WGS\n",
    "T6_s1,M,Tissue6,Sample1,WGS\n",
    "T1_s3,F,Tissue1,Sample3,WGS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting anno_obs.csv\n"
     ]
    }
   ],
   "source": [
    "%%writefile anno_obs.csv\n",
    "Gene,assembly,transcript,chromosome,start,strand\n",
    "G1,hg19,T1,chr1,1111,-\n",
    "G2,hg19,T2,chr2,2222,-\n",
    "G3,hg19,T3,chr3,3333,-"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "anndata.X ----\n",
      "[[10. 20. 30. 40.  5.  6. 23.]\n",
      " [ 5. 10. 20. 30.  5.  6. 24.]\n",
      " [ 6.  7.  8.  9.  1.  1. 25.]]\n",
      "anndata.var ----\n",
      "      Gender   Tissue Individual  Seq\n",
      "T1_s1      F  Tissue1    Sample1  WGS\n",
      "T2_s2      M  Tissue2    Sample2  WXS\n",
      "T3_s1      F  Tissue3    Sample1  WGS\n",
      "T4_s2      M  Tissue4    Sample2  WXS\n",
      "T5_s1      M  Tissue5    Sample1  WGS\n",
      "anndata.obs ----\n",
      "   assembly transcript chromosome  start strand\n",
      "G1     hg19         T1       chr1   1111      -\n",
      "G2     hg19         T2       chr2   2222      -\n",
      "G3     hg19         T3       chr3   3333      -\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# annobj is the equivalent of the summarized experiments in R. \n",
    "# It contains \n",
    "# - annobj.X - counts\n",
    "# - annobj.var - data.frame annotation object for the columns of the count matrix (in our case Tissue_Sample)\n",
    "# - annobj.obs - data.frame annotation object for the rows of the count matrix (= features that we measure, in our case Genes)\n",
    "# \n",
    "\n",
    "annobj = p.create_anndata(\"data.csv\", sep=\",\", samples_anno=\"anno_var.csv\", genes_anno=\"anno_obs.csv\")\n",
    "p.print_anndata(annobj)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Save data to h2ad format"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "... storing 'assembly' as categorical\n",
      "... storing 'strand' as categorical\n",
      "... storing 'Gender' as categorical\n",
      "... storing 'Individual' as categorical\n",
      "... storing 'Technology' as categorical\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "'/data/nasif12/home_if12/giurgiu/rep_gagneur/rep/notebooks/preprocessing/tmp1547121923.h5ad'"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "name = p.save(annobj)\n",
    "name"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 2. Prepare training sample:\n",
    "- select genes and tissue_samples for training\n",
    "- slice the anndata object\n",
    "- compute the cross tissue matrix pair (X,Y)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(['T4_s2', 'T6_s1', 'T5_s1', 'T2_s2'], ['G2', 'G1'])"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# filtering by different values in anndata.var and anndata.obs\n",
    "# stored in a json format by specifying the {column over we would like to filter : array of admitted values}\n",
    "# in this example we filter the anndata.var by gender (M) and anndata.obs should contain only G1 and G2\n",
    "(var, obs) = p.filter_anndata_by_value(annobj, {a.SAMPLES: {'Gender': ['M']}, a.GENES: {0: ['G1', 'G2']}})\n",
    "\n",
    "# these are the rows and columns selected for the training set\n",
    "(var,obs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'Sample1': ['T6_s1', 'T5_s1'], 'Sample2': ['T4_s2', 'T2_s2']}"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# since we want to impute gene expression across different tissues, we\n",
    "# would like to compute the cross tissue pairs per sample and not across samples\n",
    "\n",
    "# for this purpose we can use the group_by(df, column, index_subset) \n",
    "# function to group tissues by sample\n",
    "\n",
    "dict_samples_var = p.group_by(annobj.var, 'Individual', var)\n",
    "dict_samples_var"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total individuals: 4\n",
      "Individual split before balancing:  3 0 1\n",
      "Iteration:  1\n",
      "\tExpc counts:  [4, 1, 1]\n",
      "\tReal counts:  6 0 1\n",
      "\tExpc fraction:  [0.6, 0.2, 0.2]\n",
      "\tReal fraction:  0.8571428571428571 0.0 0.14285714285714285\n",
      "\tDiff:  [0.2571428571428571, -0.2, -0.05714285714285716]\n",
      "remove best\n",
      "Min error:  0.02857142857142858\n",
      "Sample2\n",
      "Iteration:  2\n",
      "\tExpc counts:  [4, 1, 1]\n",
      "\tReal counts:  4 2 1\n",
      "\tExpc fraction:  [0.6, 0.2, 0.2]\n",
      "\tReal fraction:  0.5714285714285714 0.2857142857142857 0.14285714285714285\n",
      "\tDiff:  [-0.02857142857142858, 0.08571428571428569, -0.05714285714285716]\n",
      "remove best\n",
      "Min error:  0.2\n",
      "Sample2\n",
      "Iteration:  3\n",
      "\tExpc counts:  [4, 1, 1]\n",
      "\tReal counts:  4 0 3\n",
      "\tExpc fraction:  [0.6, 0.2, 0.2]\n",
      "\tReal fraction:  0.5714285714285714 0.0 0.42857142857142855\n",
      "\tDiff:  [-0.02857142857142858, -0.2, 0.22857142857142854]\n",
      "remove best\n",
      "Min error:  0.05714285714285716\n",
      "Sample2\n",
      "Iteration:  4\n",
      "\tExpc counts:  [4, 1, 1]\n",
      "\tReal counts:  6 0 1\n",
      "\tExpc fraction:  [0.6, 0.2, 0.2]\n",
      "\tReal fraction:  0.8571428571428571 0.0 0.14285714285714285\n",
      "\tDiff:  [0.2571428571428571, -0.2, -0.05714285714285716]\n",
      "remove best\n",
      "Min error:  0.02857142857142858\n",
      "Sample2\n",
      "Iteration:  5\n",
      "\tExpc counts:  [4, 1, 1]\n",
      "\tReal counts:  4 2 1\n",
      "\tExpc fraction:  [0.6, 0.2, 0.2]\n",
      "\tReal fraction:  0.5714285714285714 0.2857142857142857 0.14285714285714285\n",
      "\tDiff:  [-0.02857142857142858, 0.08571428571428569, -0.05714285714285716]\n",
      "remove best\n",
      "Min error:  0.2\n",
      "Sample2\n",
      "Iteration:  6\n",
      "\tExpc counts:  [4, 1, 1]\n",
      "\tReal counts:  4 0 3\n",
      "\tExpc fraction:  [0.6, 0.2, 0.2]\n",
      "\tReal fraction:  0.5714285714285714 0.0 0.42857142857142855\n",
      "\tDiff:  [-0.02857142857142858, -0.2, 0.22857142857142854]\n",
      "remove best\n",
      "Min error:  0.05714285714285716\n",
      "Sample2\n",
      "Iteration:  7\n",
      "\tExpc counts:  [4, 1, 1]\n",
      "\tReal counts:  6 0 1\n",
      "\tExpc fraction:  [0.6, 0.2, 0.2]\n",
      "\tReal fraction:  0.8571428571428571 0.0 0.14285714285714285\n",
      "\tDiff:  [0.2571428571428571, -0.2, -0.05714285714285716]\n",
      "remove best\n",
      "Min error:  0.02857142857142858\n",
      "Sample2\n",
      "Iteration:  8\n",
      "\tExpc counts:  [4, 1, 1]\n",
      "\tReal counts:  4 2 1\n",
      "\tExpc fraction:  [0.6, 0.2, 0.2]\n",
      "\tReal fraction:  0.5714285714285714 0.2857142857142857 0.14285714285714285\n",
      "\tDiff:  [-0.02857142857142858, 0.08571428571428569, -0.05714285714285716]\n",
      "remove best\n",
      "Min error:  0.2\n",
      "Sample2\n",
      "Iteration:  9\n",
      "\tExpc counts:  [4, 1, 1]\n",
      "\tReal counts:  4 0 3\n",
      "\tExpc fraction:  [0.6, 0.2, 0.2]\n",
      "\tReal fraction:  0.5714285714285714 0.0 0.42857142857142855\n",
      "\tDiff:  [-0.02857142857142858, -0.2, 0.22857142857142854]\n",
      "remove best\n",
      "Min error:  0.05714285714285716\n",
      "Sample2\n",
      "Iteration:  10\n",
      "\tExpc counts:  [4, 1, 1]\n",
      "\tReal counts:  6 0 1\n",
      "\tExpc fraction:  [0.6, 0.2, 0.2]\n",
      "\tReal fraction:  0.8571428571428571 0.0 0.14285714285714285\n",
      "\tDiff:  [0.2571428571428571, -0.2, -0.05714285714285716]\n",
      "remove best\n",
      "Min error:  0.02857142857142858\n",
      "Sample2\n",
      "Iteration:  11\n",
      "\tExpc counts:  [4, 1, 1]\n",
      "\tReal counts:  4 2 1\n",
      "\tExpc fraction:  [0.6, 0.2, 0.2]\n",
      "\tReal fraction:  0.5714285714285714 0.2857142857142857 0.14285714285714285\n",
      "\tDiff:  [-0.02857142857142858, 0.08571428571428569, -0.05714285714285716]\n",
      "remove best\n",
      "Min error:  0.2\n",
      "Sample2\n",
      "Iteration:  12\n",
      "\tExpc counts:  [4, 1, 1]\n",
      "\tReal counts:  4 0 3\n",
      "\tExpc fraction:  [0.6, 0.2, 0.2]\n",
      "\tReal fraction:  0.5714285714285714 0.0 0.42857142857142855\n",
      "\tDiff:  [-0.02857142857142858, -0.2, 0.22857142857142854]\n",
      "remove best\n",
      "Min error:  0.05714285714285716\n",
      "Sample2\n",
      "Iteration:  13\n",
      "\tExpc counts:  [4, 1, 1]\n",
      "\tReal counts:  6 0 1\n",
      "\tExpc fraction:  [0.6, 0.2, 0.2]\n",
      "\tReal fraction:  0.8571428571428571 0.0 0.14285714285714285\n",
      "\tDiff:  [0.2571428571428571, -0.2, -0.05714285714285716]\n",
      "remove best\n",
      "Min error:  0.02857142857142858\n",
      "Sample2\n",
      "Iteration:  14\n",
      "\tExpc counts:  [4, 1, 1]\n",
      "\tReal counts:  4 2 1\n",
      "\tExpc fraction:  [0.6, 0.2, 0.2]\n",
      "\tReal fraction:  0.5714285714285714 0.2857142857142857 0.14285714285714285\n",
      "\tDiff:  [-0.02857142857142858, 0.08571428571428569, -0.05714285714285716]\n",
      "remove best\n",
      "Min error:  0.2\n",
      "Sample2\n",
      "Iteration:  15\n",
      "\tExpc counts:  [4, 1, 1]\n",
      "\tReal counts:  4 0 3\n",
      "\tExpc fraction:  [0.6, 0.2, 0.2]\n",
      "\tReal fraction:  0.5714285714285714 0.0 0.42857142857142855\n",
      "\tDiff:  [-0.02857142857142858, -0.2, 0.22857142857142854]\n",
      "remove best\n",
      "Min error:  0.05714285714285716\n",
      "Sample2\n",
      "Iteration:  16\n",
      "\tExpc counts:  [4, 1, 1]\n",
      "\tReal counts:  6 0 1\n",
      "\tExpc fraction:  [0.6, 0.2, 0.2]\n",
      "\tReal fraction:  0.8571428571428571 0.0 0.14285714285714285\n",
      "\tDiff:  [0.2571428571428571, -0.2, -0.05714285714285716]\n",
      "remove best\n",
      "Min error:  0.02857142857142858\n",
      "Sample2\n",
      "Iteration:  17\n",
      "\tExpc counts:  [4, 1, 1]\n",
      "\tReal counts:  4 2 1\n",
      "\tExpc fraction:  [0.6, 0.2, 0.2]\n",
      "\tReal fraction:  0.5714285714285714 0.2857142857142857 0.14285714285714285\n",
      "\tDiff:  [-0.02857142857142858, 0.08571428571428569, -0.05714285714285716]\n",
      "remove best\n",
      "Min error:  0.2\n",
      "Sample2\n",
      "Iteration:  18\n",
      "\tExpc counts:  [4, 1, 1]\n",
      "\tReal counts:  4 0 3\n",
      "\tExpc fraction:  [0.6, 0.2, 0.2]\n",
      "\tReal fraction:  0.5714285714285714 0.0 0.42857142857142855\n",
      "\tDiff:  [-0.02857142857142858, -0.2, 0.22857142857142854]\n",
      "remove best\n",
      "Min error:  0.05714285714285716\n",
      "Sample2\n",
      "Iteration:  19\n",
      "\tExpc counts:  [4, 1, 1]\n",
      "\tReal counts:  6 0 1\n",
      "\tExpc fraction:  [0.6, 0.2, 0.2]\n",
      "\tReal fraction:  0.8571428571428571 0.0 0.14285714285714285\n",
      "\tDiff:  [0.2571428571428571, -0.2, -0.05714285714285716]\n",
      "remove best\n",
      "Min error:  0.02857142857142858\n",
      "Sample2\n",
      "Iteration:  20\n",
      "\tExpc counts:  [4, 1, 1]\n",
      "\tReal counts:  4 2 1\n",
      "\tExpc fraction:  [0.6, 0.2, 0.2]\n",
      "\tReal fraction:  0.5714285714285714 0.2857142857142857 0.14285714285714285\n",
      "\tDiff:  [-0.02857142857142858, 0.08571428571428569, -0.05714285714285716]\n",
      "remove best\n",
      "Min error:  0.2\n",
      "Sample2\n",
      "Iteration:  21\n",
      "\tExpc counts:  [4, 1, 1]\n",
      "\tReal counts:  4 0 3\n",
      "\tExpc fraction:  [0.6, 0.2, 0.2]\n",
      "\tReal fraction:  0.5714285714285714 0.0 0.42857142857142855\n",
      "\tDiff:  [-0.02857142857142858, -0.2, 0.22857142857142854]\n",
      "remove best\n",
      "Min error:  0.05714285714285716\n",
      "Sample2\n",
      "Iteration:  22\n",
      "\tExpc counts:  [4, 1, 1]\n",
      "\tReal counts:  6 0 1\n",
      "\tExpc fraction:  [0.6, 0.2, 0.2]\n",
      "\tReal fraction:  0.8571428571428571 0.0 0.14285714285714285\n",
      "\tDiff:  [0.2571428571428571, -0.2, -0.05714285714285716]\n",
      "remove best\n",
      "Min error:  0.02857142857142858\n",
      "Sample2\n",
      "Iteration:  23\n",
      "\tExpc counts:  [4, 1, 1]\n",
      "\tReal counts:  4 2 1\n",
      "\tExpc fraction:  [0.6, 0.2, 0.2]\n",
      "\tReal fraction:  0.5714285714285714 0.2857142857142857 0.14285714285714285\n",
      "\tDiff:  [-0.02857142857142858, 0.08571428571428569, -0.05714285714285716]\n",
      "remove best\n",
      "Min error:  0.2\n",
      "Sample2\n",
      "Iteration:  24\n",
      "\tExpc counts:  [4, 1, 1]\n",
      "\tReal counts:  4 0 3\n",
      "\tExpc fraction:  [0.6, 0.2, 0.2]\n",
      "\tReal fraction:  0.5714285714285714 0.0 0.42857142857142855\n",
      "\tDiff:  [-0.02857142857142858, -0.2, 0.22857142857142854]\n",
      "remove best\n",
      "Min error:  0.05714285714285716\n",
      "Sample2\n",
      "Iteration:  25\n",
      "\tExpc counts:  [4, 1, 1]\n",
      "\tReal counts:  6 0 1\n",
      "\tExpc fraction:  [0.6, 0.2, 0.2]\n",
      "\tReal fraction:  0.8571428571428571 0.0 0.14285714285714285\n",
      "\tDiff:  [0.2571428571428571, -0.2, -0.05714285714285716]\n",
      "remove best\n",
      "Min error:  0.02857142857142858\n",
      "Sample2\n",
      "Iteration:  26\n",
      "\tExpc counts:  [4, 1, 1]\n",
      "\tReal counts:  4 2 1\n",
      "\tExpc fraction:  [0.6, 0.2, 0.2]\n",
      "\tReal fraction:  0.5714285714285714 0.2857142857142857 0.14285714285714285\n",
      "\tDiff:  [-0.02857142857142858, 0.08571428571428569, -0.05714285714285716]\n",
      "remove best\n",
      "Min error:  0.2\n",
      "Sample2\n",
      "Iteration:  27\n",
      "\tExpc counts:  [4, 1, 1]\n",
      "\tReal counts:  4 0 3\n",
      "\tExpc fraction:  [0.6, 0.2, 0.2]\n",
      "\tReal fraction:  0.5714285714285714 0.0 0.42857142857142855\n",
      "\tDiff:  [-0.02857142857142858, -0.2, 0.22857142857142854]\n",
      "remove best\n",
      "Min error:  0.05714285714285716\n",
      "Sample2\n",
      "Iteration:  28\n",
      "\tExpc counts:  [4, 1, 1]\n",
      "\tReal counts:  6 0 1\n",
      "\tExpc fraction:  [0.6, 0.2, 0.2]\n",
      "\tReal fraction:  0.8571428571428571 0.0 0.14285714285714285\n",
      "\tDiff:  [0.2571428571428571, -0.2, -0.05714285714285716]\n",
      "remove best\n",
      "Min error:  0.02857142857142858\n",
      "Sample2\n",
      "Iteration:  29\n",
      "\tExpc counts:  [4, 1, 1]\n",
      "\tReal counts:  4 2 1\n",
      "\tExpc fraction:  [0.6, 0.2, 0.2]\n",
      "\tReal fraction:  0.5714285714285714 0.2857142857142857 0.14285714285714285\n",
      "\tDiff:  [-0.02857142857142858, 0.08571428571428569, -0.05714285714285716]\n",
      "remove best\n",
      "Min error:  0.2\n",
      "Sample2\n",
      "Iteration:  30\n",
      "\tExpc counts:  [4, 1, 1]\n",
      "\tReal counts:  4 0 3\n",
      "\tExpc fraction:  [0.6, 0.2, 0.2]\n",
      "\tReal fraction:  0.5714285714285714 0.0 0.42857142857142855\n",
      "\tDiff:  [-0.02857142857142858, -0.2, 0.22857142857142854]\n",
      "remove best\n",
      "Min error:  0.05714285714285716\n",
      "Sample2\n",
      "Iteration:  31\n",
      "\tExpc counts:  [4, 1, 1]\n",
      "\tReal counts:  6 0 1\n",
      "\tExpc fraction:  [0.6, 0.2, 0.2]\n",
      "\tReal fraction:  0.8571428571428571 0.0 0.14285714285714285\n",
      "\tDiff:  [0.2571428571428571, -0.2, -0.05714285714285716]\n",
      "remove best\n",
      "Min error:  0.02857142857142858\n",
      "Sample2\n",
      "Iteration:  32\n",
      "\tExpc counts:  [4, 1, 1]\n",
      "\tReal counts:  4 2 1\n",
      "\tExpc fraction:  [0.6, 0.2, 0.2]\n",
      "\tReal fraction:  0.5714285714285714 0.2857142857142857 0.14285714285714285\n",
      "\tDiff:  [-0.02857142857142858, 0.08571428571428569, -0.05714285714285716]\n",
      "remove best\n",
      "Min error:  0.2\n",
      "Sample2\n",
      "Iteration:  33\n",
      "\tExpc counts:  [4, 1, 1]\n",
      "\tReal counts:  4 0 3\n",
      "\tExpc fraction:  [0.6, 0.2, 0.2]\n",
      "\tReal fraction:  0.5714285714285714 0.0 0.42857142857142855\n",
      "\tDiff:  [-0.02857142857142858, -0.2, 0.22857142857142854]\n",
      "remove best\n",
      "Min error:  0.05714285714285716\n",
      "Sample2\n",
      "Iteration:  34\n",
      "\tExpc counts:  [4, 1, 1]\n",
      "\tReal counts:  6 0 1\n",
      "\tExpc fraction:  [0.6, 0.2, 0.2]\n",
      "\tReal fraction:  0.8571428571428571 0.0 0.14285714285714285\n",
      "\tDiff:  [0.2571428571428571, -0.2, -0.05714285714285716]\n",
      "remove best\n",
      "Min error:  0.02857142857142858\n",
      "Sample2\n",
      "Iteration:  35\n",
      "\tExpc counts:  [4, 1, 1]\n",
      "\tReal counts:  4 2 1\n",
      "\tExpc fraction:  [0.6, 0.2, 0.2]\n",
      "\tReal fraction:  0.5714285714285714 0.2857142857142857 0.14285714285714285\n",
      "\tDiff:  [-0.02857142857142858, 0.08571428571428569, -0.05714285714285716]\n",
      "remove best\n",
      "Min error:  0.2\n",
      "Sample2\n",
      "Iteration:  36\n",
      "\tExpc counts:  [4, 1, 1]\n",
      "\tReal counts:  4 0 3\n",
      "\tExpc fraction:  [0.6, 0.2, 0.2]\n",
      "\tReal fraction:  0.5714285714285714 0.0 0.42857142857142855\n",
      "\tDiff:  [-0.02857142857142858, -0.2, 0.22857142857142854]\n",
      "remove best\n",
      "Min error:  0.05714285714285716\n",
      "Sample2\n",
      "Iteration:  37\n",
      "\tExpc counts:  [4, 1, 1]\n",
      "\tReal counts:  6 0 1\n",
      "\tExpc fraction:  [0.6, 0.2, 0.2]\n",
      "\tReal fraction:  0.8571428571428571 0.0 0.14285714285714285\n",
      "\tDiff:  [0.2571428571428571, -0.2, -0.05714285714285716]\n",
      "remove best\n",
      "Min error:  0.02857142857142858\n",
      "Sample2\n",
      "Iteration:  38\n",
      "\tExpc counts:  [4, 1, 1]\n",
      "\tReal counts:  4 2 1\n",
      "\tExpc fraction:  [0.6, 0.2, 0.2]\n",
      "\tReal fraction:  0.5714285714285714 0.2857142857142857 0.14285714285714285\n",
      "\tDiff:  [-0.02857142857142858, 0.08571428571428569, -0.05714285714285716]\n",
      "remove best\n",
      "Min error:  0.2\n",
      "Sample2\n",
      "Iteration:  39\n",
      "\tExpc counts:  [4, 1, 1]\n",
      "\tReal counts:  4 0 3\n",
      "\tExpc fraction:  [0.6, 0.2, 0.2]\n",
      "\tReal fraction:  0.5714285714285714 0.0 0.42857142857142855\n",
      "\tDiff:  [-0.02857142857142858, -0.2, 0.22857142857142854]\n",
      "remove best\n",
      "Min error:  0.05714285714285716\n",
      "Sample2\n",
      "Iteration:  40\n",
      "\tExpc counts:  [4, 1, 1]\n",
      "\tReal counts:  6 0 1\n",
      "\tExpc fraction:  [0.6, 0.2, 0.2]\n",
      "\tReal fraction:  0.8571428571428571 0.0 0.14285714285714285\n",
      "\tDiff:  [0.2571428571428571, -0.2, -0.05714285714285716]\n",
      "remove best\n",
      "Min error:  0.02857142857142858\n",
      "Sample2\n",
      "Iteration:  41\n",
      "\tExpc counts:  [4, 1, 1]\n",
      "\tReal counts:  4 2 1\n",
      "\tExpc fraction:  [0.6, 0.2, 0.2]\n",
      "\tReal fraction:  0.5714285714285714 0.2857142857142857 0.14285714285714285\n",
      "\tDiff:  [-0.02857142857142858, 0.08571428571428569, -0.05714285714285716]\n",
      "remove best\n",
      "Min error:  0.2\n",
      "Sample2\n",
      "Iteration:  42\n",
      "\tExpc counts:  [4, 1, 1]\n",
      "\tReal counts:  4 0 3\n",
      "\tExpc fraction:  [0.6, 0.2, 0.2]\n",
      "\tReal fraction:  0.5714285714285714 0.0 0.42857142857142855\n",
      "\tDiff:  [-0.02857142857142858, -0.2, 0.22857142857142854]\n",
      "remove best\n",
      "Min error:  0.05714285714285716\n",
      "Sample2\n",
      "Iteration:  43\n",
      "\tExpc counts:  [4, 1, 1]\n",
      "\tReal counts:  6 0 1\n",
      "\tExpc fraction:  [0.6, 0.2, 0.2]\n",
      "\tReal fraction:  0.8571428571428571 0.0 0.14285714285714285\n",
      "\tDiff:  [0.2571428571428571, -0.2, -0.05714285714285716]\n",
      "remove best\n",
      "Min error:  0.02857142857142858\n",
      "Sample2\n",
      "Iteration:  44\n",
      "\tExpc counts:  [4, 1, 1]\n",
      "\tReal counts:  4 2 1\n",
      "\tExpc fraction:  [0.6, 0.2, 0.2]\n",
      "\tReal fraction:  0.5714285714285714 0.2857142857142857 0.14285714285714285\n",
      "\tDiff:  [-0.02857142857142858, 0.08571428571428569, -0.05714285714285716]\n",
      "remove best\n",
      "Min error:  0.2\n",
      "Sample2\n",
      "Iteration:  45\n",
      "\tExpc counts:  [4, 1, 1]\n",
      "\tReal counts:  4 0 3\n",
      "\tExpc fraction:  [0.6, 0.2, 0.2]\n",
      "\tReal fraction:  0.5714285714285714 0.0 0.42857142857142855\n",
      "\tDiff:  [-0.02857142857142858, -0.2, 0.22857142857142854]\n",
      "remove best\n",
      "Min error:  0.05714285714285716\n",
      "Sample2\n",
      "Iteration:  46\n",
      "\tExpc counts:  [4, 1, 1]\n",
      "\tReal counts:  6 0 1\n",
      "\tExpc fraction:  [0.6, 0.2, 0.2]\n",
      "\tReal fraction:  0.8571428571428571 0.0 0.14285714285714285\n",
      "\tDiff:  [0.2571428571428571, -0.2, -0.05714285714285716]\n",
      "remove best\n",
      "Min error:  0.02857142857142858\n",
      "Sample2\n",
      "Iteration:  47\n",
      "\tExpc counts:  [4, 1, 1]\n",
      "\tReal counts:  4 2 1\n",
      "\tExpc fraction:  [0.6, 0.2, 0.2]\n",
      "\tReal fraction:  0.5714285714285714 0.2857142857142857 0.14285714285714285\n",
      "\tDiff:  [-0.02857142857142858, 0.08571428571428569, -0.05714285714285716]\n",
      "remove best\n",
      "Min error:  0.2\n",
      "Sample2\n",
      "Iteration:  48\n",
      "\tExpc counts:  [4, 1, 1]\n",
      "\tReal counts:  4 0 3\n",
      "\tExpc fraction:  [0.6, 0.2, 0.2]\n",
      "\tReal fraction:  0.5714285714285714 0.0 0.42857142857142855\n",
      "\tDiff:  [-0.02857142857142858, -0.2, 0.22857142857142854]\n",
      "remove best\n",
      "Min error:  0.05714285714285716\n",
      "Sample2\n",
      "Iteration:  49\n",
      "\tExpc counts:  [4, 1, 1]\n",
      "\tReal counts:  6 0 1\n",
      "\tExpc fraction:  [0.6, 0.2, 0.2]\n",
      "\tReal fraction:  0.8571428571428571 0.0 0.14285714285714285\n",
      "\tDiff:  [0.2571428571428571, -0.2, -0.05714285714285716]\n",
      "remove best\n",
      "Min error:  0.02857142857142858\n",
      "Sample2\n",
      "Iteration:  50\n",
      "\tExpc counts:  [4, 1, 1]\n",
      "\tReal counts:  4 2 1\n",
      "\tExpc fraction:  [0.6, 0.2, 0.2]\n",
      "\tReal fraction:  0.5714285714285714 0.2857142857142857 0.14285714285714285\n",
      "\tDiff:  [-0.02857142857142858, 0.08571428571428569, -0.05714285714285716]\n",
      "remove best\n",
      "Min error:  0.2\n",
      "Sample2\n",
      "Iteration:  51\n",
      "\tExpc counts:  [4, 1, 1]\n",
      "\tReal counts:  4 0 3\n",
      "\tExpc fraction:  [0.6, 0.2, 0.2]\n",
      "\tReal fraction:  0.5714285714285714 0.0 0.42857142857142855\n",
      "\tDiff:  [-0.02857142857142858, -0.2, 0.22857142857142854]\n",
      "remove best\n",
      "Min error:  0.05714285714285716\n",
      "Sample2\n",
      "Iteration:  52\n",
      "\tExpc counts:  [4, 1, 1]\n",
      "\tReal counts:  6 0 1\n",
      "\tExpc fraction:  [0.6, 0.2, 0.2]\n",
      "\tReal fraction:  0.8571428571428571 0.0 0.14285714285714285\n",
      "\tDiff:  [0.2571428571428571, -0.2, -0.05714285714285716]\n",
      "remove best\n",
      "Min error:  0.02857142857142858\n",
      "Sample2\n",
      "Iteration:  53\n",
      "\tExpc counts:  [4, 1, 1]\n",
      "\tReal counts:  4 2 1\n",
      "\tExpc fraction:  [0.6, 0.2, 0.2]\n",
      "\tReal fraction:  0.5714285714285714 0.2857142857142857 0.14285714285714285\n",
      "\tDiff:  [-0.02857142857142858, 0.08571428571428569, -0.05714285714285716]\n",
      "remove best\n",
      "Min error:  0.2\n",
      "Sample2\n",
      "Iteration:  54\n",
      "\tExpc counts:  [4, 1, 1]\n",
      "\tReal counts:  4 0 3\n",
      "\tExpc fraction:  [0.6, 0.2, 0.2]\n",
      "\tReal fraction:  0.5714285714285714 0.0 0.42857142857142855\n",
      "\tDiff:  [-0.02857142857142858, -0.2, 0.22857142857142854]\n",
      "remove best\n",
      "Min error:  0.05714285714285716\n",
      "Sample2\n",
      "Iteration:  55\n",
      "\tExpc counts:  [4, 1, 1]\n",
      "\tReal counts:  6 0 1\n",
      "\tExpc fraction:  [0.6, 0.2, 0.2]\n",
      "\tReal fraction:  0.8571428571428571 0.0 0.14285714285714285\n",
      "\tDiff:  [0.2571428571428571, -0.2, -0.05714285714285716]\n",
      "remove best\n",
      "Min error:  0.02857142857142858\n",
      "Sample2\n",
      "Iteration:  56\n",
      "\tExpc counts:  [4, 1, 1]\n",
      "\tReal counts:  4 2 1\n",
      "\tExpc fraction:  [0.6, 0.2, 0.2]\n",
      "\tReal fraction:  0.5714285714285714 0.2857142857142857 0.14285714285714285\n",
      "\tDiff:  [-0.02857142857142858, 0.08571428571428569, -0.05714285714285716]\n",
      "remove best\n",
      "Min error:  0.2\n",
      "Sample2\n",
      "Iteration:  57\n",
      "\tExpc counts:  [4, 1, 1]\n",
      "\tReal counts:  4 0 3\n",
      "\tExpc fraction:  [0.6, 0.2, 0.2]\n",
      "\tReal fraction:  0.5714285714285714 0.0 0.42857142857142855\n",
      "\tDiff:  [-0.02857142857142858, -0.2, 0.22857142857142854]\n",
      "remove best\n",
      "Min error:  0.05714285714285716\n",
      "Sample2\n",
      "Iteration:  58\n",
      "\tExpc counts:  [4, 1, 1]\n",
      "\tReal counts:  6 0 1\n",
      "\tExpc fraction:  [0.6, 0.2, 0.2]\n",
      "\tReal fraction:  0.8571428571428571 0.0 0.14285714285714285\n",
      "\tDiff:  [0.2571428571428571, -0.2, -0.05714285714285716]\n",
      "remove best\n",
      "Min error:  0.02857142857142858\n",
      "Sample2\n",
      "Iteration:  59\n",
      "\tExpc counts:  [4, 1, 1]\n",
      "\tReal counts:  4 2 1\n",
      "\tExpc fraction:  [0.6, 0.2, 0.2]\n",
      "\tReal fraction:  0.5714285714285714 0.2857142857142857 0.14285714285714285\n",
      "\tDiff:  [-0.02857142857142858, 0.08571428571428569, -0.05714285714285716]\n",
      "remove best\n",
      "Min error:  0.2\n",
      "Sample2\n",
      "Iteration:  60\n",
      "\tExpc counts:  [4, 1, 1]\n",
      "\tReal counts:  4 0 3\n",
      "\tExpc fraction:  [0.6, 0.2, 0.2]\n",
      "\tReal fraction:  0.5714285714285714 0.0 0.42857142857142855\n",
      "\tDiff:  [-0.02857142857142858, -0.2, 0.22857142857142854]\n",
      "remove best\n",
      "Min error:  0.05714285714285716\n",
      "Sample2\n",
      "Iteration:  61\n",
      "\tExpc counts:  [4, 1, 1]\n",
      "\tReal counts:  6 0 1\n",
      "\tExpc fraction:  [0.6, 0.2, 0.2]\n",
      "\tReal fraction:  0.8571428571428571 0.0 0.14285714285714285\n",
      "\tDiff:  [0.2571428571428571, -0.2, -0.05714285714285716]\n",
      "remove best\n",
      "Min error:  0.02857142857142858\n",
      "Sample2\n",
      "Iteration:  62\n",
      "\tExpc counts:  [4, 1, 1]\n",
      "\tReal counts:  4 2 1\n",
      "\tExpc fraction:  [0.6, 0.2, 0.2]\n",
      "\tReal fraction:  0.5714285714285714 0.2857142857142857 0.14285714285714285\n",
      "\tDiff:  [-0.02857142857142858, 0.08571428571428569, -0.05714285714285716]\n",
      "remove best\n",
      "Min error:  0.2\n",
      "Sample2\n",
      "Iteration:  63\n",
      "\tExpc counts:  [4, 1, 1]\n",
      "\tReal counts:  4 0 3\n",
      "\tExpc fraction:  [0.6, 0.2, 0.2]\n",
      "\tReal fraction:  0.5714285714285714 0.0 0.42857142857142855\n",
      "\tDiff:  [-0.02857142857142858, -0.2, 0.22857142857142854]\n",
      "remove best\n",
      "Min error:  0.05714285714285716\n",
      "Sample2\n",
      "Iteration:  64\n",
      "\tExpc counts:  [4, 1, 1]\n",
      "\tReal counts:  6 0 1\n",
      "\tExpc fraction:  [0.6, 0.2, 0.2]\n",
      "\tReal fraction:  0.8571428571428571 0.0 0.14285714285714285\n",
      "\tDiff:  [0.2571428571428571, -0.2, -0.05714285714285716]\n",
      "remove best\n",
      "Min error:  0.02857142857142858\n",
      "Sample2\n",
      "Iteration:  65\n",
      "\tExpc counts:  [4, 1, 1]\n",
      "\tReal counts:  4 2 1\n",
      "\tExpc fraction:  [0.6, 0.2, 0.2]\n",
      "\tReal fraction:  0.5714285714285714 0.2857142857142857 0.14285714285714285\n",
      "\tDiff:  [-0.02857142857142858, 0.08571428571428569, -0.05714285714285716]\n",
      "remove best\n",
      "Min error:  0.2\n",
      "Sample2\n",
      "Iteration:  66\n",
      "\tExpc counts:  [4, 1, 1]\n",
      "\tReal counts:  4 0 3\n",
      "\tExpc fraction:  [0.6, 0.2, 0.2]\n",
      "\tReal fraction:  0.5714285714285714 0.0 0.42857142857142855\n",
      "\tDiff:  [-0.02857142857142858, -0.2, 0.22857142857142854]\n",
      "remove best\n",
      "Min error:  0.05714285714285716\n",
      "Sample2\n",
      "Iteration:  67\n",
      "\tExpc counts:  [4, 1, 1]\n",
      "\tReal counts:  6 0 1\n",
      "\tExpc fraction:  [0.6, 0.2, 0.2]\n",
      "\tReal fraction:  0.8571428571428571 0.0 0.14285714285714285\n",
      "\tDiff:  [0.2571428571428571, -0.2, -0.05714285714285716]\n",
      "remove best\n",
      "Min error:  0.02857142857142858\n",
      "Sample2\n",
      "Iteration:  68\n",
      "\tExpc counts:  [4, 1, 1]\n",
      "\tReal counts:  4 2 1\n",
      "\tExpc fraction:  [0.6, 0.2, 0.2]\n",
      "\tReal fraction:  0.5714285714285714 0.2857142857142857 0.14285714285714285\n",
      "\tDiff:  [-0.02857142857142858, 0.08571428571428569, -0.05714285714285716]\n",
      "remove best\n",
      "Min error:  0.2\n",
      "Sample2\n",
      "Iteration:  69\n",
      "\tExpc counts:  [4, 1, 1]\n",
      "\tReal counts:  4 0 3\n",
      "\tExpc fraction:  [0.6, 0.2, 0.2]\n",
      "\tReal fraction:  0.5714285714285714 0.0 0.42857142857142855\n",
      "\tDiff:  [-0.02857142857142858, -0.2, 0.22857142857142854]\n",
      "remove best\n",
      "Min error:  0.05714285714285716\n",
      "Sample2\n",
      "Iteration:  70\n",
      "\tExpc counts:  [4, 1, 1]\n",
      "\tReal counts:  6 0 1\n",
      "\tExpc fraction:  [0.6, 0.2, 0.2]\n",
      "\tReal fraction:  0.8571428571428571 0.0 0.14285714285714285\n",
      "\tDiff:  [0.2571428571428571, -0.2, -0.05714285714285716]\n",
      "remove best\n",
      "Min error:  0.02857142857142858\n",
      "Sample2\n",
      "Iteration:  71\n",
      "\tExpc counts:  [4, 1, 1]\n",
      "\tReal counts:  4 2 1\n",
      "\tExpc fraction:  [0.6, 0.2, 0.2]\n",
      "\tReal fraction:  0.5714285714285714 0.2857142857142857 0.14285714285714285\n",
      "\tDiff:  [-0.02857142857142858, 0.08571428571428569, -0.05714285714285716]\n",
      "remove best\n",
      "Min error:  0.2\n",
      "Sample2\n",
      "Iteration:  72\n",
      "\tExpc counts:  [4, 1, 1]\n",
      "\tReal counts:  4 0 3\n",
      "\tExpc fraction:  [0.6, 0.2, 0.2]\n",
      "\tReal fraction:  0.5714285714285714 0.0 0.42857142857142855\n",
      "\tDiff:  [-0.02857142857142858, -0.2, 0.22857142857142854]\n",
      "remove best\n",
      "Min error:  0.05714285714285716\n",
      "Sample2\n",
      "Iteration:  73\n",
      "\tExpc counts:  [4, 1, 1]\n",
      "\tReal counts:  6 0 1\n",
      "\tExpc fraction:  [0.6, 0.2, 0.2]\n",
      "\tReal fraction:  0.8571428571428571 0.0 0.14285714285714285\n",
      "\tDiff:  [0.2571428571428571, -0.2, -0.05714285714285716]\n",
      "remove best\n",
      "Min error:  0.02857142857142858\n",
      "Sample2\n",
      "Iteration:  74\n",
      "\tExpc counts:  [4, 1, 1]\n",
      "\tReal counts:  4 2 1\n",
      "\tExpc fraction:  [0.6, 0.2, 0.2]\n",
      "\tReal fraction:  0.5714285714285714 0.2857142857142857 0.14285714285714285\n",
      "\tDiff:  [-0.02857142857142858, 0.08571428571428569, -0.05714285714285716]\n",
      "remove best\n",
      "Min error:  0.2\n",
      "Sample2\n",
      "Iteration:  75\n",
      "\tExpc counts:  [4, 1, 1]\n",
      "\tReal counts:  4 0 3\n",
      "\tExpc fraction:  [0.6, 0.2, 0.2]\n",
      "\tReal fraction:  0.5714285714285714 0.0 0.42857142857142855\n",
      "\tDiff:  [-0.02857142857142858, -0.2, 0.22857142857142854]\n",
      "remove best\n",
      "Min error:  0.05714285714285716\n",
      "Sample2\n",
      "Iteration:  76\n",
      "\tExpc counts:  [4, 1, 1]\n",
      "\tReal counts:  6 0 1\n",
      "\tExpc fraction:  [0.6, 0.2, 0.2]\n",
      "\tReal fraction:  0.8571428571428571 0.0 0.14285714285714285\n",
      "\tDiff:  [0.2571428571428571, -0.2, -0.05714285714285716]\n",
      "remove best\n",
      "Min error:  0.02857142857142858\n",
      "Sample2\n",
      "Iteration:  77\n",
      "\tExpc counts:  [4, 1, 1]\n",
      "\tReal counts:  4 2 1\n",
      "\tExpc fraction:  [0.6, 0.2, 0.2]\n",
      "\tReal fraction:  0.5714285714285714 0.2857142857142857 0.14285714285714285\n",
      "\tDiff:  [-0.02857142857142858, 0.08571428571428569, -0.05714285714285716]\n",
      "remove best\n",
      "Min error:  0.2\n",
      "Sample2\n",
      "Iteration:  78\n",
      "\tExpc counts:  [4, 1, 1]\n",
      "\tReal counts:  4 0 3\n",
      "\tExpc fraction:  [0.6, 0.2, 0.2]\n",
      "\tReal fraction:  0.5714285714285714 0.0 0.42857142857142855\n",
      "\tDiff:  [-0.02857142857142858, -0.2, 0.22857142857142854]\n",
      "remove best\n",
      "Min error:  0.05714285714285716\n",
      "Sample2\n",
      "Iteration:  79\n",
      "\tExpc counts:  [4, 1, 1]\n",
      "\tReal counts:  6 0 1\n",
      "\tExpc fraction:  [0.6, 0.2, 0.2]\n",
      "\tReal fraction:  0.8571428571428571 0.0 0.14285714285714285\n",
      "\tDiff:  [0.2571428571428571, -0.2, -0.05714285714285716]\n",
      "remove best\n",
      "Min error:  0.02857142857142858\n",
      "Sample2\n",
      "Iteration:  80\n",
      "\tExpc counts:  [4, 1, 1]\n",
      "\tReal counts:  4 2 1\n",
      "\tExpc fraction:  [0.6, 0.2, 0.2]\n",
      "\tReal fraction:  0.5714285714285714 0.2857142857142857 0.14285714285714285\n",
      "\tDiff:  [-0.02857142857142858, 0.08571428571428569, -0.05714285714285716]\n",
      "remove best\n",
      "Min error:  0.2\n",
      "Sample2\n",
      "Iteration:  81\n",
      "\tExpc counts:  [4, 1, 1]\n",
      "\tReal counts:  4 0 3\n",
      "\tExpc fraction:  [0.6, 0.2, 0.2]\n",
      "\tReal fraction:  0.5714285714285714 0.0 0.42857142857142855\n",
      "\tDiff:  [-0.02857142857142858, -0.2, 0.22857142857142854]\n",
      "remove best\n",
      "Min error:  0.05714285714285716\n",
      "Sample2\n",
      "Iteration:  82\n",
      "\tExpc counts:  [4, 1, 1]\n",
      "\tReal counts:  6 0 1\n",
      "\tExpc fraction:  [0.6, 0.2, 0.2]\n",
      "\tReal fraction:  0.8571428571428571 0.0 0.14285714285714285\n",
      "\tDiff:  [0.2571428571428571, -0.2, -0.05714285714285716]\n",
      "remove best\n",
      "Min error:  0.02857142857142858\n",
      "Sample2\n",
      "Iteration:  83\n",
      "\tExpc counts:  [4, 1, 1]\n",
      "\tReal counts:  4 2 1\n",
      "\tExpc fraction:  [0.6, 0.2, 0.2]\n",
      "\tReal fraction:  0.5714285714285714 0.2857142857142857 0.14285714285714285\n",
      "\tDiff:  [-0.02857142857142858, 0.08571428571428569, -0.05714285714285716]\n",
      "remove best\n",
      "Min error:  0.2\n",
      "Sample2\n",
      "Iteration:  84\n",
      "\tExpc counts:  [4, 1, 1]\n",
      "\tReal counts:  4 0 3\n",
      "\tExpc fraction:  [0.6, 0.2, 0.2]\n",
      "\tReal fraction:  0.5714285714285714 0.0 0.42857142857142855\n",
      "\tDiff:  [-0.02857142857142858, -0.2, 0.22857142857142854]\n",
      "remove best\n",
      "Min error:  0.05714285714285716\n",
      "Sample2\n",
      "Iteration:  85\n",
      "\tExpc counts:  [4, 1, 1]\n",
      "\tReal counts:  6 0 1\n",
      "\tExpc fraction:  [0.6, 0.2, 0.2]\n",
      "\tReal fraction:  0.8571428571428571 0.0 0.14285714285714285\n",
      "\tDiff:  [0.2571428571428571, -0.2, -0.05714285714285716]\n",
      "remove best\n",
      "Min error:  0.02857142857142858\n",
      "Sample2\n",
      "Iteration:  86\n",
      "\tExpc counts:  [4, 1, 1]\n",
      "\tReal counts:  4 2 1\n",
      "\tExpc fraction:  [0.6, 0.2, 0.2]\n",
      "\tReal fraction:  0.5714285714285714 0.2857142857142857 0.14285714285714285\n",
      "\tDiff:  [-0.02857142857142858, 0.08571428571428569, -0.05714285714285716]\n",
      "remove best\n",
      "Min error:  0.2\n",
      "Sample2\n",
      "Iteration:  87\n",
      "\tExpc counts:  [4, 1, 1]\n",
      "\tReal counts:  4 0 3\n",
      "\tExpc fraction:  [0.6, 0.2, 0.2]\n",
      "\tReal fraction:  0.5714285714285714 0.0 0.42857142857142855\n",
      "\tDiff:  [-0.02857142857142858, -0.2, 0.22857142857142854]\n",
      "remove best\n",
      "Min error:  0.05714285714285716\n",
      "Sample2\n",
      "Iteration:  88\n",
      "\tExpc counts:  [4, 1, 1]\n",
      "\tReal counts:  6 0 1\n",
      "\tExpc fraction:  [0.6, 0.2, 0.2]\n",
      "\tReal fraction:  0.8571428571428571 0.0 0.14285714285714285\n",
      "\tDiff:  [0.2571428571428571, -0.2, -0.05714285714285716]\n",
      "remove best\n",
      "Min error:  0.02857142857142858\n",
      "Sample2\n",
      "Iteration:  89\n",
      "\tExpc counts:  [4, 1, 1]\n",
      "\tReal counts:  4 2 1\n",
      "\tExpc fraction:  [0.6, 0.2, 0.2]\n",
      "\tReal fraction:  0.5714285714285714 0.2857142857142857 0.14285714285714285\n",
      "\tDiff:  [-0.02857142857142858, 0.08571428571428569, -0.05714285714285716]\n",
      "remove best\n",
      "Min error:  0.2\n",
      "Sample2\n",
      "Iteration:  90\n",
      "\tExpc counts:  [4, 1, 1]\n",
      "\tReal counts:  4 0 3\n",
      "\tExpc fraction:  [0.6, 0.2, 0.2]\n",
      "\tReal fraction:  0.5714285714285714 0.0 0.42857142857142855\n",
      "\tDiff:  [-0.02857142857142858, -0.2, 0.22857142857142854]\n",
      "remove best\n",
      "Min error:  0.05714285714285716\n",
      "Sample2\n",
      "Iteration:  91\n",
      "\tExpc counts:  [4, 1, 1]\n",
      "\tReal counts:  6 0 1\n",
      "\tExpc fraction:  [0.6, 0.2, 0.2]\n",
      "\tReal fraction:  0.8571428571428571 0.0 0.14285714285714285\n",
      "\tDiff:  [0.2571428571428571, -0.2, -0.05714285714285716]\n",
      "remove best\n",
      "Min error:  0.02857142857142858\n",
      "Sample2\n",
      "Iteration:  92\n",
      "\tExpc counts:  [4, 1, 1]\n",
      "\tReal counts:  4 2 1\n",
      "\tExpc fraction:  [0.6, 0.2, 0.2]\n",
      "\tReal fraction:  0.5714285714285714 0.2857142857142857 0.14285714285714285\n",
      "\tDiff:  [-0.02857142857142858, 0.08571428571428569, -0.05714285714285716]\n",
      "remove best\n",
      "Min error:  0.2\n",
      "Sample2\n",
      "Iteration:  93\n",
      "\tExpc counts:  [4, 1, 1]\n",
      "\tReal counts:  4 0 3\n",
      "\tExpc fraction:  [0.6, 0.2, 0.2]\n",
      "\tReal fraction:  0.5714285714285714 0.0 0.42857142857142855\n",
      "\tDiff:  [-0.02857142857142858, -0.2, 0.22857142857142854]\n",
      "remove best\n",
      "Min error:  0.05714285714285716\n",
      "Sample2\n",
      "Iteration:  94\n",
      "\tExpc counts:  [4, 1, 1]\n",
      "\tReal counts:  6 0 1\n",
      "\tExpc fraction:  [0.6, 0.2, 0.2]\n",
      "\tReal fraction:  0.8571428571428571 0.0 0.14285714285714285\n",
      "\tDiff:  [0.2571428571428571, -0.2, -0.05714285714285716]\n",
      "remove best\n",
      "Min error:  0.02857142857142858\n",
      "Sample2\n",
      "Iteration:  95\n",
      "\tExpc counts:  [4, 1, 1]\n",
      "\tReal counts:  4 2 1\n",
      "\tExpc fraction:  [0.6, 0.2, 0.2]\n",
      "\tReal fraction:  0.5714285714285714 0.2857142857142857 0.14285714285714285\n",
      "\tDiff:  [-0.02857142857142858, 0.08571428571428569, -0.05714285714285716]\n",
      "remove best\n",
      "Min error:  0.2\n",
      "Sample2\n",
      "Iteration:  96\n",
      "\tExpc counts:  [4, 1, 1]\n",
      "\tReal counts:  4 0 3\n",
      "\tExpc fraction:  [0.6, 0.2, 0.2]\n",
      "\tReal fraction:  0.5714285714285714 0.0 0.42857142857142855\n",
      "\tDiff:  [-0.02857142857142858, -0.2, 0.22857142857142854]\n",
      "remove best\n",
      "Min error:  0.05714285714285716\n",
      "Sample2\n",
      "Iteration:  97\n",
      "\tExpc counts:  [4, 1, 1]\n",
      "\tReal counts:  6 0 1\n",
      "\tExpc fraction:  [0.6, 0.2, 0.2]\n",
      "\tReal fraction:  0.8571428571428571 0.0 0.14285714285714285\n",
      "\tDiff:  [0.2571428571428571, -0.2, -0.05714285714285716]\n",
      "remove best\n",
      "Min error:  0.02857142857142858\n",
      "Sample2\n",
      "Iteration:  98\n",
      "\tExpc counts:  [4, 1, 1]\n",
      "\tReal counts:  4 2 1\n",
      "\tExpc fraction:  [0.6, 0.2, 0.2]\n",
      "\tReal fraction:  0.5714285714285714 0.2857142857142857 0.14285714285714285\n",
      "\tDiff:  [-0.02857142857142858, 0.08571428571428569, -0.05714285714285716]\n",
      "remove best\n",
      "Min error:  0.2\n",
      "Sample2\n",
      "Iteration:  99\n",
      "\tExpc counts:  [4, 1, 1]\n",
      "\tReal counts:  4 0 3\n",
      "\tExpc fraction:  [0.6, 0.2, 0.2]\n",
      "\tReal fraction:  0.5714285714285714 0.0 0.42857142857142855\n",
      "\tDiff:  [-0.02857142857142858, -0.2, 0.22857142857142854]\n",
      "remove best\n",
      "Min error:  0.05714285714285716\n",
      "Sample2\n"
     ]
    }
   ],
   "source": [
    "# compute cross tissue matrix pair (X, Y)\n",
    "(train,valid,test) = p.split_by_individuals(annobj,groupby=['Gender','Seq'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "samples_df  (7, 4)\n",
      "samples_df_sliced  (6, 4)\n",
      "compute all arrangements\n",
      "Total pairs: 14\n"
     ]
    }
   ],
   "source": [
    "(X_train, Y_train) = p.rnaseq_cross_tissue(annobj, individuals=train, gene_ids=annobj.obs_names)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>G1</th>\n",
       "      <th>G2</th>\n",
       "      <th>G3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>T1_s1_T6_s1</th>\n",
       "      <td>10.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>6.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T1_s1_T3_s1</th>\n",
       "      <td>10.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>6.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T1_s1_T5_s1</th>\n",
       "      <td>10.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>6.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T6_s1_T1_s1</th>\n",
       "      <td>6.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T6_s1_T3_s1</th>\n",
       "      <td>6.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               G1   G2   G3\n",
       "T1_s1_T6_s1  10.0  5.0  6.0\n",
       "T1_s1_T3_s1  10.0  5.0  6.0\n",
       "T1_s1_T5_s1  10.0  5.0  6.0\n",
       "T6_s1_T1_s1   6.0  6.0  1.0\n",
       "T6_s1_T3_s1   6.0  6.0  1.0"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# train matrix\n",
    "X_train[:5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>G1</th>\n",
       "      <th>G2</th>\n",
       "      <th>G3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>T1_s1_T6_s1</th>\n",
       "      <td>6.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T1_s1_T3_s1</th>\n",
       "      <td>30.0</td>\n",
       "      <td>20.0</td>\n",
       "      <td>8.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T1_s1_T5_s1</th>\n",
       "      <td>5.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T6_s1_T1_s1</th>\n",
       "      <td>10.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>6.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T6_s1_T3_s1</th>\n",
       "      <td>30.0</td>\n",
       "      <td>20.0</td>\n",
       "      <td>8.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               G1    G2   G3\n",
       "T1_s1_T6_s1   6.0   6.0  1.0\n",
       "T1_s1_T3_s1  30.0  20.0  8.0\n",
       "T1_s1_T5_s1   5.0   5.0  1.0\n",
       "T6_s1_T1_s1  10.0   5.0  6.0\n",
       "T6_s1_T3_s1  30.0  20.0  8.0"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# train matrix\n",
    "Y_train[:5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>G1</th>\n",
       "      <th>G2</th>\n",
       "      <th>G3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>T3_s1_T6_s1</th>\n",
       "      <td>30.0</td>\n",
       "      <td>20.0</td>\n",
       "      <td>8.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T3_s1_T1_s1</th>\n",
       "      <td>30.0</td>\n",
       "      <td>20.0</td>\n",
       "      <td>8.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T3_s1_T5_s1</th>\n",
       "      <td>30.0</td>\n",
       "      <td>20.0</td>\n",
       "      <td>8.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T6_s1_T3_s1</th>\n",
       "      <td>6.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T6_s1_T1_s1</th>\n",
       "      <td>6.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               G1    G2   G3\n",
       "T3_s1_T6_s1  30.0  20.0  8.0\n",
       "T3_s1_T1_s1  30.0  20.0  8.0\n",
       "T3_s1_T5_s1  30.0  20.0  8.0\n",
       "T6_s1_T3_s1   6.0   6.0  1.0\n",
       "T6_s1_T1_s1   6.0   6.0  1.0"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# train matrix\n",
    "X_train[:5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>G1</th>\n",
       "      <th>G2</th>\n",
       "      <th>G3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>T3_s1_T6_s1</th>\n",
       "      <td>0.845098</td>\n",
       "      <td>0.845098</td>\n",
       "      <td>0.301030</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T3_s1_T1_s1</th>\n",
       "      <td>1.041393</td>\n",
       "      <td>0.778151</td>\n",
       "      <td>0.845098</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T3_s1_T5_s1</th>\n",
       "      <td>0.778151</td>\n",
       "      <td>0.778151</td>\n",
       "      <td>0.301030</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T6_s1_T3_s1</th>\n",
       "      <td>1.491362</td>\n",
       "      <td>1.322219</td>\n",
       "      <td>0.954243</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T6_s1_T1_s1</th>\n",
       "      <td>1.041393</td>\n",
       "      <td>0.778151</td>\n",
       "      <td>0.845098</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   G1        G2        G3\n",
       "T3_s1_T6_s1  0.845098  0.845098  0.301030\n",
       "T3_s1_T1_s1  1.041393  0.778151  0.845098\n",
       "T3_s1_T5_s1  0.778151  0.778151  0.301030\n",
       "T6_s1_T3_s1  1.491362  1.322219  0.954243\n",
       "T6_s1_T1_s1  1.041393  0.778151  0.845098"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# label matrix\n",
    "Y_train[:5]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3. Create toy large dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "import random\n",
    "import math\n",
    "import numpy.random as nr\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "from scipy import stats\n",
    "\n",
    "import anndata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Gene_19991',\n",
       " 'Gene_19992',\n",
       " 'Gene_19993',\n",
       " 'Gene_19994',\n",
       " 'Gene_19995',\n",
       " 'Gene_19996',\n",
       " 'Gene_19997',\n",
       " 'Gene_19998',\n",
       " 'Gene_19999',\n",
       " 'Gene_20000']"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# genes\n",
    "genes = list(map(lambda x: \"Gene_\" + str(x) , list(range(1, 20001))))\n",
    "genes[19990:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([1837, 7955,  505, 4570, 1270, 4205, 5287, 1949, 1877, 5478])"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "raw_counts = nr.negative_binomial(3,0.001,size=20000)\n",
    "# sns.distplot(raw_counts)\n",
    "raw_counts[:10]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# tissues\n",
    "tissues = ['Adipose - Subcutaneous',\n",
    " 'Adipose - Visceral (Omentum)',\n",
    " 'Adrenal Gland',\n",
    " 'Artery - Aorta',\n",
    " 'Artery - Coronary',\n",
    " 'Artery - Tibial',\n",
    " 'Bladder',\n",
    " 'Brain - Amygdala',\n",
    " 'Brain - Anterior cingulate cortex (BA24)',\n",
    " 'Brain - Caudate (basal ganglia)',\n",
    " 'Brain - Cerebellar Hemisphere',\n",
    " 'Brain - Cerebellum',\n",
    " 'Brain - Cortex',\n",
    " 'Brain - Frontal Cortex (BA9)',\n",
    " 'Brain - Hippocampus',\n",
    " 'Brain - Hypothalamus',\n",
    " 'Brain - Nucleus accumbens (basal ganglia)',\n",
    " 'Brain - Putamen (basal ganglia)',\n",
    " 'Brain - Spinal cord (cervical c-1)',\n",
    " 'Brain - Substantia nigra',\n",
    " 'Breast - Mammary Tissue',\n",
    " 'Cells - EBV-transformed lymphocytes',\n",
    " 'Cells - Transformed fibroblasts',\n",
    " 'Cervix - Ectocervix',\n",
    " 'Cervix - Endocervix',\n",
    " 'Colon - Sigmoid',\n",
    " 'Colon - Transverse',\n",
    " 'Esophagus - Gastroesophageal Junction',\n",
    " 'Esophagus - Mucosa',\n",
    " 'Esophagus - Muscularis',\n",
    " 'Fallopian Tube',\n",
    " 'Heart - Atrial Appendage',\n",
    " 'Heart - Left Ventricle',\n",
    " 'Kidney - Cortex',\n",
    " 'Liver',\n",
    " 'Lung',\n",
    " 'Minor Salivary Gland',\n",
    " 'Muscle - Skeletal',\n",
    " 'Nerve - Tibial',\n",
    " 'Ovary',\n",
    " 'Pancreas',\n",
    " 'Pituitary',\n",
    " 'Prostate',\n",
    " 'Skin - Not Sun Exposed (Suprapubic)',\n",
    " 'Skin - Sun Exposed (Lower leg)',\n",
    " 'Small Intestine - Terminal Ileum',\n",
    " 'Spleen',\n",
    " 'Stomach',\n",
    " 'Testis',\n",
    " 'Thyroid',\n",
    " 'Uterus',\n",
    " 'Vagina',\n",
    " 'Whole Blood']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "((1000,),\n",
       " array([23.45844837, 32.42258621, 43.67978791, 46.51771167, 32.30568135,\n",
       "        31.01255899,  5.25819487, 25.8020789 , 28.54055117, 48.85421578]))"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# simulate how many tissues are collected per patient\n",
    "distribution_tissues = nr.uniform(1,len(tissues),1000)\n",
    "distribution_tissues.shape,distribution_tissues[:10]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "# generate samples - raw counts per gene per tissue\n",
    "frames = []\n",
    "for i in range(1,150): # patients\n",
    "  \n",
    "    # choose tissues\n",
    "    how_many_tissues = math.floor(distribution_tissues[random.randint(0,len(distribution_tissues))-1])\n",
    "    random.shuffle(tissues)\n",
    "    t_set = tissues[:how_many_tissues]  \n",
    "  \n",
    "    # initialize count matrix\n",
    "    counts = pd.DataFrame(columns=genes,index=map(lambda x: \"Patient_\" + str(i) + \"_\" + str(x), t_set))\n",
    "\n",
    "    for j in range(0,len(t_set)):\n",
    "    \n",
    "        col_name = \"Patient_\" + str(i) + \"_\" + t_set[j]\n",
    "    \n",
    "        # generate raw counts per tissue per set of genes\n",
    "        raw_counts = nr.negative_binomial(3,0.001,size=20000)\n",
    "        counts.loc[col_name] = raw_counts.reshape(1,20000)\n",
    "    \n",
    "    # append expression per tissue for the patient \n",
    "    frames.append(counts)\n",
    "      \n",
    "\n",
    "df_big = pd.concat(frames)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(20000, 4063)"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# transpose the matrix to be genes x patient_tissue\n",
    "count_matrix = df_big.transpose()\n",
    "count_matrix.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Individual</th>\n",
       "      <th>Tissue</th>\n",
       "      <th>Gender</th>\n",
       "      <th>Seq</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Patient_1_Artery - Coronary</th>\n",
       "      <td>Patient_1</td>\n",
       "      <td>Artery - Coronary</td>\n",
       "      <td>2.0</td>\n",
       "      <td>WXS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Patient_1_Artery - Tibial</th>\n",
       "      <td>Patient_1</td>\n",
       "      <td>Artery - Tibial</td>\n",
       "      <td>2.0</td>\n",
       "      <td>WXS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Patient_1_Brain - Cerebellar Hemisphere</th>\n",
       "      <td>Patient_1</td>\n",
       "      <td>Brain - Cerebellar Hemisphere</td>\n",
       "      <td>2.0</td>\n",
       "      <td>WXS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Patient_1_Skin - Sun Exposed (Lower leg)</th>\n",
       "      <td>Patient_1</td>\n",
       "      <td>Skin - Sun Exposed (Lower leg)</td>\n",
       "      <td>2.0</td>\n",
       "      <td>WXS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Patient_1_Adipose - Subcutaneous</th>\n",
       "      <td>Patient_1</td>\n",
       "      <td>Adipose - Subcutaneous</td>\n",
       "      <td>2.0</td>\n",
       "      <td>WXS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Patient_1_Uterus</th>\n",
       "      <td>Patient_1</td>\n",
       "      <td>Uterus</td>\n",
       "      <td>2.0</td>\n",
       "      <td>WXS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Patient_1_Muscle - Skeletal</th>\n",
       "      <td>Patient_1</td>\n",
       "      <td>Muscle - Skeletal</td>\n",
       "      <td>2.0</td>\n",
       "      <td>WXS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Patient_1_Brain - Substantia nigra</th>\n",
       "      <td>Patient_1</td>\n",
       "      <td>Brain - Substantia nigra</td>\n",
       "      <td>2.0</td>\n",
       "      <td>WXS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Patient_1_Ovary</th>\n",
       "      <td>Patient_1</td>\n",
       "      <td>Ovary</td>\n",
       "      <td>2.0</td>\n",
       "      <td>WXS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Patient_1_Testis</th>\n",
       "      <td>Patient_1</td>\n",
       "      <td>Testis</td>\n",
       "      <td>2.0</td>\n",
       "      <td>WXS</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                         Individual  \\\n",
       "Patient_1_Artery - Coronary               Patient_1   \n",
       "Patient_1_Artery - Tibial                 Patient_1   \n",
       "Patient_1_Brain - Cerebellar Hemisphere   Patient_1   \n",
       "Patient_1_Skin - Sun Exposed (Lower leg)  Patient_1   \n",
       "Patient_1_Adipose - Subcutaneous          Patient_1   \n",
       "Patient_1_Uterus                          Patient_1   \n",
       "Patient_1_Muscle - Skeletal               Patient_1   \n",
       "Patient_1_Brain - Substantia nigra        Patient_1   \n",
       "Patient_1_Ovary                           Patient_1   \n",
       "Patient_1_Testis                          Patient_1   \n",
       "\n",
       "                                                                  Tissue  \\\n",
       "Patient_1_Artery - Coronary                            Artery - Coronary   \n",
       "Patient_1_Artery - Tibial                                Artery - Tibial   \n",
       "Patient_1_Brain - Cerebellar Hemisphere    Brain - Cerebellar Hemisphere   \n",
       "Patient_1_Skin - Sun Exposed (Lower leg)  Skin - Sun Exposed (Lower leg)   \n",
       "Patient_1_Adipose - Subcutaneous                  Adipose - Subcutaneous   \n",
       "Patient_1_Uterus                                                  Uterus   \n",
       "Patient_1_Muscle - Skeletal                            Muscle - Skeletal   \n",
       "Patient_1_Brain - Substantia nigra              Brain - Substantia nigra   \n",
       "Patient_1_Ovary                                                    Ovary   \n",
       "Patient_1_Testis                                                  Testis   \n",
       "\n",
       "                                          Gender  Seq  \n",
       "Patient_1_Artery - Coronary                  2.0  WXS  \n",
       "Patient_1_Artery - Tibial                    2.0  WXS  \n",
       "Patient_1_Brain - Cerebellar Hemisphere      2.0  WXS  \n",
       "Patient_1_Skin - Sun Exposed (Lower leg)     2.0  WXS  \n",
       "Patient_1_Adipose - Subcutaneous             2.0  WXS  \n",
       "Patient_1_Uterus                             2.0  WXS  \n",
       "Patient_1_Muscle - Skeletal                  2.0  WXS  \n",
       "Patient_1_Brain - Substantia nigra           2.0  WXS  \n",
       "Patient_1_Ovary                              2.0  WXS  \n",
       "Patient_1_Testis                             2.0  WXS  "
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# create annotation dataframes\n",
    "# column = 0 its the index! be careful\n",
    "samples_anno = pd.DataFrame(index=count_matrix.columns,columns=['Individual','Tissue'])\n",
    "gender=[1,2]\n",
    "seq=['WGS','WXS']\n",
    "\n",
    "for i in range(0,samples_anno.shape[0]):\n",
    "    sample = samples_anno.index[i]\n",
    "    [pat,number,tissue] = sample.split(\"_\")\n",
    "  \n",
    "    samples_anno.at[sample, 'Individual'] = pat + \"_\" + str(number)\n",
    "    samples_anno.at[sample, 'Tissue'] = tissue\n",
    "    samples_anno.at[sample, 'Gender'] = gender[int(number)%2]\n",
    "    samples_anno.at[sample, 'Seq'] = seq[int(number)%2]\n",
    "\n",
    "samples_anno.drop_duplicates().shape\n",
    "samples_anno[:10]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(20000, 0)"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes_anno = pd.DataFrame(index=count_matrix.index)\n",
    "genes_anno.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "anndata.X ----\n",
      "[[2254. 2468. 5217. ... 1010.  651. 2448.]\n",
      " [4728. 2741. 9423. ... 3820. 3645. 3259.]\n",
      " [1844. 6626. 1925. ... 2535. 3043. 3148.]\n",
      " ...\n",
      " [1005. 2865. 2508. ...  370. 1902. 1089.]\n",
      " [7190. 2160.  928. ... 1992.  761. 1185.]\n",
      " [ 820. 1601. 3391. ... 3824. 2825. 2001.]]\n",
      "anndata.var ----\n",
      "                                         Individual  \\\n",
      "Patient_1_Artery - Coronary               Patient_1   \n",
      "Patient_1_Artery - Tibial                 Patient_1   \n",
      "Patient_1_Brain - Cerebellar Hemisphere   Patient_1   \n",
      "Patient_1_Skin - Sun Exposed (Lower leg)  Patient_1   \n",
      "Patient_1_Adipose - Subcutaneous          Patient_1   \n",
      "\n",
      "                                                                  Tissue  \\\n",
      "Patient_1_Artery - Coronary                            Artery - Coronary   \n",
      "Patient_1_Artery - Tibial                                Artery - Tibial   \n",
      "Patient_1_Brain - Cerebellar Hemisphere    Brain - Cerebellar Hemisphere   \n",
      "Patient_1_Skin - Sun Exposed (Lower leg)  Skin - Sun Exposed (Lower leg)   \n",
      "Patient_1_Adipose - Subcutaneous                  Adipose - Subcutaneous   \n",
      "\n",
      "                                          Gender  Seq  \n",
      "Patient_1_Artery - Coronary                  2.0  WXS  \n",
      "Patient_1_Artery - Tibial                    2.0  WXS  \n",
      "Patient_1_Brain - Cerebellar Hemisphere      2.0  WXS  \n",
      "Patient_1_Skin - Sun Exposed (Lower leg)     2.0  WXS  \n",
      "Patient_1_Adipose - Subcutaneous             2.0  WXS  \n",
      "anndata.obs ----\n",
      "Empty DataFrame\n",
      "Columns: []\n",
      "Index: [Gene_1, Gene_2, Gene_3, Gene_4, Gene_5]\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# create annobj\n",
    "annobj = anndata.AnnData(count_matrix)\n",
    "annobj.var = samples_anno\n",
    "annobj.var_names = list(samples_anno.index)\n",
    "annobj.obs = genes_anno\n",
    "annobj.obs_names = list(genes_anno.index)\n",
    "\n",
    "p.print_anndata(annobj)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total individuals: 149\n",
      "Individual split before balancing:  89 29 31\n",
      "Iteration:  1\n",
      "\tExpc counts:  [2437, 812, 812]\n",
      "\tReal counts:  2361 805 897\n",
      "\tExpc fraction:  [0.6, 0.2, 0.2]\n",
      "\tReal fraction:  0.5810977110509475 0.1981294609894167 0.22077282795963574\n",
      "\tDiff:  [-0.018902288949052437, -0.0018705390105833164, 0.020772827959635726]\n",
      "remove best\n",
      "Min error:  0.007974403150381476\n",
      "Patient_132\n",
      "Iteration:  2\n",
      "\tExpc counts:  [2437, 812, 812]\n",
      "\tReal counts:  2413 805 845\n",
      "\tExpc fraction:  [0.6, 0.2, 0.2]\n",
      "\tReal fraction:  0.5938961358602018 0.1981294609894167 0.2079744031503815\n",
      "\tDiff:  [-0.006103864139798132, -0.0018705390105833164, 0.007974403150381476]\n",
      "remove best\n",
      "Min error:  0.0045778981048486544\n",
      "Patient_118\n",
      "Iteration:  3\n",
      "\tExpc counts:  [2437, 812, 812]\n",
      "\tReal counts:  2464 805 794\n",
      "\tExpc fraction:  [0.6, 0.2, 0.2]\n",
      "\tReal fraction:  0.606448437115432 0.1981294609894167 0.19542210189515136\n",
      "\tDiff:  [0.006448437115431971, -0.0018705390105833164, -0.0045778981048486544]\n",
      "remove best\n",
      "Min error:  0.0006891459512675668\n",
      "Patient_2\n",
      "Iteration:  4\n",
      "\tExpc counts:  [2437, 812, 812]\n",
      "\tReal counts:  2435 834 794\n",
      "\tExpc fraction:  [0.6, 0.2, 0.2]\n",
      "\tReal fraction:  0.5993108540487324 0.20526704405611618 0.19542210189515136\n",
      "\tDiff:  [-0.0006891459512675668, 0.005267044056116166, -0.0045778981048486544]\n",
      "remove best\n",
      "Min error:  0.001132168348510959\n",
      "Patient_90\n",
      "Iteration:  5\n",
      "\tExpc counts:  [2437, 812, 812]\n",
      "\tReal counts:  2435 808 820\n",
      "\tExpc fraction:  [0.6, 0.2, 0.2]\n",
      "\tReal fraction:  0.5993108540487324 0.19886783165148905 0.20182131429977848\n",
      "\tDiff:  [-0.0006891459512675668, -0.001132168348510959, 0.0018213142997784704]\n",
      "CPU times: user 177 ms, sys: 9.01 ms, total: 186 ms\n",
      "Wall time: 179 ms\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "('Individual split: ', 90, 29, 30)"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%time (train,valid,test) = p.split_by_individuals(annobj,groupby=['Gender','Seq'])\n",
    "\"Individual split: \", len(train),len(valid),len(test)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "samples_df  (4063, 4)\n",
      "samples_df_sliced  (2435, 4)\n",
      "compute all arrangements\n",
      "Total pairs: 84696\n",
      "CPU times: user 21.7 s, sys: 11.8 s, total: 33.4 s\n",
      "Wall time: 15.8 s\n",
      "samples_df  (4063, 4)\n",
      "samples_df_sliced  (808, 4)\n",
      "compute all arrangements\n",
      "Total pairs: 27030\n",
      "CPU times: user 14.4 s, sys: 4.98 s, total: 19.4 s\n",
      "Wall time: 5.76 s\n",
      "samples_df  (4063, 4)\n",
      "samples_df_sliced  (820, 4)\n",
      "compute all arrangements\n",
      "Total pairs: 26526\n",
      "CPU times: user 14.3 s, sys: 5.48 s, total: 19.8 s\n",
      "Wall time: 6.05 s\n"
     ]
    }
   ],
   "source": [
    "# slice  + compute all pairs\n",
    "%time (X_train, Y_train) = p.rnaseq_cross_tissue(annobj, individuals=train,gene_ids=annobj.obs_names,input_transform='log10')\n",
    "%time (X_valid, Y_valid) = p.rnaseq_cross_tissue(annobj, individuals=valid,gene_ids=annobj.obs_names)\n",
    "%time (X_test, Y_test) = p.rnaseq_cross_tissue(annobj, individuals=test,gene_ids=annobj.obs_names)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene_1</th>\n",
       "      <th>Gene_2</th>\n",
       "      <th>Gene_3</th>\n",
       "      <th>Gene_4</th>\n",
       "      <th>Gene_5</th>\n",
       "      <th>Gene_6</th>\n",
       "      <th>Gene_7</th>\n",
       "      <th>Gene_8</th>\n",
       "      <th>Gene_9</th>\n",
       "      <th>Gene_10</th>\n",
       "      <th>...</th>\n",
       "      <th>Gene_19991</th>\n",
       "      <th>Gene_19992</th>\n",
       "      <th>Gene_19993</th>\n",
       "      <th>Gene_19994</th>\n",
       "      <th>Gene_19995</th>\n",
       "      <th>Gene_19996</th>\n",
       "      <th>Gene_19997</th>\n",
       "      <th>Gene_19998</th>\n",
       "      <th>Gene_19999</th>\n",
       "      <th>Gene_20000</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Patient_1_Liver_Patient_1_Esophagus - Muscularis</th>\n",
       "      <td>3.916033</td>\n",
       "      <td>3.713659</td>\n",
       "      <td>3.405517</td>\n",
       "      <td>3.543074</td>\n",
       "      <td>3.619406</td>\n",
       "      <td>3.448242</td>\n",
       "      <td>3.617734</td>\n",
       "      <td>3.066326</td>\n",
       "      <td>3.318689</td>\n",
       "      <td>3.703463</td>\n",
       "      <td>...</td>\n",
       "      <td>3.129045</td>\n",
       "      <td>3.50637</td>\n",
       "      <td>3.386677</td>\n",
       "      <td>3.308351</td>\n",
       "      <td>3.582972</td>\n",
       "      <td>3.700877</td>\n",
       "      <td>3.12969</td>\n",
       "      <td>3.635182</td>\n",
       "      <td>3.661055</td>\n",
       "      <td>3.289143</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Patient_1_Liver_Patient_1_Prostate</th>\n",
       "      <td>3.916033</td>\n",
       "      <td>3.713659</td>\n",
       "      <td>3.405517</td>\n",
       "      <td>3.543074</td>\n",
       "      <td>3.619406</td>\n",
       "      <td>3.448242</td>\n",
       "      <td>3.617734</td>\n",
       "      <td>3.066326</td>\n",
       "      <td>3.318689</td>\n",
       "      <td>3.703463</td>\n",
       "      <td>...</td>\n",
       "      <td>3.129045</td>\n",
       "      <td>3.50637</td>\n",
       "      <td>3.386677</td>\n",
       "      <td>3.308351</td>\n",
       "      <td>3.582972</td>\n",
       "      <td>3.700877</td>\n",
       "      <td>3.12969</td>\n",
       "      <td>3.635182</td>\n",
       "      <td>3.661055</td>\n",
       "      <td>3.289143</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Patient_1_Liver_Patient_1_Brain - Amygdala</th>\n",
       "      <td>3.916033</td>\n",
       "      <td>3.713659</td>\n",
       "      <td>3.405517</td>\n",
       "      <td>3.543074</td>\n",
       "      <td>3.619406</td>\n",
       "      <td>3.448242</td>\n",
       "      <td>3.617734</td>\n",
       "      <td>3.066326</td>\n",
       "      <td>3.318689</td>\n",
       "      <td>3.703463</td>\n",
       "      <td>...</td>\n",
       "      <td>3.129045</td>\n",
       "      <td>3.50637</td>\n",
       "      <td>3.386677</td>\n",
       "      <td>3.308351</td>\n",
       "      <td>3.582972</td>\n",
       "      <td>3.700877</td>\n",
       "      <td>3.12969</td>\n",
       "      <td>3.635182</td>\n",
       "      <td>3.661055</td>\n",
       "      <td>3.289143</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Patient_1_Liver_Patient_1_Brain - Hypothalamus</th>\n",
       "      <td>3.916033</td>\n",
       "      <td>3.713659</td>\n",
       "      <td>3.405517</td>\n",
       "      <td>3.543074</td>\n",
       "      <td>3.619406</td>\n",
       "      <td>3.448242</td>\n",
       "      <td>3.617734</td>\n",
       "      <td>3.066326</td>\n",
       "      <td>3.318689</td>\n",
       "      <td>3.703463</td>\n",
       "      <td>...</td>\n",
       "      <td>3.129045</td>\n",
       "      <td>3.50637</td>\n",
       "      <td>3.386677</td>\n",
       "      <td>3.308351</td>\n",
       "      <td>3.582972</td>\n",
       "      <td>3.700877</td>\n",
       "      <td>3.12969</td>\n",
       "      <td>3.635182</td>\n",
       "      <td>3.661055</td>\n",
       "      <td>3.289143</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Patient_1_Liver_Patient_1_Artery - Coronary</th>\n",
       "      <td>3.916033</td>\n",
       "      <td>3.713659</td>\n",
       "      <td>3.405517</td>\n",
       "      <td>3.543074</td>\n",
       "      <td>3.619406</td>\n",
       "      <td>3.448242</td>\n",
       "      <td>3.617734</td>\n",
       "      <td>3.066326</td>\n",
       "      <td>3.318689</td>\n",
       "      <td>3.703463</td>\n",
       "      <td>...</td>\n",
       "      <td>3.129045</td>\n",
       "      <td>3.50637</td>\n",
       "      <td>3.386677</td>\n",
       "      <td>3.308351</td>\n",
       "      <td>3.582972</td>\n",
       "      <td>3.700877</td>\n",
       "      <td>3.12969</td>\n",
       "      <td>3.635182</td>\n",
       "      <td>3.661055</td>\n",
       "      <td>3.289143</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Patient_1_Liver_Patient_1_Testis</th>\n",
       "      <td>3.916033</td>\n",
       "      <td>3.713659</td>\n",
       "      <td>3.405517</td>\n",
       "      <td>3.543074</td>\n",
       "      <td>3.619406</td>\n",
       "      <td>3.448242</td>\n",
       "      <td>3.617734</td>\n",
       "      <td>3.066326</td>\n",
       "      <td>3.318689</td>\n",
       "      <td>3.703463</td>\n",
       "      <td>...</td>\n",
       "      <td>3.129045</td>\n",
       "      <td>3.50637</td>\n",
       "      <td>3.386677</td>\n",
       "      <td>3.308351</td>\n",
       "      <td>3.582972</td>\n",
       "      <td>3.700877</td>\n",
       "      <td>3.12969</td>\n",
       "      <td>3.635182</td>\n",
       "      <td>3.661055</td>\n",
       "      <td>3.289143</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Patient_1_Liver_Patient_1_Adrenal Gland</th>\n",
       "      <td>3.916033</td>\n",
       "      <td>3.713659</td>\n",
       "      <td>3.405517</td>\n",
       "      <td>3.543074</td>\n",
       "      <td>3.619406</td>\n",
       "      <td>3.448242</td>\n",
       "      <td>3.617734</td>\n",
       "      <td>3.066326</td>\n",
       "      <td>3.318689</td>\n",
       "      <td>3.703463</td>\n",
       "      <td>...</td>\n",
       "      <td>3.129045</td>\n",
       "      <td>3.50637</td>\n",
       "      <td>3.386677</td>\n",
       "      <td>3.308351</td>\n",
       "      <td>3.582972</td>\n",
       "      <td>3.700877</td>\n",
       "      <td>3.12969</td>\n",
       "      <td>3.635182</td>\n",
       "      <td>3.661055</td>\n",
       "      <td>3.289143</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Patient_1_Liver_Patient_1_Uterus</th>\n",
       "      <td>3.916033</td>\n",
       "      <td>3.713659</td>\n",
       "      <td>3.405517</td>\n",
       "      <td>3.543074</td>\n",
       "      <td>3.619406</td>\n",
       "      <td>3.448242</td>\n",
       "      <td>3.617734</td>\n",
       "      <td>3.066326</td>\n",
       "      <td>3.318689</td>\n",
       "      <td>3.703463</td>\n",
       "      <td>...</td>\n",
       "      <td>3.129045</td>\n",
       "      <td>3.50637</td>\n",
       "      <td>3.386677</td>\n",
       "      <td>3.308351</td>\n",
       "      <td>3.582972</td>\n",
       "      <td>3.700877</td>\n",
       "      <td>3.12969</td>\n",
       "      <td>3.635182</td>\n",
       "      <td>3.661055</td>\n",
       "      <td>3.289143</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Patient_1_Liver_Patient_1_Skin - Sun Exposed (Lower leg)</th>\n",
       "      <td>3.916033</td>\n",
       "      <td>3.713659</td>\n",
       "      <td>3.405517</td>\n",
       "      <td>3.543074</td>\n",
       "      <td>3.619406</td>\n",
       "      <td>3.448242</td>\n",
       "      <td>3.617734</td>\n",
       "      <td>3.066326</td>\n",
       "      <td>3.318689</td>\n",
       "      <td>3.703463</td>\n",
       "      <td>...</td>\n",
       "      <td>3.129045</td>\n",
       "      <td>3.50637</td>\n",
       "      <td>3.386677</td>\n",
       "      <td>3.308351</td>\n",
       "      <td>3.582972</td>\n",
       "      <td>3.700877</td>\n",
       "      <td>3.12969</td>\n",
       "      <td>3.635182</td>\n",
       "      <td>3.661055</td>\n",
       "      <td>3.289143</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Patient_1_Liver_Patient_1_Thyroid</th>\n",
       "      <td>3.916033</td>\n",
       "      <td>3.713659</td>\n",
       "      <td>3.405517</td>\n",
       "      <td>3.543074</td>\n",
       "      <td>3.619406</td>\n",
       "      <td>3.448242</td>\n",
       "      <td>3.617734</td>\n",
       "      <td>3.066326</td>\n",
       "      <td>3.318689</td>\n",
       "      <td>3.703463</td>\n",
       "      <td>...</td>\n",
       "      <td>3.129045</td>\n",
       "      <td>3.50637</td>\n",
       "      <td>3.386677</td>\n",
       "      <td>3.308351</td>\n",
       "      <td>3.582972</td>\n",
       "      <td>3.700877</td>\n",
       "      <td>3.12969</td>\n",
       "      <td>3.635182</td>\n",
       "      <td>3.661055</td>\n",
       "      <td>3.289143</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10 rows  20000 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                      Gene_1    Gene_2  \\\n",
       "Patient_1_Liver_Patient_1_Esophagus - Muscularis    3.916033  3.713659   \n",
       "Patient_1_Liver_Patient_1_Prostate                  3.916033  3.713659   \n",
       "Patient_1_Liver_Patient_1_Brain - Amygdala          3.916033  3.713659   \n",
       "Patient_1_Liver_Patient_1_Brain - Hypothalamus      3.916033  3.713659   \n",
       "Patient_1_Liver_Patient_1_Artery - Coronary         3.916033  3.713659   \n",
       "Patient_1_Liver_Patient_1_Testis                    3.916033  3.713659   \n",
       "Patient_1_Liver_Patient_1_Adrenal Gland             3.916033  3.713659   \n",
       "Patient_1_Liver_Patient_1_Uterus                    3.916033  3.713659   \n",
       "Patient_1_Liver_Patient_1_Skin - Sun Exposed (L...  3.916033  3.713659   \n",
       "Patient_1_Liver_Patient_1_Thyroid                   3.916033  3.713659   \n",
       "\n",
       "                                                      Gene_3    Gene_4  \\\n",
       "Patient_1_Liver_Patient_1_Esophagus - Muscularis    3.405517  3.543074   \n",
       "Patient_1_Liver_Patient_1_Prostate                  3.405517  3.543074   \n",
       "Patient_1_Liver_Patient_1_Brain - Amygdala          3.405517  3.543074   \n",
       "Patient_1_Liver_Patient_1_Brain - Hypothalamus      3.405517  3.543074   \n",
       "Patient_1_Liver_Patient_1_Artery - Coronary         3.405517  3.543074   \n",
       "Patient_1_Liver_Patient_1_Testis                    3.405517  3.543074   \n",
       "Patient_1_Liver_Patient_1_Adrenal Gland             3.405517  3.543074   \n",
       "Patient_1_Liver_Patient_1_Uterus                    3.405517  3.543074   \n",
       "Patient_1_Liver_Patient_1_Skin - Sun Exposed (L...  3.405517  3.543074   \n",
       "Patient_1_Liver_Patient_1_Thyroid                   3.405517  3.543074   \n",
       "\n",
       "                                                      Gene_5    Gene_6  \\\n",
       "Patient_1_Liver_Patient_1_Esophagus - Muscularis    3.619406  3.448242   \n",
       "Patient_1_Liver_Patient_1_Prostate                  3.619406  3.448242   \n",
       "Patient_1_Liver_Patient_1_Brain - Amygdala          3.619406  3.448242   \n",
       "Patient_1_Liver_Patient_1_Brain - Hypothalamus      3.619406  3.448242   \n",
       "Patient_1_Liver_Patient_1_Artery - Coronary         3.619406  3.448242   \n",
       "Patient_1_Liver_Patient_1_Testis                    3.619406  3.448242   \n",
       "Patient_1_Liver_Patient_1_Adrenal Gland             3.619406  3.448242   \n",
       "Patient_1_Liver_Patient_1_Uterus                    3.619406  3.448242   \n",
       "Patient_1_Liver_Patient_1_Skin - Sun Exposed (L...  3.619406  3.448242   \n",
       "Patient_1_Liver_Patient_1_Thyroid                   3.619406  3.448242   \n",
       "\n",
       "                                                      Gene_7    Gene_8  \\\n",
       "Patient_1_Liver_Patient_1_Esophagus - Muscularis    3.617734  3.066326   \n",
       "Patient_1_Liver_Patient_1_Prostate                  3.617734  3.066326   \n",
       "Patient_1_Liver_Patient_1_Brain - Amygdala          3.617734  3.066326   \n",
       "Patient_1_Liver_Patient_1_Brain - Hypothalamus      3.617734  3.066326   \n",
       "Patient_1_Liver_Patient_1_Artery - Coronary         3.617734  3.066326   \n",
       "Patient_1_Liver_Patient_1_Testis                    3.617734  3.066326   \n",
       "Patient_1_Liver_Patient_1_Adrenal Gland             3.617734  3.066326   \n",
       "Patient_1_Liver_Patient_1_Uterus                    3.617734  3.066326   \n",
       "Patient_1_Liver_Patient_1_Skin - Sun Exposed (L...  3.617734  3.066326   \n",
       "Patient_1_Liver_Patient_1_Thyroid                   3.617734  3.066326   \n",
       "\n",
       "                                                      Gene_9   Gene_10  \\\n",
       "Patient_1_Liver_Patient_1_Esophagus - Muscularis    3.318689  3.703463   \n",
       "Patient_1_Liver_Patient_1_Prostate                  3.318689  3.703463   \n",
       "Patient_1_Liver_Patient_1_Brain - Amygdala          3.318689  3.703463   \n",
       "Patient_1_Liver_Patient_1_Brain - Hypothalamus      3.318689  3.703463   \n",
       "Patient_1_Liver_Patient_1_Artery - Coronary         3.318689  3.703463   \n",
       "Patient_1_Liver_Patient_1_Testis                    3.318689  3.703463   \n",
       "Patient_1_Liver_Patient_1_Adrenal Gland             3.318689  3.703463   \n",
       "Patient_1_Liver_Patient_1_Uterus                    3.318689  3.703463   \n",
       "Patient_1_Liver_Patient_1_Skin - Sun Exposed (L...  3.318689  3.703463   \n",
       "Patient_1_Liver_Patient_1_Thyroid                   3.318689  3.703463   \n",
       "\n",
       "                                                       ...      Gene_19991  \\\n",
       "Patient_1_Liver_Patient_1_Esophagus - Muscularis       ...        3.129045   \n",
       "Patient_1_Liver_Patient_1_Prostate                     ...        3.129045   \n",
       "Patient_1_Liver_Patient_1_Brain - Amygdala             ...        3.129045   \n",
       "Patient_1_Liver_Patient_1_Brain - Hypothalamus         ...        3.129045   \n",
       "Patient_1_Liver_Patient_1_Artery - Coronary            ...        3.129045   \n",
       "Patient_1_Liver_Patient_1_Testis                       ...        3.129045   \n",
       "Patient_1_Liver_Patient_1_Adrenal Gland                ...        3.129045   \n",
       "Patient_1_Liver_Patient_1_Uterus                       ...        3.129045   \n",
       "Patient_1_Liver_Patient_1_Skin - Sun Exposed (L...     ...        3.129045   \n",
       "Patient_1_Liver_Patient_1_Thyroid                      ...        3.129045   \n",
       "\n",
       "                                                    Gene_19992  Gene_19993  \\\n",
       "Patient_1_Liver_Patient_1_Esophagus - Muscularis       3.50637    3.386677   \n",
       "Patient_1_Liver_Patient_1_Prostate                     3.50637    3.386677   \n",
       "Patient_1_Liver_Patient_1_Brain - Amygdala             3.50637    3.386677   \n",
       "Patient_1_Liver_Patient_1_Brain - Hypothalamus         3.50637    3.386677   \n",
       "Patient_1_Liver_Patient_1_Artery - Coronary            3.50637    3.386677   \n",
       "Patient_1_Liver_Patient_1_Testis                       3.50637    3.386677   \n",
       "Patient_1_Liver_Patient_1_Adrenal Gland                3.50637    3.386677   \n",
       "Patient_1_Liver_Patient_1_Uterus                       3.50637    3.386677   \n",
       "Patient_1_Liver_Patient_1_Skin - Sun Exposed (L...     3.50637    3.386677   \n",
       "Patient_1_Liver_Patient_1_Thyroid                      3.50637    3.386677   \n",
       "\n",
       "                                                    Gene_19994  Gene_19995  \\\n",
       "Patient_1_Liver_Patient_1_Esophagus - Muscularis      3.308351    3.582972   \n",
       "Patient_1_Liver_Patient_1_Prostate                    3.308351    3.582972   \n",
       "Patient_1_Liver_Patient_1_Brain - Amygdala            3.308351    3.582972   \n",
       "Patient_1_Liver_Patient_1_Brain - Hypothalamus        3.308351    3.582972   \n",
       "Patient_1_Liver_Patient_1_Artery - Coronary           3.308351    3.582972   \n",
       "Patient_1_Liver_Patient_1_Testis                      3.308351    3.582972   \n",
       "Patient_1_Liver_Patient_1_Adrenal Gland               3.308351    3.582972   \n",
       "Patient_1_Liver_Patient_1_Uterus                      3.308351    3.582972   \n",
       "Patient_1_Liver_Patient_1_Skin - Sun Exposed (L...    3.308351    3.582972   \n",
       "Patient_1_Liver_Patient_1_Thyroid                     3.308351    3.582972   \n",
       "\n",
       "                                                    Gene_19996  Gene_19997  \\\n",
       "Patient_1_Liver_Patient_1_Esophagus - Muscularis      3.700877     3.12969   \n",
       "Patient_1_Liver_Patient_1_Prostate                    3.700877     3.12969   \n",
       "Patient_1_Liver_Patient_1_Brain - Amygdala            3.700877     3.12969   \n",
       "Patient_1_Liver_Patient_1_Brain - Hypothalamus        3.700877     3.12969   \n",
       "Patient_1_Liver_Patient_1_Artery - Coronary           3.700877     3.12969   \n",
       "Patient_1_Liver_Patient_1_Testis                      3.700877     3.12969   \n",
       "Patient_1_Liver_Patient_1_Adrenal Gland               3.700877     3.12969   \n",
       "Patient_1_Liver_Patient_1_Uterus                      3.700877     3.12969   \n",
       "Patient_1_Liver_Patient_1_Skin - Sun Exposed (L...    3.700877     3.12969   \n",
       "Patient_1_Liver_Patient_1_Thyroid                     3.700877     3.12969   \n",
       "\n",
       "                                                    Gene_19998  Gene_19999  \\\n",
       "Patient_1_Liver_Patient_1_Esophagus - Muscularis      3.635182    3.661055   \n",
       "Patient_1_Liver_Patient_1_Prostate                    3.635182    3.661055   \n",
       "Patient_1_Liver_Patient_1_Brain - Amygdala            3.635182    3.661055   \n",
       "Patient_1_Liver_Patient_1_Brain - Hypothalamus        3.635182    3.661055   \n",
       "Patient_1_Liver_Patient_1_Artery - Coronary           3.635182    3.661055   \n",
       "Patient_1_Liver_Patient_1_Testis                      3.635182    3.661055   \n",
       "Patient_1_Liver_Patient_1_Adrenal Gland               3.635182    3.661055   \n",
       "Patient_1_Liver_Patient_1_Uterus                      3.635182    3.661055   \n",
       "Patient_1_Liver_Patient_1_Skin - Sun Exposed (L...    3.635182    3.661055   \n",
       "Patient_1_Liver_Patient_1_Thyroid                     3.635182    3.661055   \n",
       "\n",
       "                                                    Gene_20000  \n",
       "Patient_1_Liver_Patient_1_Esophagus - Muscularis      3.289143  \n",
       "Patient_1_Liver_Patient_1_Prostate                    3.289143  \n",
       "Patient_1_Liver_Patient_1_Brain - Amygdala            3.289143  \n",
       "Patient_1_Liver_Patient_1_Brain - Hypothalamus        3.289143  \n",
       "Patient_1_Liver_Patient_1_Artery - Coronary           3.289143  \n",
       "Patient_1_Liver_Patient_1_Testis                      3.289143  \n",
       "Patient_1_Liver_Patient_1_Adrenal Gland               3.289143  \n",
       "Patient_1_Liver_Patient_1_Uterus                      3.289143  \n",
       "Patient_1_Liver_Patient_1_Skin - Sun Exposed (L...    3.289143  \n",
       "Patient_1_Liver_Patient_1_Thyroid                     3.289143  \n",
       "\n",
       "[10 rows x 20000 columns]"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# train data\n",
    "X_train[:10]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene_1</th>\n",
       "      <th>Gene_2</th>\n",
       "      <th>Gene_3</th>\n",
       "      <th>Gene_4</th>\n",
       "      <th>Gene_5</th>\n",
       "      <th>Gene_6</th>\n",
       "      <th>Gene_7</th>\n",
       "      <th>Gene_8</th>\n",
       "      <th>Gene_9</th>\n",
       "      <th>Gene_10</th>\n",
       "      <th>...</th>\n",
       "      <th>Gene_19991</th>\n",
       "      <th>Gene_19992</th>\n",
       "      <th>Gene_19993</th>\n",
       "      <th>Gene_19994</th>\n",
       "      <th>Gene_19995</th>\n",
       "      <th>Gene_19996</th>\n",
       "      <th>Gene_19997</th>\n",
       "      <th>Gene_19998</th>\n",
       "      <th>Gene_19999</th>\n",
       "      <th>Gene_20000</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Patient_101_Artery - Tibial_Patient_101_Liver</th>\n",
       "      <td>3903.0</td>\n",
       "      <td>1838.0</td>\n",
       "      <td>2472.0</td>\n",
       "      <td>2963.0</td>\n",
       "      <td>3733.0</td>\n",
       "      <td>2716.0</td>\n",
       "      <td>1810.0</td>\n",
       "      <td>1186.0</td>\n",
       "      <td>5651.0</td>\n",
       "      <td>1823.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1102.0</td>\n",
       "      <td>3748.0</td>\n",
       "      <td>2659.0</td>\n",
       "      <td>5499.0</td>\n",
       "      <td>1430.0</td>\n",
       "      <td>3045.0</td>\n",
       "      <td>6874.0</td>\n",
       "      <td>1187.0</td>\n",
       "      <td>1010.0</td>\n",
       "      <td>2281.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Patient_101_Artery - Tibial_Patient_101_Brain - Hippocampus</th>\n",
       "      <td>3903.0</td>\n",
       "      <td>1838.0</td>\n",
       "      <td>2472.0</td>\n",
       "      <td>2963.0</td>\n",
       "      <td>3733.0</td>\n",
       "      <td>2716.0</td>\n",
       "      <td>1810.0</td>\n",
       "      <td>1186.0</td>\n",
       "      <td>5651.0</td>\n",
       "      <td>1823.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1102.0</td>\n",
       "      <td>3748.0</td>\n",
       "      <td>2659.0</td>\n",
       "      <td>5499.0</td>\n",
       "      <td>1430.0</td>\n",
       "      <td>3045.0</td>\n",
       "      <td>6874.0</td>\n",
       "      <td>1187.0</td>\n",
       "      <td>1010.0</td>\n",
       "      <td>2281.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Patient_101_Artery - Tibial_Patient_101_Cells - EBV-transformed lymphocytes</th>\n",
       "      <td>3903.0</td>\n",
       "      <td>1838.0</td>\n",
       "      <td>2472.0</td>\n",
       "      <td>2963.0</td>\n",
       "      <td>3733.0</td>\n",
       "      <td>2716.0</td>\n",
       "      <td>1810.0</td>\n",
       "      <td>1186.0</td>\n",
       "      <td>5651.0</td>\n",
       "      <td>1823.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1102.0</td>\n",
       "      <td>3748.0</td>\n",
       "      <td>2659.0</td>\n",
       "      <td>5499.0</td>\n",
       "      <td>1430.0</td>\n",
       "      <td>3045.0</td>\n",
       "      <td>6874.0</td>\n",
       "      <td>1187.0</td>\n",
       "      <td>1010.0</td>\n",
       "      <td>2281.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Patient_101_Artery - Tibial_Patient_101_Brain - Nucleus accumbens (basal ganglia)</th>\n",
       "      <td>3903.0</td>\n",
       "      <td>1838.0</td>\n",
       "      <td>2472.0</td>\n",
       "      <td>2963.0</td>\n",
       "      <td>3733.0</td>\n",
       "      <td>2716.0</td>\n",
       "      <td>1810.0</td>\n",
       "      <td>1186.0</td>\n",
       "      <td>5651.0</td>\n",
       "      <td>1823.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1102.0</td>\n",
       "      <td>3748.0</td>\n",
       "      <td>2659.0</td>\n",
       "      <td>5499.0</td>\n",
       "      <td>1430.0</td>\n",
       "      <td>3045.0</td>\n",
       "      <td>6874.0</td>\n",
       "      <td>1187.0</td>\n",
       "      <td>1010.0</td>\n",
       "      <td>2281.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Patient_101_Artery - Tibial_Patient_101_Heart - Atrial Appendage</th>\n",
       "      <td>3903.0</td>\n",
       "      <td>1838.0</td>\n",
       "      <td>2472.0</td>\n",
       "      <td>2963.0</td>\n",
       "      <td>3733.0</td>\n",
       "      <td>2716.0</td>\n",
       "      <td>1810.0</td>\n",
       "      <td>1186.0</td>\n",
       "      <td>5651.0</td>\n",
       "      <td>1823.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1102.0</td>\n",
       "      <td>3748.0</td>\n",
       "      <td>2659.0</td>\n",
       "      <td>5499.0</td>\n",
       "      <td>1430.0</td>\n",
       "      <td>3045.0</td>\n",
       "      <td>6874.0</td>\n",
       "      <td>1187.0</td>\n",
       "      <td>1010.0</td>\n",
       "      <td>2281.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows  20000 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                    Gene_1  Gene_2  Gene_3  \\\n",
       "Patient_101_Artery - Tibial_Patient_101_Liver       3903.0  1838.0  2472.0   \n",
       "Patient_101_Artery - Tibial_Patient_101_Brain -...  3903.0  1838.0  2472.0   \n",
       "Patient_101_Artery - Tibial_Patient_101_Cells -...  3903.0  1838.0  2472.0   \n",
       "Patient_101_Artery - Tibial_Patient_101_Brain -...  3903.0  1838.0  2472.0   \n",
       "Patient_101_Artery - Tibial_Patient_101_Heart -...  3903.0  1838.0  2472.0   \n",
       "\n",
       "                                                    Gene_4  Gene_5  Gene_6  \\\n",
       "Patient_101_Artery - Tibial_Patient_101_Liver       2963.0  3733.0  2716.0   \n",
       "Patient_101_Artery - Tibial_Patient_101_Brain -...  2963.0  3733.0  2716.0   \n",
       "Patient_101_Artery - Tibial_Patient_101_Cells -...  2963.0  3733.0  2716.0   \n",
       "Patient_101_Artery - Tibial_Patient_101_Brain -...  2963.0  3733.0  2716.0   \n",
       "Patient_101_Artery - Tibial_Patient_101_Heart -...  2963.0  3733.0  2716.0   \n",
       "\n",
       "                                                    Gene_7  Gene_8  Gene_9  \\\n",
       "Patient_101_Artery - Tibial_Patient_101_Liver       1810.0  1186.0  5651.0   \n",
       "Patient_101_Artery - Tibial_Patient_101_Brain -...  1810.0  1186.0  5651.0   \n",
       "Patient_101_Artery - Tibial_Patient_101_Cells -...  1810.0  1186.0  5651.0   \n",
       "Patient_101_Artery - Tibial_Patient_101_Brain -...  1810.0  1186.0  5651.0   \n",
       "Patient_101_Artery - Tibial_Patient_101_Heart -...  1810.0  1186.0  5651.0   \n",
       "\n",
       "                                                    Gene_10     ...      \\\n",
       "Patient_101_Artery - Tibial_Patient_101_Liver        1823.0     ...       \n",
       "Patient_101_Artery - Tibial_Patient_101_Brain -...   1823.0     ...       \n",
       "Patient_101_Artery - Tibial_Patient_101_Cells -...   1823.0     ...       \n",
       "Patient_101_Artery - Tibial_Patient_101_Brain -...   1823.0     ...       \n",
       "Patient_101_Artery - Tibial_Patient_101_Heart -...   1823.0     ...       \n",
       "\n",
       "                                                    Gene_19991  Gene_19992  \\\n",
       "Patient_101_Artery - Tibial_Patient_101_Liver           1102.0      3748.0   \n",
       "Patient_101_Artery - Tibial_Patient_101_Brain -...      1102.0      3748.0   \n",
       "Patient_101_Artery - Tibial_Patient_101_Cells -...      1102.0      3748.0   \n",
       "Patient_101_Artery - Tibial_Patient_101_Brain -...      1102.0      3748.0   \n",
       "Patient_101_Artery - Tibial_Patient_101_Heart -...      1102.0      3748.0   \n",
       "\n",
       "                                                    Gene_19993  Gene_19994  \\\n",
       "Patient_101_Artery - Tibial_Patient_101_Liver           2659.0      5499.0   \n",
       "Patient_101_Artery - Tibial_Patient_101_Brain -...      2659.0      5499.0   \n",
       "Patient_101_Artery - Tibial_Patient_101_Cells -...      2659.0      5499.0   \n",
       "Patient_101_Artery - Tibial_Patient_101_Brain -...      2659.0      5499.0   \n",
       "Patient_101_Artery - Tibial_Patient_101_Heart -...      2659.0      5499.0   \n",
       "\n",
       "                                                    Gene_19995  Gene_19996  \\\n",
       "Patient_101_Artery - Tibial_Patient_101_Liver           1430.0      3045.0   \n",
       "Patient_101_Artery - Tibial_Patient_101_Brain -...      1430.0      3045.0   \n",
       "Patient_101_Artery - Tibial_Patient_101_Cells -...      1430.0      3045.0   \n",
       "Patient_101_Artery - Tibial_Patient_101_Brain -...      1430.0      3045.0   \n",
       "Patient_101_Artery - Tibial_Patient_101_Heart -...      1430.0      3045.0   \n",
       "\n",
       "                                                    Gene_19997  Gene_19998  \\\n",
       "Patient_101_Artery - Tibial_Patient_101_Liver           6874.0      1187.0   \n",
       "Patient_101_Artery - Tibial_Patient_101_Brain -...      6874.0      1187.0   \n",
       "Patient_101_Artery - Tibial_Patient_101_Cells -...      6874.0      1187.0   \n",
       "Patient_101_Artery - Tibial_Patient_101_Brain -...      6874.0      1187.0   \n",
       "Patient_101_Artery - Tibial_Patient_101_Heart -...      6874.0      1187.0   \n",
       "\n",
       "                                                    Gene_19999  Gene_20000  \n",
       "Patient_101_Artery - Tibial_Patient_101_Liver           1010.0      2281.0  \n",
       "Patient_101_Artery - Tibial_Patient_101_Brain -...      1010.0      2281.0  \n",
       "Patient_101_Artery - Tibial_Patient_101_Cells -...      1010.0      2281.0  \n",
       "Patient_101_Artery - Tibial_Patient_101_Brain -...      1010.0      2281.0  \n",
       "Patient_101_Artery - Tibial_Patient_101_Heart -...      1010.0      2281.0  \n",
       "\n",
       "[5 rows x 20000 columns]"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "X_valid[:5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "ArrayView([3903., 1838., 2472., 2963., 3733., 2716., 1810.], dtype=float32)"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "annobj[:,'Patient_101_Artery - Tibial'].X[:7]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python (rep)",
   "language": "python",
   "name": "rep"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
