{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Purpose: read VEG-output line-by-line, parse needed info"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "read test file \n",
    "parse info\n",
    "\n",
    "run buffered on whole file "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Testfiles if needed\n",
    "#testVEP=\"/s/project/rep/processed/minimalExample/vep.head.tsv\"\n",
    "#testVEP=\"/s/project/rep/processed/minimalExample/vep1LineTest.tsv\"\n",
    "#testVEP=\"/s/project/rep/processed/minimalExample/vep5LineTest.tsv\"\n",
    "#outFile=testVEP+\".features.tsv\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "#specifying field names for consistency\n",
    "f_subject=\"subjectID\"\n",
    "f_gene=\"Gene\"\n",
    "f_consequence=\"Consequence\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## sort command\n",
    "srun -c8 -N1 --mem=20G --pty /bin/bash ~/REP/rep/rep/data/sort_VEP.sh /s/project/scared/paper/revision/run0910-final/data/GTEx_variant_enrichment/rareModerateNHighVariantsTable.tsv ./sortedVEPoutput.tsv                                    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## create new header"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "\n",
    "#(base) [reinharj@ouga04 ~]$ wc -l  /s/project/scared/paper/revision/run0910-final/data/GTEx_variant_enrichment/rareModerateNHighVariantsTable.tsv     \n",
    "#1972142 /s/project/scared/paper/revision/run0910-final/data/GTEx_variant_enrichment/rareModerateNHighVariantsTable.tsv\n",
    "#(base) [reinharj@ouga04 ~]$ cat  /s/project/scared/paper/revision/run0910-final/data/GTEx_variant_enrichment/rareModerateNHighVariantsTable.tsv |  cut -f 40 | sort -u             \n",
    "IMPACT=\"\"\"HIGH\n",
    "MODERATE\"\"\"\n",
    "#(base) [reinharj@ouga04 ~]$ cat  /s/project/scared/paper/revision/run0910-final/data/GTEx_variant_enrichment/rareModerateNHighVariantsTable.tsv |  cut -f 7 | sort -u  \n",
    "GT=\"\"\"0/1\n",
    "1/1\"\"\"\n",
    "\n",
    "#(base) [reinharj@ouga04 ~]$ cat  /s/project/scared/paper/revision/run0910-final/data/GTEx_variant_enrichment/rareModerateNHighVariantsTable.tsv |  cut -f 39 | sort -u \n",
    "Consequence=\"\"\"frameshift_variant\n",
    "frameshift_variant&splice_region_variant\n",
    "frameshift_variant&splice_region_variant&intron_variant\n",
    "frameshift_variant&start_lost\n",
    "frameshift_variant&stop_retained_variant\n",
    "inframe_deletion\n",
    "inframe_deletion&splice_region_variant\n",
    "inframe_insertion\n",
    "inframe_insertion&incomplete_terminal_codon_variant&coding_sequence_variant\n",
    "inframe_insertion&splice_region_variant\n",
    "inframe_insertion&stop_retained_variant\n",
    "missense_variant\n",
    "missense_variant&splice_region_variant\n",
    "protein_altering_variant\n",
    "protein_altering_variant&splice_region_variant\n",
    "splice_acceptor_variant\n",
    "splice_acceptor_variant&3_prime_UTR_variant\n",
    "splice_acceptor_variant&3_prime_UTR_variant&intron_variant\n",
    "splice_acceptor_variant&5_prime_UTR_variant\n",
    "splice_acceptor_variant&5_prime_UTR_variant&intron_variant\n",
    "splice_acceptor_variant&coding_sequence_variant\n",
    "splice_acceptor_variant&coding_sequence_variant&3_prime_UTR_variant&intron_variant\n",
    "splice_acceptor_variant&coding_sequence_variant&5_prime_UTR_variant&intron_variant\n",
    "splice_acceptor_variant&coding_sequence_variant&intron_variant\n",
    "splice_acceptor_variant&intron_variant\n",
    "splice_acceptor_variant&intron_variant&non_coding_transcript_variant\n",
    "splice_acceptor_variant&non_coding_transcript_exon_variant&intron_variant&non_coding_transcript_variant\n",
    "splice_acceptor_variant&non_coding_transcript_exon_variant&non_coding_transcript_variant\n",
    "splice_acceptor_variant&non_coding_transcript_variant\n",
    "splice_acceptor_variant&splice_donor_variant&coding_sequence_variant&intron_variant\n",
    "splice_acceptor_variant&splice_donor_variant&frameshift_variant&intron_variant\n",
    "splice_acceptor_variant&splice_donor_variant&intron_variant\n",
    "splice_donor_variant\n",
    "splice_donor_variant&3_prime_UTR_variant\n",
    "splice_donor_variant&3_prime_UTR_variant&intron_variant\n",
    "splice_donor_variant&5_prime_UTR_variant\n",
    "splice_donor_variant&5_prime_UTR_variant&intron_variant\n",
    "splice_donor_variant&coding_sequence_variant\n",
    "splice_donor_variant&coding_sequence_variant&3_prime_UTR_variant\n",
    "splice_donor_variant&coding_sequence_variant&5_prime_UTR_variant&intron_variant\n",
    "splice_donor_variant&coding_sequence_variant&intron_variant\n",
    "splice_donor_variant&frameshift_variant\n",
    "splice_donor_variant&intron_variant\n",
    "splice_donor_variant&intron_variant&non_coding_transcript_variant\n",
    "splice_donor_variant&non_coding_transcript_exon_variant&intron_variant&non_coding_transcript_variant\n",
    "splice_donor_variant&non_coding_transcript_exon_variant&non_coding_transcript_variant\n",
    "splice_donor_variant&non_coding_transcript_variant\n",
    "start_lost\n",
    "start_lost&inframe_deletion\n",
    "start_lost&splice_region_variant\n",
    "stop_gained\n",
    "stop_gained&frameshift_variant\n",
    "stop_gained&frameshift_variant&splice_region_variant\n",
    "stop_gained&inframe_deletion\n",
    "stop_gained&inframe_insertion\n",
    "stop_gained&inframe_insertion&splice_region_variant\n",
    "stop_gained&protein_altering_variant\n",
    "stop_gained&protein_altering_variant&splice_region_variant\n",
    "stop_gained&splice_region_variant\n",
    "stop_gained&start_lost\n",
    "stop_lost\n",
    "stop_lost&incomplete_terminal_codon_variant\n",
    "stop_lost&inframe_deletion\n",
    "stop_lost&splice_region_variant\"\"\"\n",
    "\n",
    "#(base) [reinharj@ouga04 ~]$ cat  /s/project/scared/paper/revision/run0910-final/data/GTEx_variant_enrichment/rareModerateNHighVariantsTable.tsv |  cut -f 101 | sort -u\n",
    "LoF=\"\"\"HC\n",
    "LC\n",
    "NA\"\"\"\n",
    "\n",
    "\n",
    "stdFields=[\"subject\",\"gene\"]\n",
    "siftNpoly=[\"sift_min\",\"sift_avg\",\"polyphen_max\",\"polyphen_avg\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "IMPACT=IMPACT.split(\"\\n\")\n",
    "GT=GT.split(\"\\n\")\n",
    "Consequence=Consequence.split(\"\\n\")\n",
    "LoF=LoF.split(\"\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "impactH=[]\n",
    "for i in IMPACT:\n",
    "    for j in GT:\n",
    "        impactH.append(\"c_impact_\" + keys2str([i,j]))\n",
    "conH=[]      \n",
    "for i in Consequence:\n",
    "    for j in GT:\n",
    "        conH.append(\"c_consequence_\" + keys2str([i,j]))\n",
    "        \n",
    "lofH=[]\n",
    "for i in LoF:\n",
    "    lofH.append(\"c_lof_\" + i)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['subject',\n",
       " 'gene',\n",
       " 'c_consequence_frameshift_variant-0/1',\n",
       " 'c_consequence_frameshift_variant-1/1',\n",
       " 'c_consequence_frameshift_variant&splice_region_variant-0/1',\n",
       " 'c_consequence_frameshift_variant&splice_region_variant-1/1',\n",
       " 'c_consequence_frameshift_variant&splice_region_variant&intron_variant-0/1',\n",
       " 'c_consequence_frameshift_variant&splice_region_variant&intron_variant-1/1',\n",
       " 'c_consequence_frameshift_variant&start_lost-0/1',\n",
       " 'c_consequence_frameshift_variant&start_lost-1/1',\n",
       " 'c_consequence_frameshift_variant&stop_retained_variant-0/1',\n",
       " 'c_consequence_frameshift_variant&stop_retained_variant-1/1',\n",
       " 'c_consequence_inframe_deletion-0/1',\n",
       " 'c_consequence_inframe_deletion-1/1',\n",
       " 'c_consequence_inframe_deletion&splice_region_variant-0/1',\n",
       " 'c_consequence_inframe_deletion&splice_region_variant-1/1',\n",
       " 'c_consequence_inframe_insertion-0/1',\n",
       " 'c_consequence_inframe_insertion-1/1',\n",
       " 'c_consequence_inframe_insertion&incomplete_terminal_codon_variant&coding_sequence_variant-0/1',\n",
       " 'c_consequence_inframe_insertion&incomplete_terminal_codon_variant&coding_sequence_variant-1/1',\n",
       " 'c_consequence_inframe_insertion&splice_region_variant-0/1',\n",
       " 'c_consequence_inframe_insertion&splice_region_variant-1/1',\n",
       " 'c_consequence_inframe_insertion&stop_retained_variant-0/1',\n",
       " 'c_consequence_inframe_insertion&stop_retained_variant-1/1',\n",
       " 'c_consequence_missense_variant-0/1',\n",
       " 'c_consequence_missense_variant-1/1',\n",
       " 'c_consequence_missense_variant&splice_region_variant-0/1',\n",
       " 'c_consequence_missense_variant&splice_region_variant-1/1',\n",
       " 'c_consequence_protein_altering_variant-0/1',\n",
       " 'c_consequence_protein_altering_variant-1/1',\n",
       " 'c_consequence_protein_altering_variant&splice_region_variant-0/1',\n",
       " 'c_consequence_protein_altering_variant&splice_region_variant-1/1',\n",
       " 'c_consequence_splice_acceptor_variant-0/1',\n",
       " 'c_consequence_splice_acceptor_variant-1/1',\n",
       " 'c_consequence_splice_acceptor_variant&3_prime_UTR_variant-0/1',\n",
       " 'c_consequence_splice_acceptor_variant&3_prime_UTR_variant-1/1',\n",
       " 'c_consequence_splice_acceptor_variant&3_prime_UTR_variant&intron_variant-0/1',\n",
       " 'c_consequence_splice_acceptor_variant&3_prime_UTR_variant&intron_variant-1/1',\n",
       " 'c_consequence_splice_acceptor_variant&5_prime_UTR_variant-0/1',\n",
       " 'c_consequence_splice_acceptor_variant&5_prime_UTR_variant-1/1',\n",
       " 'c_consequence_splice_acceptor_variant&5_prime_UTR_variant&intron_variant-0/1',\n",
       " 'c_consequence_splice_acceptor_variant&5_prime_UTR_variant&intron_variant-1/1',\n",
       " 'c_consequence_splice_acceptor_variant&coding_sequence_variant-0/1',\n",
       " 'c_consequence_splice_acceptor_variant&coding_sequence_variant-1/1',\n",
       " 'c_consequence_splice_acceptor_variant&coding_sequence_variant&3_prime_UTR_variant&intron_variant-0/1',\n",
       " 'c_consequence_splice_acceptor_variant&coding_sequence_variant&3_prime_UTR_variant&intron_variant-1/1',\n",
       " 'c_consequence_splice_acceptor_variant&coding_sequence_variant&5_prime_UTR_variant&intron_variant-0/1',\n",
       " 'c_consequence_splice_acceptor_variant&coding_sequence_variant&5_prime_UTR_variant&intron_variant-1/1',\n",
       " 'c_consequence_splice_acceptor_variant&coding_sequence_variant&intron_variant-0/1',\n",
       " 'c_consequence_splice_acceptor_variant&coding_sequence_variant&intron_variant-1/1',\n",
       " 'c_consequence_splice_acceptor_variant&intron_variant-0/1',\n",
       " 'c_consequence_splice_acceptor_variant&intron_variant-1/1',\n",
       " 'c_consequence_splice_acceptor_variant&intron_variant&non_coding_transcript_variant-0/1',\n",
       " 'c_consequence_splice_acceptor_variant&intron_variant&non_coding_transcript_variant-1/1',\n",
       " 'c_consequence_splice_acceptor_variant&non_coding_transcript_exon_variant&intron_variant&non_coding_transcript_variant-0/1',\n",
       " 'c_consequence_splice_acceptor_variant&non_coding_transcript_exon_variant&intron_variant&non_coding_transcript_variant-1/1',\n",
       " 'c_consequence_splice_acceptor_variant&non_coding_transcript_exon_variant&non_coding_transcript_variant-0/1',\n",
       " 'c_consequence_splice_acceptor_variant&non_coding_transcript_exon_variant&non_coding_transcript_variant-1/1',\n",
       " 'c_consequence_splice_acceptor_variant&non_coding_transcript_variant-0/1',\n",
       " 'c_consequence_splice_acceptor_variant&non_coding_transcript_variant-1/1',\n",
       " 'c_consequence_splice_acceptor_variant&splice_donor_variant&coding_sequence_variant&intron_variant-0/1',\n",
       " 'c_consequence_splice_acceptor_variant&splice_donor_variant&coding_sequence_variant&intron_variant-1/1',\n",
       " 'c_consequence_splice_acceptor_variant&splice_donor_variant&frameshift_variant&intron_variant-0/1',\n",
       " 'c_consequence_splice_acceptor_variant&splice_donor_variant&frameshift_variant&intron_variant-1/1',\n",
       " 'c_consequence_splice_acceptor_variant&splice_donor_variant&intron_variant-0/1',\n",
       " 'c_consequence_splice_acceptor_variant&splice_donor_variant&intron_variant-1/1',\n",
       " 'c_consequence_splice_donor_variant-0/1',\n",
       " 'c_consequence_splice_donor_variant-1/1',\n",
       " 'c_consequence_splice_donor_variant&3_prime_UTR_variant-0/1',\n",
       " 'c_consequence_splice_donor_variant&3_prime_UTR_variant-1/1',\n",
       " 'c_consequence_splice_donor_variant&3_prime_UTR_variant&intron_variant-0/1',\n",
       " 'c_consequence_splice_donor_variant&3_prime_UTR_variant&intron_variant-1/1',\n",
       " 'c_consequence_splice_donor_variant&5_prime_UTR_variant-0/1',\n",
       " 'c_consequence_splice_donor_variant&5_prime_UTR_variant-1/1',\n",
       " 'c_consequence_splice_donor_variant&5_prime_UTR_variant&intron_variant-0/1',\n",
       " 'c_consequence_splice_donor_variant&5_prime_UTR_variant&intron_variant-1/1',\n",
       " 'c_consequence_splice_donor_variant&coding_sequence_variant-0/1',\n",
       " 'c_consequence_splice_donor_variant&coding_sequence_variant-1/1',\n",
       " 'c_consequence_splice_donor_variant&coding_sequence_variant&3_prime_UTR_variant-0/1',\n",
       " 'c_consequence_splice_donor_variant&coding_sequence_variant&3_prime_UTR_variant-1/1',\n",
       " 'c_consequence_splice_donor_variant&coding_sequence_variant&5_prime_UTR_variant&intron_variant-0/1',\n",
       " 'c_consequence_splice_donor_variant&coding_sequence_variant&5_prime_UTR_variant&intron_variant-1/1',\n",
       " 'c_consequence_splice_donor_variant&coding_sequence_variant&intron_variant-0/1',\n",
       " 'c_consequence_splice_donor_variant&coding_sequence_variant&intron_variant-1/1',\n",
       " 'c_consequence_splice_donor_variant&frameshift_variant-0/1',\n",
       " 'c_consequence_splice_donor_variant&frameshift_variant-1/1',\n",
       " 'c_consequence_splice_donor_variant&intron_variant-0/1',\n",
       " 'c_consequence_splice_donor_variant&intron_variant-1/1',\n",
       " 'c_consequence_splice_donor_variant&intron_variant&non_coding_transcript_variant-0/1',\n",
       " 'c_consequence_splice_donor_variant&intron_variant&non_coding_transcript_variant-1/1',\n",
       " 'c_consequence_splice_donor_variant&non_coding_transcript_exon_variant&intron_variant&non_coding_transcript_variant-0/1',\n",
       " 'c_consequence_splice_donor_variant&non_coding_transcript_exon_variant&intron_variant&non_coding_transcript_variant-1/1',\n",
       " 'c_consequence_splice_donor_variant&non_coding_transcript_exon_variant&non_coding_transcript_variant-0/1',\n",
       " 'c_consequence_splice_donor_variant&non_coding_transcript_exon_variant&non_coding_transcript_variant-1/1',\n",
       " 'c_consequence_splice_donor_variant&non_coding_transcript_variant-0/1',\n",
       " 'c_consequence_splice_donor_variant&non_coding_transcript_variant-1/1',\n",
       " 'c_consequence_start_lost-0/1',\n",
       " 'c_consequence_start_lost-1/1',\n",
       " 'c_consequence_start_lost&inframe_deletion-0/1',\n",
       " 'c_consequence_start_lost&inframe_deletion-1/1',\n",
       " 'c_consequence_start_lost&splice_region_variant-0/1',\n",
       " 'c_consequence_start_lost&splice_region_variant-1/1',\n",
       " 'c_consequence_stop_gained-0/1',\n",
       " 'c_consequence_stop_gained-1/1',\n",
       " 'c_consequence_stop_gained&frameshift_variant-0/1',\n",
       " 'c_consequence_stop_gained&frameshift_variant-1/1',\n",
       " 'c_consequence_stop_gained&frameshift_variant&splice_region_variant-0/1',\n",
       " 'c_consequence_stop_gained&frameshift_variant&splice_region_variant-1/1',\n",
       " 'c_consequence_stop_gained&inframe_deletion-0/1',\n",
       " 'c_consequence_stop_gained&inframe_deletion-1/1',\n",
       " 'c_consequence_stop_gained&inframe_insertion-0/1',\n",
       " 'c_consequence_stop_gained&inframe_insertion-1/1',\n",
       " 'c_consequence_stop_gained&inframe_insertion&splice_region_variant-0/1',\n",
       " 'c_consequence_stop_gained&inframe_insertion&splice_region_variant-1/1',\n",
       " 'c_consequence_stop_gained&protein_altering_variant-0/1',\n",
       " 'c_consequence_stop_gained&protein_altering_variant-1/1',\n",
       " 'c_consequence_stop_gained&protein_altering_variant&splice_region_variant-0/1',\n",
       " 'c_consequence_stop_gained&protein_altering_variant&splice_region_variant-1/1',\n",
       " 'c_consequence_stop_gained&splice_region_variant-0/1',\n",
       " 'c_consequence_stop_gained&splice_region_variant-1/1',\n",
       " 'c_consequence_stop_gained&start_lost-0/1',\n",
       " 'c_consequence_stop_gained&start_lost-1/1',\n",
       " 'c_consequence_stop_lost-0/1',\n",
       " 'c_consequence_stop_lost-1/1',\n",
       " 'c_consequence_stop_lost&incomplete_terminal_codon_variant-0/1',\n",
       " 'c_consequence_stop_lost&incomplete_terminal_codon_variant-1/1',\n",
       " 'c_consequence_stop_lost&inframe_deletion-0/1',\n",
       " 'c_consequence_stop_lost&inframe_deletion-1/1',\n",
       " 'c_consequence_stop_lost&splice_region_variant-0/1',\n",
       " 'c_consequence_stop_lost&splice_region_variant-1/1',\n",
       " 'c_impact_HIGH-0/1',\n",
       " 'c_impact_HIGH-1/1',\n",
       " 'c_impact_MODERATE-0/1',\n",
       " 'c_impact_MODERATE-1/1',\n",
       " 'sift_min',\n",
       " 'sift_avg',\n",
       " 'polyphen_max',\n",
       " 'polyphen_avg',\n",
       " 'c_lof_HC',\n",
       " 'c_lof_LC',\n",
       " 'c_lof_NA']"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "header = stdFields + conH + impactH + siftNpoly + lofH\n",
    "header"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## parser"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "import pandas as pd\n",
    "outputDf = pd.DataFrame(index=[\"subject\",\"gene\",\"\"])\n",
    "\n",
    "def cacheLine(line):\n",
    "    global lineCache\n",
    "    lineCache.append(line)\n",
    "    \n",
    "def processCacheNrestartWith(line: dict):\n",
    "\n",
    "    processCache()\n",
    "    restartCache(line) \n",
    "\n",
    "def processCacheNrestartWith_old(line: dict):\n",
    "\n",
    "    processCache_old()\n",
    "    restartCache(line) \n",
    "    \n",
    "def processCache(lineCache, outWriter):\n",
    "    self.lineCachhe\n",
    "\n",
    "    #continue at the start of the file\n",
    "    if len(lineCache)==0: return \n",
    "    \n",
    "    #concert to dataframe and reset list\n",
    "    df1Gene1Indiv=pd.DataFrame(lineCache)\n",
    "    lineCache=[]\n",
    " \n",
    "    #create all the dicts for all the gene/subject-pairs\n",
    "    dictList={}\n",
    "    #grouping by gene/subject pair\n",
    "    instanceGrouped=df1Gene1Indiv.groupby([f_gene,f_subject],sort=False) \n",
    "    for group in instanceGrouped:   # instanceGrouped.groups.keys(): is SORTED again! ->messes up order in frozenDict... -> would not keep input-order\n",
    "        keys=group[0]\n",
    "        idKey=keys2str(keys)\n",
    "        dictList[idKey]={\"gene\":keys[0],\"subject\":keys[1]}\n",
    "    \n",
    "    \n",
    "    #c_CONSEQUENCE_[hetero/homo] -> homo: only 1.3% of cases...\n",
    "    consCounts=df1Gene1Indiv.groupby([f_gene,f_subject,\"Consequence\",\"GT\"])[\"Consequence\"].count()\n",
    "    for keys, count in consCounts.iteritems():\n",
    "        idKey=keys2str(keys[0:2])\n",
    "        fieldKey=keys2str(keys[2:])\n",
    "        dictList[idKey][\"c_consequence_\"+fieldKey]=count\n",
    "        \n",
    "    #c_CONSEQUENCE_[hetero/homo] -> homo: only 1.3% of cases...\n",
    "    impactCounts=df1Gene1Indiv.groupby([f_gene,f_subject,\"IMPACT\",\"GT\"])[\"IMPACT\"].count()\n",
    "    for keys, count in impactCounts.iteritems():\n",
    "        idKey=keys2str(keys[0:2])\n",
    "        fieldKey=keys2str(keys[2:])\n",
    "        dictList[idKey][\"c_impact_\"+fieldKey]=count\n",
    "    \n",
    "    #[polyphen|sift|_[avg|max] # SIFT most-serious is 0, not 1 !! -) need min\n",
    "    #convert text values to probs\n",
    "    df1Gene1Indiv[\"SIFT\"]=df1Gene1Indiv[\"SIFT\"].apply(extractSiftPolyProb)\n",
    "    df1Gene1Indiv[\"PolyPhen\"]=df1Gene1Indiv[\"PolyPhen\"].apply(extractSiftPolyProb)   \n",
    "    #TODO actually use scores...\n",
    "    values=instanceGrouped[\"SIFT\"].min()\n",
    "    for keys, val in values.iteritems():\n",
    "        idKey=keys2str(keys[0:2])\n",
    "        dictList[idKey][\"sift_min\"]=val\n",
    "    values=instanceGrouped[\"SIFT\"].mean()\n",
    "    for keys, val in values.iteritems():\n",
    "        idKey=keys2str(keys[0:2])\n",
    "        dictList[idKey][\"sift_avg\"]=val   \n",
    "        \n",
    "    values=instanceGrouped[\"PolyPhen\"].max()\n",
    "    for keys, val in values.iteritems():\n",
    "        idKey=keys2str(keys[0:2])\n",
    "        dictList[idKey][\"polyphen_max\"]=val\n",
    "    values=instanceGrouped[\"PolyPhen\"].mean()\n",
    "    for keys, val in values.iteritems():\n",
    "        idKey=keys2str(keys[0:2])\n",
    "        dictList[idKey][\"polyphen_avg\"]=val\n",
    "    \n",
    "    \n",
    "    \n",
    "    \n",
    "    \n",
    "    # LoF_[LC|HC] (ignoring NA) -> this is only 3% and 0.3% of cases\n",
    "    lofCounts=df1Gene1Indiv.groupby([f_gene,f_subject,\"LoF\"])[\"LoF\"].count()\n",
    "    \n",
    "    for keys, count in lofCounts.iteritems():\n",
    "        idKey=keys2str(keys[0:2])\n",
    "        fieldKey=keys2str(keys[2:])\n",
    "        dictList[idKey][\"c_lof_\"+fieldKey]=count\n",
    "        \n",
    "        \n",
    "\n",
    "    outWriter.writerows(dictList.values())\n",
    "    \n",
    "    return "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "   \n",
    "def restartCache(line):\n",
    "    global idTuple\n",
    "    global lineCache\n",
    "    \n",
    "    idTuple=(line[f_subject],line[f_gene])\n",
    "    lineCache=[line]\n",
    "    \n",
    "def extractSiftPolyProb(string):\n",
    "    \"\"\"e.g. polyphen-field: probably_damaging(0.993)\"\"\"\n",
    "    if string==\"NA\": return np.nan #NA\n",
    "    try:\n",
    "        return float(string.split(\"(\")[1][:-1]) # get string after \"(\", remove last letter...\n",
    "    except:\n",
    "            print(string)\n",
    "    \n",
    "\n",
    "    \n",
    "\n",
    "def keys2str(keys):\n",
    "    return \"-\".join(keys)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "def filterDict(largeDict):\n",
    "    wantedKeys=[f_subject,f_gene,\"IMPACT\",\"Consequence\",\"GT\",\"SIFT\",\"PolyPhen\",\"LoF\"]\n",
    "    return { wKey:largeDict[wKey] for wKey in wantedKeys}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'lineCache' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<timed exec>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n",
      "\u001b[0;32m<ipython-input-13-90668cd3347b>\u001b[0m in \u001b[0;36mcacheLine\u001b[0;34m(line)\u001b[0m\n\u001b[1;32m      6\u001b[0m \u001b[0;32mdef\u001b[0m \u001b[0mcacheLine\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mline\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      7\u001b[0m     \u001b[0;32mglobal\u001b[0m \u001b[0mlineCache\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 8\u001b[0;31m     \u001b[0mlineCache\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mappend\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mline\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      9\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     10\u001b[0m \u001b[0;32mdef\u001b[0m \u001b[0mprocessCacheNrestartWith\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mline\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mdict\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'lineCache' is not defined"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "#%%prun -s cumulative\n",
    "#emergency exit: killall python --user reinharj\n",
    "\n",
    "import csv\n",
    "\n",
    "\"\"\"\n",
    "for each line:\n",
    "check if new gene or new indiv\n",
    "if not, add to variant-collection\n",
    "if yes: analyse cached variants, reset variant-collection\n",
    "\n",
    "analyse cached variants:\n",
    "    transform to dataframe, then group and analyze...\n",
    "    output format: Gene count_<feature_type>_[hetero|homo] count_impact_[hetero|homo] [polyphen|sift|_[avg|max] LoF_[LC|HC]\n",
    "\"\"\"\n",
    "realVEP=\"/s/project/rep/predictions/vep/sortedVEPoutput.tsv\"\n",
    "outFile=\"/s/project/rep/predictions/vep/aggregatedVEP.tsv\"\n",
    "vepRaw=open(realVEP, \"r\", buffering=8000000)\n",
    "outRaw=open(outFile, \"w\", buffering=8000000)\n",
    "processLimit=100000\n",
    "\n",
    "idTuple=(\"indiv\",\"gene\")\n",
    "lineCache = []\n",
    "\n",
    "vepDict=csv.DictReader(vepRaw,delimiter=\"\\t\")\n",
    "outWriter=csv.DictWriter(outRaw,fieldnames=header)\n",
    "outWriter.writeheader()\n",
    "\n",
    "oldSave=0\n",
    "for c, line in enumerate(vepDict):\n",
    "    if c % 100000 == 0: \n",
    "        print(c)\n",
    "    smallDict=filterDict(line) #reduces time needed from 117 to 52 s for 10k lines... (slow algo)\n",
    "    #continue until limit reached AND differnt gene/subject combo\n",
    "    if len(lineCache) < processLimit or (smallDict[f_subject]== lineCache[-1][f_subject] and smallDict[f_gene]==lineCache[-1][f_gene]): \n",
    "            cacheLine(smallDict)\n",
    "    else: \n",
    "        processCacheNrestartWith(smallDict)\n",
    "        \n",
    "processCache() #end fence-post\n",
    "            \n",
    "vepRaw.close()\n",
    "outRaw.close()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'vepDict' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-5-b46426fd53da>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0;31m# get field of ID\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      2\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0mnumpy\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0mnp\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 3\u001b[0;31m \u001b[0mnp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0margwhere\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mnp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0marray\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mvepDict\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfieldnames\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m==\u001b[0m\u001b[0;34m\"SIFT\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m: name 'vepDict' is not defined"
     ]
    }
   ],
   "source": [
    "# get field of ID\n",
    "import numpy as np\n",
    "np.argwhere(np.array(vepDict.fieldnames)==\"SIFT\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 233,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['variantID',\n",
       " 'seqnames',\n",
       " 'start',\n",
       " 'end',\n",
       " 'strand',\n",
       " 'subjectID',\n",
       " 'GT',\n",
       " 'QUAL',\n",
       " 'FILTER',\n",
       " 'REF',\n",
       " 'ALT',\n",
       " 'AlleleDepthREF',\n",
       " 'AlleleDepthALT',\n",
       " 'AC',\n",
       " 'AF',\n",
       " 'AN',\n",
       " 'BaseQRankSum',\n",
       " 'ClippingRankSum',\n",
       " 'DB',\n",
       " 'DP',\n",
       " 'DS',\n",
       " 'END',\n",
       " 'FS',\n",
       " 'HaplotypeScore',\n",
       " 'InbreedingCoeff',\n",
       " 'MLEAC',\n",
       " 'MLEAF',\n",
       " 'MQ',\n",
       " 'MQRankSum',\n",
       " 'NEGATIVE_TRAIN_SITE',\n",
       " 'POSITIVE_TRAIN_SITE',\n",
       " 'QD',\n",
       " 'ReadPosRankSum',\n",
       " 'SOR',\n",
       " 'VQSLOD',\n",
       " 'culprit',\n",
       " 'width',\n",
       " 'Allele',\n",
       " 'Consequence',\n",
       " 'IMPACT',\n",
       " 'SYMBOL',\n",
       " 'Gene',\n",
       " 'Feature_type',\n",
       " 'Feature',\n",
       " 'BIOTYPE',\n",
       " 'EXON',\n",
       " 'INTRON',\n",
       " 'HGVSc',\n",
       " 'HGVSp',\n",
       " 'cDNA_position',\n",
       " 'CDS_position',\n",
       " 'Protein_position',\n",
       " 'Amino_acids',\n",
       " 'Codons',\n",
       " 'Existing_variation',\n",
       " 'ALLELE_NUM',\n",
       " 'DISTANCE',\n",
       " 'STRAND',\n",
       " 'VARIANT_CLASS',\n",
       " 'MINIMISED',\n",
       " 'SYMBOL_SOURCE',\n",
       " 'HGNC_ID',\n",
       " 'CANONICAL',\n",
       " 'TSL',\n",
       " 'APPRIS',\n",
       " 'CCDS',\n",
       " 'ENSP',\n",
       " 'SWISSPROT',\n",
       " 'TREMBL',\n",
       " 'UNIPARC',\n",
       " 'GENE_PHENO',\n",
       " 'SIFT',\n",
       " 'PolyPhen',\n",
       " 'DOMAINS',\n",
       " 'HGVS_OFFSET',\n",
       " 'GMAF',\n",
       " 'AFR_MAF',\n",
       " 'AMR_MAF',\n",
       " 'EAS_MAF',\n",
       " 'EUR_MAF',\n",
       " 'SAS_MAF',\n",
       " 'AA_MAF',\n",
       " 'EA_MAF',\n",
       " 'ExAC_MAF',\n",
       " 'ExAC_Adj_MAF',\n",
       " 'ExAC_AFR_MAF',\n",
       " 'ExAC_AMR_MAF',\n",
       " 'ExAC_EAS_MAF',\n",
       " 'ExAC_FIN_MAF',\n",
       " 'ExAC_NFE_MAF',\n",
       " 'ExAC_OTH_MAF',\n",
       " 'ExAC_SAS_MAF',\n",
       " 'CLIN_SIG',\n",
       " 'SOMATIC',\n",
       " 'PHENO',\n",
       " 'PUBMED',\n",
       " 'MOTIF_NAME',\n",
       " 'MOTIF_POS',\n",
       " 'HIGH_INF_POS',\n",
       " 'MOTIF_SCORE_CHANGE',\n",
       " 'LoF',\n",
       " 'LoF_filter',\n",
       " 'LoF_flags',\n",
       " 'LoF_info',\n",
       " 'MAF',\n",
       " 'numTranscripts']"
      ]
     },
     "execution_count": 233,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "vepDict.fieldnames"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:florian]",
   "language": "python",
   "name": "conda-env-florian-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
