<!DOCTYPE html>
<html lang="en" class="pb-page"  data-request-id="d2e5ce0f-5486-469f-925a-e7be39a736c1"  
><head data-pb-dropzone="head"><meta name="pbContext" content=";product:product:elsevier\:product\:cell;journal:journal:ajhg;website:website:cell-site;issue:issue:pii\:S0002929718X00032;pageGroup:string:Article View;article:article:pii\:S0002929718305032;page:string:Show Full Text;wgroup:string:Default Website Group"/><meta data-pb-head="head-widgets-start"/>





<noscript>
    <style>.requiresJS {
        display: none
    }

    .dartAd {
        display: block !important
    }</style>
</noscript>
    <script rel="preconnect" async defer src="//d1bxh8uas1mnw7.cloudfront.net/assets/embed.js"></script>

        <script type="text/javascript" src="/wro/g43e~selected-legacy.js"></script>
        <link rel="stylesheet" type="text/css" href="/wro/g43e~selected-legacy.css"/>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Open+Sans"/>
        
            <link rel="stylesheet" type="text/css" href="/products/marlin/cell/releasedAssets/css/build.css"/>
        

















        <meta name="robots" content="noodp" />
    






    <meta name="viewport" content="width=device-width,initial-scale=1,maximum-scale=1, user-scalable=1"/>









        <meta property="og:title" content="Improved Pathogenic Variant Localization via a Hierarchical Model of Sub-regional Intolerance" />
    


        <meta property="og:type" content="Article" />
    


        <meta property="og:url" content="https://www.cell.com/ajhg/abstract/S0002-9297(18)30503-2" />
    



        <meta property="og:site_name" content="The American Journal of Human Genetics" />
    


        <meta property="og:image" content="https://www.cell.com/cms/attachment/1e19e53a-5717-4e17-9cb1-2f8d9e47fcbd/gr3.jpg" />
    


        <meta property="og:description" content="Different parts of a gene can be of differential importance to development and health.
This regional heterogeneity is also apparent in the distribution of disease-associated
mutations, which often cluster in particular regions of disease-associated genes.
The ability to precisely estimate functionally important sub-regions of genes will
be key in correctly deciphering relationships between genetic variation and disease.
Previous methods have had some success using standing human variation to characterize
this variability in importance by measuring sub-regional intolerance, i.e., the depletion
in functional variation from expectation within a given region of a gene." />
    







        <meta name="twitter:title" content="Improved Pathogenic Variant Localization via a Hierarchical Model of Sub-regional Intolerance" />
    




        <meta name="twitter:image" content="https://www.cell.com/cms/attachment/1e19e53a-5717-4e17-9cb1-2f8d9e47fcbd/gr3.jpg" />
    


        <meta name="twitter:description" content="Different parts of a gene can be of differential importance to development and health.
This regional heterogeneity is also apparent in the distribution of disease-associated
mutations, which often cluster in particular regions of disease-associated genes.
The ability to precisely estimate functionally important sub-regions of genes will
be key in correctly deciphering relationships between genetic variation and disease.
Previous methods have had some success using standing human variation to characterize
this variability in importance by measuring sub-regional intolerance, i.e., the depletion
in functional variation from expectation within a given region of a gene." />
    



    <meta name="apple-itunes-app" content="app-id=591217684"/>


    <meta name="google-play-app" content="app-id=com.cell.journalreader" />






<title>Improved Pathogenic Variant Localization via a Hierarchical Model of Sub-regional Intolerance: The American Journal of Human Genetics</title><meta name="citation_journal_title" content="The American Journal of Human Genetics"/> <meta name="citation_journal_abbrev" content="The American Journal of Human Genetics"/> <meta name="citation_publisher" content="Elsevier"/> <meta name="citation_title" content="Improved Pathogenic Variant Localization via a Hierarchical Model of Sub-regional Intolerance"/> <meta name="citation_date" content="2019/02/07"/> <meta name="citation_online_date" content="2019/01/24"/> <meta name="citation_volume" content="104"/> <meta name="citation_issue" content="2"/> <meta name="citation_firstpage" content="299"/> <meta name="citation_lastpage" content="309"/> <meta name="citation_pmid" content="30686509"/> <meta name="citation_abstract_html_url" content="http://www.cell.com/article/S0002929718305032/abstract"/> <meta name="citation_fulltext_html_url" content="http://www.cell.com/article/S0002929718305032/fulltext"/> <meta name="citation_pdf_url" content="http://www.cell.com/article/S0002929718305032/pdf"/> <meta name="citation_issn" content="0002-9297"/> <meta name="citation_issn" content="1537-6605"/> <meta name="citation_language" content="English"/> <meta name="citation_doi" content="10.1016/j.ajhg.2018.12.020"/> <meta name="citation_pmid" content="30686509"/> <meta name="citation_author" content="Tristan J. Hayeck"></meta> <meta name="citation_author_email" content="tristan.hayeck@gmail.com"></meta> <meta name="citation_author_institution" content="Institute for Genomic Medicine, Columbia University, New York, NY 10032, USA"></meta> <meta name="citation_author" content="Nicholas Stong"></meta> <meta name="citation_author_institution" content="Institute for Genomic Medicine, Columbia University, New York, NY 10032, USA"></meta> <meta name="citation_author" content="Charles J. Wolock"></meta> <meta name="citation_author_institution" content="Institute for Genomic Medicine, Columbia University, New York, NY 10032, USA"></meta> <meta name="citation_author" content="Brett Copeland"></meta> <meta name="citation_author_institution" content="Institute for Genomic Medicine, Columbia University, New York, NY 10032, USA"></meta> <meta name="citation_author" content="Sitharthan Kamalakaran"></meta> <meta name="citation_author_institution" content="Institute for Genomic Medicine, Columbia University, New York, NY 10032, USA"></meta> <meta name="citation_author" content="David B. Goldstein"></meta> <meta name="citation_author_institution" content="Institute for Genomic Medicine, Columbia University, New York, NY 10032, USA"></meta> <meta name="citation_author" content="Andrew S. Allen"></meta> <meta name="citation_author_email" content="andrew.s.allen@duke.edu"></meta> <meta name="citation_author_institution" content="Institute for Genomic Medicine, Columbia University, New York, NY 10032, USA"></meta> <meta name="citation_author_institution" content="Division of Integrative Genomics, Department of Biostatistics and Bioinformatics, Duke University, Durham, NC 27710, USA"></meta> <meta name="citation_author_institution" content="Duke Center for Statistical Genetics and Genomics, Duke University, Durham, NC 27710, USA"></meta> <meta name="citation_keywords" content="LIMBR; negative selection; conservation; constraint; intolerance; domains; exons; pathogenic; genic sub-region; RVIS"></meta> <meta name="citation_reference" content="citation_title=Sequencing studies in human genetics: design and interpretation; citation_author=D.B. Goldstein; citation_author=A. Allen; citation_author=J. Keebler; citation_author=E.H. Margulies; citation_author=S. Petrou; citation_author=S. Petrovski; citation_author=S. Sunyaev; citation_journal_title=Nat. Rev. Genet.; citation_volume=14; citation_firstpage=460; citation_lastpage=470; citation_publication_date=2013; "></meta> <meta name="citation_reference" content="citation_title=Settling the score: variant prioritization and Mendelian disease; citation_author=K. Eilbeck; citation_author=A. Quinlan; citation_author=M. Yandell; citation_journal_title=Nat. Rev. Genet.; citation_volume=18; citation_firstpage=599; citation_lastpage=612; citation_publication_date=2017; "></meta> <meta name="citation_reference" content="citation_title=Identifying a high fraction of the human genome to be under selective constraint using GERP++; citation_author=E.V. Davydov; citation_author=D.L. Goode; citation_author=M. Sirota; citation_author=G.M. Cooper; citation_author=A. Sidow; citation_author=S. Batzoglou; citation_journal_title=PLoS Comput. Biol.; citation_volume=6; citation_firstpage=e1001025; citation_publication_date=2010; "></meta> <meta name="citation_reference" content="citation_title=Genic intolerance to functional variation and the interpretation of personal genomes; citation_author=S. Petrovski; citation_author=Q. Wang; citation_author=E.L. Heinzen; citation_author=A.S. Allen; citation_author=D.B. Goldstein; citation_journal_title=PLoS Genet.; citation_volume=9; citation_firstpage=e1003709; citation_publication_date=2013; "></meta> <meta name="citation_reference" content="citation_title=A framework for the interpretation of de novo mutation in human disease; citation_author=K.E. Samocha; citation_author=E.B. Robinson; citation_author=S.J. Sanders; citation_author=C. Stevens; citation_author=A. Sabo; citation_author=L.M. McGrath; citation_author=J.A. Kosmicki; citation_author=K. RehnstrÃ¶m; citation_author=S. Mallick; citation_author=A. Kirby; citation_author=et al.; citation_journal_title=Nat. Genet.; citation_volume=46; citation_firstpage=944; citation_lastpage=950; citation_publication_date=2014; "></meta> <meta name="citation_reference" content="citation_title=De novo gene disruptions in children on the autistic spectrum; citation_author=I. Iossifov; citation_author=M. Ronemus; citation_author=D. Levy; citation_author=Z. Wang; citation_author=I. Hakker; citation_author=J. Rosenbaum; citation_author=B. Yamrom; citation_author=Y.H. Lee; citation_author=G. Narzisi; citation_author=A. Leotta; citation_author=et al.; citation_journal_title=Neuron; citation_volume=74; citation_firstpage=285; citation_lastpage=299; citation_publication_date=2012; "></meta> <meta name="citation_reference" content="citation_title=The intolerance to functional genetic variation of protein domains predicts the localization of pathogenic mutations within genes; citation_author=A.B. Gussow; citation_author=S. Petrovski; citation_author=Q. Wang; citation_author=A.S. Allen; citation_author=D.B. Goldstein; citation_journal_title=Genome Biol.; citation_volume=17; citation_firstpage=9; citation_publication_date=2016; "></meta> <meta name="citation_reference" content="citation_title=Analysis of protein-coding genetic variation in 60,706 humans; citation_author=M. Lek; citation_author=K.J. Karczewski; citation_author=E.V. Minikel; citation_author=K.E. Samocha; citation_author=E. Banks; citation_author=T. Fennell; citation_author=A.H. OâDonnell-Luria; citation_author=J.S. Ware; citation_author=A.J. Hill; citation_author=B.B. Cummings; citation_author=Exome Aggregation Consortium; citation_author=et al.; citation_journal_title=Nature; citation_volume=536; citation_firstpage=285; citation_lastpage=291; citation_publication_date=2016; "></meta> <meta name="citation_reference" content="citation_title=Human gene essentiality; citation_author=I. Bartha; citation_author=J. di Iulio; citation_author=J.C. Venter; citation_author=A. Telenti; citation_journal_title=Nat. Rev. Genet.; citation_volume=19; citation_firstpage=51; citation_lastpage=62; citation_publication_date=2018; "></meta> <meta name="citation_reference" content="citation_title=Non-centered parameterisations for hierarchical models and data augmentation; citation_author=O. Papaspiliopoulos; citation_author=G.O. Roberts; citation_journal_title=Bayesian Stat.; citation_volume=7; citation_firstpage=307; citation_lastpage=326; citation_publication_date=2003; "></meta> <meta name="citation_reference" content="citation_title=Stability of the Gibbs sampler for Bayesian hierarchical models; citation_author=O. Papaspiliopoulos; citation_author=G. Roberts; citation_journal_title=Ann. Stat.; citation_volume=36; citation_firstpage=95; citation_lastpage=117; citation_publication_date=2008; "></meta> <meta name="citation_reference" content="citation_title=A general metric for Riemannian Manifold Hamiltonian Monte Carlo; citation_author=M. Betancourt; citation_journal_title=Nature; citation_volume=9; citation_firstpage=327; citation_lastpage=334; citation_publication_date=2013; "></meta> <meta name="citation_reference" content="citation_title=To center or not to center: that is not the questionâAn ancillarityâsufficiency interweaving strategy (ASIS) for boosting MCMC efficiency; citation_author=Y. Yu; citation_author=X.-L. Meng; citation_journal_title=J.Â Comput. Graph. Stat.; citation_volume=20; citation_firstpage=531; citation_lastpage=570; citation_publication_date=2011; "></meta> <meta name="citation_reference" content="citation_title=Scaling up data augmentation MCMC via calibration; citation_author=L.L. Duan; citation_author=J.E. Johndrow; citation_author=D.B. Dunson; citation_journal_title=arXiv; citation_publication_date=2017; "></meta> <meta name="citation_reference" content="citation_title=The collapsed Gibbs sampler with applications to a gene regulation problem; citation_author=J.S. Liu; citation_journal_title=J.Â Am. Stat. Assoc.; citation_volume=89; citation_firstpage=958; citation_lastpage=966; citation_publication_date=1994; "></meta> <meta name="citation_reference" content="citation_title=Hybrid Monte Carlo; citation_author=S. Duane; citation_author=A.D. Kennedy; citation_author=B.J. Pendleton; citation_author=D. Roweth; citation_journal_title=Phys. Lett. B; citation_volume=195; citation_firstpage=216; citation_lastpage=222; citation_publication_date=1987; "></meta> <meta name="citation_reference" content="citation_title=A groupwise association test for rare mutations using a weighted sum statistic; citation_author=B.E. Madsen; citation_author=S.R. Browning; citation_journal_title=PLoS Genet.; citation_volume=5; citation_firstpage=e1000384; citation_publication_date=2009; "></meta> <meta name="citation_reference" content="citation_title=Analysis of protein-coding genetic variation in 60,706 humans; citation_author=M. Lek; citation_author=K.J. Karczewski; citation_author=E.V. Minikel; citation_author=K.E. Samocha; citation_author=E. Banks; citation_author=T. Fennell; citation_author=A.H. OâDonnell-Luria; citation_author=J.S. Ware; citation_author=A.J. Hill; citation_author=B.B. Cummings; citation_author=Exome Aggregation Consortium; citation_author=et al.; citation_journal_title=Nature; citation_volume=536; citation_firstpage=285; citation_lastpage=291; citation_publication_date=2016; "></meta> <meta name="citation_reference" content="citation_title=Optimizing genomic medicine in epilepsy through a gene-customized approach to missense variant interpretation; citation_author=J. Traynelis; citation_author=M. Silk; citation_author=Q. Wang; citation_author=S.F. Berkovic; citation_author=L. Liu; citation_author=D.B. Ascher; citation_author=D.J. Balding; citation_author=S. Petrovski; citation_journal_title=Genome Res.; citation_volume=27; citation_firstpage=1715; citation_lastpage=1729; citation_publication_date=2017; "></meta> <meta name="citation_reference" content="citation_title=Comprehensive analysis of constraint on the spatial distribution of missense variants in human protein structures; citation_author=R.M. Sivley; citation_author=X. Dou; citation_author=J. Meiler; citation_author=W.S. Bush; citation_author=J.A. Capra; citation_author=T.S. Blackwell; citation_author=J.A. Phillips; citation_author=J. Meiler; citation_author=W.S. Bush; citation_author=J.A. Capra; citation_journal_title=Am. J. Hum. Genet.; citation_volume=102; citation_firstpage=415; citation_lastpage=426; citation_publication_date=2018; "></meta> <meta name="citation_reference" content="citation_title=CDD: conserved domains and protein three-dimensional structure; citation_author=A. Marchler-Bauer; citation_author=C. Zheng; citation_author=F. Chitsaz; citation_author=M.K. Derbyshire; citation_author=L.Y. Geer; citation_author=R.C. Geer; citation_author=N.R. Gonzales; citation_author=M. Gwadz; citation_author=D.I. Hurwitz; citation_author=C.J. Lanczycki; citation_author=et al.; citation_journal_title=Nucleic Acids Res.; citation_volume=41; citation_firstpage=D348; citation_lastpage=D352; citation_publication_date=2013; "></meta> <meta name="citation_reference" content="citation_title=The consensus coding sequence (CCDS) project: Identifying a common protein-coding gene set for the human and mouse genomes; citation_author=K.D. Pruitt; citation_author=J. Harrow; citation_author=R.A. Harte; citation_author=C. Wallin; citation_author=M. Diekhans; citation_author=D.R. Maglott; citation_author=S. Searle; citation_author=C.M. Farrell; citation_author=J.E. Loveland; citation_author=B.J. Ruef; citation_author=et al.; citation_journal_title=Genome Res.; citation_volume=19; citation_firstpage=1316; citation_lastpage=1323; citation_publication_date=2009; "></meta> <meta name="citation_reference" content="citation_title=ClinVar: public archive of relationships among sequence variation and human phenotype; citation_author=M.J. Landrum; citation_author=J.M. Lee; citation_author=G.R. Riley; citation_author=W. Jang; citation_author=W.S. Rubinstein; citation_author=D.M. Church; citation_author=D.R. Maglott; citation_journal_title=Nucleic Acids Res.; citation_volume=42; citation_firstpage=D980; citation_lastpage=D985; citation_publication_date=2014; "></meta> <meta name="citation_reference" content="citation_title=Distribution and clinical impact of functional variants in 50,726 whole-exome sequences from the DiscovEHR study; citation_author=F.E. Dewey; citation_author=M.F. Murray; citation_author=J.D. Overton; citation_author=L. Habegger; citation_author=J.B. Leader; citation_author=S.N. Fetterolf; citation_author=C. OâDushlaine; citation_author=C.V. Van Hout; citation_author=J. Staples; citation_author=C. Gonzaga-Jauregui; citation_author=et al.; citation_journal_title=Science; citation_volume=354; citation_doi=10.1126/science.aaf6814; citation_publication_date=2016; "></meta> <meta name="citation_reference" content="citation_title=The intolerance of regulatory sequence to genetic variation predicts gene dosage sensitivity; citation_author=S. Petrovski; citation_author=A.B. Gussow; citation_author=Q. Wang; citation_author=M. Halvorsen; citation_author=Y. Han; citation_author=W.H. Weir; citation_author=A.S. Allen; citation_author=D.B. Goldstein; citation_journal_title=PLoS Genet.; citation_volume=11; citation_firstpage=e1005492; citation_publication_date=2015; "></meta> <meta name="citation_reference" content="citation_title=DECIPHER: Database of Chromosomal Imbalance and Phenotype in Humans Using Ensembl Resources; citation_author=H.V. Firth; citation_author=S.M. Richards; citation_author=A.P. Bevan; citation_author=S. Clayton; citation_author=M. Corpas; citation_author=D. Rajan; citation_author=S. Van Vooren; citation_author=Y. Moreau; citation_author=R.M. Pettett; citation_author=N.P. Carter; citation_journal_title=Am. J. Hum. Genet.; citation_volume=84; citation_firstpage=524; citation_lastpage=533; citation_publication_date=2009; "></meta> <meta name="citation_reference" content="citation_title=A case-control collapsing analysis identifies epilepsy genes implicated in trio sequencing studies focused on de novo mutations; citation_author=X. Zhu; citation_author=R. Padmanabhan; citation_author=B. Copeland; citation_author=J. Bridgers; citation_author=Z. Ren; citation_author=S. Kamalakaran; citation_author=A.O. Driscoll-collins; citation_author=S.F. Berkovic; citation_author=I.E. Scheffer; citation_author=A. Poduri; citation_author=et al.; citation_journal_title=PLoS Genet.; citation_volume=13; citation_firstpage=e1007104; citation_publication_date=2017; "></meta> <meta name="citation_reference" content="citation_title=The known unknown: the challenges of genetic variants of uncertain significance in clinical practice; citation_author=L. Hoffman-Andrews; citation_journal_title=J.Â Law Biosci.; citation_volume=4; citation_firstpage=648; citation_lastpage=657; citation_publication_date=2018; "></meta> <meta name="citation_reference" content="citation_title=Regional missense constraint improves variant deleteriousness prediction; citation_author=K.E. Samocha; citation_author=J.A. Kosmicki; citation_author=K.J. Karczewski; citation_author=A.H. OâDonnell-Luria; citation_author=E. Pierce-Hoffman; citation_author=D.G. MacArthur; citation_author=B.M. Neale; citation_author=M.J. Daly; citation_journal_title=bioRxiv; citation_doi=10.1101/148353; citation_publication_date=2017; "></meta> <meta name="citation_reference" content="citation_title=A method and server for predicting damaging missense mutations; citation_author=I.A. Adzhubei; citation_author=S. Schmidt; citation_author=L. Peshkin; citation_author=V.E. Ramensky; citation_author=A. Gerasimova; citation_author=P. Bork; citation_author=A.S. Kondrashov; citation_author=S.R. Sunyaev; citation_journal_title=Nat. Methods; citation_volume=7; citation_firstpage=248; citation_lastpage=249; citation_publication_date=2010; "></meta> <meta name="citation_reference" content="citation_title=A general framework for estimating the relative pathogenicity of human genetic variants; citation_author=M. Kircher; citation_author=D.M. Witten; citation_author=P. Jain; citation_author=B.J. OâRoak; citation_author=G.M. Cooper; citation_author=J. Shendure; citation_journal_title=Nat. Genet.; citation_volume=46; citation_firstpage=310; citation_lastpage=315; citation_publication_date=2014; "></meta> <meta name="citation_reference" content="citation_title=REVEL: An Ensemble Method for Predicting the Pathogenicity of Rare Missense Variants; citation_author=N.M. Ioannidis; citation_author=J.H. Rothstein; citation_author=V. Pejaver; citation_author=S. Middha; citation_author=S.K. McDonnell; citation_author=S. Baheti; citation_author=A. Musolf; citation_author=Q. Li; citation_author=E. Holzinger; citation_author=D. Karyadi; citation_author=et al.; citation_journal_title=Am. J. Hum. Genet.; citation_volume=99; citation_firstpage=877; citation_lastpage=885; citation_publication_date=2016; "></meta> <meta name="citation_abstract" content="&lt;p&gt;Different parts of a gene can be of differential importance to development and health. This regional heterogeneity is also apparent in the distribution of disease-associated mutations, which often cluster in particular regions of disease-associated genes. The ability to precisely estimate functionally important sub-regions of genes will be key in correctly deciphering relationships between genetic variation and disease. Previous methods have had some success using standing human variation to characterize this variability in importance by measuring sub-regional intolerance, i.e., the depletion in functional variation from expectation within a given region of a gene. However, the ability to precisely estimate local intolerance was restricted by the fact that only information within a given sub-region is used, leading to instability in local estimates, especially for small regions. We show that borrowing information across regions using a Bayesian hierarchical model stabilizes estimates, leading to lower variability and improved predictive utility. Specifically, our approach more effectively identifies regions enriched for ClinVar pathogenic variants. We also identify significant correlations between sub-region intolerance and the distribution of pathogenic variation in disease-associated genes, with AUCs for classifying &lt;i&gt;de novo&lt;/i&gt; missense variants in Online Mendelian Inheritance in Man (OMIM) genes of up to 0.86 using exonic sub-regions and 0.91 using sub-regions defined by protein domains. This result immediately suggests that considering the intolerance of regions in which variants are found may improve diagnostic interpretation. We also illustrate the utility of integrating regional intolerance into gene-level disease association tests with a study of known disease-associated genes for epileptic encephalopathy.&lt;/p&gt;"></meta><meta data-pb-head="head-widgets-end"/><meta charset="UTF-8" /><!--

    Pagebuild admin UI

    Explain mode styling
-->


<!---->
<meta http-equiv="X-UA-Compatible" content="IE=edge">








    
        
    








    
        <script type="text/javascript">
            function optimizelyTimeout() {
                window.optimizely = window.optimizely || [];
                if (!window.optimizely.data) {
                    window.optimizely.push("timeout")
                }
            }
            (function () {
                var e = "https:" == document.location.protocol ? "https://" : "http://";
                var t = document.createElement("script");
                t.type = "text/javascript";
                t.async = true;
                t.src = e + "cdn.optimizely.com/js/537390469.js";
                var n = document.getElementsByTagName("script")[0];
                n.parentNode.insertBefore(t, n)
            })();
            setTimeout(optimizelyTimeout, 1e3)
        </script>
    



<script type="text/javascript" src="//s7.addthis.com/js/250/addthis_widget.js#pubid=xa-4faab26f2cff13a7"></script></head>
<body class="pb-ui">
    <div class="skip">
        <a href="#pageBody"
                  id="skipNavigationLink"
                  title="Skip to Main Content" tabIndex="0">Skip to Main Content</a>
    </div>


<script type="text/javascript">

        if(true) {
            document.getElementById("skipNavigationLink").onclick =function skipElement () {
                document.getElementById('pageBody').setAttribute('tabindex','0');
                document.getElementById('pageBody').focus();
            }

        }

</script>
    <div id="pb-page-content" data-ng-non-bindable>
        <div data-pb-dropzone="main" data-pb-dropzone-name="Main">










        <div data-widget-def="ux3-layout-widget" data-widget-id="832ad3f3-97ac-45ab-9af1-90c6f7bff54a" class="ux3">
        
    

        


    

            

            
                <div class="ajhg">
            

            <div data-widget-id="832ad3f3-97ac-45ab-9af1-90c6f7bff54a" data-widget-name="ux3-layout-widget"><div>
    

        


    

            

            
                <div class="ux3-ads">
            

            <div class="pb-ad">
            <!-- Start: GPT Async -->
<script type="text/javascript">

	var gptadslots=[];
	var googletag = googletag ||
	{
	};

	googletag.cmd = googletag.cmd || [];

	(function()
		{
			var gads = document.createElement('script');

			gads.async = true; gads.type = 'text/javascript';

			var useSSL = 'https:' == document.location.protocol;

			gads.src = (useSSL ? 'https:' : 'http:') + '//www.googletagservices.com/tag/js/gpt.js';

			var node = document.getElementsByTagName('script')[0];

			node.parentNode.insertBefore(gads, node);

		})();

</script>

<script type="text/javascript">
	//--------------------------------------------------------------------------
	$(document).ready(function ()
		{
			//Once the dom has loaded, begin the process to init the ads.
			
			gpt_InitAds();
		});
	//--------------------------------------------------------------------------	
	function GPT_CheckForNULL(x, valueifnull) {
	    if (typeof valueifnull == "undefined") { valueifnull = null; }

	    if (typeof x == 'undefined') {
	        return valueifnull;
	    }
	    if (x == null) {
	        return valueifnull;
	    }
	    if (x == "null") {
	        return valueifnull;
	    }
	    if ($.trim(x) == '') {
	        return valueifnull;
	    }
	    else {
	        return x;
	    }
	}		
	//--------------------------------------------------------------------------
	function gpt_GetAdditionalTargeting()
	{
		//adds page level targeting of custom variables
		//the id (minus gpt) and the value will be used as a key-value pair
		//eg: <input type="hidden" id="gptissuepii" value="x" class="gpt-Target">
		//eg: target issuepii=x
		$(".gpt-Target").each(function( index )
			{
				var gptTargetId = $(this).attr('id');
				gptTargetId = gptTargetId.replace('gpt','').toLowerCase();
				if(gptTargetId!='')
				{
					gpt_Target(gptTargetId);
				}
			})
	}
	//--------------------------------------------------------------------------
	function enableStickySideBar(){
		var stickyStatus = 0;
		var scrollx, skytop, voffset;
		skytop = $(".widget.literatumAd.alignCenter.GPT-ad-placeholder.widget-verticalAd").offset().top;

		$(window).scroll(function() {
			if (stickyStatus == 1) {
				if (rightColHeight != $(".widget.literatumAd.alignCenter.GPT-ad-placeholder.widget-verticalAd").parent().height()) {
			        rightColHeight = $(".widget.literatumAd.alignCenter.GPT-ad-placeholder.widget-verticalAd").parent().height();
			        skytop = $(".widget.literatumAd.alignCenter.GPT-ad-placeholder.widget-verticalAd").offset().top;
				}
			}

			scrollx = $(window).scrollTop();

			// IF THE LEADERBOARD AD IS FIXED
			if ($(".widget.literatumAd.alignCenter.GPT-ad-placeholder.widget-horizontalAd").css("position") == "fixed") {
				voffset = 110;
			} else {
				voffset = 0;
			}
			
			if ($("#jbs-fixed-hdr").length!=0){
				if ($("#jbs-fixed-hdr").css("display") == "block") {
				    voffset += $("#jbs-fixed-hdr").outerHeight();
				}			
			}

			if (scrollx >= (skytop-voffset)) {   // MAKE AD IN RIGHT COLUMN STICKY
				var tmpWidth;
				tmpWidth = $(".widget.literatumAd.alignCenter.GPT-ad-placeholder.widget-verticalAd").width();
				$(".widget.literatumAd.alignCenter.GPT-ad-placeholder.widget-verticalAd").attr("style","top: " + voffset + "px !important;position: fixed !important;display:block !important;width:" + tmpWidth + "px !important;");
			} else {       // REVERT BACK TO ORIGINAL AD PLACEMENT
				$(".widget.literatumAd.alignCenter.GPT-ad-placeholder.widget-verticalAd").attr("style","position: static !important; display:block !important;");
				skytop = $(".widget.literatumAd.alignCenter.GPT-ad-placeholder.widget-verticalAd").offset().top;
				stickyStatus = 0;
			}	
		});
	}
	//--------------------------------------------------------------------------
	function handleStickyLeaderboard()
	{
		var forceStickyLeaderboard = gpt_GetQuerystring('forcestickyleaderboard',0);
		if(forceStickyLeaderboard==0){
			// THIS FUNCTION REMOVES THE STICKY FEATURE OF THE LEADERBOARD
			$("#pb-page-content").attr("style","margin-top:0 !important;");
			$(".widget.literatumAd.alignCenter.GPT-ad-placeholder.widget-horizontalAd").attr("style","margin-top: 0px !important;position: static !important;display:block !important;background-color: transparent;padding:0;");

			$("#jbs-fixed-hdr").attr("style","top: 0 !important;");

			if ($(".denialInfo").length === 0 && $("#jbs-fixed-hdr").length > 0)
			{
				var topY = window.pageYOffset;
				if (topY > 300)
				{
					$("#jbs-fixed-hdr").attr("style","top: 0 !important;display:block !important");
				}
			}
			$(window).scroll();						
		}
	}
	//--------------------------------------------------------------------------
	function handleStickyLeaderboard_CELL()
	{
		var forceStickyLeaderboard = gpt_GetQuerystring('forcestickyleaderboard',0);
		if(forceStickyLeaderboard==0){		
			$("body").attr("style","padding-top:10px !important;");
			$(".cell-top-head").first().attr("style","margin-top: 0px !important;position: static !important;display:block !important;background-color: transparent;padding:0;");
			$(".article.enhanced .tab-nav").addClass("resetTop0");
			$(".article.enhanced #fixedPane").addClass("resetTop38");
			$(".article.enhanced .fixedArticle > .tab-nav").css("top","0px");
			$(".article.enhanced .fixedArticle #fixedPane").css("top","38px");
			var tmpStyle;
			tmpStyle = $(".article.enhanced .tab-nav").attr("style");
			tmpStyle = "top:0px;" + tmpStyle;
			$(".article.enhanced .tab-nav").removeAttr("style");
			$(".article.enhanced .tab-nav").attr("style",tmpStyle);
			tmpStyle = $(".article.enhanced #fixedPane").attr("style");
			tmpStyle = "top:38px;" + tmpStyle;
			$(".article.enhanced #fixedPane").removeAttr("style");
			$(".article.enhanced #fixedPane").attr("style",tmpStyle);
			$(".skip").addClass("skip-post-stickyleaderboard");
		}
	}	
	//--------------------------------------------------------------------------
	function gpt_GetQuerystring(key, default_)
	{
		//function to read querystring variables
		if (default_ == null) default_ = '';
		key = key.replace(/[\[]/, "\\\[").replace(/[\]]/, "\\\]");
		var regex = new RegExp("[\\?&]" + key + "=([^&#]*)");
		var qs = regex.exec(window.location.href);
		if (qs == null)
		return default_;
		else
		return qs[1];
	}
	//--------------------------------------------------------------------------
	function gpt_GetDFPTestId()
	{
		//Get any request for a test id from querystring and set custom targeting parameter
		var gptDFPTestId = gpt_GetQuerystring('dfptestid','');
		if(gptDFPTestId != '')
		{
			$('#gpt-Variables').append('<input type="hidden" id="gptDFPTestId">');
			$('#gptDFPTestId').val(gptDFPTestId);
			gpt_Target("DFPTestId");
		}
	}
	//--------------------------------------------------------------------------
	function gpt_Target(targetName)
	{
		//Set page level target if value is found
		if($('#gpt' + targetName).length>0)
		{
			var targetValue = $('#gpt' + targetName).val();

			if(targetValue!='')
			{
				targetValue = targetValue.replace(/[^-a-z0-9]/ig,'');	//Only Allow a-z, 0-9
				googletag.pubads().setTargeting(targetName.toLowerCase(),targetValue.toLowerCase());
			}
		}
	}
	//--------------------------------------------------------------------------
	function gpt_InitAds()
	{
		var disableStickyLeaderboard = 0;
		var disableStickySidebar = 0;
		var responsiveDesign = 0;
		
		// ux3 Class will only exist if the page is Responsive Design
		if ($('.ux3').length!=0) {	
			//Responsive Design Found, so disable the sticky Leaderboard
			responsiveDesign = 1;
			disableStickyLeaderboard = 1;
		}
		else{
			//Page is not responsive so check for Disable on Querystring
			disableStickyLeaderboard = gpt_GetQuerystring('disableStickyLeaderboard',0);
		}
		
		if(disableStickyLeaderboard!=0){
			//Disable the sticky leaderbard
			//Check for page flag, if it does not exist, create it.
			if($('#disableStickyLeaderboard').length==0){
				$("body").append("<div id='disableStickyLeaderboard'></div>");
			}
		}		

		if(responsiveDesign==1){
			//Page is responsive design, check if it is an article page
			if($("[data-widget-def='UX3articleContent']").length!=0){
				//Its an article page, disable the sticky sidebar
				disableStickySidebar = 1;			
			}			
		}
		if(disableStickySidebar==0){
			//Check for disableStickySidebar in Querystring
			disableStickySidebar = gpt_GetQuerystring('disableStickySidebar',0);
		}
		
		if(disableStickySidebar!=0){
			//Disable the sticky leaderbard
			//Check for page flag, if it does not exist, create it.
			if($('#disableStickySidebar').length==0){
				$("body").append("<div id='disableStickySidebar'></div>");
			}
		}

		//Load common stylesheet for all sites except Cell & Responsive Design
		if($('.cell-site-header').length==0 && responsiveDesign==0){
			$("body").append('<link href="/pb-assets/shared/leaderboards/gpt_jbs-1504880817473.css" rel="stylesheet" />');
		}	
		
		googletag.cmd.push(function()
			{
				var gptAdSlots = [];

				//***********************************************
				//Start Define Size mapping based on client Viewport
				//***********************************************
				var gptLeaderboardSizeMapping = googletag.sizeMapping().
				addSize([768, 0], [[728, 90]]). 								// Leaderboard
				addSize([0, 0], [[320, 50], [300, 50]]). 						// Mobile Leaderboard
				build();

				if ($("#jbs-fixed-hdr").length!=0){
					//Download Bar Exists
					var gptSideBarSizeMapping = googletag.sizeMapping().
					addSize([1140, 330], [[336,280],[300, 250]]). 				// Desktop BoomBox
					addSize([0, 0], []). 										// Phone-Tablet 
					build();					
				}
				else{
					//Download Bar does not exist
					var gptSideBarSizeMapping = googletag.sizeMapping().
					addSize([1140, 280], [[336,280],[300, 250]]).			 	// Desktop
					addSize([0, 0], []). 										// Phone-Tablet
					build();
				}

				var gptSideBarNotStickySizeMapping = googletag.sizeMapping().
				addSize([1140, 160], [[336,280],[300, 250]]). 					// Desktop
				addSize([0, 0], []). 											// Phone-Tablet
				build();

				var gptBoomBoxSizeMapping = googletag.sizeMapping().
				addSize([1140, 0], [[336,280],[300, 250]]). 					// Desktop
				addSize([0, 0], []). 											// Phone-Tablet
                build();

                var gptSkyscraperNotResponsiveSizeMapping = googletag.sizeMapping().
                addSize([1140, 800], [[160, 600], [120, 600]]). 		        // SkyScraper - Desktop (Require min 800 vertical)
                addSize([0, 0], []). 											// Phone-Tablet
                build();

				
				//***********************************************
				//End Define Size mapping based on client Viewport
				//***********************************************

				//***********************************************
				//Start detection for ad slots on page
				//***********************************************
				$(".gpt-ad").each(function( index )
					{					
						var gptSite = GPT_CheckForNULL($('#gptSite').val());
						var gptPage = '';
						var adSizeMapping='';
						var adId;
						var gptPOS;
						var gptMasterPage = GPT_CheckForNULL($('#gptPage').val());
						
						if (gptSite == null) { 
							gptSite = GPT_CheckForNULL($('#gptAltSite').val());
							if (gptSite == null) { 	
								gptSite = GPT_CheckForNULL($(this).data('site'));
							}
						}
						if (gptSite == null) { 	
							gptSite = GPT_CheckForNULL($('#gptDefaultSite').val());
						}
						if(gptSite!=null){
							adId = 'gpt-ad-' + index + 1;
						
							gptPOS = parseInt($(this).data('pos')) || -1;
							if(gptPOS == -1){
								gptPOS = index + 1;
							}

                            if (gptMasterPage != null) {
                                gptPage = '/' + gptMasterPage;
                            }
                            else {
                                gptPage = GPT_CheckForNULL($(this).data('page'));
                                if (gptPage != null) {
                                    gptPage = '/' + gptPage;
                                }
                                else {
                                    gptPage = '';
                                }
                            }
							
							//Determine sizemapping based on data-adsize
							switch($(this).data('adsize'))
							{
								case 'cell-leaderboard':
									adSizeMapping = gptLeaderboardSizeMapping;
									if(gptPOS==1){
										$(this).addClass('GPT-Leaderboard-Cell');
									}
									break;
								case 'leaderboard':
									adSizeMapping = gptLeaderboardSizeMapping;
									$(this).addClass('GPT-Leaderboard');
									break;
								case 'boombox':
									adSizeMapping = gptBoomBoxSizeMapping;
									break;
								case 'skyscraper':
									if($(this).hasClass("disableStickySidebar")){
										adSizeMapping = gptSideBarNotStickySizeMapping;
									}
									else{
										$(this).addClass('GPT-StickySideBar');
										adSizeMapping = gptSideBarSizeMapping;	
									}									
                                    break;
                                case 'skyscrapernotresponsive':
                                    adSizeMapping = gptSkyscraperNotResponsiveSizeMapping;
                                    break;
							}		
						}
					
						//Ignore any invalid adsizes that could be on the page
						if(adSizeMapping!='')
						{
							$(this).addClass('gpt-ad-defined');
							$(this).attr('id',adId);
							gptAdSlots[index] = googletag.defineSlot('/6053/els.' + gptSite + gptPage, [], adId).defineSizeMapping(adSizeMapping).addService(googletag.pubads()).setTargeting("pos", gptPOS);
						}
						else
						{
							$(this).remove();
						}
					});
				//***********************************************
				//End detection for ad slots on page
				//***********************************************

				//***********************************************
				//Start page level targeting
				//***********************************************
				gpt_Target("LoggedIn");
				gpt_Target("PII");

				//Check for DFP Test
				gpt_GetDFPTestId();

				//Check for Additional Targeting
				gpt_GetAdditionalTargeting()

				//***********************************************
				//End page level targeting
				//***********************************************

				//Enable google dfp services
				googletag.pubads().enableSingleRequest();
				googletag.pubads().collapseEmptyDivs();
				googletag.enableServices();

				//Add Listener for Leaderboard
				googletag.pubads().addEventListener('slotRenderEnded', function(event)
					{
						if (!event.isEmpty)
						{
							$('#' + event.slot.getSlotElementId()).closest(".GPT-ad-placeholder").css("display", "block");


							//Sticky LeaderBoard JBS
							if ($('#' + event.slot.getSlotElementId()).hasClass("GPT-Leaderboard"))
							{
								if($('#disableStickyLeaderboard').length==0){
									$('body').addClass("hasLeaderboard");

									// INJECT SOME CSS FOR STICKY LEADERBOARD
									// THIS INCLUDES CHANGING THE TOP OF THE STICKY HEADER #jbs-fixed-hdr
									$("body").append("<style>.prevNextHeader:first-of-type,.skip{height:0}.widget.literatumAd.alignCenter.GPT-ad-placeholder.widget-horizontalAd{position:fixed;top:0;left:0;width:100%;background-color:#d8d8d8;padding:10px 0 0;margin:0px !important}#pb-page-content{margin-top:120px}#jbs-fixed-hdr{top:108px!important;z-index:999999!important;height:26px!important}.horizontalAd iframe{}div.horizontalAdContainer{height:95px!important}.iHubLinkContainerX{margin:0!important} div.horizontalAdContainer{height:90px !important;}</style>");

									setTimeout(handleStickyLeaderboard, 8000);									
								}
							}
							
							//Sticky LeaderBoard Cell
							if ($('#' + event.slot.getSlotElementId()).hasClass("GPT-Leaderboard-Cell"))
							{
								if($('#disableStickyLeaderboard').length==0){						
									$("body").append('<link href="/pb-assets/shared/leaderboards/gpt_cell-1504880817443.css" rel="stylesheet" />');
									$("div.cell-top-head").attr("style","display:block !important;"); $("body").append("");
									var tmp;
									tmp = $(".article.enhanced .tab-nav").attr("style");
									tmp = "top:110px;" + tmp;
									$(".article.enhanced .tab-nav").removeAttr("style");
									$(".article.enhanced .tab-nav").attr("style",tmp);
									tmp = $(".article.enhanced #fixedPane").attr("style");
									tmp = "top:148px;" + tmp;
									$(".article.enhanced #fixedPane").removeAttr("style");
									$(".article.enhanced #fixedPane").attr("style",tmp);
									setTimeout(handleStickyLeaderboard_CELL, 8000);
								}

								// THE FOLLOWING MAKES SURE THAT THE FIGURES PANE POPULATES 
								// (there might be a bug on the live site) 
								$("li.showFullText a").click();
							}
							
							
							//Sticky SideBar
							if ($('#' + event.slot.getSlotElementId()).hasClass("GPT-StickySideBar")){
								if($('#disableStickySidebar').length==0){
									enableStickySideBar();										
								}
							}							
						}
					});

				//Populate Defined Ad Slots
				$(".gpt-ad-defined").each(function( index )
					{
						var adId = $(this).attr('id');
						googletag.cmd.push(function() {googletag.display(adId);});
					});

			});
	}
</script>
<div id="gpt-Variables" data-adfile="journal" data-version="10152018">
	<input type="hidden" id="gptDefaultSite" value="cell">
	<input type="hidden" id="gptSite" value="ajhg">
	<input type="hidden" id="gptLoggedIn" value="no">
	<input type="hidden" id="gptPII" value="S0002929718305032">
</div>

<!-- End: GPT -->



        </div>

            
                </div>
            

        

    

        


    

            

            

            <div data-widget-id="912135d6-7135-4d45-b457-b503cacc3c8b" data-widget-name="ux3-layout-widget"><header class="header base fixed">
    

        


    

                
                
                
                
                
                
                    
                
                    
                
                    
                
                    
                
                    
                
                    
                
                    
                
                    
                
                    
                
                    
                
                    
                
                
                    
                
                    
                
                    
                
                    
                
                    
                
                    
                        
                        
                        
                    
                
                    
                
                    
                
                    
                
                    
                
                    
                
                
                
                <div class="widget literatumAd none adplaceholder adplaceholder--header js__adToHide widget-horizontalAd" id="9d6062d7-fad6-4e64-81cc-7dd25d99b23d">
                
                

                <div class="wrapped " >
                    
                    <div class="widget-body body body-horizontalAd "><div class="horizontalAdContainer">
    <div class="horizontalAdLabelContainer">
        <div class="horizontalAdLabel">
                <img src="/templates/jsp/_style2/_marlin/images/bg_adHoriz.gif" alt="Advertisement"/>
        </div>
    </div>
    <div class="horizontalAd">
        




        <div class="pb-ad">
            <!-- Beginning Async AdSlot 1 for Ad unit els.cellpress/cell  ### size: [[728,90]] -->
<div class="gpt-ad" data-adsize="cell-leaderboard" data-page="" data-site="cell" data-pos="1"></div>
<!-- End AdSlot 1 -->
        </div>
    

    </div>
</div></div>
                </div>
                </div>

        










        <div data-widget-def="ux3-layout-widget" data-widget-id="3a83e2a6-2e7e-4931-892c-4ef43b6e3312" class="header__wrapper clearfix" data-db-parent-of="quickSearch">
        
    

        


    

            

            

            <div data-widget-id="3a83e2a6-2e7e-4931-892c-4ef43b6e3312" data-widget-name="ux3-layout-widget"><div class="col-xs-6">
    

        


    

            

            
                <div class="journal-logos">
            

            <span data-widget-id="35d8d734-3aa0-4ec8-9e2d-14babe803594" data-widget-name="ux3-general-image"><a href="/ajhg/home" title="AJHG"><img alt="AJHG" src="/pb-assets/ux3/logos/cell/Ajhg_white-1532516612747.svg"/></a></span>

            
                </div>
            

        
</div><div class="col-xs-6">
    

        


    

            

            

            <div data-widget-id="2555f6db-b333-4edb-8b15-ef9213f1f0a5" data-widget-name="ux3-layout-widget"><div class="header__nav row">
    

        


    

            

            

            <div data-widget-id="ae520d74-989f-4741-a72a-da74d1984826" data-widget-name="UX3MenuWidget" class="main-nav menu-drawer"><div class="menu-drawer__ctrl"><a href="#main-menu" data-target="main-menu" data-toggle="nav" title="Explore Cell.com" aria-controls="main-menu" aria-expanded="false"><span class="menu-drawer--open"><i aria-hidden="true" class="icon-gizmo-hamburger"></i></span><span class="menu-drawer--close"><i aria-hidden="true" class="icon-close_thin"><span>Close</span></i></span></a></div><nav role="navigation" id="main-menu" data-ctrl-res="screen-sm" class="menu-drawer__nav drawer__nav gutterless"><span class="menu-drawer--holder hidden-md hidden-lg"></span><ul id="menubar" role="menubar" class="menubar rlist--inline"><li role="menuitem" aria-haspopup="true" aria-label="This Journal" id="menu-item-main-menu-0" class="dropdown menu-parent is--current"><a href="#" title="This Journal" id="main-menu-0" data-toggle="dropdown" class="dropdown__toggle"><span>This Journal</span><i aria-hidden="true" class="icon-arrow_r pull-right"></i></a><ul id="menu-items-0" role="menu" class="rlist dropdown__menu"><li role="menuitem" aria-haspopup="true" class="menu-drawer__back hidden-md hidden-lg"><a href="#" class="menu-drawer__back__ctrl"><i aria-hidden="true" class="icon-arrow_l"></i><span>Back</span></a></li><li role="menuitem" tabindex="-1" aria-label="Submit" class="menu-item"><a href="https://www.editorialmanager.com/ajhg/default.aspx" title="Submit"><span>Submit</span></a></li><li role="menuitem" tabindex="-1" aria-label="Online Now" class="menu-item"><a href="/ajhg/newarticles" title="Online Now"><span>Online Now</span></a></li><li role="menuitem" tabindex="-1" aria-label="Current Issue" class="menu-item"><a href="/ajhg/current" title="Current Issue"><span>Current Issue</span></a></li><li role="menuitem" tabindex="-1" aria-label="Archive" class="menu-item"><a href="/ajhg/archive" title="Archive"><span>Archive</span></a></li><li role="menuitem" tabindex="-1" aria-label="Aims and Scope" class="menu-item"><a href="/ajhg/aims" title="Aims and Scope"><span>Aims and Scope</span></a></li><li role="menuitem" tabindex="-1" aria-label="Contact" class="menu-item"><a href="/ajhg/contact" title="Contact"><span>Contact</span></a></li><li role="menuitem" tabindex="-1" aria-label="Staff and Editorial Board" class="menu-item"><a href="/ajhg/editorial-board" title="Staff and Editorial Board"><span>Staff and Editorial Board</span></a></li><li role="menuitem" tabindex="-1" aria-label="Reviewer Instructions" class="menu-item"><a href="/reviewers" title="Reviewer Instructions"><span>Reviewer Instructions</span></a></li><li role="menuitem" tabindex="-1" aria-label="Author Instructions" class="menu-item"><a href="/ajhg/authors" title="Author Instructions"><span>Author Instructions</span></a></li><li role="menuitem" tabindex="-1" aria-label="Figure Guidelines" class="menu-item"><a href="/figureguidelines" title="Figure Guidelines"><span>Figure Guidelines</span></a></li><li role="menuitem" tabindex="-1" aria-label="Funding Body Policies" class="menu-item"><a href="https://www.elsevier.com/about/open-science/open-access/funding-body-agreements" title="Funding Body Policies"><span>Funding Body Policies</span></a></li><li role="menuitem" tabindex="-1" aria-label="Information for Librarians" class="menu-item"><a href="/librarians" title="Information for Librarians"><span>Information for Librarians</span></a></li><li role="menuitem" tabindex="-1" aria-label="Rights, Sharing, Embargoes" class="menu-item"><a href="/rights-sharing-embargoes" title="Rights, Sharing, Embargoes"><span>Rights, Sharing, Embargoes</span></a></li></ul></li><li role="menuitem" aria-haspopup="true" aria-label="ASHG" id="menu-item-main-menu-1" class="dropdown menu-parent"><a href="#" title="ASHG" id="main-menu-1" data-toggle="dropdown" class="dropdown__toggle"><span>ASHG</span><i aria-hidden="true" class="icon-arrow_r pull-right"></i></a><ul id="menu-items-1" role="menu" class="rlist dropdown__menu level-2"><li role="menuitem" aria-haspopup="true" class="menu-drawer__back hidden-md hidden-lg"><a href="#" class="menu-drawer__back__ctrl"><i aria-hidden="true" class="icon-arrow_l"></i><span>Back</span></a></li><li role="menuitem" tabindex="-1" aria-label="ASHG Home" class="menu-item"><a href="https://www.ashg.org/" title="ASHG Home"><span>ASHG Home</span></a></li><li role="menuitem" tabindex="-1" aria-label="Join ASHG" class="menu-item"><a href="https://www.ashg.org/pages/member_overview.shtml" title="Join ASHG"><span>Join ASHG</span></a></li><li role="menuitem" tabindex="-1" aria-label="Annual Meeting" class="menu-item"><a href="https://www.ashg.org/meetings/profile.shtml" title="Annual Meeting"><span>Annual Meeting</span></a></li><li role="menuitem" tabindex="-1" aria-label="Meetings and Events of Interest" class="menu-item"><a href="http://www.ashg.org/meetings/interest.shtml" title="Meetings and Events of Interest"><span>Meetings and Events of Interest</span></a></li><li role="menuitem" tabindex="-1" aria-label="Education: K-12" class="menu-item"><a href="https://www.ashg.org/education/k12.shtml" title="Education: K-12"><span>Education: K-12</span></a></li><li role="menuitem" tabindex="-1" aria-label="Education: Training &amp;amp; Faculty" class="menu-item"><a href="https://www.ashg.org/education/Trainees_and_Faculty.shtml" title="Education: Training &amp;amp; Faculty"><span>Education: Training &amp; Faculty</span></a></li><li role="menuitem" tabindex="-1" aria-label="Education:Health Professionals" class="menu-item"><a href="https://www.ashg.org/education/Health_Professionals.shtml" title="Education:Health Professionals"><span>Education:Health Professionals</span></a></li><li role="menuitem" tabindex="-1" aria-label="Education: General Public" class="menu-item"><a href="https://www.ashg.org/education/General_Public.shtml" title="Education: General Public"><span>Education: General Public</span></a></li></ul></li><li role="menuitem" aria-haspopup="true" aria-label="Research Journals" id="menu-item-main-menu-2" class="dropdown menu-parent"><a href="#" title="Research Journals" id="main-menu-2" data-toggle="dropdown" class="dropdown__toggle"><span>Research Journals</span><i aria-hidden="true" class="icon-arrow_r pull-right"></i></a><ul id="menu-items-2" role="menu" class="rlist dropdown__menu level-3"><li role="menuitem" aria-haspopup="true" class="menu-drawer__back hidden-md hidden-lg"><a href="#" class="menu-drawer__back__ctrl"><i aria-hidden="true" class="icon-arrow_l"></i><span>Back</span></a></li><li role="menuitem" tabindex="-1" aria-label="Cell" class="menu-item"><a href="/cell/home" title="Cell"><span>Cell</span></a></li><li role="menuitem" tabindex="-1" aria-label="Cancer Cell" class="menu-item"><a href="/cancer-cell/home" title="Cancer Cell"><span>Cancer Cell</span></a></li><li role="menuitem" tabindex="-1" aria-label="Cell Chemical Biology" class="menu-item"><a href="/cell-chemical-biology/home" title="Cell Chemical Biology"><span>Cell Chemical Biology</span></a></li><li role="menuitem" tabindex="-1" aria-label="Cell Host &amp; Microbe" class="menu-item"><a href="/cell-host-microbe/home" title="Cell Host &amp; Microbe"><span>Cell Host & Microbe</span></a></li><li role="menuitem" tabindex="-1" aria-label="Cell Metabolism" class="menu-item"><a href="/cell-metabolism/home" title="Cell Metabolism"><span>Cell Metabolism</span></a></li><li role="menuitem" tabindex="-1" aria-label="Cell Reports" class="menu-item"><a href="/cell-reports/home" title="Cell Reports"><span>Cell Reports</span></a></li><li role="menuitem" tabindex="-1" aria-label="Cell Stem Cell" class="menu-item"><a href="/cell-stem-cell/home" title="Cell Stem Cell"><span>Cell Stem Cell</span></a></li><li role="menuitem" tabindex="-1" aria-label="Cell Systems" class="menu-item"><a href="/cell-systems/home" title="Cell Systems"><span>Cell Systems</span></a></li><li role="menuitem" tabindex="-1" aria-label="Chem" class="menu-item"><a href="/chem/home" title="Chem"><span>Chem</span></a></li><li role="menuitem" tabindex="-1" aria-label="Current Biology" class="menu-item"><a href="/current-biology/home" title="Current Biology"><span>Current Biology</span></a></li><li role="menuitem" tabindex="-1" aria-label="Developmental Cell" class="menu-item"><a href="/developmental-cell/home" title="Developmental Cell"><span>Developmental Cell</span></a></li><li role="menuitem" tabindex="-1" aria-label="Immunity" class="menu-item"><a href="/immunity/home" title="Immunity"><span>Immunity</span></a></li><li role="menuitem" tabindex="-1" aria-label="iScience" class="menu-item"><a href="/iscience/home" title="iScience"><span>iScience</span></a></li><li role="menuitem" tabindex="-1" aria-label="Joule" class="menu-item"><a href="/joule/home" title="Joule"><span>Joule</span></a></li><li role="menuitem" tabindex="-1" aria-label="Matter" class="menu-item"><a href="/matter/home" title="Matter"><span>Matter</span></a></li><li role="menuitem" tabindex="-1" aria-label="Molecular Cell" class="menu-item"><a href="/molecular-cell/home" title="Molecular Cell"><span>Molecular Cell</span></a></li><li role="menuitem" tabindex="-1" aria-label="Neuron" class="menu-item"><a href="/neuron/home" title="Neuron"><span>Neuron</span></a></li><li role="menuitem" tabindex="-1" aria-label="One Earth" class="menu-item"><a href="/one-earth/home" title="One Earth"><span>One Earth</span></a></li><li role="menuitem" tabindex="-1" aria-label="Structure" class="menu-item"><a href="/structure/home" title="Structure"><span>Structure</span></a></li></ul></li><li role="menuitem" aria-haspopup="true" aria-label="Trends Reviews Journals" id="menu-item-main-menu-3" class="dropdown menu-parent"><a href="#" title="Trends Reviews Journals" id="main-menu-3" data-toggle="dropdown" class="dropdown__toggle"><span>Trends Reviews Journals</span><i aria-hidden="true" class="icon-arrow_r pull-right"></i></a><ul id="menu-items-3" role="menu" class="rlist dropdown__menu level-4"><li role="menuitem" aria-haspopup="true" class="menu-drawer__back hidden-md hidden-lg"><a href="#" class="menu-drawer__back__ctrl"><i aria-hidden="true" class="icon-arrow_l"></i><span>Back</span></a></li><li role="menuitem" tabindex="-1" aria-label="Biochemical Sciences" class="menu-item"><a href="/trends/biochemical-sciences/home" title="Biochemical Sciences"><span>Biochemical Sciences</span></a></li><li role="menuitem" tabindex="-1" aria-label="Biotechnology" class="menu-item"><a href="/trends/biotechnology/home" title="Biotechnology"><span>Biotechnology</span></a></li><li role="menuitem" tabindex="-1" aria-label="Cancer" class="menu-item"><a href="/trends/cancer/home" title="Cancer"><span>Cancer</span></a></li><li role="menuitem" tabindex="-1" aria-label="Cell Biology" class="menu-item"><a href="/trends/cell-biology/home" title="Cell Biology"><span>Cell Biology</span></a></li><li role="menuitem" tabindex="-1" aria-label="Chemistry" class="menu-item"><a href="/trends/chemistry/home" title="Chemistry"><span>Chemistry</span></a></li><li role="menuitem" tabindex="-1" aria-label="Cognitive Sciences" class="menu-item"><a href="/trends/cognitive-sciences/home" title="Cognitive Sciences"><span>Cognitive Sciences</span></a></li><li role="menuitem" tabindex="-1" aria-label="Ecology &amp; Evolution" class="menu-item"><a href="/trends/ecology-evolution/home" title="Ecology &amp; Evolution"><span>Ecology & Evolution</span></a></li><li role="menuitem" tabindex="-1" aria-label="Endocrinology &amp; Metabolism" class="menu-item"><a href="/trends/endocrinology-metabolism/home" title="Endocrinology &amp; Metabolism"><span>Endocrinology & Metabolism</span></a></li><li role="menuitem" tabindex="-1" aria-label="Genetics" class="menu-item"><a href="/trends/genetics/home" title="Genetics"><span>Genetics</span></a></li><li role="menuitem" tabindex="-1" aria-label="Immunology" class="menu-item"><a href="/trends/immunology/home" title="Immunology"><span>Immunology</span></a></li><li role="menuitem" tabindex="-1" aria-label="Microbiology" class="menu-item"><a href="/trends/microbiology/home" title="Microbiology"><span>Microbiology</span></a></li><li role="menuitem" tabindex="-1" aria-label="Molecular Medicine" class="menu-item"><a href="/trends/molecular-medicine/home" title="Molecular Medicine"><span>Molecular Medicine</span></a></li><li role="menuitem" tabindex="-1" aria-label="Neurosciences" class="menu-item"><a href="/trends/neurosciences/home" title="Neurosciences"><span>Neurosciences</span></a></li><li role="menuitem" tabindex="-1" aria-label="Parasitology" class="menu-item"><a href="/trends/parasitology/home" title="Parasitology"><span>Parasitology</span></a></li><li role="menuitem" tabindex="-1" aria-label="Pharmacological Sciences" class="menu-item"><a href="/trends/pharmacological-sciences/home" title="Pharmacological Sciences"><span>Pharmacological Sciences</span></a></li><li role="menuitem" tabindex="-1" aria-label="Plant Science" class="menu-item"><a href="/trends/plant-science/home" title="Plant Science"><span>Plant Science</span></a></li></ul></li><li role="menuitem" aria-haspopup="true" aria-label="Partner Journals" id="menu-item-main-menu-4" class="dropdown menu-parent"><a href="#" title="Partner Journals" id="main-menu-4" data-toggle="dropdown" class="dropdown__toggle"><span>Partner Journals</span><i aria-hidden="true" class="icon-arrow_r pull-right"></i></a><ul id="menu-items-4" role="menu" class="rlist dropdown__menu level-5"><li role="menuitem" aria-haspopup="true" class="menu-drawer__back hidden-md hidden-lg"><a href="#" class="menu-drawer__back__ctrl"><i aria-hidden="true" class="icon-arrow_l"></i><span>Back</span></a></li><li role="menuitem" tabindex="-1" aria-label="AJHG" class="menu-item"><a href="/ajhg/home" title="AJHG"><span>AJHG</span></a></li><li role="menuitem" tabindex="-1" aria-label="Biophysical Journal" class="menu-item"><a href="/biophysj/home" title="Biophysical Journal"><span>Biophysical Journal</span></a></li><li role="menuitem" tabindex="-1" aria-label="EBioMedicine" class="menu-item"><a href="https://www.ebiomedicine.com/" title="EBioMedicine"><span>EBioMedicine</span></a></li><li role="menuitem" tabindex="-1" aria-label="Heliyon" class="menu-item"><a href="https://www.heliyon.com/" title="Heliyon"><span>Heliyon</span></a></li><li role="menuitem" tabindex="-1" aria-label="Molecular Plant" class="menu-item"><a href="/molecular-plant/home" title="Molecular Plant"><span>Molecular Plant</span></a></li><li role="menuitem" tabindex="-1" aria-label="Molecular Therapy Family" class="menu-item"><a href="/molecular-therapy-family/home" title="Molecular Therapy Family"><span>Molecular Therapy Family</span></a></li><li role="menuitem" tabindex="-1" aria-label="Stem Cell Reports" class="menu-item"><a href="/stem-cell-reports/home" title="Stem Cell Reports"><span>Stem Cell Reports</span></a></li></ul></li><li role="menuitem" aria-label="Cell Press Home" id="menu-item-main-menu-5" class="menu-item"><a href="/home" title="Cell Press Home"><span>Cell Press Home</span></a></li></ul></nav></div><div class="header__dropzone"><div class="pb-dropzone" data-pb-dropzone="menuDropzone"></div></div>

            

        

    

        


    

            

            

            <div data-widget-id="49f1edd6-fb35-4b49-9f18-39a90e37669b" data-widget-name="UX3QuickSearchWidget" class="quick-search"><div class="dropBlock"><a href="#" title="Search" data-db-target-for="quickSearch" aria-expanded="false" aria-controls="quick-search-holder" class="search-drawer"><i aria-hidden="true" class="icon-search"></i></a><div data-db-target-of="quickSearch" id="quick-search-holder" class="dropBlock__holder quick-search__holder"><div class="tab row clearfix quick-search__wrapper"><form action="https://www.cell.com/action/doSearch?journalCode=ajhg" name="quickSearchBar" method="get" title="Quick Search" onSubmit="return validateQuickSearchForm();"><div class="clearfix"><div class="col-sm-5 col-md-6 col-lg-7"><input type="hidden" name="searchType" value="quick" aria-hidden="true"/><div class="input-group quick-search__searchbar"><input type="text" name="searchText" placeholder="Search for..."/></div><button type="submit" title="Search" aria-label="Search" class="btn quick-search__button icon-search"><span>Go search</span></button></div><div class="col-sm-4 col-md-4 col-lg-3"><select class="quick-search__dropdown"><option value="all" selected="selected">All Content</option><option value="articleTitleAbstractKeywords">Article Title, Abstract, Keywords</option><option value="author">Authors</option><option value="articleTitle">Article Title</option><option value="abstract">Abstract</option></select><div class="quick-search__scope-options clearfix"><label for="searchScopeSeries" title="This Journal"><span>This Journal</span><input id="searchScopeSeries" type="radio" name="searchScope" title="This Journal" value="series" checked="checked"/></label><label for="searchScopeFullSite" title="Full Site"><span>Full Site</span><input id="searchScopeFullSite" type="radio" name="searchScope" value="fullSite" title="Full Site"/></label><input type="hidden" name="journalCode" value="ajhg" aria-hidden="true"/><input type="hidden" name="seriesISSNFltraddfilter" value="0002-9297" aria-hidden="true"/></div></div><input type="hidden" name="occurrences" value="all" aria-hidden="true"/><div class="col-sm-3 col-md-2 col-lg-2"><div class="quick-search__advanced-region"><a href="https://www.cell.com/ajhg/search?searchType=advanced&amp;journalCode=ajhg&amp;seriesIssn=0002-9297" title="Advanced search" class="quick-search__advanced-link">Advanced Search</a></div></div><input type="hidden" name="code" value="cell-site" aria-hidden="true"/><input type="hidden" name="journalCode" value="ajhg" aria-hidden="true"/></div></form><a href="#" title="Close search bar" class="quick-search__ctrl quick-search__close hidden-xs hidden-sm"><i aria-hidden="true" class="icon-tools_close"></i></a></div></div></div></div>

            

        

    

        


    

            

            

            <ul data-widget-id="52432d24-0331-4c9d-85f3-2033edc7208b" data-widget-name="UX3ProfileLinksWidget" class="rlist rlist--inline profile-links"><li><a href="/action/showLogin?redirectUri=%2Fajhg%2Ffulltext%2FS0002-9297%2818%2930503-2&amp;pii=S0002-9297%2818%2930503-2" title="Log in to Cell.com">Log in</a></li><li class="profile-links__register"><a href="/action/registration?journalCode=ajhg&amp;redirectUri=%2Fajhg%2Ffulltext%2FS0002-9297%2818%2930503-2" title="Register a username and password for free">Register</a></li><li class="profile-links__claim"><a href="/ajhg/activate" title="Claim your online subscription">Claim</a></li></ul>

            

        
</div></div>

            

        
</div></div>

            

        

        </div>
    
</header></div>

            

        










        <main data-widget-def="UX3articleContent" data-widget-id="7f1f4a57-23b9-4395-ab00-2ba3b187e1d7" class="content">
        
    

        


    

            

            

            <article data-widget-id="7f1f4a57-23b9-4395-ab00-2ba3b187e1d7" data-widget-name="UX3articleContent" data-pii="S0002-9297(18)30503-2"><div id="articleHeader" class="article__header"><div class="article-header"><div class="article-header__overlay"></div><div class="article-header__wrapper"><div class="article-header__top"><div class="row"><div class="col-sm-6"><div class="article-header__meta"><span class="article-header__journal">Article</span><span class="article-header__sep">|</span><a href="https://www.cell.com/ajhg/issue?pii=S0002-9297(18)X0003-2" class="article-header__vol faded">
                                                    Volume 104, ISSUE 2</a><span>, </span><span class="article-header__pages faded">P299-309, </span><span class="article-header__date faded">February 07, 2019</span></div></div><div class="col-sm-6"><div class="article-header__dropZone pb-dropzone" data-pb-dropzone="articleTools-dropzone">
    

        


    

            

            

            <div data-widget-id="183a17a2-87b5-4e6f-bb85-6b5cda068c89" data-widget-name="UX3ArticleToolsWidget" class="article-tools"><div class="article-tools__wrapper"><span class="article-tools__header"><!--.transplant(data-target="article h1" data-remove="false" data-target-class="hidden-xs hidden-sm" data-toggle="transplant" data-transplant="content")--></span><div class="article-tools__holder pull-right"><ul class="rlist rlist--inline article-tools__list"><li class="article-tools__item article-tools__pdf"><div aria-hidden="false" class="dropBlock article-tools__dropblock article-tools__pdf"><a href="#" title="open PDF menu" data-db-target-for="article-tools-pdf" aria-expanded="false" class="article-tools__ctrl article-tools__item__pdf"><i aria-hidden="true" class="icon-gizmo-pdf-file"></i><span class="basic">PDF</span><span class="alternative">Download PDF</span></a><div data-db-target-of="article-tools-pdf" class="dropBlock__holder article-tools__dropblock__holder"><ul class="rlist article-tools__sub-list"><li class="article-tools__sub-item"><a href="/action/showPdf?pii=S0002-9297%2818%2930503-2" title="Standard PDF (opens in a new tab)" target="_blank" class="article-tools__item__displayStandardPdf">Standard PDF [2 MB]</a></li><li class="article-tools__sub-item"><a href="/ajhg/pdfExtended/S0002-9297(18)30503-2" title="Extended PDF (opens in a new tab)" target="_blank" class="article-tools__item__displayExtendedPdf">Extended PDF [3 MB]</a></li></ul></div></div></li><li class="article-tools__item article-tools__figures"><div aria-hidden="true" class="dropBlock article-tools__dropblock article-tools__figures"><a href="#" title="open Figures menu" data-db-target-for="article-tools-figures" aria-expanded="false" class="article-tools__ctrl article-tools__item__figures"><i aria-hidden="true" class="icon-gizmo-image_figure"></i><span>Figures</span></a><div data-db-target-of="article-tools-figures" class="dropBlock__holder article-tools__dropblock__holder"><ul class="rlist article-tools__sub-list"><li class="article-tools__sub-item"><a href="#" title="Figure Viewer (opens in a new tab)" target="_blank" class="article-tools__item__displayFigureViewer js--openFigureViewer">Figure Viewer</a></li><li class="article-tools__sub-item"><a href="/ajhg/ppt/S0002-9297(18)30503-2.ppt" title="Download Figures (PPT) (opens in a new tab)" target="_blank" class="article-tools__item__displayDownloadFigures">Download Figures (PPT)</a></li></ul></div></div></li><li class="article-tools__item article-tools__save"><div aria-hidden="false" class="dropBlock article-tools__dropblock article-tools__save"><a href="#" title="open Save menu" data-db-target-for="article-tools-save" aria-expanded="false" class="article-tools__ctrl article-tools__item__save"><i aria-hidden="true" class="icon-gizmo-folder-in"></i><span>Save</span></a><div data-db-target-of="article-tools-save" class="dropBlock__holder article-tools__dropblock__holder"><ul class="rlist article-tools__sub-list"><li class="article-tools__sub-item"><a href="#" title="Add to Online Library" class="article-tools__item__displayAddToOnlineLibrary mendeley">Add to Online Library</a><span class="powered-by"><span>Powered By</span><i class="icon-mendeley"></i><span>Mendeley</span></span></li><li class="article-tools__sub-item"><a href="https://www.cell.com/personalize/addFavoritePublication?doi=10.1016%2Fj.ajhg.2018.12.020" title="Add to My Reading List (opens in a new tab)" target="_blank" class="article-tools__item__displayAddToMyReadingList">Add to My Reading List</a></li><li class="article-tools__sub-item"><a href="https://www.cell.com/action/showCitFormats?doi=10.1016%2Fj.ajhg.2018.12.020&amp;pii=S0002-9297%2818%2930503-2" title="Export Citation (opens in a new tab)" target="_blank" class="article-tools__item__displayExportCitation">Export Citation</a></li><li class="article-tools__sub-item"><a href="https://www.cell.com/action/addCitationAlert?doi=10.1016%2Fj.ajhg.2018.12.020" title="Create Citation Alert (opens in a new tab)" target="_blank" class="article-tools__item__displayCreateCitationAlert">Create Citation Alert</a></li></ul></div></div></li><li class="article-tools__item">
    

        


    

            

            

            <!-- Go to www.addthis.com/dashboard to customize your tools --><script type="text/javascript" src="//s7.addthis.com/js/300/addthis_widget.js#pubid=xa-4faab26f2cff13a7"></script><div data-widget-id="6f130e25-c0e4-42df-bb7d-35ca8d725c64" data-widget-name="UX3share" class="dropBlock article-tools__dropblock article-tools__share"><a href="#" data-db-target-for="6f130e25-c0e4-42df-bb7d-35ca8d725c64" aria-expanded="false" class="article-tools__ctrl share__ctrl"><i aria-hidden="true" class="icon-gizmo-share"></i><span>Share</span></a><div data-db-target-of="6f130e25-c0e4-42df-bb7d-35ca8d725c64" id="6f130e25-c0e4-42df-bb7d-35ca8d725c64_Pop" class="share__block dropBlock__holder article-tools__dropblock__holder"><ul class="rlist w-slide--list addthis addthis_toolbox addthis_default_style addthis_32x32_style"><li><a class="addthis_button_email"><i aria-hidden="true" class="at-icon-wrapper icon-gizmo-email"></i><span>Email</span></a></li><li><a class="addthis_button_twitter"><i aria-hidden="true" class="at-icon-wrapper icon-twitter3"></i><span>Twitter</span></a></li><li><a class="addthis_button_facebook"><i aria-hidden="true" class="at-icon-wrapper icon-facebook3"></i><span>Facebook</span></a></li><li><a class="addthis_button_linkedin"><i aria-hidden="true" class="at-icon-wrapper icon-linkedin3"></i><span>LinkedIn</span></a></li><li><a class="addthis_button_sinaweibo"><i aria-hidden="true" class="at-icon-wrapper icon-weibo"></i><span>Sina Weibo</span></a></li></ul></div></div>

            

        
</li><li><div class="dropBlock article-tools__dropblock article-tools__more"><a href="#" title="Show more tools menu" data-db-target-for="article-tools-more" aria-expanded="false" class="article-tools__ctrl"><i aria-hidden="true" class="icon-dots-three-horizontal"></i><span>more</span></a><div data-db-target-of="article-tools-more" class="dropBlock__holder article-tools__dropblock__holder"><ul class="rlist"><li><a href="http://reprints.elsevier.com/index.php?option=com_oscommerce&amp;amp;osMod=raq&amp;amp;Itemid=7&amp;amp;publicationName=The American Journal of Human Genetics&amp;amp;title=Improved Pathogenic Variant Localization via a Hierarchical Model of Sub-regional Intolerance&amp;amp;creator=Tristan+J.+Hayeck+and+colleagues&amp;amp;volume=104&amp;amp;number=2&amp;amp;startingPage=299&amp;amp;endingPage=309" title="Reprints" class="article-tools__item__reprints"><i aria-hidden="true" class="icon-gizmo-reprints"></i><span>Reprints</span></a></li><li><a href="https://s100.copyright.com/AppDispatchServlet?publisherName=ELS&amp;contentID=S0002929718305032&amp;orderBeanReset=true&amp;orderSource=Phoenix" title="Request" class="article-tools__item__request"><i aria-hidden="true" class="icon-copyright"></i><span>Request</span></a></li></ul></div></div></li><li><a href="#" class="back-to-top"><i aria-hidden="true" class="icon-arrow_u_p"></i><span>Top</span></a></li></ul></div></div></div>

            

        
</div></div></div></div><div class="article-header__middle"><h1 class="article-header__title">Improved Pathogenic Variant Localization via a Hierarchical Model of Sub-regional Intolerance</h1><span class="article-header__sub-title"></span><ul class="rlist loa inline-bullet-list ellipsis-dot" data-number-of-author="6"><li class="loa__item author"><div class="dropBlock article-header__info"><a href="#" data-db-target-for="au1-authorcard" aria-expanded="false" aria-controls="au1" class="loa__item__name article-header__info__ctrl loa__item__email" title="Correspondence information about the author Tristan J. Hayeck">Tristan J. Hayeck<i aria-hidden="true" class="icon-gizmo-person faded"></i></a><div data-db-target-of="au1-authorcard" class="dropBlock__holder article-header__info__holder lefted"><div class="article-header__info__scrollable"><div class="article-header__info__label">Tristan J. Hayeck</div><span class="article-header__info__group__label">Correspondence</span><div class="article-header__info__sub-label">Corresponding author</div><div class="article-header__info__body"><div class="article-header__info__social"><a class="article-header__info__email default" href="mailto:tristan.hayeck@gmail.com" title="Send email to Tristan J. Hayeck"><i aria-hidden="true" class="icon-Email"></i><span>
                            Contact
                        </span></a></div><div class="article-header__info__group"><span class="article-header__info__group__label">Affiliations</span><div class="article-header__info__group__body">Institute for Genomic Medicine, Columbia University, New York, NY 10032, USA</div></div><a href="/action/doSearch?searchType=authorLookUp&amp;author=Hayeck,&#xA;            Tristan J.&amp;prod=CP" class="author__search" target="_blank"><i class="icon-arrow_r"></i><span>Search for articles by this author</span></a></div></div><a href="#" title="Close info" class="dropBlock__ctrl--close"><i aria-hidden="true" class="icon-close_thin"></i></a></div></div><div class="dropBlock article-header__info"><a title="Send email to Tristan J. Hayeck" class="loa__item__name  loa__item__email" href="mailto:tristan.hayeck@gmail.com"><i aria-hidden="true" class="icon-Email faded"></i></a></div></li><li class="loa__item author"><div class="dropBlock article-header__info"><a href="#" data-db-target-for="au2-authorcard" aria-expanded="false" aria-controls="au2" class="loa__item__name article-header__info__ctrl loa__item__email" title="Correspondence information about the author Nicholas Stong">Nicholas Stong</a><div data-db-target-of="au2-authorcard" class="dropBlock__holder article-header__info__holder lefted"><div class="article-header__info__scrollable"><div class="article-header__info__label">Nicholas Stong</div><div class="article-header__info__body"><div class="article-header__info__social"><ul class="rlist inline-bullet-list"></ul></div><div class="article-header__info__group"><span class="article-header__info__group__label">Affiliations</span><div class="article-header__info__group__body">Institute for Genomic Medicine, Columbia University, New York, NY 10032, USA</div></div><a href="/action/doSearch?searchType=authorLookUp&amp;author=Stong,&#xA;            Nicholas&amp;prod=CP" class="author__search" target="_blank"><i class="icon-arrow_r"></i><span>Search for articles by this author</span></a></div></div><a href="#" title="Close info" class="dropBlock__ctrl--close"><i aria-hidden="true" class="icon-close_thin"></i></a></div></div></li><li class="loa__item author"><div class="dropBlock article-header__info"><a href="#" data-db-target-for="au3-authorcard" aria-expanded="false" aria-controls="au3" class="loa__item__name article-header__info__ctrl loa__item__email" title="Correspondence information about the author Charles J. Wolock">Charles J. Wolock</a><div data-db-target-of="au3-authorcard" class="dropBlock__holder article-header__info__holder lefted"><div class="article-header__info__scrollable"><div class="article-header__info__label">Charles J. Wolock</div><div class="article-header__info__body"><div class="article-header__info__social"><ul class="rlist inline-bullet-list"></ul></div><div class="article-header__info__group"><span class="article-header__info__group__label">Affiliations</span><div class="article-header__info__group__body">Institute for Genomic Medicine, Columbia University, New York, NY 10032, USA</div></div><a href="/action/doSearch?searchType=authorLookUp&amp;author=Wolock,&#xA;            Charles J.&amp;prod=CP" class="author__search" target="_blank"><i class="icon-arrow_r"></i><span>Search for articles by this author</span></a></div></div><a href="#" title="Close info" class="dropBlock__ctrl--close"><i aria-hidden="true" class="icon-close_thin"></i></a></div></div></li><li class="loa__item author"><div class="dropBlock article-header__info"><a href="#" data-db-target-for="au4-authorcard" aria-expanded="false" aria-controls="au4" class="loa__item__name article-header__info__ctrl loa__item__email" title="Correspondence information about the author Brett Copeland">Brett Copeland</a><div data-db-target-of="au4-authorcard" class="dropBlock__holder article-header__info__holder lefted"><div class="article-header__info__scrollable"><div class="article-header__info__label">Brett Copeland</div><div class="article-header__info__body"><div class="article-header__info__social"><ul class="rlist inline-bullet-list"></ul></div><div class="article-header__info__group"><span class="article-header__info__group__label">Affiliations</span><div class="article-header__info__group__body">Institute for Genomic Medicine, Columbia University, New York, NY 10032, USA</div></div><a href="/action/doSearch?searchType=authorLookUp&amp;author=Copeland,&#xA;            Brett&amp;prod=CP" class="author__search" target="_blank"><i class="icon-arrow_r"></i><span>Search for articles by this author</span></a></div></div><a href="#" title="Close info" class="dropBlock__ctrl--close"><i aria-hidden="true" class="icon-close_thin"></i></a></div></div></li><li class="loa__item author"><div class="dropBlock article-header__info"><a href="#" data-db-target-for="au5-authorcard" aria-expanded="false" aria-controls="au5" class="loa__item__name article-header__info__ctrl loa__item__email" title="Correspondence information about the author Sitharthan Kamalakaran">Sitharthan Kamalakaran</a><div data-db-target-of="au5-authorcard" class="dropBlock__holder article-header__info__holder lefted"><div class="article-header__info__scrollable"><div class="article-header__info__label">Sitharthan Kamalakaran</div><div class="article-header__info__body"><div class="article-header__info__social"><ul class="rlist inline-bullet-list"></ul></div><div class="article-header__info__group"><span class="article-header__info__group__label">Affiliations</span><div class="article-header__info__group__body">Institute for Genomic Medicine, Columbia University, New York, NY 10032, USA</div></div><a href="/action/doSearch?searchType=authorLookUp&amp;author=Kamalakaran,&#xA;            Sitharthan&amp;prod=CP" class="author__search" target="_blank"><i class="icon-arrow_r"></i><span>Search for articles by this author</span></a></div></div><a href="#" title="Close info" class="dropBlock__ctrl--close"><i aria-hidden="true" class="icon-close_thin"></i></a></div></div></li><li class="loa__item author"><div class="dropBlock article-header__info"><a href="#" data-db-target-for="au6-authorcard" aria-expanded="false" aria-controls="au6" class="loa__item__name article-header__info__ctrl loa__item__email" title="Correspondence information about the author David B. Goldstein">David B. Goldstein</a><div data-db-target-of="au6-authorcard" class="dropBlock__holder article-header__info__holder lefted"><div class="article-header__info__scrollable"><div class="article-header__info__label">David B. Goldstein</div><div class="article-header__info__body"><div class="article-header__info__social"><ul class="rlist inline-bullet-list"></ul></div><div class="article-header__info__group"><span class="article-header__info__group__label">Affiliations</span><div class="article-header__info__group__body">Institute for Genomic Medicine, Columbia University, New York, NY 10032, USA</div></div><a href="/action/doSearch?searchType=authorLookUp&amp;author=Goldstein,&#xA;            David B.&amp;prod=CP" class="author__search" target="_blank"><i class="icon-arrow_r"></i><span>Search for articles by this author</span></a></div></div><a href="#" title="Close info" class="dropBlock__ctrl--close"><i aria-hidden="true" class="icon-close_thin"></i></a></div></div></li><li class="loa__item author"><div class="dropBlock article-header__info"><a href="#" data-db-target-for="au7-authorcard" aria-expanded="false" aria-controls="au7" class="loa__item__name article-header__info__ctrl loa__item__email" title="Correspondence information about the author Andrew S. Allen">Andrew S. Allen<i aria-hidden="true" class="icon-gizmo-person faded"></i></a><div data-db-target-of="au7-authorcard" class="dropBlock__holder article-header__info__holder lefted"><div class="article-header__info__scrollable"><div class="article-header__info__label">Andrew S. Allen</div><span class="article-header__info__group__label">Correspondence</span><div class="article-header__info__sub-label">Corresponding author</div><div class="article-header__info__body"><div class="article-header__info__social"><a class="article-header__info__email default" href="mailto:andrew.s.allen@duke.edu" title="Send email to Andrew S. Allen"><i aria-hidden="true" class="icon-Email"></i><span>
                            Contact
                        </span></a></div><div class="article-header__info__group"><span class="article-header__info__group__label">Affiliations</span><div class="article-header__info__group__body">Institute for Genomic Medicine, Columbia University, New York, NY 10032, USA<p></p>Division of Integrative Genomics, Department of Biostatistics and Bioinformatics, Duke University, Durham, NC 27710, USA<p></p>Duke Center for Statistical Genetics and Genomics, Duke University, Durham, NC 27710, USA</div></div><a href="/action/doSearch?searchType=authorLookUp&amp;author=Allen,&#xA;            Andrew S.&amp;prod=CP" class="author__search" target="_blank"><i class="icon-arrow_r"></i><span>Search for articles by this author</span></a></div></div><a href="#" title="Close info" class="dropBlock__ctrl--close"><i aria-hidden="true" class="icon-close_thin"></i></a></div></div><div class="dropBlock article-header__info"><a title="Send email to Andrew S. Allen" class="loa__item__name  loa__item__email" href="mailto:andrew.s.allen@duke.edu"><i aria-hidden="true" class="icon-Email faded"></i></a></div></li> </ul><div class="inline-it"><span class="article-header__publish-date bulleted"><span class="article-header__publish-date__value">Published: January 24, 2019</span></span><span class="article-header__doi bulleted"><span class="article-header__doi__label">DOI:</span><a class="article-header__doi__value" href="https://doi.org/10.1016/j.ajhg.2018.12.020">https://doi.org/10.1016/j.ajhg.2018.12.020</a></span></div></div><div class="article-header__bottom clearfix row"><div class="col-md-6 static-md"></div><div class="col-md-6 static-md"><script type="text/javascript" src="//d39af2mgp1pqhg.cloudfront.net/widget-popup.js"></script><a href="https://plu.mx/plum/a/?doi=10.1016/j.ajhg.2018.12.020" class="plumx-plum-print-popup plum-jbs-theme article-header__metrics" data-popup="left" data-hide-when-empty="true" data-hide-usage="true" data-pass-hidden-categories="true" data-site="plum" data-badge="true" data-size="small">Improved Pathogenic Variant Localization via a Hierarchical Model of Sub-regional Intolerance</a></div></div></div><div class="article-header__dropZone pb-dropzone" data-pb-dropzone="articleNavigation-Dropzone">
    

        


    

            

            

            <div data-widget-id="54733c81-009b-4b22-b588-f003dedf997f" data-widget-name="UX3ArticleNavigation" class="content-navigation clearfix"><div class="content-navigation__btn__wrapper--prev"><a href="/ajhg/fulltext/S0002-9297(18)30470-1" title="Previous Article" class="content-navigation__btn--pre default"><span class="content-navigation__body pre"><span class="content-navigation__body__top">Previous Article</span><span class="content-navigation__body__title">Recessive Rare Variants in Deoxyhypusine Synthase, an Enzyme Involved in the Synthesis of Hypusine, Are Associated with a Neurodevelopmental Disorder</span></span><i aria-hidden="true" class="icon-arrow_l"></i></a></div><div class="content-navigation__extra"></div><div class="content-navigation__btn__wrapper--next"><a href="/ajhg/fulltext/S0002-9297(18)30502-0" title="Next Article" class="content-navigation__btn--next default"><span class="content-navigation__body next"><span class="content-navigation__body__top">Next Article</span><span class="content-navigation__body__title">Gene Augmentation and Readthrough Rescue Channelopathy in an iPSC-RPE Model of Congenital Blindness</span></span><i aria-hidden="true" class="icon-arrow_r"></i></a></div></div>

            

        

    

        


    

            

            

            <div data-widget-id="8633d695-b258-4908-b86f-e2c268ec3c6b" data-widget-name="UX3HTMLWidget"><script type="text/javascript">
if($('.content-navigation__body__title').length){
$('.content-navigation__body__title').each(function(){
var html=this.innerHTML;
var div = document.createElement("div");
div.innerHTML = html;
var text = div.textContent || div.innerText || "";
this.innerHTML=text;
});
}
</script></div>

            

        
</div></div></div><div class="article__body"><button title="open table of sections" aria-label="open table of sections" type="button" class="left-side-nav__ctrl"><i aria-hidden="true" class="icon-sub-menu"></i></button><div class="left-side-nav col-md-2 col-lg-1"><ul class="rlist left-side-nav__upper"></ul><ul class="rlist left-side-nav__bottom"></ul></div><div class="adplaceholders hidden-xs hidden-sm not-sticky">

</div><div class="container"><div class="row"><div class="col-md-2 col-lg-1 hidden-sm"></div><div class="col-md-7 col-lg-9 article__sections"><section><div class="section-paragraph"><div class="section-paragraph">Different parts of a gene can be of differential importance to development and health. This regional heterogeneity is also apparent in the distribution of disease-associated mutations, which often cluster in particular regions of disease-associated genes. The ability to precisely estimate functionally important sub-regions of genes will be key in correctly deciphering relationships between genetic variation and disease. Previous methods have had some success using standing human variation to characterize this variability in importance by measuring sub-regional intolerance, i.e., the depletion in functional variation from expectation within a given region of a gene. However, the ability to precisely estimate local intolerance was restricted by the fact that only information within a given sub-region is used, leading to instability in local estimates, especially for small regions. We show that borrowing information across regions using a Bayesian hierarchical model stabilizes estimates, leading to lower variability and improved predictive utility. Specifically, our approach more effectively identifies regions enriched for ClinVar pathogenic variants. We also identify significant correlations between sub-region intolerance and the distribution of pathogenic variation in disease-associated genes, with AUCs for classifying <em>de novo</em> missense variants in Online Mendelian Inheritance in Man (OMIM) genes of up to 0.86 using exonic sub-regions and 0.91 using sub-regions defined by protein domains. This result immediately suggests that considering the intolerance of regions in which variants are found may improve diagnostic interpretation. We also illustrate the utility of integrating regional intolerance into gene-level disease association tests with a study of known disease-associated genes for epileptic encephalopathy.</div></div></section><section><h2 class="top" id="secsectitle0010"><span class="top__sub"></span><span class="top__text">Keywords</span></h2><ul class="rlist keywords-list inline-bullet-list"><li><a href="/action/doSearch?searchType=quick&amp;occurrences=all&amp;ltrlSrch=true&amp;searchScope=series&amp;searchText=LIMBR&amp;seriesISSN=0002-9297">LIMBR</a></li><li><a href="/action/doSearch?searchType=quick&amp;occurrences=all&amp;ltrlSrch=true&amp;searchScope=series&amp;searchText=negative selection&amp;seriesISSN=0002-9297">negative selection</a></li><li><a href="/action/doSearch?searchType=quick&amp;occurrences=all&amp;ltrlSrch=true&amp;searchScope=series&amp;searchText=conservation&amp;seriesISSN=0002-9297">conservation</a></li><li><a href="/action/doSearch?searchType=quick&amp;occurrences=all&amp;ltrlSrch=true&amp;searchScope=series&amp;searchText=constraint&amp;seriesISSN=0002-9297">constraint</a></li><li><a href="/action/doSearch?searchType=quick&amp;occurrences=all&amp;ltrlSrch=true&amp;searchScope=series&amp;searchText=intolerance&amp;seriesISSN=0002-9297">intolerance</a></li><li><a href="/action/doSearch?searchType=quick&amp;occurrences=all&amp;ltrlSrch=true&amp;searchScope=series&amp;searchText=domains&amp;seriesISSN=0002-9297">domains</a></li><li><a href="/action/doSearch?searchType=quick&amp;occurrences=all&amp;ltrlSrch=true&amp;searchScope=series&amp;searchText=exons&amp;seriesISSN=0002-9297">exons</a></li><li><a href="/action/doSearch?searchType=quick&amp;occurrences=all&amp;ltrlSrch=true&amp;searchScope=series&amp;searchText=pathogenic&amp;seriesISSN=0002-9297">pathogenic</a></li><li><a href="/action/doSearch?searchType=quick&amp;occurrences=all&amp;ltrlSrch=true&amp;searchScope=series&amp;searchText=genic sub-region&amp;seriesISSN=0002-9297">genic sub-region</a></li><li><a href="/action/doSearch?searchType=quick&amp;occurrences=all&amp;ltrlSrch=true&amp;searchScope=series&amp;searchText=RVIS&amp;seriesISSN=0002-9297">RVIS</a></li></ul></section><div data-pb-dropzone="relatedItems-dropzone" class="relatedItems-dropzone pb-dropzone"></div><section id="sec1" data-locator="sec1"><h2 class="top" id="secsectitle0015"><span class="top__sub"></span><span class="top__text">Introduction</span></h2><div class="section-paragraph">Accurate identification of pathogenic mutations is a central challenge in medical genetics and key to establishing accurate genotype-phenotype associations. Such an accurate classification requires one to consider not only the characteristics of the variants themselves, but also the genomic context in which they are found. Not all parts of a gene are of equal importance to viability and health. A variant that severely disrupts the function of one part of a gene may have a very different impact from a variant that severely disrupts another part of the same gene. This fact is also borne out when one looks at distributions of disease-associated mutations, which often cluster in specific sub-regions of genes.<div class="dropBlock reference-citations"><sup><a href="#" id="back-bib1" data-db-target-for="bib1-d176679441e564" class="reference-citations__ctrl" aria-expanded="false" aria-controls="bib1">1</a></sup><div data-db-target-of="bib1-d176679441e564" class="dropBlock__holder"><div class="dropBlock__body"><ul title="list of authors" class="rlist reference-citations__loa inline-bullet-list"><li class="loa__item">Goldstein D.B. </li><li class="loa__item">Allen A. </li><li class="loa__item">Keebler J. </li><li class="loa__item">Margulies E.H. </li><li class="loa__item">Petrou S. </li><li class="loa__item">Petrovski S. </li><li class="loa__item">Sunyaev S. </li></ul><div class="reference-citations__detials">Sequencing studies in human genetics: design and interpretation.<span class="ref__series"><i>Nat. Rev. Genet.</i> <span class="ref__seriesDate">2013; </span><span class="ref__seriesVolume">14</span></span>: <span class="ref__seriesPages">460-470</span></div><div class="reference-citations__links"><ul class="rlist inline-bullet-list"><li><a href="/servlet/linkout?suffix=e_1_5_1_2_1_2&amp;dbid=16&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=10.1038%2Fnrg3455&amp;cf=" onclick="newWindow(this.href);return false">Crossref</a></li>
<li><a href="/servlet/linkout?suffix=e_1_5_1_2_1_2&amp;dbid=8&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=23752795&amp;cf=" onclick="newWindow(this.href);return false">PubMed</a></li>
<li><a href="/servlet/linkout?suffix=e_1_5_1_2_1_2&amp;dbid=137438953472&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=2-s2.0-84879411643&amp;cf=" onclick="newWindow(this.href);return false">Scopus (141)</a></li>
<li><a class="google-scholar" href="http://scholar.google.com/scholar_lookup?hl=en&volume=14&publication_year=2013&pages=460-470&journal=Nat.+Rev.+Genet.&author=Goldstein+D.B.+%2C+Allen+A.+%2C+Keebler+J.+%2C+Margulies+E.H.+%2C+Petrou+S.+%2C+Petrovski+S.+%2C++and+Sunyaev+S.+&title=Sequencing+studies+in+human+genetics%3A+design+and+interpretation&">Google Scholar</a></li></ul></div></div><a href="#" class="dropBlock__ctrl--close" title="Close info"><i aria-hidden="true" class="icon-close_thin"></i></a></div></div><sup>,Â </sup> <div class="dropBlock reference-citations"><sup><a href="#" id="back-bib2" data-db-target-for="bib2-d176679441e564" class="reference-citations__ctrl" aria-expanded="false" aria-controls="bib2">2</a></sup><div data-db-target-of="bib2-d176679441e564" class="dropBlock__holder"><div class="dropBlock__body"><ul title="list of authors" class="rlist reference-citations__loa inline-bullet-list"><li class="loa__item">Eilbeck K. </li><li class="loa__item">Quinlan A. </li><li class="loa__item">Yandell M. </li></ul><div class="reference-citations__detials">Settling the score: variant prioritization and Mendelian disease.<span class="ref__series"><i>Nat. Rev. Genet.</i> <span class="ref__seriesDate">2017; </span><span class="ref__seriesVolume">18</span></span>: <span class="ref__seriesPages">599-612</span></div><div class="reference-citations__links"><ul class="rlist inline-bullet-list"><li><a href="/servlet/linkout?suffix=e_1_5_1_2_2_2&amp;dbid=16&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=10.1038%2Fnrg.2017.52&amp;cf=" onclick="newWindow(this.href);return false">Crossref</a></li>
<li><a href="/servlet/linkout?suffix=e_1_5_1_2_2_2&amp;dbid=8&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=28804138&amp;cf=" onclick="newWindow(this.href);return false">PubMed</a></li>
<li><a href="/servlet/linkout?suffix=e_1_5_1_2_2_2&amp;dbid=137438953472&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=2-s2.0-85029470345&amp;cf=" onclick="newWindow(this.href);return false">Scopus (28)</a></li>
<li><a class="google-scholar" href="http://scholar.google.com/scholar_lookup?hl=en&volume=18&publication_year=2017&pages=599-612&journal=Nat.+Rev.+Genet.&author=Eilbeck+K.+%2C+Quinlan+A.+%2C++and+Yandell+M.+&title=Settling+the+score%3A+variant+prioritization+and+Mendelian+disease&">Google Scholar</a></li></ul></div></div><a href="#" class="dropBlock__ctrl--close" title="Close info"><i aria-hidden="true" class="icon-close_thin"></i></a></div></div> While variant level measures of functional impact are useful, their value can be increased by integration with information about the surrounding sequence context.</div><div class="section-paragraph">Cross species conservation<div class="dropBlock reference-citations"><sup><a href="#" id="back-bib3" data-db-target-for="bib3-d176679441e570" class="reference-citations__ctrl" aria-expanded="false" aria-controls="bib3">3</a></sup><div data-db-target-of="bib3-d176679441e570" class="dropBlock__holder"><div class="dropBlock__body"><ul title="list of authors" class="rlist reference-citations__loa inline-bullet-list"><li class="loa__item">Davydov E.V. </li><li class="loa__item">Goode D.L. </li><li class="loa__item">Sirota M. </li><li class="loa__item">Cooper G.M. </li><li class="loa__item">Sidow A. </li><li class="loa__item">Batzoglou S. </li></ul><div class="reference-citations__detials">Identifying a high fraction of the human genome to be under selective constraint using GERP++.<span class="ref__series"><i>PLoS Comput. Biol.</i> <span class="ref__seriesDate">2010; </span><span class="ref__seriesVolume">6</span></span>: <span class="ref__seriesPages">e1001025</span></div><div class="reference-citations__links"><ul class="rlist inline-bullet-list"><li><a href="/servlet/linkout?suffix=e_1_5_1_2_3_2&amp;dbid=16&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=10.1371%2Fjournal.pcbi.1001025&amp;cf=" onclick="newWindow(this.href);return false">Crossref</a></li>
<li><a href="/servlet/linkout?suffix=e_1_5_1_2_3_2&amp;dbid=8&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=21152010&amp;cf=" onclick="newWindow(this.href);return false">PubMed</a></li>
<li><a href="/servlet/linkout?suffix=e_1_5_1_2_3_2&amp;dbid=137438953472&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=2-s2.0-78651237647&amp;cf=" onclick="newWindow(this.href);return false">Scopus (510)</a></li>
<li><a class="google-scholar" href="http://scholar.google.com/scholar_lookup?hl=en&volume=6&publication_year=2010&pages=e1001025&journal=PLoS+Comput.+Biol.&author=Davydov+E.V.+%2C+Goode+D.L.+%2C+Sirota+M.+%2C+Cooper+G.M.+%2C+Sidow+A.+%2C++and+Batzoglou+S.+&title=Identifying+a+high+fraction+of+the+human+genome+to+be+under+selective+constraint+using+GERP%2B%2B&">Google Scholar</a></li></ul></div></div><a href="#" class="dropBlock__ctrl--close" title="Close info"><i aria-hidden="true" class="icon-close_thin"></i></a></div></div> has often been used to quantify sequence importance. A genomic region that is strongly conserved across species is likely subject to purifying selection, and therefore more likely to result in disease when disrupted. However, the conservation approach breaks down when the functionality of the sequence is human specific. To overcome this, Petrovski etÂ al.<div class="dropBlock reference-citations"><sup><a href="#" id="back-bib4" data-db-target-for="bib4-d176679441e574" class="reference-citations__ctrl" aria-expanded="false" aria-controls="bib4">4</a></sup><div data-db-target-of="bib4-d176679441e574" class="dropBlock__holder"><div class="dropBlock__body"><ul title="list of authors" class="rlist reference-citations__loa inline-bullet-list"><li class="loa__item">Petrovski S. </li><li class="loa__item">Wang Q. </li><li class="loa__item">Heinzen E.L. </li><li class="loa__item">Allen A.S. </li><li class="loa__item">Goldstein D.B. </li></ul><div class="reference-citations__detials">Genic intolerance to functional variation and the interpretation of personal genomes.<span class="ref__series"><i>PLoS Genet.</i> <span class="ref__seriesDate">2013; </span><span class="ref__seriesVolume">9</span></span>: <span class="ref__seriesPages">e1003709</span></div><div class="reference-citations__links"><ul class="rlist inline-bullet-list"><li><a href="/servlet/linkout?suffix=e_1_5_1_2_4_2&amp;dbid=16&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=10.1371%2Fjournal.pgen.1003709&amp;cf=" onclick="newWindow(this.href);return false">Crossref</a></li>
<li><a href="/servlet/linkout?suffix=e_1_5_1_2_4_2&amp;dbid=8&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=23990802&amp;cf=" onclick="newWindow(this.href);return false">PubMed</a></li>
<li><a href="/servlet/linkout?suffix=e_1_5_1_2_4_2&amp;dbid=137438953472&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=2-s2.0-84884592445&amp;cf=" onclick="newWindow(this.href);return false">Scopus (379)</a></li>
<li><a class="google-scholar" href="http://scholar.google.com/scholar_lookup?hl=en&volume=9&publication_year=2013&pages=e1003709&journal=PLoS+Genet.&author=Petrovski+S.+%2C+Wang+Q.+%2C+Heinzen+E.L.+%2C+Allen+A.S.+%2C++and+Goldstein+D.B.+&title=Genic+intolerance+to+functional+variation+and+the+interpretation+of+personal+genomes&">Google Scholar</a></li></ul></div></div><a href="#" class="dropBlock__ctrl--close" title="Close info"><i aria-hidden="true" class="icon-close_thin"></i></a></div></div> advanced an empirical measure of purifying selection in the human lineage. Using standing variation in the human population, they quantified the âintoleranceâ of genes by the observed depletion of common functional variation relative to expectation given the total amount of variation in the gene. Others have advanced similar gene-level scores, including âconstraintâ<div class="dropBlock reference-citations"><sup><a href="#" id="back-bib5" data-db-target-for="bib5-d176679441e578" class="reference-citations__ctrl" aria-expanded="false" aria-controls="bib5">5</a></sup><div data-db-target-of="bib5-d176679441e578" class="dropBlock__holder"><div class="dropBlock__body"><ul title="list of authors" class="rlist reference-citations__loa inline-bullet-list"><li class="loa__item">Samocha K.E. </li><li class="loa__item">Robinson E.B. </li><li class="loa__item">Sanders S.J. </li><li class="loa__item">Stevens C. </li><li class="loa__item">Sabo A. </li><li class="loa__item">McGrath L.M. </li><li class="loa__item">Kosmicki J.A. </li><li class="loa__item">RehnstrÃ¶m K. </li><li class="loa__item">Mallick S. </li><li class="loa__item">Kirby A. </li><li>et al.</li></ul><div class="reference-citations__detials">A framework for the interpretation of de novo mutation in human disease.<span class="ref__series"><i>Nat. Genet.</i> <span class="ref__seriesDate">2014; </span><span class="ref__seriesVolume">46</span></span>: <span class="ref__seriesPages">944-950</span></div><div class="reference-citations__links"><ul class="rlist inline-bullet-list"><li><a href="/servlet/linkout?suffix=e_1_5_1_2_5_2&amp;dbid=16&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=10.1038%2Fng.3050&amp;cf=" onclick="newWindow(this.href);return false">Crossref</a></li>
<li><a href="/servlet/linkout?suffix=e_1_5_1_2_5_2&amp;dbid=8&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=25086666&amp;cf=" onclick="newWindow(this.href);return false">PubMed</a></li>
<li><a href="/servlet/linkout?suffix=e_1_5_1_2_5_2&amp;dbid=137438953472&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=2-s2.0-84905309176&amp;cf=" onclick="newWindow(this.href);return false">Scopus (322)</a></li>
<li><a class="google-scholar" href="http://scholar.google.com/scholar_lookup?hl=en&volume=46&publication_year=2014&pages=944-950&journal=Nat.+Genet.&author=Samocha+K.E.+%2C+Robinson+E.B.+%2C+Sanders+S.J.+%2C+Stevens+C.+%2C+Sabo+A.+%2C+McGrath+L.M.+%2C+Kosmicki+J.A.+%2C+Rehnstr%C3%B6m+K.+%2C+Mallick+S.+%2C+Kirby+A.++et+al.&title=A+framework+for+the+interpretation+of+de+novo+mutation+in+human+disease&">Google Scholar</a></li></ul></div></div><a href="#" class="dropBlock__ctrl--close" title="Close info"><i aria-hidden="true" class="icon-close_thin"></i></a></div></div> and âvulnerabilityâ<div class="dropBlock reference-citations"><sup><a href="#" id="back-bib6" data-db-target-for="bib6-d176679441e582" class="reference-citations__ctrl" aria-expanded="false" aria-controls="bib6">6</a></sup><div data-db-target-of="bib6-d176679441e582" class="dropBlock__holder"><div class="dropBlock__body"><ul title="list of authors" class="rlist reference-citations__loa inline-bullet-list"><li class="loa__item">Iossifov I. </li><li class="loa__item">Ronemus M. </li><li class="loa__item">Levy D. </li><li class="loa__item">Wang Z. </li><li class="loa__item">Hakker I. </li><li class="loa__item">Rosenbaum J. </li><li class="loa__item">Yamrom B. </li><li class="loa__item">Lee Y.H. </li><li class="loa__item">Narzisi G. </li><li class="loa__item">Leotta A. </li><li>et al.</li></ul><div class="reference-citations__detials">De novo gene disruptions in children on the autistic spectrum.<span class="ref__series"><i>Neuron.</i> <span class="ref__seriesDate">2012; </span><span class="ref__seriesVolume">74</span></span>: <span class="ref__seriesPages">285-299</span></div><div class="reference-citations__links"><ul class="rlist inline-bullet-list"><li><a href="/servlet/linkout?suffix=e_1_5_1_2_6_2&amp;dbid=4&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=10.1016%2Fj.neuron.2012.04.009&amp;cf=abstract&amp;site=cell-site" onclick="newWindow(this.href);return false">Abstract</a></li>
<li><a href="/servlet/linkout?suffix=e_1_5_1_2_6_2&amp;dbid=4&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=10.1016%2Fj.neuron.2012.04.009&amp;cf=fulltext&amp;site=cell-site" onclick="newWindow(this.href);return false">Full Text</a></li>
<li><a href="/servlet/linkout?suffix=e_1_5_1_2_6_2&amp;dbid=4&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=10.1016%2Fj.neuron.2012.04.009&amp;cf=pdf&amp;site=cell-site" onclick="newWindow(this.href);return false">Full Text PDF</a></li>
<li><a href="/servlet/linkout?suffix=e_1_5_1_2_6_2&amp;dbid=8&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=22542183&amp;cf=" onclick="newWindow(this.href);return false">PubMed</a></li>
<li><a href="/servlet/linkout?suffix=e_1_5_1_2_6_2&amp;dbid=137438953472&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=2-s2.0-84860297457&amp;cf=" onclick="newWindow(this.href);return false">Scopus (735)</a></li>
<li><a class="google-scholar" href="http://scholar.google.com/scholar_lookup?hl=en&volume=74&publication_year=2012&pages=285-299&journal=Neuron&author=Iossifov+I.+%2C+Ronemus+M.+%2C+Levy+D.+%2C+Wang+Z.+%2C+Hakker+I.+%2C+Rosenbaum+J.+%2C+Yamrom+B.+%2C+Lee+Y.H.+%2C+Narzisi+G.+%2C+Leotta+A.++et+al.&title=De+novo+gene+disruptions+in+children+on+the+autistic+spectrum&">Google Scholar</a></li></ul></div></div><a href="#" class="dropBlock__ctrl--close" title="Close info"><i aria-hidden="true" class="icon-close_thin"></i></a></div></div> scores. These scores have proven useful in prioritizing genes by likelihood of being disease related. However, they do not aid interpreting regions within genes, since they score the entire gene and not its sub-units.</div><div class="section-paragraph">Sub-regional versions of intolerance and constraint scores have been proposed.<div class="dropBlock reference-citations"><sup><a href="#" id="back-bib3" data-db-target-for="bib3-d176679441e588" class="reference-citations__ctrl" aria-expanded="false" aria-controls="bib3">3</a></sup><div data-db-target-of="bib3-d176679441e588" class="dropBlock__holder"><div class="dropBlock__body"><ul title="list of authors" class="rlist reference-citations__loa inline-bullet-list"><li class="loa__item">Davydov E.V. </li><li class="loa__item">Goode D.L. </li><li class="loa__item">Sirota M. </li><li class="loa__item">Cooper G.M. </li><li class="loa__item">Sidow A. </li><li class="loa__item">Batzoglou S. </li></ul><div class="reference-citations__detials">Identifying a high fraction of the human genome to be under selective constraint using GERP++.<span class="ref__series"><i>PLoS Comput. Biol.</i> <span class="ref__seriesDate">2010; </span><span class="ref__seriesVolume">6</span></span>: <span class="ref__seriesPages">e1001025</span></div><div class="reference-citations__links"><ul class="rlist inline-bullet-list"><li><a href="/servlet/linkout?suffix=e_1_5_1_2_3_2&amp;dbid=16&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=10.1371%2Fjournal.pcbi.1001025&amp;cf=" onclick="newWindow(this.href);return false">Crossref</a></li>
<li><a href="/servlet/linkout?suffix=e_1_5_1_2_3_2&amp;dbid=8&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=21152010&amp;cf=" onclick="newWindow(this.href);return false">PubMed</a></li>
<li><a href="/servlet/linkout?suffix=e_1_5_1_2_3_2&amp;dbid=137438953472&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=2-s2.0-78651237647&amp;cf=" onclick="newWindow(this.href);return false">Scopus (510)</a></li>
<li><a class="google-scholar" href="http://scholar.google.com/scholar_lookup?hl=en&volume=6&publication_year=2010&pages=e1001025&journal=PLoS+Comput.+Biol.&author=Davydov+E.V.+%2C+Goode+D.L.+%2C+Sirota+M.+%2C+Cooper+G.M.+%2C+Sidow+A.+%2C++and+Batzoglou+S.+&title=Identifying+a+high+fraction+of+the+human+genome+to+be+under+selective+constraint+using+GERP%2B%2B&">Google Scholar</a></li></ul></div></div><a href="#" class="dropBlock__ctrl--close" title="Close info"><i aria-hidden="true" class="icon-close_thin"></i></a></div></div><sup>,Â </sup> <div class="dropBlock reference-citations"><sup><a href="#" id="back-bib4" data-db-target-for="bib4-d176679441e588" class="reference-citations__ctrl" aria-expanded="false" aria-controls="bib4">4</a></sup><div data-db-target-of="bib4-d176679441e588" class="dropBlock__holder"><div class="dropBlock__body"><ul title="list of authors" class="rlist reference-citations__loa inline-bullet-list"><li class="loa__item">Petrovski S. </li><li class="loa__item">Wang Q. </li><li class="loa__item">Heinzen E.L. </li><li class="loa__item">Allen A.S. </li><li class="loa__item">Goldstein D.B. </li></ul><div class="reference-citations__detials">Genic intolerance to functional variation and the interpretation of personal genomes.<span class="ref__series"><i>PLoS Genet.</i> <span class="ref__seriesDate">2013; </span><span class="ref__seriesVolume">9</span></span>: <span class="ref__seriesPages">e1003709</span></div><div class="reference-citations__links"><ul class="rlist inline-bullet-list"><li><a href="/servlet/linkout?suffix=e_1_5_1_2_4_2&amp;dbid=16&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=10.1371%2Fjournal.pgen.1003709&amp;cf=" onclick="newWindow(this.href);return false">Crossref</a></li>
<li><a href="/servlet/linkout?suffix=e_1_5_1_2_4_2&amp;dbid=8&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=23990802&amp;cf=" onclick="newWindow(this.href);return false">PubMed</a></li>
<li><a href="/servlet/linkout?suffix=e_1_5_1_2_4_2&amp;dbid=137438953472&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=2-s2.0-84884592445&amp;cf=" onclick="newWindow(this.href);return false">Scopus (379)</a></li>
<li><a class="google-scholar" href="http://scholar.google.com/scholar_lookup?hl=en&volume=9&publication_year=2013&pages=e1003709&journal=PLoS+Genet.&author=Petrovski+S.+%2C+Wang+Q.+%2C+Heinzen+E.L.+%2C+Allen+A.S.+%2C++and+Goldstein+D.B.+&title=Genic+intolerance+to+functional+variation+and+the+interpretation+of+personal+genomes&">Google Scholar</a></li></ul></div></div><a href="#" class="dropBlock__ctrl--close" title="Close info"><i aria-hidden="true" class="icon-close_thin"></i></a></div></div><sup>,Â </sup> <div class="dropBlock reference-citations"><sup><a href="#" id="back-bib5" data-db-target-for="bib5-d176679441e588" class="reference-citations__ctrl" aria-expanded="false" aria-controls="bib5">5</a></sup><div data-db-target-of="bib5-d176679441e588" class="dropBlock__holder"><div class="dropBlock__body"><ul title="list of authors" class="rlist reference-citations__loa inline-bullet-list"><li class="loa__item">Samocha K.E. </li><li class="loa__item">Robinson E.B. </li><li class="loa__item">Sanders S.J. </li><li class="loa__item">Stevens C. </li><li class="loa__item">Sabo A. </li><li class="loa__item">McGrath L.M. </li><li class="loa__item">Kosmicki J.A. </li><li class="loa__item">RehnstrÃ¶m K. </li><li class="loa__item">Mallick S. </li><li class="loa__item">Kirby A. </li><li>et al.</li></ul><div class="reference-citations__detials">A framework for the interpretation of de novo mutation in human disease.<span class="ref__series"><i>Nat. Genet.</i> <span class="ref__seriesDate">2014; </span><span class="ref__seriesVolume">46</span></span>: <span class="ref__seriesPages">944-950</span></div><div class="reference-citations__links"><ul class="rlist inline-bullet-list"><li><a href="/servlet/linkout?suffix=e_1_5_1_2_5_2&amp;dbid=16&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=10.1038%2Fng.3050&amp;cf=" onclick="newWindow(this.href);return false">Crossref</a></li>
<li><a href="/servlet/linkout?suffix=e_1_5_1_2_5_2&amp;dbid=8&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=25086666&amp;cf=" onclick="newWindow(this.href);return false">PubMed</a></li>
<li><a href="/servlet/linkout?suffix=e_1_5_1_2_5_2&amp;dbid=137438953472&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=2-s2.0-84905309176&amp;cf=" onclick="newWindow(this.href);return false">Scopus (322)</a></li>
<li><a class="google-scholar" href="http://scholar.google.com/scholar_lookup?hl=en&volume=46&publication_year=2014&pages=944-950&journal=Nat.+Genet.&author=Samocha+K.E.+%2C+Robinson+E.B.+%2C+Sanders+S.J.+%2C+Stevens+C.+%2C+Sabo+A.+%2C+McGrath+L.M.+%2C+Kosmicki+J.A.+%2C+Rehnstr%C3%B6m+K.+%2C+Mallick+S.+%2C+Kirby+A.++et+al.&title=A+framework+for+the+interpretation+of+de+novo+mutation+in+human+disease&">Google Scholar</a></li></ul></div></div><a href="#" class="dropBlock__ctrl--close" title="Close info"><i aria-hidden="true" class="icon-close_thin"></i></a></div></div><sup>,Â </sup> <div class="dropBlock reference-citations"><sup><a href="#" id="back-bib7" data-db-target-for="bib7-d176679441e588" class="reference-citations__ctrl" aria-expanded="false" aria-controls="bib7">7</a></sup><div data-db-target-of="bib7-d176679441e588" class="dropBlock__holder"><div class="dropBlock__body"><ul title="list of authors" class="rlist reference-citations__loa inline-bullet-list"><li class="loa__item">Gussow A.B. </li><li class="loa__item">Petrovski S. </li><li class="loa__item">Wang Q. </li><li class="loa__item">Allen A.S. </li><li class="loa__item">Goldstein D.B. </li></ul><div class="reference-citations__detials">The intolerance to functional genetic variation of protein domains predicts the localization of pathogenic mutations within genes.<span class="ref__series"><i>Genome Biol.</i> <span class="ref__seriesDate">2016; </span><span class="ref__seriesVolume">17</span></span>: <span class="ref__seriesPages">9</span></div><div class="reference-citations__links"><ul class="rlist inline-bullet-list"><li><a href="/servlet/linkout?suffix=e_1_5_1_2_7_2&amp;dbid=16&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=10.1186%2Fs13059-016-0869-4&amp;cf=" onclick="newWindow(this.href);return false">Crossref</a></li>
<li><a href="/servlet/linkout?suffix=e_1_5_1_2_7_2&amp;dbid=8&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=26781712&amp;cf=" onclick="newWindow(this.href);return false">PubMed</a></li>
<li><a class="google-scholar" href="http://scholar.google.com/scholar_lookup?hl=en&volume=17&publication_year=2016&pages=9&journal=Genome+Biol.&author=Gussow+A.B.+%2C+Petrovski+S.+%2C+Wang+Q.+%2C+Allen+A.S.+%2C++and+Goldstein+D.B.+&title=The+intolerance+to+functional+genetic+variation+of+protein+domains+predicts+the+localization+of+pathogenic+mutations+within+genes&">Google Scholar</a></li></ul></div></div><a href="#" class="dropBlock__ctrl--close" title="Close info"><i aria-hidden="true" class="icon-close_thin"></i></a></div></div><sup>,Â </sup> <div class="dropBlock reference-citations"><sup><a href="#" id="back-bib8" data-db-target-for="bib8-d176679441e588" class="reference-citations__ctrl" aria-expanded="false" aria-controls="bib8">8</a></sup><div data-db-target-of="bib8-d176679441e588" class="dropBlock__holder"><div class="dropBlock__body"><ul title="list of authors" class="rlist reference-citations__loa inline-bullet-list"><li class="loa__item">Lek M. </li><li class="loa__item">Karczewski K.J. </li><li class="loa__item">Minikel E.V. </li><li class="loa__item">Samocha K.E. </li><li class="loa__item">Banks E. </li><li class="loa__item">Fennell T. </li><li class="loa__item">OâDonnell-Luria A.H. </li><li class="loa__item">Ware J.S. </li><li class="loa__item">Hill A.J. </li><li class="loa__item">Cummings B.B. </li><li>et al.</li></ul><div class="reference-citations__detials">M.LekK.J.KarczewskiE.V.MinikelK.E.SamochaE.BanksT.FennellA.H.OâDonnell-LuriaJ.S.WareA.J.HillB.B.Cummings<span>Exome Aggregation Consortium</span>Analysis of protein-coding genetic variation in 60,706 humans.<span class="ref__series"><i>Nature.</i> <span class="ref__seriesDate">2016; </span><span class="ref__seriesVolume">536</span></span>: <span class="ref__seriesPages">285-291</span></div><div class="reference-citations__links"><ul class="rlist inline-bullet-list"><li><a href="/servlet/linkout?suffix=e_1_5_1_2_8_2&amp;dbid=16&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=10.1038%2Fnature19057&amp;cf=" onclick="newWindow(this.href);return false">Crossref</a></li>
<li><a href="/servlet/linkout?suffix=e_1_5_1_2_8_2&amp;dbid=8&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=27535533&amp;cf=" onclick="newWindow(this.href);return false">PubMed</a></li>
<li><a class="google-scholar" href="http://scholar.google.com/scholar_lookup?hl=en&volume=536&publication_year=2016&pages=285-291&journal=Nature&author=Exome+Aggregation+Consortium&title=Analysis+of+protein-coding+genetic+variation+in+60%2C706+humans&">Google Scholar</a></li></ul></div></div><a href="#" class="dropBlock__ctrl--close" title="Close info"><i aria-hidden="true" class="icon-close_thin"></i></a></div></div> Sub-RVIS,<div class="dropBlock reference-citations"><sup><a href="#" id="back-bib7" data-db-target-for="bib7-d176679441e592" class="reference-citations__ctrl" aria-expanded="false" aria-controls="bib7">7</a></sup><div data-db-target-of="bib7-d176679441e592" class="dropBlock__holder"><div class="dropBlock__body"><ul title="list of authors" class="rlist reference-citations__loa inline-bullet-list"><li class="loa__item">Gussow A.B. </li><li class="loa__item">Petrovski S. </li><li class="loa__item">Wang Q. </li><li class="loa__item">Allen A.S. </li><li class="loa__item">Goldstein D.B. </li></ul><div class="reference-citations__detials">The intolerance to functional genetic variation of protein domains predicts the localization of pathogenic mutations within genes.<span class="ref__series"><i>Genome Biol.</i> <span class="ref__seriesDate">2016; </span><span class="ref__seriesVolume">17</span></span>: <span class="ref__seriesPages">9</span></div><div class="reference-citations__links"><ul class="rlist inline-bullet-list"><li><a href="/servlet/linkout?suffix=e_1_5_1_2_7_2&amp;dbid=16&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=10.1186%2Fs13059-016-0869-4&amp;cf=" onclick="newWindow(this.href);return false">Crossref</a></li>
<li><a href="/servlet/linkout?suffix=e_1_5_1_2_7_2&amp;dbid=8&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=26781712&amp;cf=" onclick="newWindow(this.href);return false">PubMed</a></li>
<li><a class="google-scholar" href="http://scholar.google.com/scholar_lookup?hl=en&volume=17&publication_year=2016&pages=9&journal=Genome+Biol.&author=Gussow+A.B.+%2C+Petrovski+S.+%2C+Wang+Q.+%2C+Allen+A.S.+%2C++and+Goldstein+D.B.+&title=The+intolerance+to+functional+genetic+variation+of+protein+domains+predicts+the+localization+of+pathogenic+mutations+within+genes&">Google Scholar</a></li></ul></div></div><a href="#" class="dropBlock__ctrl--close" title="Close info"><i aria-hidden="true" class="icon-close_thin"></i></a></div></div> for example, partitions genes into functional subunits (e.g., exons or domains) and then measures the depletion of common missense variation, relative to expectation, within them. Missense variation is used since other types of variants, including loss of function (LoF) variants, may damage the resulting protein regardless of where they are found in the gene, whereas missense variation may result in localized, structural changes to the protein. However, a limitation of sub-RVIS is that it does not consider the fact that each sub-region lies within a gene and therefore has natural relationships to other sub-regions in the same gene. It is already difficult to detect functional depletion on a gene-wide level, so analyzing smaller regions of genes adds to the challenge.<div class="dropBlock reference-citations"><sup><a href="#" id="back-bib9" data-db-target-for="bib9-d176679441e596" class="reference-citations__ctrl" aria-expanded="false" aria-controls="bib9">9</a></sup><div data-db-target-of="bib9-d176679441e596" class="dropBlock__holder"><div class="dropBlock__body"><ul title="list of authors" class="rlist reference-citations__loa inline-bullet-list"><li class="loa__item">Bartha I. </li><li class="loa__item">di Iulio J. </li><li class="loa__item">Venter J.C. </li><li class="loa__item">Telenti A. </li></ul><div class="reference-citations__detials">Human gene essentiality.<span class="ref__series"><i>Nat. Rev. Genet.</i> <span class="ref__seriesDate">2018; </span><span class="ref__seriesVolume">19</span></span>: <span class="ref__seriesPages">51-62</span></div><div class="reference-citations__links"><ul class="rlist inline-bullet-list"><li><a href="/servlet/linkout?suffix=e_1_5_1_2_9_2&amp;dbid=16&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=10.1038%2Fnrg.2017.75&amp;cf=" onclick="newWindow(this.href);return false">Crossref</a></li>
<li><a href="/servlet/linkout?suffix=e_1_5_1_2_9_2&amp;dbid=8&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=29082913&amp;cf=" onclick="newWindow(this.href);return false">PubMed</a></li>
<li><a href="/servlet/linkout?suffix=e_1_5_1_2_9_2&amp;dbid=137438953472&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=2-s2.0-85039852688&amp;cf=" onclick="newWindow(this.href);return false">Scopus (4)</a></li>
<li><a class="google-scholar" href="http://scholar.google.com/scholar_lookup?hl=en&volume=19&publication_year=2018&pages=51-62&journal=Nat.+Rev.+Genet.&author=Bartha+I.+%2C+di+Iulio+J.+%2C+Venter+J.C.+%2C++and+Telenti+A.+&title=Human+gene+essentiality&">Google Scholar</a></li></ul></div></div><a href="#" class="dropBlock__ctrl--close" title="Close info"><i aria-hidden="true" class="icon-close_thin"></i></a></div></div> If a sub-region is small or the amount of variation is low, estimates of intolerance are highly variable and often inaccurate.</div><div class="section-paragraph">Since domains and exons are nested within genes, it isÂ natural to model the relationship between subunits using a hierarchical model. Here, we improve existing methods and develop a statistical intolerance framework that is able to jointly model genome-wide, genic, and sub-regional effects. The hierarchical model, a localized intolerance model using Bayesian regression (LIMBR), facilitates the borrowing of information across a gene, leading to stabilization of estimates, lower variability, and improved predictive utility. As a result of this approach, LIMBR shows much less bias due to sub-region size than previous methods. To evaluate the utility of our approach, we show that the intolerance scores are significantly associated with whether regions carry pathogenic mutations and are highly predictive, in particular, of <em>deÂ novo</em> missense variants in Online Mendelian Inheritance in Man (OMIM) disease-associated genes. Finally, we illustrate the utility of integrating regional intolerance into gene-level disease association tests with a study of known disease-associated genes for epileptic encephalopathy, where results suggest that considering the intolerance of regions in which variants are found may improve power of gene-disease association testing.</div></section><section id="sec2" data-locator="sec2"><h2 class="top" id="secsectitle0020"><span class="top__sub"></span><span class="top__text">Material and Methods</span></h2><section id="sec2.1" data-locator="sec2.1"><h3 class="sectionTitle"><span class="top__sub">Â </span><span class="top__text">Bayesian Hierarchical Model</span></h3><div class="section-paragraph">We establish a Bayesian hierarchical model explicitly characterizing depletion in functional variation at both the gene and sub-regional level. As with sub-RVIS, we estimate intolerance within a genic sub-region by the departure, within the region, from the expected number of missense variants given the total number of variants (synonymous and functional) within the sub-region. However, unlike sub-RVIS, we use a hierarchical model that explicitly nests sub-regions within genes to further decompose this residual variation into gene-level and sub-region components. <a data-link="modal" data-target="#image-S0002-9297(18)30503-2gr1" href="#fig1" data-locator="fig1" id="back-gr1" class="figure-link scroll-into-link">FigureÂ 1</a> depicts the approximate geometric interpretation of this approach. The hierarchical model allows information to be shared across sub-regions, stabilizing intolerance estimates.</div><div class="floatDisplay"><figure id="fig1" class=""><img class="b-lazy" data-src="//els-jbs-prod-cdn.literatumonline.com/cms/attachment/7cce8be4-cec3-43bc-a7e6-58b2b193e7a3/gr1.jpg" alt="Figure thumbnail gr1" data-locator="fig1" data-high-res="//els-jbs-prod-cdn.literatumonline.com/cms/attachment/18aa28c5-683f-4dd9-9501-031cafb43e48/gr1_lrg.jpg" data-large-image="//els-jbs-prod-cdn.literatumonline.com/cms/attachment/7cce8be4-cec3-43bc-a7e6-58b2b193e7a3/gr1.jpg"></img><figcaption><div class="figure__caption__body"><div class="figure__title"><span data-fv-label="Fig. 1" class="label figure__label">FigureÂ 1</span><span class="figure__title__text">Relative Missense Variation versus Total Variation across Domains</span></div><div class="accordion"><a href="#" class="accordion__control" aria-expanded="false"><i class="icon-section_arrow_d"></i><span>Show full caption</span></a><div class="accordion__content"><div class="figure__caption__text"><div class="section-paragraph">Regions (domains) are plotted with the number of total SNV variants in each coding region on the x axis versus the number of missense variants on the y axis. The genome-wide average missense variation versus total variation is plotted as a black solid line (A). Highlighted in blue are the SCN1A domains as an example. The offset average gene level trend for SCN1A is plotted as a blue dotted line (B) and can be seen more clearly in the exploded panel. Fitting a Bayesian hierarchical model allows for sharing of information across sub-regions, pulling the sub-region level terms toward the genic average.</div></div></div></div><ul class="rlist--inline separator download-links"><li><a target="_blank" href="//els-jbs-prod-cdn.literatumonline.com/cms/attachment/7cce8be4-cec3-43bc-a7e6-58b2b193e7a3/gr1.jpg" title="View Large Image in a new tab">View Large
                                            Image
                                        </a></li><li class="download-links__figureViewer"><a href="#" class="figure__open__ctrl">Figure Viewer</a></li><li><a href="//els-jbs-prod-cdn.literatumonline.com/cms/attachment/18aa28c5-683f-4dd9-9501-031cafb43e48/gr1_lrg.jpg" class="download-links__download-Hi-res" target="_blank" download="">Download Hi-res
                                            image
                                        </a></li><li class="hidden-xs"><a href="/action/downloadFigures?pii=S0002-9297(18)30503-2&amp;id=gr1.jpg" class="download-links__download-This">Download (PPT)</a></li></ul></div></figcaption></figure></div><div class="section-paragraph">We use a two-level linear mixed effects model with a single intercept for the genetic effect and two levels of random effects. Specifically, we model the number of missense variants within the <em>i</em><sup>th</sup> sub-region of gene <em>j</em>, <em>y</em><sub>ij</sub>, as a function of the total number of variants within the <em>i</em><sup>th</sup> sub-region of gene <em>j</em>, <em>x</em><sub>ij</sub>, through the following regression model,<div class="section-paragraph"><img data-src="//els-jbs-prod-cdn.literatumonline.com/cms/attachment/6ab57b88-51f7-4e20-82c5-8b087b83f1ae/si1.gif" alt="Math Eq" class="b-lazy" id="ufd1"></img></div>where <img data-src="//els-jbs-prod-cdn.literatumonline.com/cms/attachment/66ff9abb-4f16-42f5-8097-e0262c2ef0e2/si28.gif" alt="Math Eq" class="b-lazy" id=""></img> is a genome-wide intercept, <img data-src="//els-jbs-prod-cdn.literatumonline.com/cms/attachment/ccb3b551-8f6e-4e68-a7a3-8ea5e2448e3d/si3.gif" alt="Math Eq" class="b-lazy" id=""></img> is the average deviation from <img data-src="//els-jbs-prod-cdn.literatumonline.com/cms/attachment/83631bbb-aebf-4f0e-82e8-096f42c55566/si2.gif" alt="Math Eq" class="b-lazy" id=""></img> for gene <em>j</em>, and <img data-src="//els-jbs-prod-cdn.literatumonline.com/cms/attachment/49f7d56c-d4da-4be5-9fa9-4eda6d924b33/si4.gif" alt="Math Eq" class="b-lazy" id=""></img> is a deviation for the <em>i</em><sup>th</sup> sub-region nested within gene <em>j</em> and <img data-src="//els-jbs-prod-cdn.literatumonline.com/cms/attachment/acf0f231-e677-432f-8d2b-402eaa4e7817/si5.gif" alt="Math Eq" class="b-lazy" id=""></img></div><div class="section-paragraph">We model <img data-src="//els-jbs-prod-cdn.literatumonline.com/cms/attachment/ccb3b551-8f6e-4e68-a7a3-8ea5e2448e3d/si3.gif" alt="Math Eq" class="b-lazy" id=""></img> and <img data-src="//els-jbs-prod-cdn.literatumonline.com/cms/attachment/49f7d56c-d4da-4be5-9fa9-4eda6d924b33/si4.gif" alt="Math Eq" class="b-lazy" id=""></img> as random effects within a Bayesian framework. We assume standard normal priors for <img data-src="//els-jbs-prod-cdn.literatumonline.com/cms/attachment/f10a73b5-25fb-45c7-83c6-e3f203b35002/si6.gif" alt="Math Eq" class="b-lazy" id=""></img> butÂ allow for separate prior variances for <img data-src="//els-jbs-prod-cdn.literatumonline.com/cms/attachment/49f7d56c-d4da-4be5-9fa9-4eda6d924b33/si4.gif" alt="Math Eq" class="b-lazy" id=""></img> for each gene <em>j.</em> WeÂ begin by assuming an inverse gamma prior for the variances. Specifically, we choose <img data-src="//els-jbs-prod-cdn.literatumonline.com/cms/attachment/cb0d8f39-4987-40ce-80b1-46b4ea4ef59c/si7.gif" alt="Math Eq" class="b-lazy" id=""></img> with hyper-parameters <img data-src="//els-jbs-prod-cdn.literatumonline.com/cms/attachment/cdd0eccb-9925-4b10-89e1-ed0d77d939e4/si8.gif" alt="Math Eq" class="b-lazy" id=""></img> and <img data-src="//els-jbs-prod-cdn.literatumonline.com/cms/attachment/08fe8b6e-4e44-45b1-9a33-5a73f26d9829/si9.gif" alt="Math Eq" class="b-lazy" id=""></img>, where <img data-src="//els-jbs-prod-cdn.literatumonline.com/cms/attachment/e3304660-3b99-48bc-b351-a3b5b1be18a1/si10.gif" alt="Math Eq" class="b-lazy" id=""></img> is a small positive constant and <img data-src="//els-jbs-prod-cdn.literatumonline.com/cms/attachment/ff637214-8bac-45b7-95ea-65002e28b7bf/si11.gif" alt="Math Eq" class="b-lazy" id=""></img> is a large positive constant to induce a diffuse prior. For the sub-region parameters, we use a similar structure but with a separate variance for each gene, i.e., we choose <img data-src="//els-jbs-prod-cdn.literatumonline.com/cms/attachment/ffabb5fb-310a-4ca7-8a64-666ae2336edc/si12.gif" alt="Math Eq" class="b-lazy" id=""></img> with hyper-parameters <img data-src="//els-jbs-prod-cdn.literatumonline.com/cms/attachment/2aca0c2b-a6ac-4f08-86c6-59ac17f36633/si13.gif" alt="Math Eq" class="b-lazy" id=""></img> and <img data-src="//els-jbs-prod-cdn.literatumonline.com/cms/attachment/aa86ef02-396b-43c9-8e78-d9d2b2116390/si14.gif" alt="Math Eq" class="b-lazy" id=""></img>. Note that by allowing for a gene-level variance, the <img data-src="//els-jbs-prod-cdn.literatumonline.com/cms/attachment/d9658cd5-791a-407e-9105-284868eef626/si15.gif" alt="Math Eq" class="b-lazy" id=""></img>s can be shrunken back to the gene-level intolerance when there are no large differences between sub-region or when data is sparse. This will decrease the variability of the <img data-src="//els-jbs-prod-cdn.literatumonline.com/cms/attachment/6ffd05c3-de82-454e-924b-65206a511881/si16.gif" alt="Math Eq" class="b-lazy" id=""></img>s, leading to more stable intolerance estimation.</div><div class="section-paragraph">After initial testing the posteriors across different chains appeared to be volatile in certain settings, where the within versus across chain variance for some variables was above 1.1. To improve the ergodicity, <img data-src="//els-jbs-prod-cdn.literatumonline.com/cms/attachment/6ffd05c3-de82-454e-924b-65206a511881/si16.gif" alt="Math Eq" class="b-lazy" id=""></img> was set to zero for genes with two or fewer sub-regions, which is effectively just collapsing genes with only two sub-regions eliminating any inflated within versus across chain variance. Further, it is known that the hierarchical model canÂ beÂ augmented, sometimes referred to as noncentral<div class="dropBlock reference-citations"><sup><a href="#" id="back-bib10" data-db-target-for="bib10-d176679441e1138" class="reference-citations__ctrl" aria-expanded="false" aria-controls="bib10">10</a></sup><div data-db-target-of="bib10-d176679441e1138" class="dropBlock__holder"><div class="dropBlock__body"><ul title="list of authors" class="rlist reference-citations__loa inline-bullet-list"><li class="loa__item">Papaspiliopoulos O. </li><li class="loa__item">Roberts G.O. </li></ul><div class="reference-citations__detials">Non-centered parameterisations for hierarchical models and data augmentation.<span class="ref__series"><i>Bayesian Stat.</i> <span class="ref__seriesDate">2003; </span><span class="ref__seriesVolume">7</span></span>: <span class="ref__seriesPages">307-326</span></div><div class="reference-citations__links"><ul class="rlist inline-bullet-list"><li><a class="google-scholar" href="http://scholar.google.com/scholar_lookup?hl=en&volume=7&publication_year=2003&pages=307-326&journal=Bayesian+Stat.&author=Papaspiliopoulos+O.++and+Roberts+G.O.+&title=Non-centered+parameterisations+for+hierarchical+models+and+data+augmentation&">Google Scholar</a></li></ul></div></div><a href="#" class="dropBlock__ctrl--close" title="Close info"><i aria-hidden="true" class="icon-close_thin"></i></a></div></div><sup>,Â </sup> <div class="dropBlock reference-citations"><sup><a href="#" id="back-bib11" data-db-target-for="bib11-d176679441e1138" class="reference-citations__ctrl" aria-expanded="false" aria-controls="bib11">11</a></sup><div data-db-target-of="bib11-d176679441e1138" class="dropBlock__holder"><div class="dropBlock__body"><ul title="list of authors" class="rlist reference-citations__loa inline-bullet-list"><li class="loa__item">Papaspiliopoulos O. </li><li class="loa__item">Roberts G. </li></ul><div class="reference-citations__detials">Stability of the Gibbs sampler for Bayesian hierarchical models.<span class="ref__series"><i>Ann. Stat.</i> <span class="ref__seriesDate">2008; </span><span class="ref__seriesVolume">36</span></span>: <span class="ref__seriesPages">95-117</span></div><div class="reference-citations__links"><ul class="rlist inline-bullet-list"><li><a href="/servlet/linkout?suffix=e_1_5_1_2_11_2&amp;dbid=16&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=10.1214%2F009053607000000749&amp;cf=" onclick="newWindow(this.href);return false">Crossref</a></li>
<li><a href="/servlet/linkout?suffix=e_1_5_1_2_11_2&amp;dbid=137438953472&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=2-s2.0-50649109863&amp;cf=" onclick="newWindow(this.href);return false">Scopus (24)</a></li>
<li><a class="google-scholar" href="http://scholar.google.com/scholar_lookup?hl=en&volume=36&publication_year=2008&pages=95-117&journal=Ann.+Stat.&author=Papaspiliopoulos+O.++and+Roberts+G.+&title=Stability+of+the+Gibbs+sampler+for+Bayesian+hierarchical+models&">Google Scholar</a></li></ul></div></div><a href="#" class="dropBlock__ctrl--close" title="Close info"><i aria-hidden="true" class="icon-close_thin"></i></a></div></div><sup>,Â </sup> <div class="dropBlock reference-citations"><sup><a href="#" id="back-bib12" data-db-target-for="bib12-d176679441e1138" class="reference-citations__ctrl" aria-expanded="false" aria-controls="bib12">12</a></sup><div data-db-target-of="bib12-d176679441e1138" class="dropBlock__holder"><div class="dropBlock__body"><ul title="list of authors" class="rlist reference-citations__loa inline-bullet-list"><li class="loa__item">Betancourt M. </li></ul><div class="reference-citations__detials">A general metric for Riemannian Manifold Hamiltonian Monte Carlo.<span class="ref__series"><i>Nature.</i> <span class="ref__seriesDate">2013; </span><span class="ref__seriesVolume">9</span></span>: <span class="ref__seriesPages">327-334</span></div><div class="reference-citations__links"><ul class="rlist inline-bullet-list"><li><a class="google-scholar" href="http://scholar.google.com/scholar_lookup?hl=en&volume=9&publication_year=2013&pages=327-334&journal=Nature&author=Betancourt+M.+&title=A+general+metric+for+Riemannian+Manifold+Hamiltonian+Monte+Carlo&">Google Scholar</a></li></ul></div></div><a href="#" class="dropBlock__ctrl--close" title="Close info"><i aria-hidden="true" class="icon-close_thin"></i></a></div></div> or ancillary<div class="dropBlock reference-citations"><sup><a href="#" id="back-bib13" data-db-target-for="bib13-d176679441e1142" class="reference-citations__ctrl" aria-expanded="false" aria-controls="bib13">13</a></sup><div data-db-target-of="bib13-d176679441e1142" class="dropBlock__holder"><div class="dropBlock__body"><ul title="list of authors" class="rlist reference-citations__loa inline-bullet-list"><li class="loa__item">Yu Y. </li><li class="loa__item">Meng X.-L. </li></ul><div class="reference-citations__detials">To center or not to center: that is not the questionâAn ancillarityâsufficiency interweaving strategy (ASIS) for boosting MCMC efficiency.<span class="ref__series"><i>J.Â Comput. Graph. Stat.</i> <span class="ref__seriesDate">2011; </span><span class="ref__seriesVolume">20</span></span>: <span class="ref__seriesPages">531-570</span></div><div class="reference-citations__links"><ul class="rlist inline-bullet-list"><li><a href="/servlet/linkout?suffix=e_1_5_1_2_13_2&amp;dbid=16&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=10.1198%2Fjcgs.2011.203main&amp;cf=" onclick="newWindow(this.href);return false">Crossref</a></li>
<li><a href="/servlet/linkout?suffix=e_1_5_1_2_13_2&amp;dbid=137438953472&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=2-s2.0-80053346023&amp;cf=" onclick="newWindow(this.href);return false">Scopus (55)</a></li>
<li><a class="google-scholar" href="http://scholar.google.com/scholar_lookup?hl=en&volume=20&publication_year=2011&pages=531-570&journal=J.%C2%A0Comput.+Graph.+Stat.&author=Yu+Y.++and+Meng+X.-L.+&title=To+center+or+not+to+center%3A+that+is+not+the+question%E2%80%94An+ancillarity%E2%80%93sufficiency+interweaving+strategy+%28ASIS%29+for+boosting+MCMC+efficiency&">Google Scholar</a></li></ul></div></div><a href="#" class="dropBlock__ctrl--close" title="Close info"><i aria-hidden="true" class="icon-close_thin"></i></a></div></div> augmentation, and similar methods are known to improve performance.<div class="dropBlock reference-citations"><sup><a href="#" id="back-bib14" data-db-target-for="bib14-d176679441e1146" class="reference-citations__ctrl" aria-expanded="false" aria-controls="bib14">14</a></sup><div data-db-target-of="bib14-d176679441e1146" class="dropBlock__holder"><div class="dropBlock__body"><ul title="list of authors" class="rlist reference-citations__loa inline-bullet-list"><li class="loa__item">Duan L.L. </li><li class="loa__item">Johndrow J.E. </li><li class="loa__item">Dunson D.B. </li></ul><div class="reference-citations__detials">Scaling up data augmentation MCMC via calibration.<span class="ref__series"><i>arXiv.</i> <span class="ref__seriesDate">2017; </span><span class="ref__seriesVolume"></span></span><div class="citation-host"><a href="https://arxiv.org/abs/1703.03123" target="_blank">https://arxiv.org/abs/1703.03123</a></div></div><div class="reference-citations__links"><ul class="rlist inline-bullet-list"><li><a class="google-scholar" href="http://scholar.google.com/scholar_lookup?hl=en&publication_year=2017&journal=arXiv&author=Duan+L.L.+%2C+Johndrow+J.E.+%2C++and+Dunson+D.B.+&title=Scaling+up+data+augmentation+MCMC+via+calibration&">Google Scholar</a></li></ul></div></div><a href="#" class="dropBlock__ctrl--close" title="Close info"><i aria-hidden="true" class="icon-close_thin"></i></a></div></div><sup>,Â </sup> <div class="dropBlock reference-citations"><sup><a href="#" id="back-bib15" data-db-target-for="bib15-d176679441e1146" class="reference-citations__ctrl" aria-expanded="false" aria-controls="bib15">15</a></sup><div data-db-target-of="bib15-d176679441e1146" class="dropBlock__holder"><div class="dropBlock__body"><ul title="list of authors" class="rlist reference-citations__loa inline-bullet-list"><li class="loa__item">Liu J.S. </li></ul><div class="reference-citations__detials">The collapsed Gibbs sampler with applications to a gene regulation problem.<span class="ref__series"><i>J.Â Am. Stat. Assoc.</i> <span class="ref__seriesDate">1994; </span><span class="ref__seriesVolume">89</span></span>: <span class="ref__seriesPages">958-966</span></div><div class="reference-citations__links"><ul class="rlist inline-bullet-list"><li><a href="/servlet/linkout?suffix=e_1_5_1_2_15_2&amp;dbid=16&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=10.1080%2F01621459.1994.10476829&amp;cf=" onclick="newWindow(this.href);return false">Crossref</a></li>
<li><a class="google-scholar" href="http://scholar.google.com/scholar_lookup?hl=en&volume=89&publication_year=1994&pages=958-966&journal=J.%C2%A0Am.+Stat.+Assoc.&author=Liu+J.S.+&title=The+collapsed+Gibbs+sampler+with+applications+to+a+gene+regulation+problem&">Google Scholar</a></li></ul></div></div><a href="#" class="dropBlock__ctrl--close" title="Close info"><i aria-hidden="true" class="icon-close_thin"></i></a></div></div> So, we introduced an additional hyper parameter <img data-src="//els-jbs-prod-cdn.literatumonline.com/cms/attachment/fa888b71-e0b9-4de7-bb8b-6b9a6a709d99/si17.gif" alt="Math Eq" class="b-lazy" id=""></img> for <img data-src="//els-jbs-prod-cdn.literatumonline.com/cms/attachment/8a75093d-30e8-4b7d-bbf3-4abd5f44ae13/si18.gif" alt="Math Eq" class="b-lazy" id=""></img> to reduce autocorrelation:<div class="math-formula"><span class="equationTd"><img data-src="//els-jbs-prod-cdn.literatumonline.com/cms/attachment/6513d893-cc6c-4757-bb2f-37cc66ddd898/si19.gif" alt="Math Eq" class="b-lazy" id="fd1"></img></span><span class="formulaLabel">(Equation 1)</span></div></div><div class="section-paragraph">By introducing an auxiliary hyper parameter, the conditional variance structure is maintained while decoupling the random variables we wish to make inference on, <img data-src="//els-jbs-prod-cdn.literatumonline.com/cms/attachment/dbfadfa2-cc49-43b1-b73b-a864bb604728/si20.gif" alt="Math Eq" class="b-lazy" id=""></img>. The final score that is used for the analysis is the posterior mode of the combined genic and sub-region terms<img data-src="//els-jbs-prod-cdn.literatumonline.com/cms/attachment/06c97d75-ac1d-4b46-a05e-6341bb880973/si21.gif" alt="Math Eq" class="b-lazy" id=""></img></div><div class="section-paragraph">The software was implemented in RStan (<a href="#app3" class="scroll-into-link">Web Resources</a>) interface of the Stan statistical modeling program, specifying the model in C++. Stanâs underlying software engine is a C++ program designed to be an efficient implementation of a Hamiltonian Monte Carlo (HMC) algorithm.<div class="dropBlock reference-citations"><sup><a href="#" id="back-bib16" data-db-target-for="bib16-d176679441e1334" class="reference-citations__ctrl" aria-expanded="false" aria-controls="bib16">16</a></sup><div data-db-target-of="bib16-d176679441e1334" class="dropBlock__holder"><div class="dropBlock__body"><ul title="list of authors" class="rlist reference-citations__loa inline-bullet-list"><li class="loa__item">Duane S. </li><li class="loa__item">Kennedy A.D. </li><li class="loa__item">Pendleton B.J. </li><li class="loa__item">Roweth D. </li></ul><div class="reference-citations__detials">Hybrid Monte Carlo.<span class="ref__series"><i>Phys. Lett. B.</i> <span class="ref__seriesDate">1987; </span><span class="ref__seriesVolume">195</span></span>: <span class="ref__seriesPages">216-222</span></div><div class="reference-citations__links"><ul class="rlist inline-bullet-list"><li><a href="/servlet/linkout?suffix=e_1_5_1_2_16_2&amp;dbid=16&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=10.1016%2F0370-2693%2887%2991197-X&amp;cf=" onclick="newWindow(this.href);return false">Crossref</a></li>
<li><a href="/servlet/linkout?suffix=e_1_5_1_2_16_2&amp;dbid=137438953472&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=2-s2.0-4243137056&amp;cf=" onclick="newWindow(this.href);return false">Scopus (0)</a></li>
<li><a class="google-scholar" href="http://scholar.google.com/scholar_lookup?hl=en&volume=195&publication_year=1987&pages=216-222&journal=Phys.+Lett.+B&author=Duane+S.+%2C+Kennedy+A.D.+%2C+Pendleton+B.J.+%2C++and+Roweth+D.+&title=Hybrid+Monte+Carlo&">Google Scholar</a></li></ul></div></div><a href="#" class="dropBlock__ctrl--close" title="Close info"><i aria-hidden="true" class="icon-close_thin"></i></a></div></div> HMC uses gradient information based on Hamiltonian dynamics, borrowing from physics, to better hone in on states of higher acceptance rates. Since the proposal step is leveraging more information, it results in a faster convergence of the algorithm. A burn in of 1,000 with an additional 10,000 HMC steps across 5 chains was run for both domains and exons.</div></section><section id="sec2.3" data-locator="sec2.3"><h3 class="sectionTitle"><span class="top__sub">Â </span><span class="top__text">Assessing Prediction of Pathogenic Variants and Sensitivity to Sub-region Length</span></h3><div class="section-paragraph">After calculating the scores, we perform assessment of the methods ability to predict pathogenic variants in several different scenarios. We directly compare the LIMBR scores against sub-RVIS to better understand whether extending from a frequentist model to the more involved and computationally intensive Bayesian model improves prediction of pathogenic variance. It is also important to assess how sensitive the methods are to the relative sub-region lengths. For both sub-RVIS and LIMBR, the scores are regressed against the presence of ClinVar pathogenic mutations within each sub-region while controlling for mutation rate using logistics regression.<div class="math-formula"><span class="equationTd"><img data-src="//els-jbs-prod-cdn.literatumonline.com/cms/attachment/e1eef09b-df67-48d5-b4bd-fd7d27d41aee/si22.gif" alt="Math Eq" class="b-lazy" id="fd2"></img></span><span class="formulaLabel">(Equation 2)</span></div>where the sub-regions here considered pathogenic if they had at least one ClinVar pathogenic variant, <img data-src="//els-jbs-prod-cdn.literatumonline.com/cms/attachment/4c3f5496-ced1-4953-9892-f90bc849540d/si23.gif" alt="Math Eq" class="b-lazy" id=""></img>. The log mutation rate, <em>M</em><sub>ij</sub>, was included as a covariate and the model was fit both with and without sub-region length, <em>L</em><sub>ij</sub>, as a covariate. The change in <img data-src="//els-jbs-prod-cdn.literatumonline.com/cms/attachment/ebeba76e-4e82-40c6-99ca-1e4ccea2e7cf/si24.gif" alt="Math Eq" class="b-lazy" id=""></img> was observed fitting the models with and without the sub-region length term in the model to see the potential effects of sub-region size.</div><div class="section-paragraph">Next, again using the calculated scores, a gene-by-gene assessment is run to test per-gene how well LIMBR predicts where within each gene pathogenic mutations are most likely to fall. For each gene, the expected distribution of the pathogenic variants within the gene were calculated, based on the geneâs sub-regionsâ mutation rates. First, we regressed pathogenic variant counts within the sub-regions of genes against LIMBR scores and noted the observed slope.<div class="math-formula"><span class="equationTd"><img data-src="//els-jbs-prod-cdn.literatumonline.com/cms/attachment/a04ce57a-4dfd-462f-94bf-c7638e402e31/si25.gif" alt="Math Eq" class="b-lazy" id="fd3"></img></span><span class="formulaLabel">(Equation 3)</span></div></div><div class="section-paragraph">Regressing the model again but this time redistributing (permuting) the pathogenic variants into sub-regions by multinomial sampling, where variants are allocated to sub-regions proportional to the cumulative mutation rate of the sub-region. This effectively breaks any association between the location of the pathogenic mutations and intolerance. We repeat this process 1e4 times, characterizing the null distribution of the slope, and thereby compute p values by comparing the observed slope to permuted setsâ slopes.</div><div class="section-paragraph">Finally, we assess LIMBRâs utility in case-control studies to improve ability to detect gene-disease associations. The scores could be used as a weighting factor to increase power in disease related gene detection. A group-wise association test, similar to Madsen and Browningâs,<div class="dropBlock reference-citations"><sup><a href="#" id="back-bib17" data-db-target-for="bib17-d176679441e1571" class="reference-citations__ctrl" aria-expanded="false" aria-controls="bib17">17</a></sup><div data-db-target-of="bib17-d176679441e1571" class="dropBlock__holder"><div class="dropBlock__body"><ul title="list of authors" class="rlist reference-citations__loa inline-bullet-list"><li class="loa__item">Madsen B.E. </li><li class="loa__item">Browning S.R. </li></ul><div class="reference-citations__detials">A groupwise association test for rare mutations using a weighted sum statistic.<span class="ref__series"><i>PLoS Genet.</i> <span class="ref__seriesDate">2009; </span><span class="ref__seriesVolume">5</span></span>: <span class="ref__seriesPages">e1000384</span></div><div class="reference-citations__links"><ul class="rlist inline-bullet-list"><li><a href="/servlet/linkout?suffix=e_1_5_1_2_17_2&amp;dbid=16&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=10.1371%2Fjournal.pgen.1000384&amp;cf=" onclick="newWindow(this.href);return false">Crossref</a></li>
<li><a href="/servlet/linkout?suffix=e_1_5_1_2_17_2&amp;dbid=8&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=19214210&amp;cf=" onclick="newWindow(this.href);return false">PubMed</a></li>
<li><a href="/servlet/linkout?suffix=e_1_5_1_2_17_2&amp;dbid=137438953472&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=2-s2.0-61449168010&amp;cf=" onclick="newWindow(this.href);return false">Scopus (648)</a></li>
<li><a class="google-scholar" href="http://scholar.google.com/scholar_lookup?hl=en&volume=5&publication_year=2009&pages=e1000384&journal=PLoS+Genet.&author=Madsen+B.E.++and+Browning+S.R.+&title=A+groupwise+association+test+for+rare+mutations+using+a+weighted+sum+statistic&">Google Scholar</a></li></ul></div></div><a href="#" class="dropBlock__ctrl--close" title="Close info"><i aria-hidden="true" class="icon-close_thin"></i></a></div></div> was constructed using the intolerance scores to weight the variants. An additive genetic model was used, where a genotype value of 0 corresponds to homozygous for the major allele, 1Â to heterozygous, and 2 to homozygous for the minor allele. Assuming Hardy-Weinberg equilibrium, missing genotypes were imputed assigning each missing genotype a value of <em>2p</em>, where <em>p</em> is the frequency of the minor allele in the cohort. For a gene with <em>n</em>Â exons, let <em>s</em><sub>i</sub> be the intolerance score percentile of exon <em>i</em>. For all variants in exon <em>i</em>, the corresponding weight <em>w</em><sub>i</sub><img data-src="//els-jbs-prod-cdn.literatumonline.com/cms/attachment/c4886a82-3991-4d91-bf03-860aaca6be51/si26.gif" alt="Math Eq" class="b-lazy" id=""></img>is given by:<div class="math-formula"><span class="equationTd"><img data-src="//els-jbs-prod-cdn.literatumonline.com/cms/attachment/889e621e-7704-4d45-bb83-6815678569f1/si27.gif" alt="Math Eq" class="b-lazy" id="fd4"></img></span><span class="formulaLabel">(Equation 4)</span></div></div><div class="section-paragraph">For the weighted model, each individualâs genotype vector was multiplied by the weight vector and 1 for the unweighted model. A weight of 1 was given to all LoF variants since the intolerance score applies only to missense variants and LoF variants are expected to have the greatest relative effect on protein function.</div><div class="section-paragraph">The sum across each case individualâs weighted genotype vector yields an individual-level score, which is then summed to get the cohort-wide score. Then, 10 million permutations were generated of the score, where each time the case-controlÂ status of each individual was permuted. The permutation p value corresponds to the proportion of permutations that wereÂ larger than the observed score with the true case-control labels.</div></section><section id="sec2.2" data-locator="sec2.2"><h3 class="sectionTitle"><span class="top__sub">Â </span><span class="top__text">Data and Quality Control</span></h3><div class="section-paragraph">The model is fit using 123,136 whole-exome sequences from the genome Aggregation Database (gnomAD) across 17,765 genes.<div class="dropBlock reference-citations"><sup><a href="#" id="back-bib18" data-db-target-for="bib18-d176679441e1666" class="reference-citations__ctrl" aria-expanded="false" aria-controls="bib18">18</a></sup><div data-db-target-of="bib18-d176679441e1666" class="dropBlock__holder"><div class="dropBlock__body"><ul title="list of authors" class="rlist reference-citations__loa inline-bullet-list"><li class="loa__item">Lek M. </li><li class="loa__item">Karczewski K.J. </li><li class="loa__item">Minikel E.V. </li><li class="loa__item">Samocha K.E. </li><li class="loa__item">Banks E. </li><li class="loa__item">Fennell T. </li><li class="loa__item">OâDonnell-Luria A.H. </li><li class="loa__item">Ware J.S. </li><li class="loa__item">Hill A.J. </li><li class="loa__item">Cummings B.B. </li><li>et al.</li></ul><div class="reference-citations__detials">M.LekK.J.KarczewskiE.V.MinikelK.E.SamochaE.BanksT.FennellA.H.OâDonnell-LuriaJ.S.WareA.J.HillB.B.Cummings<span>Exome Aggregation Consortium</span>Analysis of protein-coding genetic variation in 60,706 humans.<span class="ref__series"><i>Nature.</i> <span class="ref__seriesDate">2016; </span><span class="ref__seriesVolume">536</span></span>: <span class="ref__seriesPages">285-291</span></div><div class="reference-citations__links"><ul class="rlist inline-bullet-list"><li><a href="/servlet/linkout?suffix=e_1_5_1_2_18_2&amp;dbid=16&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=10.1038%2Fnature19057&amp;cf=" onclick="newWindow(this.href);return false">Crossref</a></li>
<li><a href="/servlet/linkout?suffix=e_1_5_1_2_18_2&amp;dbid=8&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=27535533&amp;cf=" onclick="newWindow(this.href);return false">PubMed</a></li>
<li><a class="google-scholar" href="http://scholar.google.com/scholar_lookup?hl=en&volume=536&publication_year=2016&pages=285-291&journal=Nature&author=Exome+Aggregation+Consortium&title=Analysis+of+protein-coding+genetic+variation+in+60%2C706+humans&">Google Scholar</a></li></ul></div></div><a href="#" class="dropBlock__ctrl--close" title="Close info"><i aria-hidden="true" class="icon-close_thin"></i></a></div></div> LoF variants are excluded from this analysis since LoFs can damage the resulting protein regardless of the sub-region in which they fall.<div class="dropBlock reference-citations"><sup><a href="#" id="back-bib5" data-db-target-for="bib5-d176679441e1670" class="reference-citations__ctrl" aria-expanded="false" aria-controls="bib5">5</a></sup><div data-db-target-of="bib5-d176679441e1670" class="dropBlock__holder"><div class="dropBlock__body"><ul title="list of authors" class="rlist reference-citations__loa inline-bullet-list"><li class="loa__item">Samocha K.E. </li><li class="loa__item">Robinson E.B. </li><li class="loa__item">Sanders S.J. </li><li class="loa__item">Stevens C. </li><li class="loa__item">Sabo A. </li><li class="loa__item">McGrath L.M. </li><li class="loa__item">Kosmicki J.A. </li><li class="loa__item">RehnstrÃ¶m K. </li><li class="loa__item">Mallick S. </li><li class="loa__item">Kirby A. </li><li>et al.</li></ul><div class="reference-citations__detials">A framework for the interpretation of de novo mutation in human disease.<span class="ref__series"><i>Nat. Genet.</i> <span class="ref__seriesDate">2014; </span><span class="ref__seriesVolume">46</span></span>: <span class="ref__seriesPages">944-950</span></div><div class="reference-citations__links"><ul class="rlist inline-bullet-list"><li><a href="/servlet/linkout?suffix=e_1_5_1_2_5_2&amp;dbid=16&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=10.1038%2Fng.3050&amp;cf=" onclick="newWindow(this.href);return false">Crossref</a></li>
<li><a href="/servlet/linkout?suffix=e_1_5_1_2_5_2&amp;dbid=8&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=25086666&amp;cf=" onclick="newWindow(this.href);return false">PubMed</a></li>
<li><a href="/servlet/linkout?suffix=e_1_5_1_2_5_2&amp;dbid=137438953472&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=2-s2.0-84905309176&amp;cf=" onclick="newWindow(this.href);return false">Scopus (322)</a></li>
<li><a class="google-scholar" href="http://scholar.google.com/scholar_lookup?hl=en&volume=46&publication_year=2014&pages=944-950&journal=Nat.+Genet.&author=Samocha+K.E.+%2C+Robinson+E.B.+%2C+Sanders+S.J.+%2C+Stevens+C.+%2C+Sabo+A.+%2C+McGrath+L.M.+%2C+Kosmicki+J.A.+%2C+Rehnstr%C3%B6m+K.+%2C+Mallick+S.+%2C+Kirby+A.++et+al.&title=A+framework+for+the+interpretation+of+de+novo+mutation+in+human+disease&">Google Scholar</a></li></ul></div></div><a href="#" class="dropBlock__ctrl--close" title="Close info"><i aria-hidden="true" class="icon-close_thin"></i></a></div></div><sup>,Â </sup> <div class="dropBlock reference-citations"><sup><a href="#" id="back-bib19" data-db-target-for="bib19-d176679441e1670" class="reference-citations__ctrl" aria-expanded="false" aria-controls="bib19">19</a></sup><div data-db-target-of="bib19-d176679441e1670" class="dropBlock__holder"><div class="dropBlock__body"><ul title="list of authors" class="rlist reference-citations__loa inline-bullet-list"><li class="loa__item">Traynelis J. </li><li class="loa__item">Silk M. </li><li class="loa__item">Wang Q. </li><li class="loa__item">Berkovic S.F. </li><li class="loa__item">Liu L. </li><li class="loa__item">Ascher D.B. </li><li class="loa__item">Balding D.J. </li><li class="loa__item">Petrovski S. </li></ul><div class="reference-citations__detials">Optimizing genomic medicine in epilepsy through a gene-customized approach to missense variant interpretation.<span class="ref__series"><i>Genome Res.</i> <span class="ref__seriesDate">2017; </span><span class="ref__seriesVolume">27</span></span>: <span class="ref__seriesPages">1715-1729</span></div><div class="reference-citations__links"><ul class="rlist inline-bullet-list"><li><a href="/servlet/linkout?suffix=e_1_5_1_2_19_2&amp;dbid=16&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=10.1101%2Fgr.226589.117&amp;cf=" onclick="newWindow(this.href);return false">Crossref</a></li>
<li><a href="/servlet/linkout?suffix=e_1_5_1_2_19_2&amp;dbid=8&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=28864458&amp;cf=" onclick="newWindow(this.href);return false">PubMed</a></li>
<li><a href="/servlet/linkout?suffix=e_1_5_1_2_19_2&amp;dbid=137438953472&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=2-s2.0-85030645609&amp;cf=" onclick="newWindow(this.href);return false">Scopus (9)</a></li>
<li><a class="google-scholar" href="http://scholar.google.com/scholar_lookup?hl=en&volume=27&publication_year=2017&pages=1715-1729&journal=Genome+Res.&author=Traynelis+J.+%2C+Silk+M.+%2C+Wang+Q.+%2C+Berkovic+S.F.+%2C+Liu+L.+%2C+Ascher+D.B.+%2C+Balding+D.J.+%2C++and+Petrovski+S.+&title=Optimizing+genomic+medicine+in+epilepsy+through+a+gene-customized+approach+to+missense+variant+interpretation&">Google Scholar</a></li></ul></div></div><a href="#" class="dropBlock__ctrl--close" title="Close info"><i aria-hidden="true" class="icon-close_thin"></i></a></div></div><sup>,Â </sup> <div class="dropBlock reference-citations"><sup><a href="#" id="back-bib20" data-db-target-for="bib20-d176679441e1670" class="reference-citations__ctrl" aria-expanded="false" aria-controls="bib20">20</a></sup><div data-db-target-of="bib20-d176679441e1670" class="dropBlock__holder"><div class="dropBlock__body"><ul title="list of authors" class="rlist reference-citations__loa inline-bullet-list"><li class="loa__item">Sivley R.M. </li><li class="loa__item">Dou X. </li><li class="loa__item">Meiler J. </li><li class="loa__item">Bush W.S. </li><li class="loa__item">Capra J.A. </li><li class="loa__item">Blackwell T.S. </li><li class="loa__item">Phillips J.A. </li><li class="loa__item">Meiler J. </li><li class="loa__item">Bush W.S. </li><li class="loa__item">Capra J.A. </li></ul><div class="reference-citations__detials">Comprehensive analysis of constraint on the spatial distribution of missense variants in human protein structures.<span class="ref__series"><i>Am. J. Hum. Genet.</i> <span class="ref__seriesDate">2018; </span><span class="ref__seriesVolume">102</span></span>: <span class="ref__seriesPages">415-426</span></div><div class="reference-citations__links"><ul class="rlist inline-bullet-list"><li><a href="/servlet/linkout?suffix=e_1_5_1_2_20_2&amp;dbid=4&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=10.1016%2Fj.ajhg.2018.01.017&amp;cf=abstract&amp;site=cell-site" onclick="newWindow(this.href);return false">Abstract</a></li>
<li><a href="/servlet/linkout?suffix=e_1_5_1_2_20_2&amp;dbid=4&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=10.1016%2Fj.ajhg.2018.01.017&amp;cf=fulltext&amp;site=cell-site" onclick="newWindow(this.href);return false">Full Text</a></li>
<li><a href="/servlet/linkout?suffix=e_1_5_1_2_20_2&amp;dbid=4&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=10.1016%2Fj.ajhg.2018.01.017&amp;cf=pdf&amp;site=cell-site" onclick="newWindow(this.href);return false">Full Text PDF</a></li>
<li><a href="/servlet/linkout?suffix=e_1_5_1_2_20_2&amp;dbid=8&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=29455857&amp;cf=" onclick="newWindow(this.href);return false">PubMed</a></li>
<li><a href="/servlet/linkout?suffix=e_1_5_1_2_20_2&amp;dbid=137438953472&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=2-s2.0-85042045525&amp;cf=" onclick="newWindow(this.href);return false">Scopus (0)</a></li>
<li><a class="google-scholar" href="http://scholar.google.com/scholar_lookup?hl=en&volume=102&publication_year=2018&pages=415-426&journal=Am.+J.+Hum.+Genet.&author=Sivley+R.M.+%2C+Dou+X.+%2C+Meiler+J.+%2C+Bush+W.S.+%2C+Capra+J.A.+%2C+Blackwell+T.S.+%2C+Phillips+J.A.+%2C+Meiler+J.+%2C+Bush+W.S.+%2C++and+Capra+J.A.+&title=Comprehensive+analysis+of+constraint+on+the+spatial+distribution+of+missense+variants+in+human+protein+structures&">Google Scholar</a></li></ul></div></div><a href="#" class="dropBlock__ctrl--close" title="Close info"><i aria-hidden="true" class="icon-close_thin"></i></a></div></div> Two sets of scores are calculated by fitting different definitions of sub-regions across genes, once with genic sub-regions defined by 172,123 exon boundaries and then againÂ with sub-regions defined by 82,265 functional domains (<a href="/cms/10.1016/j.ajhg.2018.12.020/attachment/8d42c908-07db-4f91-995c-87a0307e0d79/mmc2" class="scroll-into-link">TablesÂ S1</a> and <a href="/cms/10.1016/j.ajhg.2018.12.020/attachment/04592b59-2b21-44fc-aa10-05228ba69e55/mmc3" class="scroll-into-link">S2</a>), both using the Conserved Domain Database (CDD).<div class="dropBlock reference-citations"><sup><a href="#" id="back-bib7" data-db-target-for="bib7-d176679441e1680" class="reference-citations__ctrl" aria-expanded="false" aria-controls="bib7">7</a></sup><div data-db-target-of="bib7-d176679441e1680" class="dropBlock__holder"><div class="dropBlock__body"><ul title="list of authors" class="rlist reference-citations__loa inline-bullet-list"><li class="loa__item">Gussow A.B. </li><li class="loa__item">Petrovski S. </li><li class="loa__item">Wang Q. </li><li class="loa__item">Allen A.S. </li><li class="loa__item">Goldstein D.B. </li></ul><div class="reference-citations__detials">The intolerance to functional genetic variation of protein domains predicts the localization of pathogenic mutations within genes.<span class="ref__series"><i>Genome Biol.</i> <span class="ref__seriesDate">2016; </span><span class="ref__seriesVolume">17</span></span>: <span class="ref__seriesPages">9</span></div><div class="reference-citations__links"><ul class="rlist inline-bullet-list"><li><a href="/servlet/linkout?suffix=e_1_5_1_2_7_2&amp;dbid=16&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=10.1186%2Fs13059-016-0869-4&amp;cf=" onclick="newWindow(this.href);return false">Crossref</a></li>
<li><a href="/servlet/linkout?suffix=e_1_5_1_2_7_2&amp;dbid=8&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=26781712&amp;cf=" onclick="newWindow(this.href);return false">PubMed</a></li>
<li><a class="google-scholar" href="http://scholar.google.com/scholar_lookup?hl=en&volume=17&publication_year=2016&pages=9&journal=Genome+Biol.&author=Gussow+A.B.+%2C+Petrovski+S.+%2C+Wang+Q.+%2C+Allen+A.S.+%2C++and+Goldstein+D.B.+&title=The+intolerance+to+functional+genetic+variation+of+protein+domains+predicts+the+localization+of+pathogenic+mutations+within+genes&">Google Scholar</a></li></ul></div></div><a href="#" class="dropBlock__ctrl--close" title="Close info"><i aria-hidden="true" class="icon-close_thin"></i></a></div></div><sup>,Â </sup> <div class="dropBlock reference-citations"><sup><a href="#" id="back-bib21" data-db-target-for="bib21-d176679441e1680" class="reference-citations__ctrl" aria-expanded="false" aria-controls="bib21">21</a></sup><div data-db-target-of="bib21-d176679441e1680" class="dropBlock__holder"><div class="dropBlock__body"><ul title="list of authors" class="rlist reference-citations__loa inline-bullet-list"><li class="loa__item">Marchler-Bauer A. </li><li class="loa__item">Zheng C. </li><li class="loa__item">Chitsaz F. </li><li class="loa__item">Derbyshire M.K. </li><li class="loa__item">Geer L.Y. </li><li class="loa__item">Geer R.C. </li><li class="loa__item">Gonzales N.R. </li><li class="loa__item">Gwadz M. </li><li class="loa__item">Hurwitz D.I. </li><li class="loa__item">Lanczycki C.J. </li><li>et al.</li></ul><div class="reference-citations__detials">CDD: conserved domains and protein three-dimensional structure.<span class="ref__series"><i>Nucleic Acids Res.</i> <span class="ref__seriesDate">2013; </span><span class="ref__seriesVolume">41</span></span>: <span class="ref__seriesPages">D348-D352</span></div><div class="reference-citations__links"><ul class="rlist inline-bullet-list"><li><a href="/servlet/linkout?suffix=e_1_5_1_2_21_2&amp;dbid=16&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=10.1093%2Fnar%2Fgks1243&amp;cf=" onclick="newWindow(this.href);return false">Crossref</a></li>
<li><a href="/servlet/linkout?suffix=e_1_5_1_2_21_2&amp;dbid=8&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=23197659&amp;cf=" onclick="newWindow(this.href);return false">PubMed</a></li>
<li><a href="/servlet/linkout?suffix=e_1_5_1_2_21_2&amp;dbid=137438953472&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=2-s2.0-84875445608&amp;cf=" onclick="newWindow(this.href);return false">Scopus (549)</a></li>
<li><a class="google-scholar" href="http://scholar.google.com/scholar_lookup?hl=en&volume=41&publication_year=2013&pages=D348-D352&journal=Nucleic+Acids+Res.&author=Marchler-Bauer+A.+%2C+Zheng+C.+%2C+Chitsaz+F.+%2C+Derbyshire+M.K.+%2C+Geer+L.Y.+%2C+Geer+R.C.+%2C+Gonzales+N.R.+%2C+Gwadz+M.+%2C+Hurwitz+D.I.+%2C+Lanczycki+C.J.++et+al.&title=CDD%3A+conserved+domains+and+protein+three-dimensional+structure&">Google Scholar</a></li></ul></div></div><a href="#" class="dropBlock__ctrl--close" title="Close info"><i aria-hidden="true" class="icon-close_thin"></i></a></div></div> The scores have also been added to our in-house software Analysis Tool for Annotated Variants (ATAV) which communicates with our in-house relational database along with other resources (<a href="https://redmine.igm.cumc.columbia.edu/projects/atav" target="_blank">https://redmine.igm.cumc.columbia.edu/projects/atav</a>). The filtered data from gnomAD first had to go through âPASSâ gnomAD criteria and further restricted to regions with at least 10Ã coverage in at least 70% of the samples. The model was regressed on missense variation versus all classes of functionality. Variants causing loss-of-function damage the resulting protein, regardless of the sub-region, so we removed those predicted to be LoF. Variants that were annotated as either splice acceptor variant, splice donor variant, stop gained, or stopÂ lost were considered LoF. Indels were ignored. Exons were defined based on consensus coding sequence project release 20 (CCDS20).<div class="dropBlock reference-citations"><sup><a href="#" id="back-bib22" data-db-target-for="bib22-d176679441e1688" class="reference-citations__ctrl" aria-expanded="false" aria-controls="bib22">22</a></sup><div data-db-target-of="bib22-d176679441e1688" class="dropBlock__holder"><div class="dropBlock__body"><ul title="list of authors" class="rlist reference-citations__loa inline-bullet-list"><li class="loa__item">Pruitt K.D. </li><li class="loa__item">Harrow J. </li><li class="loa__item">Harte R.A. </li><li class="loa__item">Wallin C. </li><li class="loa__item">Diekhans M. </li><li class="loa__item">Maglott D.R. </li><li class="loa__item">Searle S. </li><li class="loa__item">Farrell C.M. </li><li class="loa__item">Loveland J.E. </li><li class="loa__item">Ruef B.J. </li><li>et al.</li></ul><div class="reference-citations__detials">The consensus coding sequence (CCDS) project: Identifying a common protein-coding gene set for the human and mouse genomes.<span class="ref__series"><i>Genome Res.</i> <span class="ref__seriesDate">2009; </span><span class="ref__seriesVolume">19</span></span>: <span class="ref__seriesPages">1316-1323</span></div><div class="reference-citations__links"><ul class="rlist inline-bullet-list"><li><a href="/servlet/linkout?suffix=e_1_5_1_2_22_2&amp;dbid=16&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=10.1101%2Fgr.080531.108&amp;cf=" onclick="newWindow(this.href);return false">Crossref</a></li>
<li><a href="/servlet/linkout?suffix=e_1_5_1_2_22_2&amp;dbid=8&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=19498102&amp;cf=" onclick="newWindow(this.href);return false">PubMed</a></li>
<li><a href="/servlet/linkout?suffix=e_1_5_1_2_22_2&amp;dbid=137438953472&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=2-s2.0-67650064594&amp;cf=" onclick="newWindow(this.href);return false">Scopus (326)</a></li>
<li><a class="google-scholar" href="http://scholar.google.com/scholar_lookup?hl=en&volume=19&publication_year=2009&pages=1316-1323&journal=Genome+Res.&author=Pruitt+K.D.+%2C+Harrow+J.+%2C+Harte+R.A.+%2C+Wallin+C.+%2C+Diekhans+M.+%2C+Maglott+D.R.+%2C+Searle+S.+%2C+Farrell+C.M.+%2C+Loveland+J.E.+%2C+Ruef+B.J.++et+al.&title=The+consensus+coding+sequence+%28CCDS%29+project%3A+Identifying+a+common+protein-coding+gene+set+for+the+human+and+mouse+genomes&">Google Scholar</a></li></ul></div></div><a href="#" class="dropBlock__ctrl--close" title="Close info"><i aria-hidden="true" class="icon-close_thin"></i></a></div></div> All coding regions of a gene were considered across all possible CCDS transcripts of a gene. Domains were defined inÂ the Conserved Domain Database (CDD) and aligned to CCDS15.<div class="dropBlock reference-citations"><sup><a href="#" id="back-bib7" data-db-target-for="bib7-d176679441e1692" class="reference-citations__ctrl" aria-expanded="false" aria-controls="bib7">7</a></sup><div data-db-target-of="bib7-d176679441e1692" class="dropBlock__holder"><div class="dropBlock__body"><ul title="list of authors" class="rlist reference-citations__loa inline-bullet-list"><li class="loa__item">Gussow A.B. </li><li class="loa__item">Petrovski S. </li><li class="loa__item">Wang Q. </li><li class="loa__item">Allen A.S. </li><li class="loa__item">Goldstein D.B. </li></ul><div class="reference-citations__detials">The intolerance to functional genetic variation of protein domains predicts the localization of pathogenic mutations within genes.<span class="ref__series"><i>Genome Biol.</i> <span class="ref__seriesDate">2016; </span><span class="ref__seriesVolume">17</span></span>: <span class="ref__seriesPages">9</span></div><div class="reference-citations__links"><ul class="rlist inline-bullet-list"><li><a href="/servlet/linkout?suffix=e_1_5_1_2_7_2&amp;dbid=16&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=10.1186%2Fs13059-016-0869-4&amp;cf=" onclick="newWindow(this.href);return false">Crossref</a></li>
<li><a href="/servlet/linkout?suffix=e_1_5_1_2_7_2&amp;dbid=8&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=26781712&amp;cf=" onclick="newWindow(this.href);return false">PubMed</a></li>
<li><a class="google-scholar" href="http://scholar.google.com/scholar_lookup?hl=en&volume=17&publication_year=2016&pages=9&journal=Genome+Biol.&author=Gussow+A.B.+%2C+Petrovski+S.+%2C+Wang+Q.+%2C+Allen+A.S.+%2C++and+Goldstein+D.B.+&title=The+intolerance+to+functional+genetic+variation+of+protein+domains+predicts+the+localization+of+pathogenic+mutations+within+genes&">Google Scholar</a></li></ul></div></div><a href="#" class="dropBlock__ctrl--close" title="Close info"><i aria-hidden="true" class="icon-close_thin"></i></a></div></div><sup>,Â </sup> <div class="dropBlock reference-citations"><sup><a href="#" id="back-bib22" data-db-target-for="bib22-d176679441e1692" class="reference-citations__ctrl" aria-expanded="false" aria-controls="bib22">22</a></sup><div data-db-target-of="bib22-d176679441e1692" class="dropBlock__holder"><div class="dropBlock__body"><ul title="list of authors" class="rlist reference-citations__loa inline-bullet-list"><li class="loa__item">Pruitt K.D. </li><li class="loa__item">Harrow J. </li><li class="loa__item">Harte R.A. </li><li class="loa__item">Wallin C. </li><li class="loa__item">Diekhans M. </li><li class="loa__item">Maglott D.R. </li><li class="loa__item">Searle S. </li><li class="loa__item">Farrell C.M. </li><li class="loa__item">Loveland J.E. </li><li class="loa__item">Ruef B.J. </li><li>et al.</li></ul><div class="reference-citations__detials">The consensus coding sequence (CCDS) project: Identifying a common protein-coding gene set for the human and mouse genomes.<span class="ref__series"><i>Genome Res.</i> <span class="ref__seriesDate">2009; </span><span class="ref__seriesVolume">19</span></span>: <span class="ref__seriesPages">1316-1323</span></div><div class="reference-citations__links"><ul class="rlist inline-bullet-list"><li><a href="/servlet/linkout?suffix=e_1_5_1_2_22_2&amp;dbid=16&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=10.1101%2Fgr.080531.108&amp;cf=" onclick="newWindow(this.href);return false">Crossref</a></li>
<li><a href="/servlet/linkout?suffix=e_1_5_1_2_22_2&amp;dbid=8&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=19498102&amp;cf=" onclick="newWindow(this.href);return false">PubMed</a></li>
<li><a href="/servlet/linkout?suffix=e_1_5_1_2_22_2&amp;dbid=137438953472&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=2-s2.0-67650064594&amp;cf=" onclick="newWindow(this.href);return false">Scopus (326)</a></li>
<li><a class="google-scholar" href="http://scholar.google.com/scholar_lookup?hl=en&volume=19&publication_year=2009&pages=1316-1323&journal=Genome+Res.&author=Pruitt+K.D.+%2C+Harrow+J.+%2C+Harte+R.A.+%2C+Wallin+C.+%2C+Diekhans+M.+%2C+Maglott+D.R.+%2C+Searle+S.+%2C+Farrell+C.M.+%2C+Loveland+J.E.+%2C+Ruef+B.J.++et+al.&title=The+consensus+coding+sequence+%28CCDS%29+project%3A+Identifying+a+common+protein-coding+gene+set+for+the+human+and+mouse+genomes&">Google Scholar</a></li></ul></div></div><a href="#" class="dropBlock__ctrl--close" title="Close info"><i aria-hidden="true" class="icon-close_thin"></i></a></div></div> ClinVar<div class="dropBlock reference-citations"><sup><a href="#" id="back-bib23" data-db-target-for="bib23-d176679441e1696" class="reference-citations__ctrl" aria-expanded="false" aria-controls="bib23">23</a></sup><div data-db-target-of="bib23-d176679441e1696" class="dropBlock__holder"><div class="dropBlock__body"><ul title="list of authors" class="rlist reference-citations__loa inline-bullet-list"><li class="loa__item">Landrum M.J. </li><li class="loa__item">Lee J.M. </li><li class="loa__item">Riley G.R. </li><li class="loa__item">Jang W. </li><li class="loa__item">Rubinstein W.S. </li><li class="loa__item">Church D.M. </li><li class="loa__item">Maglott D.R. </li></ul><div class="reference-citations__detials">ClinVar: public archive of relationships among sequence variation and human phenotype.<span class="ref__series"><i>Nucleic Acids Res.</i> <span class="ref__seriesDate">2014; </span><span class="ref__seriesVolume">42</span></span>: <span class="ref__seriesPages">D980-D985</span></div><div class="reference-citations__links"><ul class="rlist inline-bullet-list"><li><a href="/servlet/linkout?suffix=e_1_5_1_2_23_2&amp;dbid=16&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=10.1093%2Fnar%2Fgkt1113&amp;cf=" onclick="newWindow(this.href);return false">Crossref</a></li>
<li><a href="/servlet/linkout?suffix=e_1_5_1_2_23_2&amp;dbid=8&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=24234437&amp;cf=" onclick="newWindow(this.href);return false">PubMed</a></li>
<li><a href="/servlet/linkout?suffix=e_1_5_1_2_23_2&amp;dbid=137438953472&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=2-s2.0-84891809093&amp;cf=" onclick="newWindow(this.href);return false">Scopus (707)</a></li>
<li><a class="google-scholar" href="http://scholar.google.com/scholar_lookup?hl=en&volume=42&publication_year=2014&pages=D980-D985&journal=Nucleic+Acids+Res.&author=Landrum+M.J.+%2C+Lee+J.M.+%2C+Riley+G.R.+%2C+Jang+W.+%2C+Rubinstein+W.S.+%2C+Church+D.M.+%2C++and+Maglott+D.R.+&title=ClinVar%3A+public+archive+of+relationships+among+sequence+variation+and+human+phenotype&">Google Scholar</a></li></ul></div></div><a href="#" class="dropBlock__ctrl--close" title="Close info"><i aria-hidden="true" class="icon-close_thin"></i></a></div></div> pathogenic variants that were labeled asÂ either pathogenic or likely pathogenic were used for the validation analysis and control variants were taken from 50,726 DiscovEHR<div class="dropBlock reference-citations"><sup><a href="#" id="back-bib24" data-db-target-for="bib24-d176679441e1700" class="reference-citations__ctrl" aria-expanded="false" aria-controls="bib24">24</a></sup><div data-db-target-of="bib24-d176679441e1700" class="dropBlock__holder"><div class="dropBlock__body"><ul title="list of authors" class="rlist reference-citations__loa inline-bullet-list"><li class="loa__item">Dewey F.E. </li><li class="loa__item">Murray M.F. </li><li class="loa__item">Overton J.D. </li><li class="loa__item">Habegger L. </li><li class="loa__item">Leader J.B. </li><li class="loa__item">Fetterolf S.N. </li><li class="loa__item">OâDushlaine C. </li><li class="loa__item">Van Hout C.V. </li><li class="loa__item">Staples J. </li><li class="loa__item">Gonzaga-Jauregui C. </li><li>et al.</li></ul><div class="reference-citations__detials">Distribution and clinical impact of functional variants in 50,726 whole-exome sequences from the DiscovEHR study.<span class="ref__series"><i>Science.</i> <span class="ref__seriesDate">2016; </span><span class="ref__seriesVolume">354</span></span></div><div class="reference-citations__links"><ul class="rlist inline-bullet-list"><li><a href="/servlet/linkout?suffix=e_1_5_1_2_24_2&amp;dbid=16&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=10.1126%2Fscience.aaf6814&amp;cf=" onclick="newWindow(this.href);return false">Crossref</a></li>
<li><a href="/servlet/linkout?suffix=e_1_5_1_2_24_2&amp;dbid=137438953472&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=2-s2.0-85007028930&amp;cf=" onclick="newWindow(this.href);return false">Scopus (87)</a></li>
<li><a class="google-scholar" href="http://scholar.google.com/scholar_lookup?hl=en&volume=354&publication_year=2016&journal=Science&author=Dewey+F.E.+%2C+Murray+M.F.+%2C+Overton+J.D.+%2C+Habegger+L.+%2C+Leader+J.B.+%2C+Fetterolf+S.N.+%2C+O%E2%80%99Dushlaine+C.+%2C+Van+Hout+C.V.+%2C+Staples+J.+%2C+Gonzaga-Jauregui+C.++et+al.&title=Distribution+and+clinical+impact+of+functional+variants+in+50%2C726+whole-exome+sequences+from+the+DiscovEHR+study&">Google Scholar</a></li></ul></div></div><a href="#" class="dropBlock__ctrl--close" title="Close info"><i aria-hidden="true" class="icon-close_thin"></i></a></div></div> samples, excluding variants also found in gnomAD and ClinVar.</div></section></section><section id="sec3" data-locator="sec3"><h2 class="top" id="secsectitle0040"><span class="top__sub"></span><span class="top__text">Results</span></h2><section id="sec3.1" data-locator="sec3.1"><h3 class="sectionTitle"><span class="top__sub">Â </span><span class="top__text">Exome-wide Predictive Ability of Intolerance Scores on Pathogenic Mutations</span></h3><div class="section-paragraph">To better understand the advantage of the hierarchical approach, we examine the top and bottom 10% of the intolerance scores from both LIMBR and sub-RVIS and the number of bases spanned by these regions (<a data-link="modal" data-target="#image-S0002-9297(18)30503-2gr2" href="#fig2" data-locator="fig2" id="back-gr2" class="figure-link scroll-into-link">FigureÂ 2</a>). For exon-level sub-RVIS, the average bases spanned forÂ the top 10% intolerant exons is 452, compared to anÂ average of 214 bases for the bottom 10% (<a data-link="modal" data-target="#image-S0002-9297(18)30503-2gr2" href="#fig2" data-locator="fig2" id="back-gr2" class="figure-link scroll-into-link">FigureÂ 2</a>A: t<sub>df = 25397</sub> = 44.01 and p &lt; 2.2eâ16). For domains, the top intolerant domains average 1,124 bases versus 559 bases for the most tolerant regions (<a data-link="modal" data-target="#image-S0002-9297(18)30503-2gr2" href="#fig2" data-locator="fig2" id="back-gr2" class="figure-link scroll-into-link">FigureÂ 2</a>C: t<sub>df = 25397</sub> = 40.9 and pÂ &lt;Â 2.2eâ16). Using LIMBR, the top 10% intolerant exons average 297 bases versus 286 bases for theÂ most tolerant exons (<a data-link="modal" data-target="#image-S0002-9297(18)30503-2gr2" href="#fig2" data-locator="fig2" id="back-gr2" class="figure-link scroll-into-link">FigureÂ 2</a>B: t<sub>df = 33322</sub> = 2.2, pÂ valueÂ = 0.034) and 717 bases versus 654 bases for domains (<a data-link="modal" data-target="#image-S0002-9297(18)30503-2gr2" href="#fig2" data-locator="fig2" id="back-gr2" class="figure-link scroll-into-link">FigureÂ 2</a>D: t<sub>df = 16145</sub> = 4.4, p value = 1.3eâ5). LIMBR appears considerably less sensitive to sub-regional length; there is around a 10-base difference on average between the most and least intolerant exons, which is only marginally significant at an Î± = 0.05. In contrast, exon-level sub-RVIS has a highly significant difference of more than 200 bases spanned between the most and least intolerant exons. As shown in separation of the sub-region length, histograms for top 10% and bottom 10% scores points to a potential association between longer regions and higher intolerance scores in sub-RVIS.</div><div class="floatDisplay"><figure id="fig2" class=""><img class="b-lazy" data-src="//els-jbs-prod-cdn.literatumonline.com/cms/attachment/91873dec-c81d-4b60-a398-8d094d3472d1/gr2.jpg" alt="Figure thumbnail gr2" data-locator="fig2" data-high-res="//els-jbs-prod-cdn.literatumonline.com/cms/attachment/fd6bd781-8c8e-4ba9-89fb-c88c1f6fdb50/gr2_lrg.jpg" data-large-image="//els-jbs-prod-cdn.literatumonline.com/cms/attachment/91873dec-c81d-4b60-a398-8d094d3472d1/gr2.jpg"></img><figcaption><div class="figure__caption__body"><div class="figure__title"><span data-fv-label="Fig. 2" class="label figure__label">FigureÂ 2</span><span class="figure__title__text">Comparing Sub-RVIS versus LIMBR Length Distributions Looking at the Top and Bottom 10% Intolerant Sub-regions</span></div><div class="accordion"><a href="#" class="accordion__control" aria-expanded="false"><i class="icon-section_arrow_d"></i><span>Show full caption</span></a><div class="accordion__content"><div class="figure__caption__text"><div class="section-paragraph">The histograms show the distribution of the number of bases spanned in sub-regions in top and bottom 10% of intolerance scores sets for (A) sub-RVIS exon scores, (B) LIMBR exon scores, (C) sub-RVIS domain scores, and (D) LIMBR domain scores.</div></div></div></div><ul class="rlist--inline separator download-links"><li><a target="_blank" href="//els-jbs-prod-cdn.literatumonline.com/cms/attachment/91873dec-c81d-4b60-a398-8d094d3472d1/gr2.jpg" title="View Large Image in a new tab">View Large
                                            Image
                                        </a></li><li class="download-links__figureViewer"><a href="#" class="figure__open__ctrl">Figure Viewer</a></li><li><a href="//els-jbs-prod-cdn.literatumonline.com/cms/attachment/fd6bd781-8c8e-4ba9-89fb-c88c1f6fdb50/gr2_lrg.jpg" class="download-links__download-Hi-res" target="_blank" download="">Download Hi-res
                                            image
                                        </a></li><li class="hidden-xs"><a href="/action/downloadFigures?pii=S0002-9297(18)30503-2&amp;id=gr2.jpg" class="download-links__download-This">Download (PPT)</a></li></ul></div></figcaption></figure></div><div class="section-paragraph">To explore whether LIMBR provides a benefit over sub-RVIS in terms of classification, we consider how well the resulting intolerance scores predict the presence of pathogenic mutations. Specifically, for both sub-RVIS and LIMBR, we regress the sub-region scores on the presence of ClinVar pathogenic mutations within these sub-regions while controlling for mutation rate using logistic regression (<a href="#fd2" data-locator="fd2" class="scroll-into-link">Equation 2</a>). We see a significant association between exon-based intolerance scores and the presence of pathogenic variants for both LIMBR (pÂ = 1.3eâ29) and sub-RVIS (p = 2.8eâ13) and less significant association for domains (LIMBR p = 1.6eâ28, sub-RVIS p = 6.6eâ67). Then we want to understand the scores ability to classify pathogenic regions in the context of region size and whether there is some potential bias, as implied by <a data-link="modal" data-target="#image-S0002-9297(18)30503-2gr2" href="#fig2" data-locator="fig2" id="back-gr2" class="figure-link scroll-into-link">FigureÂ 2</a>. To directly test for this, we fit a logistic regression model as described above, but with the addition of sub-region bases spanned as a covariate. An observed 49% (exons) and 11% (domains) changes in the estimate of sub-RVIS effect size after adding the sub-region length as a covariate, as opposed to around 14% (exons) and 2% (domains) changes for LIMBR. This indicates the possibility of strong influence of the size of the sub-region in the relative intolerance score using the sub-RVIS model.</div><div class="section-paragraph">We then consider how well a given intolerance threshold captures pathogenic mutations. Specifically, a given intolerance threshold defines both the total number of pathogenic mutations found in sub-regions as intolerant (or more) as the threshold as well as the total number of bases spanned by these same sub-regions. LIMBR appears to perform consistently better than sub-RVIS, capturing more pathogenic variants per base spanned (<a data-link="modal" data-target="#image-S0002-9297(18)30503-2gr3" href="#fig3" data-locator="fig3" id="back-gr3" class="figure-link scroll-into-link">FigureÂ 3</a>). The analysis indicates similar results between exon-level LIMBR and domain-level LIMBR in terms of the number of pathogenic variants captured versus bases spanned (<a data-link="modal" data-target="#image-S0002-9297(18)30503-2gr3" href="#fig3" data-locator="fig3" id="back-gr3" class="figure-link scroll-into-link">FigureÂ 3</a>).</div><div class="floatDisplay"><figure id="fig3" class=""><img class="b-lazy" data-src="//els-jbs-prod-cdn.literatumonline.com/cms/attachment/1e19e53a-5717-4e17-9cb1-2f8d9e47fcbd/gr3.jpg" alt="Figure thumbnail gr3" data-locator="fig3" data-high-res="//els-jbs-prod-cdn.literatumonline.com/cms/attachment/b7512473-dba4-4a85-a47d-58be35ce903a/gr3_lrg.jpg" data-large-image="//els-jbs-prod-cdn.literatumonline.com/cms/attachment/1e19e53a-5717-4e17-9cb1-2f8d9e47fcbd/gr3.jpg"></img><figcaption><div class="figure__caption__body"><div class="figure__title"><span data-fv-label="Fig. 3" class="label figure__label">FigureÂ 3</span><span class="figure__title__text">Comparing LIMBR and Sub-RVIS Ability to Capture ClinVar Pathogenic Variants versus Percent of Bases Spanned, Restricting to OMIM Genes</span></div><div class="accordion"><a href="#" class="accordion__control" aria-expanded="false"><i class="icon-section_arrow_d"></i><span>Show full caption</span></a><div class="accordion__content"><div class="figure__caption__text"><div class="section-paragraph">The genes are either broken up into exons (blue) or domains (black), then sub-regions are sorted based on either their sub-RVIS (dotted lines) or LIMBR scores (solid lines). At each percentile, the percentage of ClinVar pathogenic variants captured relative to the percent of bases covered is compared between the two methods.</div></div></div></div><ul class="rlist--inline separator download-links"><li><a target="_blank" href="//els-jbs-prod-cdn.literatumonline.com/cms/attachment/1e19e53a-5717-4e17-9cb1-2f8d9e47fcbd/gr3.jpg" title="View Large Image in a new tab">View Large
                                            Image
                                        </a></li><li class="download-links__figureViewer"><a href="#" class="figure__open__ctrl">Figure Viewer</a></li><li><a href="//els-jbs-prod-cdn.literatumonline.com/cms/attachment/b7512473-dba4-4a85-a47d-58be35ce903a/gr3_lrg.jpg" class="download-links__download-Hi-res" target="_blank" download="">Download Hi-res
                                            image
                                        </a></li><li class="hidden-xs"><a href="/action/downloadFigures?pii=S0002-9297(18)30503-2&amp;id=gr3.jpg" class="download-links__download-This">Download (PPT)</a></li></ul></div></figcaption></figure></div></section><section id="sec3.2" data-locator="sec3.2"><h3 class="sectionTitle"><span class="top__sub">Â </span><span class="top__text">Evaluation of Impact of Incorporating Sub-regions using ClinVar Pathogenic Variants</span></h3><div class="section-paragraph">We compare LIMBR to existing gene-level scores, including genic RVIS, pLI, and missense Z,<div class="dropBlock reference-citations"><sup><a href="#" id="back-bib5" data-db-target-for="bib5-d176679441e1763" class="reference-citations__ctrl" aria-expanded="false" aria-controls="bib5">5</a></sup><div data-db-target-of="bib5-d176679441e1763" class="dropBlock__holder"><div class="dropBlock__body"><ul title="list of authors" class="rlist reference-citations__loa inline-bullet-list"><li class="loa__item">Samocha K.E. </li><li class="loa__item">Robinson E.B. </li><li class="loa__item">Sanders S.J. </li><li class="loa__item">Stevens C. </li><li class="loa__item">Sabo A. </li><li class="loa__item">McGrath L.M. </li><li class="loa__item">Kosmicki J.A. </li><li class="loa__item">RehnstrÃ¶m K. </li><li class="loa__item">Mallick S. </li><li class="loa__item">Kirby A. </li><li>et al.</li></ul><div class="reference-citations__detials">A framework for the interpretation of de novo mutation in human disease.<span class="ref__series"><i>Nat. Genet.</i> <span class="ref__seriesDate">2014; </span><span class="ref__seriesVolume">46</span></span>: <span class="ref__seriesPages">944-950</span></div><div class="reference-citations__links"><ul class="rlist inline-bullet-list"><li><a href="/servlet/linkout?suffix=e_1_5_1_2_5_2&amp;dbid=16&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=10.1038%2Fng.3050&amp;cf=" onclick="newWindow(this.href);return false">Crossref</a></li>
<li><a href="/servlet/linkout?suffix=e_1_5_1_2_5_2&amp;dbid=8&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=25086666&amp;cf=" onclick="newWindow(this.href);return false">PubMed</a></li>
<li><a href="/servlet/linkout?suffix=e_1_5_1_2_5_2&amp;dbid=137438953472&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=2-s2.0-84905309176&amp;cf=" onclick="newWindow(this.href);return false">Scopus (322)</a></li>
<li><a class="google-scholar" href="http://scholar.google.com/scholar_lookup?hl=en&volume=46&publication_year=2014&pages=944-950&journal=Nat.+Genet.&author=Samocha+K.E.+%2C+Robinson+E.B.+%2C+Sanders+S.J.+%2C+Stevens+C.+%2C+Sabo+A.+%2C+McGrath+L.M.+%2C+Kosmicki+J.A.+%2C+Rehnstr%C3%B6m+K.+%2C+Mallick+S.+%2C+Kirby+A.++et+al.&title=A+framework+for+the+interpretation+of+de+novo+mutation+in+human+disease&">Google Scholar</a></li></ul></div></div><a href="#" class="dropBlock__ctrl--close" title="Close info"><i aria-hidden="true" class="icon-close_thin"></i></a></div></div><sup>,Â </sup> <div class="dropBlock reference-citations"><sup><a href="#" id="back-bib18" data-db-target-for="bib18-d176679441e1763" class="reference-citations__ctrl" aria-expanded="false" aria-controls="bib18">18</a></sup><div data-db-target-of="bib18-d176679441e1763" class="dropBlock__holder"><div class="dropBlock__body"><ul title="list of authors" class="rlist reference-citations__loa inline-bullet-list"><li class="loa__item">Lek M. </li><li class="loa__item">Karczewski K.J. </li><li class="loa__item">Minikel E.V. </li><li class="loa__item">Samocha K.E. </li><li class="loa__item">Banks E. </li><li class="loa__item">Fennell T. </li><li class="loa__item">OâDonnell-Luria A.H. </li><li class="loa__item">Ware J.S. </li><li class="loa__item">Hill A.J. </li><li class="loa__item">Cummings B.B. </li><li>et al.</li></ul><div class="reference-citations__detials">M.LekK.J.KarczewskiE.V.MinikelK.E.SamochaE.BanksT.FennellA.H.OâDonnell-LuriaJ.S.WareA.J.HillB.B.Cummings<span>Exome Aggregation Consortium</span>Analysis of protein-coding genetic variation in 60,706 humans.<span class="ref__series"><i>Nature.</i> <span class="ref__seriesDate">2016; </span><span class="ref__seriesVolume">536</span></span>: <span class="ref__seriesPages">285-291</span></div><div class="reference-citations__links"><ul class="rlist inline-bullet-list"><li><a href="/servlet/linkout?suffix=e_1_5_1_2_18_2&amp;dbid=16&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=10.1038%2Fnature19057&amp;cf=" onclick="newWindow(this.href);return false">Crossref</a></li>
<li><a href="/servlet/linkout?suffix=e_1_5_1_2_18_2&amp;dbid=8&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=27535533&amp;cf=" onclick="newWindow(this.href);return false">PubMed</a></li>
<li><a class="google-scholar" href="http://scholar.google.com/scholar_lookup?hl=en&volume=536&publication_year=2016&pages=285-291&journal=Nature&author=Exome+Aggregation+Consortium&title=Analysis+of+protein-coding+genetic+variation+in+60%2C706+humans&">Google Scholar</a></li></ul></div></div><a href="#" class="dropBlock__ctrl--close" title="Close info"><i aria-hidden="true" class="icon-close_thin"></i></a></div></div><sup>,Â </sup> <div class="dropBlock reference-citations"><sup><a href="#" id="back-bib25" data-db-target-for="bib25-d176679441e1763" class="reference-citations__ctrl" aria-expanded="false" aria-controls="bib25">25</a></sup><div data-db-target-of="bib25-d176679441e1763" class="dropBlock__holder"><div class="dropBlock__body"><ul title="list of authors" class="rlist reference-citations__loa inline-bullet-list"><li class="loa__item">Petrovski S. </li><li class="loa__item">Gussow A.B. </li><li class="loa__item">Wang Q. </li><li class="loa__item">Halvorsen M. </li><li class="loa__item">Han Y. </li><li class="loa__item">Weir W.H. </li><li class="loa__item">Allen A.S. </li><li class="loa__item">Goldstein D.B. </li></ul><div class="reference-citations__detials">The intolerance of regulatory sequence to genetic variation predicts gene dosage sensitivity.<span class="ref__series"><i>PLoS Genet.</i> <span class="ref__seriesDate">2015; </span><span class="ref__seriesVolume">11</span></span>: <span class="ref__seriesPages">e1005492</span></div><div class="reference-citations__links"><ul class="rlist inline-bullet-list"><li><a href="/servlet/linkout?suffix=e_1_5_1_2_25_2&amp;dbid=16&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=10.1371%2Fjournal.pgen.1005492&amp;cf=" onclick="newWindow(this.href);return false">Crossref</a></li>
<li><a href="/servlet/linkout?suffix=e_1_5_1_2_25_2&amp;dbid=8&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=26332131&amp;cf=" onclick="newWindow(this.href);return false">PubMed</a></li>
<li><a class="google-scholar" href="http://scholar.google.com/scholar_lookup?hl=en&volume=11&publication_year=2015&pages=e1005492&journal=PLoS+Genet.&author=Petrovski+S.+%2C+Gussow+A.B.+%2C+Wang+Q.+%2C+Halvorsen+M.+%2C+Han+Y.+%2C+Weir+W.H.+%2C+Allen+A.S.+%2C++and+Goldstein+D.B.+&title=The+intolerance+of+regulatory+sequence+to+genetic+variation+predicts+gene+dosage+sensitivity&">Google Scholar</a></li></ul></div></div><a href="#" class="dropBlock__ctrl--close" title="Close info"><i aria-hidden="true" class="icon-close_thin"></i></a></div></div> beginning by focusing on OMIM genes. We focus on OMIM genes as they are known to be associated with disease (we expand these comparisons to all genes in the supplementary materials; <a href="/cms/10.1016/j.ajhg.2018.12.020/attachment/8d42c908-07db-4f91-995c-87a0307e0d79/mmc2" class="scroll-into-link">Table S1</a> and <a href="/cms/10.1016/j.ajhg.2018.12.020/attachment/42ba8291-bbf5-4e37-8292-534540b90141/mmc1" class="scroll-into-link">FigureÂ S5</a>). In making these comparisons, when considering a gene-level score, all sub-regions of a given gene are annotated with the same score. This, in fact, corresponds to how such scores are used in practice, as all variants found within a gene would be annotated with the same gene-level score. Since there is no explicit way of knowing whether regions are benign, we perform several comparisons against ClinVar pathogenic variants to help understand how well the methods classify variants: this includes comparing classification of pathogenic variation relative to regions that are devoid of pathogenic variants, relative to total bases covered and relative to variants seen in control subjects. First, the percentile-sorted scores are used as thresholds to assess the ability of the methods to capture ClinVar pathogenic <em>de novo</em> missense variants versus sub-regions devoid of pathogenic variants (<a data-link="modal" data-target="#image-S0002-9297(18)30503-2gr4" href="#fig4" data-locator="fig4" id="back-gr4" class="figure-link scroll-into-link">Figures 4</a>Aâ4C for exons and <a data-link="modal" data-target="#image-S0002-9297(18)30503-2gr4" href="#fig4" data-locator="fig4" id="back-gr4" class="figure-link scroll-into-link">Figures 4</a>Dâ4F for domains). In this analysis, LIMBR has the highest AUC (0.86) followed by missense Z (0.81), RVIS (0.78), pLI (0.76), and sub-RVIS (0.59) (<a data-link="modal" data-target="#image-S0002-9297(18)30503-2gr4" href="#fig4" data-locator="fig4" id="back-gr4" class="figure-link scroll-into-link">FigureÂ 4</a>A). Similarly, for domains LIMBR has the highest AUC (0. 91) followed by sub-RVIS (0.81), missense Z (0.81), pLI (0.78), and RVIS (0.76) (<a data-link="modal" data-target="#image-S0002-9297(18)30503-2gr4" href="#fig4" data-locator="fig4" id="back-gr4" class="figure-link scroll-into-link">FigureÂ 4</a>D). LIMBR demonstrates at least a 5% and up to 10% improvement in AUC for classification of pathogenic missense <em>de novo</em> versus other methods. We observe similar results when using gene sets associated with epilepsy and neurodevelopmental AD.</div><div class="floatDisplay"><figure id="fig4" class=""><img class="b-lazy" data-src="//els-jbs-prod-cdn.literatumonline.com/cms/attachment/a759cbea-77e4-4dcf-b331-bf8b7de84724/gr4.jpg" alt="Figure thumbnail gr4" data-locator="fig4" data-high-res="//els-jbs-prod-cdn.literatumonline.com/cms/attachment/47dfd24d-344a-49e5-82f9-19d36afb7d41/gr4_lrg.jpg" data-large-image="//els-jbs-prod-cdn.literatumonline.com/cms/attachment/a759cbea-77e4-4dcf-b331-bf8b7de84724/gr4.jpg"></img><figcaption><div class="figure__caption__body"><div class="figure__title"><span data-fv-label="Fig. 4" class="label figure__label">FigureÂ 4</span><span class="figure__title__text">Performance of Different Methodsâ Ability to Capture <em>De Novo</em> Pathogenicity across Exons Relative to Benign Variation</span></div><div class="accordion"><a href="#" class="accordion__control" aria-expanded="false"><i class="icon-section_arrow_d"></i><span>Show full caption</span></a><div class="accordion__content"><div class="figure__caption__text"><div class="section-paragraph">(AâC) The methods are compared looking first at exons and the percent <em>de novo</em> missense variants versus benign exons restricting to (A) all OMIM genes, (B) epilepsy gene set, and (C) neurodevelopmental autosomal-dominant genes.</div><div class="section-paragraph">(DâF) For domain groupings, the methods are compared again with the percent <em>de novo</em> missense variants versus benign exons restricting to (D) all OMIM genes, (E) epilepsy gene set, and (F) neurodevelopmental autosomal-dominant genes.</div></div></div></div><ul class="rlist--inline separator download-links"><li><a target="_blank" href="//els-jbs-prod-cdn.literatumonline.com/cms/attachment/a759cbea-77e4-4dcf-b331-bf8b7de84724/gr4.jpg" title="View Large Image in a new tab">View Large
                                            Image
                                        </a></li><li class="download-links__figureViewer"><a href="#" class="figure__open__ctrl">Figure Viewer</a></li><li><a href="//els-jbs-prod-cdn.literatumonline.com/cms/attachment/47dfd24d-344a-49e5-82f9-19d36afb7d41/gr4_lrg.jpg" class="download-links__download-Hi-res" target="_blank" download="">Download Hi-res
                                            image
                                        </a></li><li class="hidden-xs"><a href="/action/downloadFigures?pii=S0002-9297(18)30503-2&amp;id=gr4.jpg" class="download-links__download-This">Download (PPT)</a></li></ul></div></figcaption></figure></div><div class="section-paragraph">Then, no longer restricting to <em>de novo</em>, we classify sub-regions as pathogenic or benign, where a sub-region is pathogenic if it contains at least one ClinVar missense pathogenic variant and benign if it does not contain any pathogenic variants. Scores are sorted by percentile and used as a threshold to examine the percentage of pathogenic sub-regions versus benign sub-regions captured. We examine all OMIM genes, then restrict to gene sets associated with either epilepsy or neurodevelopmental autosomal dominance (AD) (<a data-link="modal" data-target="#image-S0002-9297(18)30503-2gr5" href="#fig5" data-locator="fig5" id="back-gr5" class="figure-link scroll-into-link">Figures 5</a>Aâ5C and <a href="/cms/10.1016/j.ajhg.2018.12.020/attachment/42ba8291-bbf5-4e37-8292-534540b90141/mmc1" class="scroll-into-link">S1</a>AâS1C for domains). For exons across all OMIM genes, LIMBR has the highest AUC (0.58) followed by sub-RVIS (0.56), missense Z (0.54), pLI (0.52), and RVIS (0.49) (<a data-link="modal" data-target="#image-S0002-9297(18)30503-2gr5" href="#fig5" data-locator="fig5" id="back-gr5" class="figure-link scroll-into-link">FigureÂ 5</a>A) whereas for domains sub-RVIS performs best (0.62) followed by LIMBR (0.59), missense Z (0.56), RVIS (0.56), and pLI (0.53).</div><div class="floatDisplay"><figure id="fig5" class=""><img class="b-lazy" data-src="//els-jbs-prod-cdn.literatumonline.com/cms/attachment/01a83ae3-78a6-4457-bd65-d1d27544d674/gr5.jpg" alt="Figure thumbnail gr5" data-locator="fig5" data-high-res="//els-jbs-prod-cdn.literatumonline.com/cms/attachment/ab40d846-8a84-4f9e-a218-b7d7e746482c/gr5_lrg.jpg" data-large-image="//els-jbs-prod-cdn.literatumonline.com/cms/attachment/01a83ae3-78a6-4457-bd65-d1d27544d674/gr5.jpg"></img><figcaption><div class="figure__caption__body"><div class="figure__title"><span data-fv-label="Fig. 5" class="label figure__label">FigureÂ 5</span><span class="figure__title__text">Performance of Different Methodsâ Ability to Capture Pathogenicity across Exons Relative to Benign Variation</span></div><div class="accordion"><a href="#" class="accordion__control" aria-expanded="false"><i class="icon-section_arrow_d"></i><span>Show full caption</span></a><div class="accordion__content"><div class="figure__caption__text"><div class="section-paragraph">(AâC) LIMBR is compared against the other methods to see how well it classifies pathogenic exons versus benign exons restricting to (A) all OMIM, (B) epilepsy gene set, or (C) neurodevelopmental autosomal-dominant gene set.</div><div class="section-paragraph">(DâF) Then the percent pathogenic variants versus control variants captured by the different methods is compared restricting to (D) all OMIM, (E) epilepsy gene set, or (F) neurodevelopmental autosomal-dominant gene set.</div></div></div></div><ul class="rlist--inline separator download-links"><li><a target="_blank" href="//els-jbs-prod-cdn.literatumonline.com/cms/attachment/01a83ae3-78a6-4457-bd65-d1d27544d674/gr5.jpg" title="View Large Image in a new tab">View Large
                                            Image
                                        </a></li><li class="download-links__figureViewer"><a href="#" class="figure__open__ctrl">Figure Viewer</a></li><li><a href="//els-jbs-prod-cdn.literatumonline.com/cms/attachment/ab40d846-8a84-4f9e-a218-b7d7e746482c/gr5_lrg.jpg" class="download-links__download-Hi-res" target="_blank" download="">Download Hi-res
                                            image
                                        </a></li><li class="hidden-xs"><a href="/action/downloadFigures?pii=S0002-9297(18)30503-2&amp;id=gr5.jpg" class="download-links__download-This">Download (PPT)</a></li></ul></div></figcaption></figure></div><div class="section-paragraph">Then the percentile-sorted scores are used as a threshold to examine the percentage of missense pathogenic ClinVar variants captured versus control variants, where control variants are defined as variants found in DiscovEHR<div class="dropBlock reference-citations"><sup><a href="#" id="back-bib24" data-db-target-for="bib24-d176679441e1810" class="reference-citations__ctrl" aria-expanded="false" aria-controls="bib24">24</a></sup><div data-db-target-of="bib24-d176679441e1810" class="dropBlock__holder"><div class="dropBlock__body"><ul title="list of authors" class="rlist reference-citations__loa inline-bullet-list"><li class="loa__item">Dewey F.E. </li><li class="loa__item">Murray M.F. </li><li class="loa__item">Overton J.D. </li><li class="loa__item">Habegger L. </li><li class="loa__item">Leader J.B. </li><li class="loa__item">Fetterolf S.N. </li><li class="loa__item">OâDushlaine C. </li><li class="loa__item">Van Hout C.V. </li><li class="loa__item">Staples J. </li><li class="loa__item">Gonzaga-Jauregui C. </li><li>et al.</li></ul><div class="reference-citations__detials">Distribution and clinical impact of functional variants in 50,726 whole-exome sequences from the DiscovEHR study.<span class="ref__series"><i>Science.</i> <span class="ref__seriesDate">2016; </span><span class="ref__seriesVolume">354</span></span></div><div class="reference-citations__links"><ul class="rlist inline-bullet-list"><li><a href="/servlet/linkout?suffix=e_1_5_1_2_24_2&amp;dbid=16&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=10.1126%2Fscience.aaf6814&amp;cf=" onclick="newWindow(this.href);return false">Crossref</a></li>
<li><a href="/servlet/linkout?suffix=e_1_5_1_2_24_2&amp;dbid=137438953472&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=2-s2.0-85007028930&amp;cf=" onclick="newWindow(this.href);return false">Scopus (87)</a></li>
<li><a class="google-scholar" href="http://scholar.google.com/scholar_lookup?hl=en&volume=354&publication_year=2016&journal=Science&author=Dewey+F.E.+%2C+Murray+M.F.+%2C+Overton+J.D.+%2C+Habegger+L.+%2C+Leader+J.B.+%2C+Fetterolf+S.N.+%2C+O%E2%80%99Dushlaine+C.+%2C+Van+Hout+C.V.+%2C+Staples+J.+%2C+Gonzaga-Jauregui+C.++et+al.&title=Distribution+and+clinical+impact+of+functional+variants+in+50%2C726+whole-exome+sequences+from+the+DiscovEHR+study&">Google Scholar</a></li></ul></div></div><a href="#" class="dropBlock__ctrl--close" title="Close info"><i aria-hidden="true" class="icon-close_thin"></i></a></div></div> control samples, excluding variants also found in gnomAD and ClinVar (<a data-link="modal" data-target="#image-S0002-9297(18)30503-2gr5" href="#fig5" data-locator="fig5" id="back-gr5" class="figure-link scroll-into-link">Figures 5</a>D, 5E, <a href="/cms/10.1016/j.ajhg.2018.12.020/attachment/42ba8291-bbf5-4e37-8292-534540b90141/mmc1" class="scroll-into-link">S1</a>D, and S1E for domains). Despite sub-RVIS performing nearly as well as LIMBR in terms of classifying sub-regions, when we examine the ability to classify pathogenic versus control variants, sub-RVIS performs the worst and LIMBR performs the best (<a data-link="modal" data-target="#image-S0002-9297(18)30503-2gr5" href="#fig5" data-locator="fig5" id="back-gr5" class="figure-link scroll-into-link">Figures 5</a>D and <a href="/cms/10.1016/j.ajhg.2018.12.020/attachment/42ba8291-bbf5-4e37-8292-534540b90141/mmc1" class="scroll-into-link">S1</a>D). We also considered the percent ClinVar pathogenic variants captured versus benign sub-regions (for exons <a href="/cms/10.1016/j.ajhg.2018.12.020/attachment/42ba8291-bbf5-4e37-8292-534540b90141/mmc1" class="scroll-into-link">Figures S2</a>AâS2C and for domains <a href="/cms/10.1016/j.ajhg.2018.12.020/attachment/42ba8291-bbf5-4e37-8292-534540b90141/mmc1" class="scroll-into-link">Figures S3</a>AâS3C) and the percent ClinVar pathogenic variants captured versus bases covered (for exons <a href="/cms/10.1016/j.ajhg.2018.12.020/attachment/42ba8291-bbf5-4e37-8292-534540b90141/mmc1" class="scroll-into-link">Figures S2</a>D and S2E and for domains <a href="/cms/10.1016/j.ajhg.2018.12.020/attachment/42ba8291-bbf5-4e37-8292-534540b90141/mmc1" class="scroll-into-link">Figures S3</a>D and S3E). Similar results are observed when restricting to genes associated with epilepsy (<a data-link="modal" data-target="#image-S0002-9297(18)30503-2gr5" href="#fig5" data-locator="fig5" id="back-gr5" class="figure-link scroll-into-link">Figures 5</a>B, 5E, <a href="/cms/10.1016/j.ajhg.2018.12.020/attachment/42ba8291-bbf5-4e37-8292-534540b90141/mmc1" class="scroll-into-link">S1</a>B, S1E, <a href="/cms/10.1016/j.ajhg.2018.12.020/attachment/42ba8291-bbf5-4e37-8292-534540b90141/mmc1" class="scroll-into-link">S2</a>B, and S2E) and neurodevelopmental AD (<a data-link="modal" data-target="#image-S0002-9297(18)30503-2gr5" href="#fig5" data-locator="fig5" id="back-gr5" class="figure-link scroll-into-link">Figures 5</a>C, 5F, <a href="/cms/10.1016/j.ajhg.2018.12.020/attachment/42ba8291-bbf5-4e37-8292-534540b90141/mmc1" class="scroll-into-link">S1</a>C, S1F, <a href="/cms/10.1016/j.ajhg.2018.12.020/attachment/42ba8291-bbf5-4e37-8292-534540b90141/mmc1" class="scroll-into-link">S2</a>C, and S2F), where LIMBRÂ appears to demonstrate consistent improvement in capturing pathogenic variation. LIMBR results across different gene sets related to epilepsy, neurodevelopmental, AD, neurodevelopmental AD, haploinsufficiency, andÂ recessive genes (<a data-link="modal" data-target="#image-S0002-9297(18)30503-2gr6" href="#fig6" data-locator="fig6" id="back-gr6" class="figure-link scroll-into-link">FigureÂ 6</a> for exons and <a href="/cms/10.1016/j.ajhg.2018.12.020/attachment/42ba8291-bbf5-4e37-8292-534540b90141/mmc1" class="scroll-into-link">FigureÂ S4</a> for domains) are plotted for further comparisons of the gene groupings. The curation of ClinVar pathogenic variants isÂ generally more rigorously restricted to OMIM genes. However, we also extend to all genes, not restricting to OMIM (<a href="/cms/10.1016/j.ajhg.2018.12.020/attachment/42ba8291-bbf5-4e37-8292-534540b90141/mmc1" class="scroll-into-link">FigureÂ S5</a>), and observe similar results to the OMIM-only set.</div><div class="floatDisplay"><figure id="fig6" class=""><img class="b-lazy" data-src="//els-jbs-prod-cdn.literatumonline.com/cms/attachment/716e1385-057b-4927-a01e-b1c691a40dea/gr6.jpg" alt="Figure thumbnail gr6" data-locator="fig6" data-high-res="//els-jbs-prod-cdn.literatumonline.com/cms/attachment/efc36a4c-642c-4587-9c4f-622e9fde0340/gr6_lrg.jpg" data-large-image="//els-jbs-prod-cdn.literatumonline.com/cms/attachment/716e1385-057b-4927-a01e-b1c691a40dea/gr6.jpg"></img><figcaption><div class="figure__caption__body"><div class="figure__title"><span data-fv-label="Fig. 6" class="label figure__label">FigureÂ 6</span><span class="figure__title__text">Performance of LIMBR across Different Gene Sets</span></div><div class="accordion"><a href="#" class="accordion__control" aria-expanded="false"><i class="icon-section_arrow_d"></i><span>Show full caption</span></a><div class="accordion__content"><div class="figure__caption__text"><div class="section-paragraph">The LIMBR classification (A) plotting exons with at least one pathogenic variant versus benign exons restricting to different OMIM genes sets (overlapping with set in <a data-locator="fig4" aria-expanded="false" aria-controls="crosref0405" class="figure-link caption__link scroll-into-link" href="&#xA;                            #fig4">FigureÂ 4</a>). Then similarly using to LIMBR percentile rankings of exons to see the (B) percent pathogenic <em>de novo</em> variants relative to benign exons again restricting to different OMIM genes sets.</div></div></div></div><ul class="rlist--inline separator download-links"><li><a target="_blank" href="//els-jbs-prod-cdn.literatumonline.com/cms/attachment/716e1385-057b-4927-a01e-b1c691a40dea/gr6.jpg" title="View Large Image in a new tab">View Large
                                            Image
                                        </a></li><li class="download-links__figureViewer"><a href="#" class="figure__open__ctrl">Figure Viewer</a></li><li><a href="//els-jbs-prod-cdn.literatumonline.com/cms/attachment/efc36a4c-642c-4587-9c4f-622e9fde0340/gr6_lrg.jpg" class="download-links__download-Hi-res" target="_blank" download="">Download Hi-res
                                            image
                                        </a></li><li class="hidden-xs"><a href="/action/downloadFigures?pii=S0002-9297(18)30503-2&amp;id=gr6.jpg" class="download-links__download-This">Download (PPT)</a></li></ul></div></figcaption></figure></div></section><section id="sec3.3" data-locator="sec3.3"><h3 class="sectionTitle"><span class="top__sub">Â </span><span class="top__text">Genes with Significant Association between Sub-Region Level Intolerance and Pathogenic Variant Placement</span></h3><div class="section-paragraph">To test whether the intolerance scores are significantly predictive of where pathogenic mutations fall within a given gene, a permutation test is constructed. Using this permutation test (<a href="#fd3" data-locator="fd3" class="scroll-into-link">Equation 3</a>), there are 259 genes that demonstrate significance at Î± = 0.05 (<a href="/cms/10.1016/j.ajhg.2018.12.020/attachment/04592b59-2b21-44fc-aa10-05228ba69e55/mmc3" class="scroll-into-link">Table S2</a>). Of the 259 significant genes, we see enrichment in several different gene sets where 41 are related to haploinsufficiency, 76 are recessive, 101 are associated with AD, 103Â are associated neurodevelopmental disorders, 41 are both associated with neurodevelopmental disorders and AD, and 15 are associated with epilepsy (<a href="#tbl1" id="back-tbl1" class="scroll-into-link">Table 1</a> and <a data-link="modal" data-target="#image-S0002-9297(18)30503-2gr6" href="#fig6" data-locator="fig6" id="back-gr6" class="figure-link scroll-into-link">FigureÂ 6</a>). We examined gene sets in different organ groupsÂ from Database of Chromosomal Imbalance andÂ Phenotype in Humans Using Ensembl Resources (DECIPHER)<div class="dropBlock reference-citations"><sup><a href="#" id="back-bib26" data-db-target-for="bib26-d176679441e1887" class="reference-citations__ctrl" aria-expanded="false" aria-controls="bib26">26</a></sup><div data-db-target-of="bib26-d176679441e1887" class="dropBlock__holder"><div class="dropBlock__body"><ul title="list of authors" class="rlist reference-citations__loa inline-bullet-list"><li class="loa__item">Firth H.V. </li><li class="loa__item">Richards S.M. </li><li class="loa__item">Bevan A.P. </li><li class="loa__item">Clayton S. </li><li class="loa__item">Corpas M. </li><li class="loa__item">Rajan D. </li><li class="loa__item">Van Vooren S. </li><li class="loa__item">Moreau Y. </li><li class="loa__item">Pettett R.M. </li><li class="loa__item">Carter N.P. </li></ul><div class="reference-citations__detials">DECIPHER: Database of Chromosomal Imbalance and Phenotype in Humans Using Ensembl Resources.<span class="ref__series"><i>Am. J. Hum. Genet.</i> <span class="ref__seriesDate">2009; </span><span class="ref__seriesVolume">84</span></span>: <span class="ref__seriesPages">524-533</span></div><div class="reference-citations__links"><ul class="rlist inline-bullet-list"><li><a href="/servlet/linkout?suffix=e_1_5_1_2_26_2&amp;dbid=4&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=10.1016%2Fj.ajhg.2009.03.010&amp;cf=abstract&amp;site=cell-site" onclick="newWindow(this.href);return false">Abstract</a></li>
<li><a href="/servlet/linkout?suffix=e_1_5_1_2_26_2&amp;dbid=4&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=10.1016%2Fj.ajhg.2009.03.010&amp;cf=fulltext&amp;site=cell-site" onclick="newWindow(this.href);return false">Full Text</a></li>
<li><a href="/servlet/linkout?suffix=e_1_5_1_2_26_2&amp;dbid=4&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=10.1016%2Fj.ajhg.2009.03.010&amp;cf=pdf&amp;site=cell-site" onclick="newWindow(this.href);return false">Full Text PDF</a></li>
<li><a href="/servlet/linkout?suffix=e_1_5_1_2_26_2&amp;dbid=8&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=19344873&amp;cf=" onclick="newWindow(this.href);return false">PubMed</a></li>
<li><a href="/servlet/linkout?suffix=e_1_5_1_2_26_2&amp;dbid=137438953472&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=2-s2.0-64149099583&amp;cf=" onclick="newWindow(this.href);return false">Scopus (550)</a></li>
<li><a class="google-scholar" href="http://scholar.google.com/scholar_lookup?hl=en&volume=84&publication_year=2009&pages=524-533&journal=Am.+J.+Hum.+Genet.&author=Firth+H.V.+%2C+Richards+S.M.+%2C+Bevan+A.P.+%2C+Clayton+S.+%2C+Corpas+M.+%2C+Rajan+D.+%2C+Van+Vooren+S.+%2C+Moreau+Y.+%2C+Pettett+R.M.+%2C++and+Carter+N.P.+&title=DECIPHER%3A+Database+of+Chromosomal+Imbalance+and+Phenotype+in+Humans+Using+Ensembl+Resources&">Google Scholar</a></li></ul></div></div><a href="#" class="dropBlock__ctrl--close" title="Close info"><i aria-hidden="true" class="icon-close_thin"></i></a></div></div> and found, for face-related genes, 29 out of 252 demonstrate significant permutation tests (exact test pÂ =Â 0.004), and for heart-related cardiovascular/lymphatic genes, 32 out of 240 demonstrate significance (exact test pÂ = 2.7eâ4) (<a href="#tbl1" id="back-tbl1" class="scroll-into-link">Table 1</a>). This also includes 5 genes known to be associated with early-onset epilepsy: SCN1A (pÂ &lt;Â 1eâ5), SCN8A (p = 2eâ4), CDKL5 (p = 0.011), PCDH19 (p = 4eâ5), and KCNT1 (p = 0.006) (<a data-link="modal" data-target="#image-S0002-9297(18)30503-2gr7" href="#fig7" data-locator="fig7" id="back-gr7" class="figure-link scroll-into-link">FigureÂ 7</a>). We also investigated whether there was enrichment in transmembrane and ion channel genes. We did not find enrichment in transmembrane gene but did see enrichment in ion channels: 23 out of 124 ion channel genes demonstrate significance (p = 1.1eâ5).</div><div class="floatDisplay"><div id="tbl1" class="inline-table"><div class="inline-table__head"><span class="inline-table__label">Table 1</span><span class="inline-table__title">Gene Set Enrichment using Gene-Level Permutation Test Demonstrating Significant Association between Exon-Level Intolerance and Pathogenic Variant Localization</span></div><table width="100%" cellpadding="0" cellspacing="0" border="0" class="scrollable bordered"><thead><tr><th class="rowsep"><strong>Gene Set</strong></th><th class="rowsep"><strong>Number of Genes in Set</strong></th><th class="rowsep"><strong>Number Genes Significant Permutation Test</strong></th><th class="rowsep"><strong>Exact p Value</strong></th></tr></thead><tbody><tr><td>Haploinsufficiency</td><td>175</td><td>41</td><td>2.04Eâ12<sup>â</sup></td></tr><tr><td>Recessive</td><td>817</td><td>76</td><td>0.027</td></tr><tr><td>Autosomal dominant</td><td>1,083</td><td>101</td><td>1.56Eâ11<sup>â</sup></td></tr><tr><td>Neurodevelopmental</td><td>1,010</td><td>103</td><td>1.21Eâ07<sup>â</sup></td></tr><tr><td>Neurodevelopmental AD</td><td>252</td><td>41</td><td>2.08Eâ10<sup>â</sup></td></tr><tr><td>Epilepsy</td><td>100</td><td>15</td><td>2.66Eâ04<sup>â</sup></td></tr><tr><td>Bone marrow/immune</td><td>143</td><td>7</td><td>0.705</td></tr><tr><td>Brain/cognition</td><td>1,314</td><td>96</td><td>0.131</td></tr><tr><td>Ear</td><td>134</td><td>10</td><td>0.332</td></tr><tr><td>Endocrine metabolic</td><td>452</td><td>25</td><td>0.937</td></tr><tr><td>Eye</td><td>347</td><td>17</td><td>0.924</td></tr><tr><td>Face</td><td>252</td><td>29</td><td>0.004</td></tr><tr><td>GI tract</td><td>159</td><td>11</td><td>0.441</td></tr><tr><td>Genitalia</td><td>67</td><td>8</td><td>0.062</td></tr><tr><td>Heart cardiovasculature/lymphatic</td><td>240</td><td>32</td><td>2.62Eâ04<sup>â</sup></td></tr><tr><td>Kidney/renal tract</td><td>148</td><td>11</td><td>0.511</td></tr><tr><td>Musculature</td><td>124</td><td>11</td><td>0.162</td></tr><tr><td>Respiratory tract</td><td>33</td><td>2</td><td>0.291</td></tr><tr><td>Skeleton</td><td>687</td><td>47</td><td>0.577</td></tr><tr><td>Skin</td><td>241</td><td>23</td><td>0.054</td></tr><tr><td>Spinal cord/peripheral nerves</td><td>57</td><td>2</td><td>0.764</td></tr><tr><td>Teeth and dentition</td><td>50</td><td>0</td><td>0.966</td></tr><tr><td>Ion channel</td><td>124</td><td>23</td><td>1.10eâ5<sup>â</sup></td></tr><tr><td>Transmembrane</td><td>992</td><td>76</td><td>0.403</td></tr><tr><td>All OMIM</td><td>3,433</td><td>259</td><td></td></tr></tbody></table><div class="inline-table__tail"><div class="section-paragraph">Shown are different gene sets with the number of genes in each set, of those genes the number of them that are significant using the permutation test, and the exact test for observing that many significant genes in the sub-category relative to the total OMIM disease-associated genes. Asterisks indicate significant p value.</div></div><ul class="rlist--inline separator download-links"><li><a title="Open Full Table" class="table__open__ctrl" target="_blank" href="/action/showFullTableHTML?isHtml=true&amp;tableId=tbl1&amp;pii=S0002-9297%2818%2930503-2">
                            Open table in a new tab
                        </a></li></ul></div></div><div class="floatDisplay"><figure id="fig7" class=""><img class="b-lazy" data-src="//els-jbs-prod-cdn.literatumonline.com/cms/attachment/65810e5d-1fef-4ac2-90d8-cd859da69cb8/gr7.jpg" alt="Figure thumbnail gr7" data-locator="fig7" data-high-res="//els-jbs-prod-cdn.literatumonline.com/cms/attachment/7588efad-f3a3-4192-a3e4-6ed7c5cf9bb0/gr7_lrg.jpg" data-large-image="//els-jbs-prod-cdn.literatumonline.com/cms/attachment/65810e5d-1fef-4ac2-90d8-cd859da69cb8/gr7.jpg"></img><figcaption><div class="figure__caption__body"><div class="figure__title"><span data-fv-label="Fig. 7" class="label figure__label">FigureÂ 7</span><span class="figure__title__text">Localized Genic Intolerance to Variation in Key Epileptic Encephalopathy Genes</span></div><div class="accordion"><a href="#" class="accordion__control" aria-expanded="false"><i class="icon-section_arrow_d"></i><span>Show full caption</span></a><div class="accordion__content"><div class="figure__caption__text"><div class="section-paragraph">(AâE) The plots above are of the intolerance scores for (A) SCN1A, (B) SC81A, (C) CDKL5, (D) PCDH19, and (E) KCNT1 with 95% credibility in gray across combined coding positions in all transcripts. The bar strip below that plot indicates when the start and end of an exon occurs. Below are the densities of ClinVar variants matched up at the corresponding genomic positions with the intolerance scores.</div><div class="section-paragraph">(F) Table that depicts a group-wise association test, both unweighted and weighted with the inverse percentile of the LIMBR intolerance, using a cohort of 488 epileptic encephalopathy case subjects and 12,151 unrelated control subjects from a previous rare variant collapsing analysis.</div></div></div></div><ul class="rlist--inline separator download-links"><li><a target="_blank" href="//els-jbs-prod-cdn.literatumonline.com/cms/attachment/65810e5d-1fef-4ac2-90d8-cd859da69cb8/gr7.jpg" title="View Large Image in a new tab">View Large
                                            Image
                                        </a></li><li class="download-links__figureViewer"><a href="#" class="figure__open__ctrl">Figure Viewer</a></li><li><a href="//els-jbs-prod-cdn.literatumonline.com/cms/attachment/7588efad-f3a3-4192-a3e4-6ed7c5cf9bb0/gr7_lrg.jpg" class="download-links__download-Hi-res" target="_blank" download="">Download Hi-res
                                            image
                                        </a></li><li class="hidden-xs"><a href="/action/downloadFigures?pii=S0002-9297(18)30503-2&amp;id=gr7.jpg" class="download-links__download-This">Download (PPT)</a></li></ul></div></figcaption></figure></div></section><section id="sec3.4" data-locator="sec3.4"><h3 class="sectionTitle"><span class="top__sub">Â </span><span class="top__text">Using Sub-Region Level Intolerance in Analyses of whether the Burden of Rare Variation Is Associated with Disease</span></h3><div class="section-paragraph">We investigated whether the LIMBR scores can be used inÂ a case-control study to improve our ability to detect gene-disease associations. When associating rare variation with disease, it is common to aggregate the effects of individual variants across a gene as individual variant-level tests will likely be underpowered. This works well as long as the set of variants one is aggregating over is enriched for variants that disrupt the gene. Thus, inÂ practice, one usually filters or weights variants by predictions of the variantsâ likely pathogenicity. Here we investigate using LIMBR in this role. Using a cohort of 488 epileptic encephalopathy-affected case subjects and 12,151 unrelated control subjects from a previous rare variant collapsing analysis,<div class="dropBlock reference-citations"><sup><a href="#" id="back-bib27" data-db-target-for="bib27-d176679441e1904" class="reference-citations__ctrl" aria-expanded="false" aria-controls="bib27">27</a></sup><div data-db-target-of="bib27-d176679441e1904" class="dropBlock__holder"><div class="dropBlock__body"><ul title="list of authors" class="rlist reference-citations__loa inline-bullet-list"><li class="loa__item">Zhu X. </li><li class="loa__item">Padmanabhan R. </li><li class="loa__item">Copeland B. </li><li class="loa__item">Bridgers J. </li><li class="loa__item">Ren Z. </li><li class="loa__item">Kamalakaran S. </li><li class="loa__item">Driscoll-collins A.O. </li><li class="loa__item">Berkovic S.F. </li><li class="loa__item">Scheffer I.E. </li><li class="loa__item">Poduri A. </li><li>et al.</li></ul><div class="reference-citations__detials">A case-control collapsing analysis identifies epilepsy genes implicated in trio sequencing studies focused on de novo mutations.<span class="ref__series"><i>PLoS Genet.</i> <span class="ref__seriesDate">2017; </span><span class="ref__seriesVolume">13</span></span>: <span class="ref__seriesPages">e1007104</span></div><div class="reference-citations__links"><ul class="rlist inline-bullet-list"><li><a href="/servlet/linkout?suffix=e_1_5_1_2_27_2&amp;dbid=16&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=10.1371%2Fjournal.pgen.1007104&amp;cf=" onclick="newWindow(this.href);return false">Crossref</a></li>
<li><a href="/servlet/linkout?suffix=e_1_5_1_2_27_2&amp;dbid=8&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=29186148&amp;cf=" onclick="newWindow(this.href);return false">PubMed</a></li>
<li><a href="/servlet/linkout?suffix=e_1_5_1_2_27_2&amp;dbid=137438953472&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=2-s2.0-85036612433&amp;cf=" onclick="newWindow(this.href);return false">Scopus (0)</a></li>
<li><a class="google-scholar" href="http://scholar.google.com/scholar_lookup?hl=en&volume=13&publication_year=2017&pages=e1007104&journal=PLoS+Genet.&author=Zhu+X.+%2C+Padmanabhan+R.+%2C+Copeland+B.+%2C+Bridgers+J.+%2C+Ren+Z.+%2C+Kamalakaran+S.+%2C+Driscoll-collins+A.O.+%2C+Berkovic+S.F.+%2C+Scheffer+I.E.+%2C+Poduri+A.++et+al.&title=A+case-control+collapsing+analysis+identifies+epilepsy+genes+implicated+in+trio+sequencing+studies+focused+on+de+novo+mutations&">Google Scholar</a></li></ul></div></div><a href="#" class="dropBlock__ctrl--close" title="Close info"><i aria-hidden="true" class="icon-close_thin"></i></a></div></div> we performed a group-wise association test on the five epilepsy-associated genes mentioned above (<a href="#fd4" data-locator="fd4" class="scroll-into-link">Equation 4</a>).<div class="dropBlock reference-citations"><sup><a href="#" id="back-bib17" data-db-target-for="bib17-d176679441e1911" class="reference-citations__ctrl" aria-expanded="false" aria-controls="bib17">17</a></sup><div data-db-target-of="bib17-d176679441e1911" class="dropBlock__holder"><div class="dropBlock__body"><ul title="list of authors" class="rlist reference-citations__loa inline-bullet-list"><li class="loa__item">Madsen B.E. </li><li class="loa__item">Browning S.R. </li></ul><div class="reference-citations__detials">A groupwise association test for rare mutations using a weighted sum statistic.<span class="ref__series"><i>PLoS Genet.</i> <span class="ref__seriesDate">2009; </span><span class="ref__seriesVolume">5</span></span>: <span class="ref__seriesPages">e1000384</span></div><div class="reference-citations__links"><ul class="rlist inline-bullet-list"><li><a href="/servlet/linkout?suffix=e_1_5_1_2_17_2&amp;dbid=16&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=10.1371%2Fjournal.pgen.1000384&amp;cf=" onclick="newWindow(this.href);return false">Crossref</a></li>
<li><a href="/servlet/linkout?suffix=e_1_5_1_2_17_2&amp;dbid=8&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=19214210&amp;cf=" onclick="newWindow(this.href);return false">PubMed</a></li>
<li><a href="/servlet/linkout?suffix=e_1_5_1_2_17_2&amp;dbid=137438953472&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=2-s2.0-61449168010&amp;cf=" onclick="newWindow(this.href);return false">Scopus (648)</a></li>
<li><a class="google-scholar" href="http://scholar.google.com/scholar_lookup?hl=en&volume=5&publication_year=2009&pages=e1000384&journal=PLoS+Genet.&author=Madsen+B.E.++and+Browning+S.R.+&title=A+groupwise+association+test+for+rare+mutations+using+a+weighted+sum+statistic&">Google Scholar</a></li></ul></div></div><a href="#" class="dropBlock__ctrl--close" title="Close info"><i aria-hidden="true" class="icon-close_thin"></i></a></div></div> In the first set of tests, we assign equal weights to all variants; in the second, we weight the variants by the inverse percentile of their exon-level intolerance scores (details in <a href="#sec2" class="scroll-into-link">Material and Methods</a>). Four out of the five genes appear to demonstrate lower p values under the weighting scheme, while the weighted p value for KCNT1 increases but has overlapping 95% confidence intervals with the unweighted p value (<a data-link="modal" data-target="#image-S0002-9297(18)30503-2gr7" href="#fig7" data-locator="fig7" id="back-gr7" class="figure-link scroll-into-link">FigureÂ 7</a>F). This serves as a proof of concept for the potential to use the scores to aid in identifying key genes associated with diseases.</div></section></section><section id="sec4" data-locator="sec4"><h2 class="top" id="secsectitle0065"><span class="top__sub"></span><span class="top__text">Discussion</span></h2><div class="section-paragraph">We demonstrate that the approach to intolerance scoring implement in LIMBR results in a significant improvement over existing methods in the prediction of pathogenic variants. We believe that the primary reason for this improvement is that LIMBR is able to capture patterns of variation in intolerance among genic regions while minimizing the noise introduced by sub-region size. Of particular relevance to diagnostic sequencing, LIMBR also identifies a set of 259 genes in which there is a significant relationship between intolerance scores and the location of pathogenic missense mutations. This work therefore identifies a set of existing genes for which diagnostic labs should consider not only location in interpreting variants but also the intolerance of the regions within which variants fall. We anticipate that more effective use of the location of missense variants in genes will help to address substantial challenge presented by variants of unknown significance (VUSs). Interestingly, some of the genes that have high rates of variants of unknown significance show significant regional variation in intolerance including one gene central to a high-profileÂ court case resulting from misclassification of a pathogenic variant as a variant of unknown significance.<div class="dropBlock reference-citations"><sup><a href="#" id="back-bib28" data-db-target-for="bib28-d176679441e1926" class="reference-citations__ctrl" aria-expanded="false" aria-controls="bib28">28</a></sup><div data-db-target-of="bib28-d176679441e1926" class="dropBlock__holder"><div class="dropBlock__body"><ul title="list of authors" class="rlist reference-citations__loa inline-bullet-list"><li class="loa__item">Hoffman-Andrews L. </li></ul><div class="reference-citations__detials">The known unknown: the challenges of genetic variants of uncertain significance in clinical practice.<span class="ref__series"><i>J.Â Law Biosci.</i> <span class="ref__seriesDate">2018; </span><span class="ref__seriesVolume">4</span></span>: <span class="ref__seriesPages">648-657</span></div><div class="reference-citations__links"><ul class="rlist inline-bullet-list"><li><a href="/servlet/linkout?suffix=e_1_5_1_2_28_2&amp;dbid=16&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=10.1093%2Fjlb%2Flsx038&amp;cf=" onclick="newWindow(this.href);return false">Crossref</a></li>
<li><a href="/servlet/linkout?suffix=e_1_5_1_2_28_2&amp;dbid=8&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=29868193&amp;cf=" onclick="newWindow(this.href);return false">PubMed</a></li>
<li><a href="/servlet/linkout?suffix=e_1_5_1_2_28_2&amp;dbid=137438953472&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=2-s2.0-85056700984&amp;cf=" onclick="newWindow(this.href);return false">Scopus (2)</a></li>
<li><a class="google-scholar" href="http://scholar.google.com/scholar_lookup?hl=en&volume=4&publication_year=2018&pages=648-657&journal=J.%C2%A0Law+Biosci.&author=Hoffman-Andrews+L.+&title=The+known+unknown%3A+the+challenges+of+genetic+variants+of+uncertain+significance+in+clinical+practice&">Google Scholar</a></li></ul></div></div><a href="#" class="dropBlock__ctrl--close" title="Close info"><i aria-hidden="true" class="icon-close_thin"></i></a></div></div><sup>,Â </sup> <div class="dropBlock reference-citations"><sup><a href="#" id="back-bib29" data-db-target-for="bib29-d176679441e1926" class="reference-citations__ctrl" aria-expanded="false" aria-controls="bib29">29</a></sup><div data-db-target-of="bib29-d176679441e1926" class="dropBlock__holder"><div class="dropBlock__body"><p>Ray, T. (2016). Motherâs Negligence Suit Against Questâs Athena Could Broadly Impact Genetic Testing Labs. GenomeWeb, March 14, 2016. <a href="https://www.genomeweb.com/molecular-diagnostics/mothers-negligence-suit-against-quests-athena-could-broadly-impact-genetic#.XCy8t5NKjIU" target="_blank">https://www.genomeweb.com/molecular-diagnostics/mothers-negligence-suit-against-quests-athena-could-broadly-impact-genetic#.XCy8t5NKjIU</a>.</p><div class="reference-citations__links"><ul class="rlist inline-bullet-list"><li><a class="google-scholar" href="http://scholar.google.com/scholar?hl=en&q=29">Google Scholar</a></li></ul></div></div><a href="#" class="dropBlock__ctrl--close" title="Close info"><i aria-hidden="true" class="icon-close_thin"></i></a></div></div> For initial analysis we compare the LIMBR scores associated with VUSs to the LIMBR scores of TOPMed missense singletons, restricting both sets to variants found in the 259 diseases genes where LIMBR was significantly predictive of pathogenicity. We observe that the VUSs tend to land in more tolerant sub-regionsÂ than rare missense variants found in the general population. This suggests that many of these VUSs are likely benign. We are currently investigating approaches using LIMBR to better classify VUSs as benign or pathogenic.</div><div class="section-paragraph">As with any analytical model, the analysis is only as good as the constraints of the data. As control datasets increase in size, so will the power to detect intolerance. Furthermore, improvement in genic sub-region definitions and the increasing availability of trait-specificÂ dataÂ will also improve the ability to understand sub-regional intolerance. We doÂ not compare to alternative sub-regional definitions in Samocha<div class="dropBlock reference-citations"><sup><a href="#" id="back-bib30" data-db-target-for="bib30-d176679441e1932" class="reference-citations__ctrl" aria-expanded="false" aria-controls="bib30">30</a></sup><div data-db-target-of="bib30-d176679441e1932" class="dropBlock__holder"><div class="dropBlock__body"><ul title="list of authors" class="rlist reference-citations__loa inline-bullet-list"><li class="loa__item">Samocha K.E. </li><li class="loa__item">Kosmicki J.A. </li><li class="loa__item">Karczewski K.J. </li><li class="loa__item">OâDonnell-Luria A.H. </li><li class="loa__item">Pierce-Hoffman E. </li><li class="loa__item">MacArthur D.G. </li><li class="loa__item">Neale B.M. </li><li class="loa__item">Daly M.J. </li></ul><div class="reference-citations__detials">Regional missense constraint improves variant deleteriousness prediction.<span class="ref__series"><i>bioRxiv.</i> <span class="ref__seriesDate">2017; </span><span class="ref__seriesVolume"></span></span></div><div class="reference-citations__links"><ul class="rlist inline-bullet-list"><li><a href="/servlet/linkout?suffix=e_1_5_1_2_30_2&amp;dbid=16&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=10.1101%2F148353&amp;cf=" onclick="newWindow(this.href);return false">Crossref</a></li>
<li><a class="google-scholar" href="http://scholar.google.com/scholar_lookup?hl=en&publication_year=2017&journal=bioRxiv&author=Samocha+K.E.+%2C+Kosmicki+J.A.+%2C+Karczewski+K.J.+%2C+O%E2%80%99Donnell-Luria+A.H.+%2C+Pierce-Hoffman+E.+%2C+MacArthur+D.G.+%2C+Neale+B.M.+%2C++and+Daly+M.J.+&title=Regional+missense+constraint+improves+variant+deleteriousness+prediction&">Google Scholar</a></li></ul></div></div><a href="#" class="dropBlock__ctrl--close" title="Close info"><i aria-hidden="true" class="icon-close_thin"></i></a></div></div> because the aimÂ was to characterize exon and domain groupings. Additionally, the Samocha<div class="dropBlock reference-citations"><sup><a href="#" id="back-bib30" data-db-target-for="bib30-d176679441e1936" class="reference-citations__ctrl" aria-expanded="false" aria-controls="bib30">30</a></sup><div data-db-target-of="bib30-d176679441e1936" class="dropBlock__holder"><div class="dropBlock__body"><ul title="list of authors" class="rlist reference-citations__loa inline-bullet-list"><li class="loa__item">Samocha K.E. </li><li class="loa__item">Kosmicki J.A. </li><li class="loa__item">Karczewski K.J. </li><li class="loa__item">OâDonnell-Luria A.H. </li><li class="loa__item">Pierce-Hoffman E. </li><li class="loa__item">MacArthur D.G. </li><li class="loa__item">Neale B.M. </li><li class="loa__item">Daly M.J. </li></ul><div class="reference-citations__detials">Regional missense constraint improves variant deleteriousness prediction.<span class="ref__series"><i>bioRxiv.</i> <span class="ref__seriesDate">2017; </span><span class="ref__seriesVolume"></span></span></div><div class="reference-citations__links"><ul class="rlist inline-bullet-list"><li><a href="/servlet/linkout?suffix=e_1_5_1_2_30_2&amp;dbid=16&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=10.1101%2F148353&amp;cf=" onclick="newWindow(this.href);return false">Crossref</a></li>
<li><a class="google-scholar" href="http://scholar.google.com/scholar_lookup?hl=en&publication_year=2017&journal=bioRxiv&author=Samocha+K.E.+%2C+Kosmicki+J.A.+%2C+Karczewski+K.J.+%2C+O%E2%80%99Donnell-Luria+A.H.+%2C+Pierce-Hoffman+E.+%2C+MacArthur+D.G.+%2C+Neale+B.M.+%2C++and+Daly+M.J.+&title=Regional+missense+constraint+improves+variant+deleteriousness+prediction&">Google Scholar</a></li></ul></div></div><a href="#" class="dropBlock__ctrl--close" title="Close info"><i aria-hidden="true" class="icon-close_thin"></i></a></div></div> approach results in partitioning a small fraction of genes into different sub-regional definitions, resulting in the majority of genes defaulting to gene-level missense Z scores. We believe this is due to low power to detect a depletion of variation in small sub-regions. However, we do believe that with increasingly larger datasets, such methods will be quite useful in refining the distribution of intolerant sub-regions within genes. Beyond the immediate contribution to variant interpretation, the implicit sharing of information across lower levels in the hierarchy allows the statistical framework utilized here to be extended to address a number of topics that have not yet been investigated. The regression formulation we propose is flexible and can accommodate additional covariates; for example, one can include additional levels in the hierarchy to model intolerance within and between populations or species. This would result in improved precision via theÂ sharing of information across groups, but also the ability to estimate population specific effects. This creates exciting opportunities, for example to develop an intolerance scoring system trained on human and mouse population genetic data which would identify genes that are under common and divergent patterns of purifying selection. LIMBR provides a natural framework for such studies which could provide important insights in the use of animal models to study human genetic disease. The model could be adapted to jointly examine missense and LoF variants as different outcomes or to incorporate prediction of how damaging individual variants are.<div class="dropBlock reference-citations"><sup><a href="#" id="back-bib31" data-db-target-for="bib31-d176679441e1940" class="reference-citations__ctrl" aria-expanded="false" aria-controls="bib31">31</a></sup><div data-db-target-of="bib31-d176679441e1940" class="dropBlock__holder"><div class="dropBlock__body"><ul title="list of authors" class="rlist reference-citations__loa inline-bullet-list"><li class="loa__item">Adzhubei I.A. </li><li class="loa__item">Schmidt S. </li><li class="loa__item">Peshkin L. </li><li class="loa__item">Ramensky V.E. </li><li class="loa__item">Gerasimova A. </li><li class="loa__item">Bork P. </li><li class="loa__item">Kondrashov A.S. </li><li class="loa__item">Sunyaev S.R. </li></ul><div class="reference-citations__detials">A method and server for predicting damaging missense mutations.<span class="ref__series"><i>Nat. Methods.</i> <span class="ref__seriesDate">2010; </span><span class="ref__seriesVolume">7</span></span>: <span class="ref__seriesPages">248-249</span></div><div class="reference-citations__links"><ul class="rlist inline-bullet-list"><li><a href="/servlet/linkout?suffix=e_1_5_1_2_31_2&amp;dbid=16&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=10.1038%2Fnmeth0410-248&amp;cf=" onclick="newWindow(this.href);return false">Crossref</a></li>
<li><a href="/servlet/linkout?suffix=e_1_5_1_2_31_2&amp;dbid=8&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=20354512&amp;cf=" onclick="newWindow(this.href);return false">PubMed</a></li>
<li><a href="/servlet/linkout?suffix=e_1_5_1_2_31_2&amp;dbid=137438953472&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=2-s2.0-77951640946&amp;cf=" onclick="newWindow(this.href);return false">Scopus (5843)</a></li>
<li><a class="google-scholar" href="http://scholar.google.com/scholar_lookup?hl=en&volume=7&publication_year=2010&pages=248-249&journal=Nat.+Methods&author=Adzhubei+I.A.+%2C+Schmidt+S.+%2C+Peshkin+L.+%2C+Ramensky+V.E.+%2C+Gerasimova+A.+%2C+Bork+P.+%2C+Kondrashov+A.S.+%2C++and+Sunyaev+S.R.+&title=A+method+and+server+for+predicting+damaging+missense+mutations&">Google Scholar</a></li></ul></div></div><a href="#" class="dropBlock__ctrl--close" title="Close info"><i aria-hidden="true" class="icon-close_thin"></i></a></div></div><sup>,Â </sup> <div class="dropBlock reference-citations"><sup><a href="#" id="back-bib32" data-db-target-for="bib32-d176679441e1940" class="reference-citations__ctrl" aria-expanded="false" aria-controls="bib32">32</a></sup><div data-db-target-of="bib32-d176679441e1940" class="dropBlock__holder"><div class="dropBlock__body"><ul title="list of authors" class="rlist reference-citations__loa inline-bullet-list"><li class="loa__item">Kircher M. </li><li class="loa__item">Witten D.M. </li><li class="loa__item">Jain P. </li><li class="loa__item">OâRoak B.J. </li><li class="loa__item">Cooper G.M. </li><li class="loa__item">Shendure J. </li></ul><div class="reference-citations__detials">A general framework for estimating the relative pathogenicity of human genetic variants.<span class="ref__series"><i>Nat. Genet.</i> <span class="ref__seriesDate">2014; </span><span class="ref__seriesVolume">46</span></span>: <span class="ref__seriesPages">310-315</span></div><div class="reference-citations__links"><ul class="rlist inline-bullet-list"><li><a href="/servlet/linkout?suffix=e_1_5_1_2_32_2&amp;dbid=16&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=10.1038%2Fng.2892&amp;cf=" onclick="newWindow(this.href);return false">Crossref</a></li>
<li><a href="/servlet/linkout?suffix=e_1_5_1_2_32_2&amp;dbid=8&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=24487276&amp;cf=" onclick="newWindow(this.href);return false">PubMed</a></li>
<li><a href="/servlet/linkout?suffix=e_1_5_1_2_32_2&amp;dbid=137438953472&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=2-s2.0-84895858942&amp;cf=" onclick="newWindow(this.href);return false">Scopus (1621)</a></li>
<li><a class="google-scholar" href="http://scholar.google.com/scholar_lookup?hl=en&volume=46&publication_year=2014&pages=310-315&journal=Nat.+Genet.&author=Kircher+M.+%2C+Witten+D.M.+%2C+Jain+P.+%2C+O%E2%80%99Roak+B.J.+%2C+Cooper+G.M.+%2C++and+Shendure+J.+&title=A+general+framework+for+estimating+the+relative+pathogenicity+of+human+genetic+variants&">Google Scholar</a></li></ul></div></div><a href="#" class="dropBlock__ctrl--close" title="Close info"><i aria-hidden="true" class="icon-close_thin"></i></a></div></div> We do not go into a detailed comparison of variant-level predictors of pathogenicity<div class="dropBlock reference-citations"><sup><a href="#" id="back-bib19" data-db-target-for="bib19-d176679441e1944" class="reference-citations__ctrl" aria-expanded="false" aria-controls="bib19">19</a></sup><div data-db-target-of="bib19-d176679441e1944" class="dropBlock__holder"><div class="dropBlock__body"><ul title="list of authors" class="rlist reference-citations__loa inline-bullet-list"><li class="loa__item">Traynelis J. </li><li class="loa__item">Silk M. </li><li class="loa__item">Wang Q. </li><li class="loa__item">Berkovic S.F. </li><li class="loa__item">Liu L. </li><li class="loa__item">Ascher D.B. </li><li class="loa__item">Balding D.J. </li><li class="loa__item">Petrovski S. </li></ul><div class="reference-citations__detials">Optimizing genomic medicine in epilepsy through a gene-customized approach to missense variant interpretation.<span class="ref__series"><i>Genome Res.</i> <span class="ref__seriesDate">2017; </span><span class="ref__seriesVolume">27</span></span>: <span class="ref__seriesPages">1715-1729</span></div><div class="reference-citations__links"><ul class="rlist inline-bullet-list"><li><a href="/servlet/linkout?suffix=e_1_5_1_2_19_2&amp;dbid=16&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=10.1101%2Fgr.226589.117&amp;cf=" onclick="newWindow(this.href);return false">Crossref</a></li>
<li><a href="/servlet/linkout?suffix=e_1_5_1_2_19_2&amp;dbid=8&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=28864458&amp;cf=" onclick="newWindow(this.href);return false">PubMed</a></li>
<li><a href="/servlet/linkout?suffix=e_1_5_1_2_19_2&amp;dbid=137438953472&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=2-s2.0-85030645609&amp;cf=" onclick="newWindow(this.href);return false">Scopus (9)</a></li>
<li><a class="google-scholar" href="http://scholar.google.com/scholar_lookup?hl=en&volume=27&publication_year=2017&pages=1715-1729&journal=Genome+Res.&author=Traynelis+J.+%2C+Silk+M.+%2C+Wang+Q.+%2C+Berkovic+S.F.+%2C+Liu+L.+%2C+Ascher+D.B.+%2C+Balding+D.J.+%2C++and+Petrovski+S.+&title=Optimizing+genomic+medicine+in+epilepsy+through+a+gene-customized+approach+to+missense+variant+interpretation&">Google Scholar</a></li></ul></div></div><a href="#" class="dropBlock__ctrl--close" title="Close info"><i aria-hidden="true" class="icon-close_thin"></i></a></div></div><sup>,Â </sup> <div class="dropBlock reference-citations"><sup><a href="#" id="back-bib30" data-db-target-for="bib30-d176679441e1944" class="reference-citations__ctrl" aria-expanded="false" aria-controls="bib30">30</a></sup><div data-db-target-of="bib30-d176679441e1944" class="dropBlock__holder"><div class="dropBlock__body"><ul title="list of authors" class="rlist reference-citations__loa inline-bullet-list"><li class="loa__item">Samocha K.E. </li><li class="loa__item">Kosmicki J.A. </li><li class="loa__item">Karczewski K.J. </li><li class="loa__item">OâDonnell-Luria A.H. </li><li class="loa__item">Pierce-Hoffman E. </li><li class="loa__item">MacArthur D.G. </li><li class="loa__item">Neale B.M. </li><li class="loa__item">Daly M.J. </li></ul><div class="reference-citations__detials">Regional missense constraint improves variant deleteriousness prediction.<span class="ref__series"><i>bioRxiv.</i> <span class="ref__seriesDate">2017; </span><span class="ref__seriesVolume"></span></span></div><div class="reference-citations__links"><ul class="rlist inline-bullet-list"><li><a href="/servlet/linkout?suffix=e_1_5_1_2_30_2&amp;dbid=16&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=10.1101%2F148353&amp;cf=" onclick="newWindow(this.href);return false">Crossref</a></li>
<li><a class="google-scholar" href="http://scholar.google.com/scholar_lookup?hl=en&publication_year=2017&journal=bioRxiv&author=Samocha+K.E.+%2C+Kosmicki+J.A.+%2C+Karczewski+K.J.+%2C+O%E2%80%99Donnell-Luria+A.H.+%2C+Pierce-Hoffman+E.+%2C+MacArthur+D.G.+%2C+Neale+B.M.+%2C++and+Daly+M.J.+&title=Regional+missense+constraint+improves+variant+deleteriousness+prediction&">Google Scholar</a></li></ul></div></div><a href="#" class="dropBlock__ctrl--close" title="Close info"><i aria-hidden="true" class="icon-close_thin"></i></a></div></div><sup>,Â </sup> <div class="dropBlock reference-citations"><sup><a href="#" id="back-bib33" data-db-target-for="bib33-d176679441e1944" class="reference-citations__ctrl" aria-expanded="false" aria-controls="bib33">33</a></sup><div data-db-target-of="bib33-d176679441e1944" class="dropBlock__holder"><div class="dropBlock__body"><ul title="list of authors" class="rlist reference-citations__loa inline-bullet-list"><li class="loa__item">Ioannidis N.M. </li><li class="loa__item">Rothstein J.H. </li><li class="loa__item">Pejaver V. </li><li class="loa__item">Middha S. </li><li class="loa__item">McDonnell S.K. </li><li class="loa__item">Baheti S. </li><li class="loa__item">Musolf A. </li><li class="loa__item">Li Q. </li><li class="loa__item">Holzinger E. </li><li class="loa__item">Karyadi D. </li><li>et al.</li></ul><div class="reference-citations__detials">REVEL: An Ensemble Method for Predicting the Pathogenicity of Rare Missense Variants.<span class="ref__series"><i>Am. J. Hum. Genet.</i> <span class="ref__seriesDate">2016; </span><span class="ref__seriesVolume">99</span></span>: <span class="ref__seriesPages">877-885</span></div><div class="reference-citations__links"><ul class="rlist inline-bullet-list"><li><a href="/servlet/linkout?suffix=e_1_5_1_2_33_2&amp;dbid=4&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=10.1016%2Fj.ajhg.2016.08.016&amp;cf=abstract&amp;site=cell-site" onclick="newWindow(this.href);return false">Abstract</a></li>
<li><a href="/servlet/linkout?suffix=e_1_5_1_2_33_2&amp;dbid=4&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=10.1016%2Fj.ajhg.2016.08.016&amp;cf=fulltext&amp;site=cell-site" onclick="newWindow(this.href);return false">Full Text</a></li>
<li><a href="/servlet/linkout?suffix=e_1_5_1_2_33_2&amp;dbid=4&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=10.1016%2Fj.ajhg.2016.08.016&amp;cf=pdf&amp;site=cell-site" onclick="newWindow(this.href);return false">Full Text PDF</a></li>
<li><a href="/servlet/linkout?suffix=e_1_5_1_2_33_2&amp;dbid=8&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=27666373&amp;cf=" onclick="newWindow(this.href);return false">PubMed</a></li>
<li><a class="google-scholar" href="http://scholar.google.com/scholar_lookup?hl=en&volume=99&publication_year=2016&pages=877-885&journal=Am.+J.+Hum.+Genet.&author=Ioannidis+N.M.+%2C+Rothstein+J.H.+%2C+Pejaver+V.+%2C+Middha+S.+%2C+McDonnell+S.K.+%2C+Baheti+S.+%2C+Musolf+A.+%2C+Li+Q.+%2C+Holzinger+E.+%2C+Karyadi+D.++et+al.&title=REVEL%3A+An+Ensemble+Method+for+Predicting+the+Pathogenicity+of+Rare+Missense+Variants&">Google Scholar</a></li></ul></div></div><a href="#" class="dropBlock__ctrl--close" title="Close info"><i aria-hidden="true" class="icon-close_thin"></i></a></div></div> in this paper because many of these methods use information on pathogenic variants to develop their scores, and they also may not account for gene-level effects. These scores may still be of interest in conjunction with LIMBR for variant interpretation, especially by integrating methods that use orthogonal information to characterize how damaging a variant is. We are currently working on expanding our tests to explore these avenues, and as more data become available we will be able to directly address regional variation in more specialized settings.</div><div class="section-paragraph">The Bayesian framework also allows for different priors to be put on the model, potentially leading to shrinkage of genes or sub-regions to improve classification of novel variants. Overall, the Bayesian hierarchical intolerance model is a flexible and robust way to model sub-regional intolerance to variation, showing strong classification of previously identified pathogenic variants while demonstrating less susceptibility to variability in estimates due to sub-region size.</div></section><section id="sec5" data-locator="sec5"><h2 class="top" id="secsectitle0070"><span class="top__sub"></span><span class="top__text">Declaration of Interests</span></h2><div class="section-paragraph">D.B.G. is a founder of and holds equity in Pairnomix and PraxisÂ and has research supported by Janssen, Gilead, Biogen, AstraZeneca, and UCB. A.S.A. and T.J.H. have research supported by AstraZeneca. D.B.G. received consultancy payments from AstraZeneca ending August 2017. A.S.A. received consultancy payments from AstraZeneca ending June 2018. There are no patents, products in development, or marketed products to declare related to this work. This does not alter the authorsâ adherence to all <em>American Journal of Human Genetics</em> policies on sharing data and materials.</div></section><section><h2 class="top" id="secsectitle0075"><span class="top__sub"></span><span class="top__text">Acknowledgments</span></h2><div class="section-paragraph">This work was supported by funding from Biogen and AstraZeneca.</div></section><section id="app2" data-locator="app2"><h2 class="top" id="secsectitle0085"><span class="top__sub"></span><span class="top__text">Supplemental Data</span></h2><div class="section-paragraph supplemental-information"><ul class="rlist"><li class="supplemental-information__item"><div class="supplemental-information__icon icon-gizmo-pdf-file2"></div><div class="supplemental-information__links clearfix"><div class="col-sm-6 gutterless"><a href="/cms/10.1016/j.ajhg.2018.12.020/attachment/42ba8291-bbf5-4e37-8292-534540b90141/mmc1.pdf" class="supplemental-information__download " target="_blank">
                                    Download .pdf (1.29
                                    MB)
                                </a></div><div class="col-sm-6 gutterless"><a href="https://service.elsevier.com/app/answers/detail/a_id/19286/supporthub/sciencedirect/" class="supplemental-information__help " target="_blank">
                                    Help with
                                    pdf files
                                </a></div></div><div class="supplemental-information__title"><div>Document S1. Figures S1âS5</div></div></li></ul><ul class="rlist"><li class="supplemental-information__item"><div class="supplemental-information__icon icon-gizmo-interactive-case-insights"></div><div class="supplemental-information__links clearfix"><div class="col-sm-6 gutterless"><a href="/cms/10.1016/j.ajhg.2018.12.020/attachment/8d42c908-07db-4f91-995c-87a0307e0d79/mmc2.xlsx" class="supplemental-information__download " target="_blank">
                                    Download .xlsx (20.28
                                    MB)
                                </a></div><div class="col-sm-6 gutterless"><a href="https://service.elsevier.com/app/answers/detail/a_id/19286/supporthub/sciencedirect/" class="supplemental-information__help " target="_blank">
                                    Help with
                                    xlsx files
                                </a></div></div><div class="supplemental-information__title"><div>Table S1. LIMBR Statistics</div></div></li></ul><ul class="rlist"><li class="supplemental-information__item"><div class="supplemental-information__icon icon-gizmo-interactive-case-insights"></div><div class="supplemental-information__links clearfix"><div class="col-sm-6 gutterless"><a href="/cms/10.1016/j.ajhg.2018.12.020/attachment/04592b59-2b21-44fc-aa10-05228ba69e55/mmc3.xlsx" class="supplemental-information__download " target="_blank">
                                    Download .xlsx (.02
                                    MB)
                                </a></div><div class="col-sm-6 gutterless"><a href="https://service.elsevier.com/app/answers/detail/a_id/19286/supporthub/sciencedirect/" class="supplemental-information__help " target="_blank">
                                    Help with
                                    xlsx files
                                </a></div></div><div class="supplemental-information__title"><div>Table S2. Gene Permutation Test</div></div></li></ul></div></section><section id="app3" data-locator="app3"><h2 class="top" id="secsectitle0090"><span class="top__sub"></span><span class="top__text">Web Resources</span></h2><div class="section-paragraph"><ul class="ce-list--remove-bullets" id="ulist0010"><li class="ce-list--remove-bullets__list-item" id="u0010"><div class="ce-list--remove-bullets__list-item__text">ClinVar, <a href="https://www.ncbi.nlm.nih.gov/clinvar/" target="_blank">https://www.ncbi.nlm.nih.gov/clinvar/</a></div></li><li class="ce-list--remove-bullets__list-item" id="u0015"><div class="ce-list--remove-bullets__list-item__text">gnomAD Browser, <a href="http://gnomad.broadinstitute.org/" target="_blank">http://gnomad.broadinstitute.org/</a></div></li><li class="ce-list--remove-bullets__list-item" id="u0020"><div class="ce-list--remove-bullets__list-item__text">OMIM, <a href="http://www.omim.org/" target="_blank">http://www.omim.org/</a></div></li><li class="ce-list--remove-bullets__list-item" id="u0025"><div class="ce-list--remove-bullets__list-item__text">Stan Userâs Guide, <a href="https://mc-stan.org/users/documentation/" target="_blank">https://mc-stan.org/users/documentation/</a></div></li></ul></div></section><div data-pb-dropzone="extra-dropzone" class="extra-dropzone pb-dropzone"></div><section class="refs"><h2 class="top" id="secsectitle0095"><span class="top__sub"></span><span class="top__text">References</span></h2><div class="ref"><ul title="list of authors" class="rlist inline-bullet-list ref__authors ref__inline-block"><li class="ref__authors__name">Goldstein D.B. </li><li class="ref__authors__name">Allen A. </li><li class="ref__authors__name">Keebler J. </li><li class="ref__authors__name">Margulies E.H. </li><li class="ref__authors__name">Petrou S. </li><li class="ref__authors__name">Petrovski S. </li><li class="ref__authors__name">Sunyaev S. </li></ul><div class="ref__title">Sequencing studies in human genetics: design and interpretation.</div><span class="ref__series"><i>Nat. Rev. Genet.</i> <span class="ref__seriesDate">2013; </span><span class="ref__seriesVolume">14</span></span>: <span class="ref__seriesPages">460-470</span><div class="ref__controls"><a href="#back-bib1" title="View in Article" class="ref__controls__viewInArticle scroll-into-link">View
                        in Article
                        <i class="icon-section_arrow_u"></i></a><ul class="rlist inline-bullet-list"><li><a href="/servlet/linkout?suffix=e_1_5_1_2_1_2&amp;dbid=137438953472&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=2-s2.0-84879411643&amp;cf=" onclick="newWindow(this.href);return false">Scopus (141)</a></li>
<li><a href="/servlet/linkout?suffix=e_1_5_1_2_1_2&amp;dbid=8&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=23752795&amp;cf=" onclick="newWindow(this.href);return false">PubMed</a></li>
<li><a href="/servlet/linkout?suffix=e_1_5_1_2_1_2&amp;dbid=16&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=10.1038%2Fnrg3455&amp;cf=" onclick="newWindow(this.href);return false">Crossref</a></li>
<li><a class="google-scholar" href="http://scholar.google.com/scholar_lookup?hl=en&volume=14&publication_year=2013&pages=460-470&journal=Nat.+Rev.+Genet.&author=Goldstein+D.B.+%2C+Allen+A.+%2C+Keebler+J.+%2C+Margulies+E.H.+%2C+Petrou+S.+%2C+Petrovski+S.+%2C++and+Sunyaev+S.+&title=Sequencing+studies+in+human+genetics%3A+design+and+interpretation&">Google Scholar</a></li></ul></div></div><div class="ref"><ul title="list of authors" class="rlist inline-bullet-list ref__authors ref__inline-block"><li class="ref__authors__name">Eilbeck K. </li><li class="ref__authors__name">Quinlan A. </li><li class="ref__authors__name">Yandell M. </li></ul><div class="ref__title">Settling the score: variant prioritization and Mendelian disease.</div><span class="ref__series"><i>Nat. Rev. Genet.</i> <span class="ref__seriesDate">2017; </span><span class="ref__seriesVolume">18</span></span>: <span class="ref__seriesPages">599-612</span><div class="ref__controls"><a href="#back-bib2" title="View in Article" class="ref__controls__viewInArticle scroll-into-link">View
                        in Article
                        <i class="icon-section_arrow_u"></i></a><ul class="rlist inline-bullet-list"><li><a href="/servlet/linkout?suffix=e_1_5_1_2_2_2&amp;dbid=137438953472&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=2-s2.0-85029470345&amp;cf=" onclick="newWindow(this.href);return false">Scopus (28)</a></li>
<li><a href="/servlet/linkout?suffix=e_1_5_1_2_2_2&amp;dbid=8&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=28804138&amp;cf=" onclick="newWindow(this.href);return false">PubMed</a></li>
<li><a href="/servlet/linkout?suffix=e_1_5_1_2_2_2&amp;dbid=16&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=10.1038%2Fnrg.2017.52&amp;cf=" onclick="newWindow(this.href);return false">Crossref</a></li>
<li><a class="google-scholar" href="http://scholar.google.com/scholar_lookup?hl=en&volume=18&publication_year=2017&pages=599-612&journal=Nat.+Rev.+Genet.&author=Eilbeck+K.+%2C+Quinlan+A.+%2C++and+Yandell+M.+&title=Settling+the+score%3A+variant+prioritization+and+Mendelian+disease&">Google Scholar</a></li></ul></div></div><div class="ref"><ul title="list of authors" class="rlist inline-bullet-list ref__authors ref__inline-block"><li class="ref__authors__name">Davydov E.V. </li><li class="ref__authors__name">Goode D.L. </li><li class="ref__authors__name">Sirota M. </li><li class="ref__authors__name">Cooper G.M. </li><li class="ref__authors__name">Sidow A. </li><li class="ref__authors__name">Batzoglou S. </li></ul><div class="ref__title">Identifying a high fraction of the human genome to be under selective constraint using GERP++.</div><span class="ref__series"><i>PLoS Comput. Biol.</i> <span class="ref__seriesDate">2010; </span><span class="ref__seriesVolume">6</span></span>: <span class="ref__seriesPages">e1001025</span><div class="ref__controls"><a href="#back-bib3" title="View in Article" class="ref__controls__viewInArticle scroll-into-link">View
                        in Article
                        <i class="icon-section_arrow_u"></i></a><ul class="rlist inline-bullet-list"><li><a href="/servlet/linkout?suffix=e_1_5_1_2_3_2&amp;dbid=137438953472&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=2-s2.0-78651237647&amp;cf=" onclick="newWindow(this.href);return false">Scopus (510)</a></li>
<li><a href="/servlet/linkout?suffix=e_1_5_1_2_3_2&amp;dbid=8&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=21152010&amp;cf=" onclick="newWindow(this.href);return false">PubMed</a></li>
<li><a href="/servlet/linkout?suffix=e_1_5_1_2_3_2&amp;dbid=16&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=10.1371%2Fjournal.pcbi.1001025&amp;cf=" onclick="newWindow(this.href);return false">Crossref</a></li>
<li><a class="google-scholar" href="http://scholar.google.com/scholar_lookup?hl=en&volume=6&publication_year=2010&pages=e1001025&journal=PLoS+Comput.+Biol.&author=Davydov+E.V.+%2C+Goode+D.L.+%2C+Sirota+M.+%2C+Cooper+G.M.+%2C+Sidow+A.+%2C++and+Batzoglou+S.+&title=Identifying+a+high+fraction+of+the+human+genome+to+be+under+selective+constraint+using+GERP%2B%2B&">Google Scholar</a></li></ul></div></div><div class="ref"><ul title="list of authors" class="rlist inline-bullet-list ref__authors ref__inline-block"><li class="ref__authors__name">Petrovski S. </li><li class="ref__authors__name">Wang Q. </li><li class="ref__authors__name">Heinzen E.L. </li><li class="ref__authors__name">Allen A.S. </li><li class="ref__authors__name">Goldstein D.B. </li></ul><div class="ref__title">Genic intolerance to functional variation and the interpretation of personal genomes.</div><span class="ref__series"><i>PLoS Genet.</i> <span class="ref__seriesDate">2013; </span><span class="ref__seriesVolume">9</span></span>: <span class="ref__seriesPages">e1003709</span><div class="ref__controls"><a href="#back-bib4" title="View in Article" class="ref__controls__viewInArticle scroll-into-link">View
                        in Article
                        <i class="icon-section_arrow_u"></i></a><ul class="rlist inline-bullet-list"><li><a href="/servlet/linkout?suffix=e_1_5_1_2_4_2&amp;dbid=137438953472&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=2-s2.0-84884592445&amp;cf=" onclick="newWindow(this.href);return false">Scopus (379)</a></li>
<li><a href="/servlet/linkout?suffix=e_1_5_1_2_4_2&amp;dbid=8&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=23990802&amp;cf=" onclick="newWindow(this.href);return false">PubMed</a></li>
<li><a href="/servlet/linkout?suffix=e_1_5_1_2_4_2&amp;dbid=16&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=10.1371%2Fjournal.pgen.1003709&amp;cf=" onclick="newWindow(this.href);return false">Crossref</a></li>
<li><a class="google-scholar" href="http://scholar.google.com/scholar_lookup?hl=en&volume=9&publication_year=2013&pages=e1003709&journal=PLoS+Genet.&author=Petrovski+S.+%2C+Wang+Q.+%2C+Heinzen+E.L.+%2C+Allen+A.S.+%2C++and+Goldstein+D.B.+&title=Genic+intolerance+to+functional+variation+and+the+interpretation+of+personal+genomes&">Google Scholar</a></li></ul></div></div><div class="ref"><ul title="list of authors" class="rlist inline-bullet-list ref__authors ref__inline-block"><li class="ref__authors__name">Samocha K.E. </li><li class="ref__authors__name">Robinson E.B. </li><li class="ref__authors__name">Sanders S.J. </li><li class="ref__authors__name">Stevens C. </li><li class="ref__authors__name">Sabo A. </li><li class="ref__authors__name">McGrath L.M. </li><li class="ref__authors__name">Kosmicki J.A. </li><li class="ref__authors__name">RehnstrÃ¶m K. </li><li class="ref__authors__name">Mallick S. </li><li class="ref__authors__name">Kirby A. </li><li>et al.</li></ul><div class="ref__title">A framework for the interpretation of de novo mutation in human disease.</div><span class="ref__series"><i>Nat. Genet.</i> <span class="ref__seriesDate">2014; </span><span class="ref__seriesVolume">46</span></span>: <span class="ref__seriesPages">944-950</span><div class="ref__controls"><a href="#back-bib5" title="View in Article" class="ref__controls__viewInArticle scroll-into-link">View
                        in Article
                        <i class="icon-section_arrow_u"></i></a><ul class="rlist inline-bullet-list"><li><a href="/servlet/linkout?suffix=e_1_5_1_2_5_2&amp;dbid=137438953472&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=2-s2.0-84905309176&amp;cf=" onclick="newWindow(this.href);return false">Scopus (322)</a></li>
<li><a href="/servlet/linkout?suffix=e_1_5_1_2_5_2&amp;dbid=8&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=25086666&amp;cf=" onclick="newWindow(this.href);return false">PubMed</a></li>
<li><a href="/servlet/linkout?suffix=e_1_5_1_2_5_2&amp;dbid=16&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=10.1038%2Fng.3050&amp;cf=" onclick="newWindow(this.href);return false">Crossref</a></li>
<li><a class="google-scholar" href="http://scholar.google.com/scholar_lookup?hl=en&volume=46&publication_year=2014&pages=944-950&journal=Nat.+Genet.&author=Samocha+K.E.+%2C+Robinson+E.B.+%2C+Sanders+S.J.+%2C+Stevens+C.+%2C+Sabo+A.+%2C+McGrath+L.M.+%2C+Kosmicki+J.A.+%2C+Rehnstr%C3%B6m+K.+%2C+Mallick+S.+%2C+Kirby+A.++et+al.&title=A+framework+for+the+interpretation+of+de+novo+mutation+in+human+disease&">Google Scholar</a></li></ul></div></div><div class="ref"><ul title="list of authors" class="rlist inline-bullet-list ref__authors ref__inline-block"><li class="ref__authors__name">Iossifov I. </li><li class="ref__authors__name">Ronemus M. </li><li class="ref__authors__name">Levy D. </li><li class="ref__authors__name">Wang Z. </li><li class="ref__authors__name">Hakker I. </li><li class="ref__authors__name">Rosenbaum J. </li><li class="ref__authors__name">Yamrom B. </li><li class="ref__authors__name">Lee Y.H. </li><li class="ref__authors__name">Narzisi G. </li><li class="ref__authors__name">Leotta A. </li><li>et al.</li></ul><div class="ref__title">De novo gene disruptions in children on the autistic spectrum.</div><span class="ref__series"><i>Neuron.</i> <span class="ref__seriesDate">2012; </span><span class="ref__seriesVolume">74</span></span>: <span class="ref__seriesPages">285-299</span><div class="ref__controls"><a href="#back-bib6" title="View in Article" class="ref__controls__viewInArticle scroll-into-link">View
                        in Article
                        <i class="icon-section_arrow_u"></i></a><ul class="rlist inline-bullet-list"><li><a href="/servlet/linkout?suffix=e_1_5_1_2_6_2&amp;dbid=137438953472&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=2-s2.0-84860297457&amp;cf=" onclick="newWindow(this.href);return false">Scopus (735)</a></li>
<li><a href="/servlet/linkout?suffix=e_1_5_1_2_6_2&amp;dbid=8&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=22542183&amp;cf=" onclick="newWindow(this.href);return false">PubMed</a></li>
<li><a href="/servlet/linkout?suffix=e_1_5_1_2_6_2&amp;dbid=4&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=10.1016%2Fj.neuron.2012.04.009&amp;cf=abstract&amp;site=cell-site" onclick="newWindow(this.href);return false">Abstract</a></li>
<li><a href="/servlet/linkout?suffix=e_1_5_1_2_6_2&amp;dbid=4&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=10.1016%2Fj.neuron.2012.04.009&amp;cf=fulltext&amp;site=cell-site" onclick="newWindow(this.href);return false">Full Text</a></li>
<li><a href="/servlet/linkout?suffix=e_1_5_1_2_6_2&amp;dbid=4&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=10.1016%2Fj.neuron.2012.04.009&amp;cf=pdf&amp;site=cell-site" onclick="newWindow(this.href);return false">Full Text PDF</a></li>
<li><a class="google-scholar" href="http://scholar.google.com/scholar_lookup?hl=en&volume=74&publication_year=2012&pages=285-299&journal=Neuron&author=Iossifov+I.+%2C+Ronemus+M.+%2C+Levy+D.+%2C+Wang+Z.+%2C+Hakker+I.+%2C+Rosenbaum+J.+%2C+Yamrom+B.+%2C+Lee+Y.H.+%2C+Narzisi+G.+%2C+Leotta+A.++et+al.&title=De+novo+gene+disruptions+in+children+on+the+autistic+spectrum&">Google Scholar</a></li></ul></div></div><div class="ref"><ul title="list of authors" class="rlist inline-bullet-list ref__authors ref__inline-block"><li class="ref__authors__name">Gussow A.B. </li><li class="ref__authors__name">Petrovski S. </li><li class="ref__authors__name">Wang Q. </li><li class="ref__authors__name">Allen A.S. </li><li class="ref__authors__name">Goldstein D.B. </li></ul><div class="ref__title">The intolerance to functional genetic variation of protein domains predicts the localization of pathogenic mutations within genes.</div><span class="ref__series"><i>Genome Biol.</i> <span class="ref__seriesDate">2016; </span><span class="ref__seriesVolume">17</span></span>: <span class="ref__seriesPages">9</span><div class="ref__controls"><a href="#back-bib7" title="View in Article" class="ref__controls__viewInArticle scroll-into-link">View
                        in Article
                        <i class="icon-section_arrow_u"></i></a><ul class="rlist inline-bullet-list"><li><a href="/servlet/linkout?suffix=e_1_5_1_2_7_2&amp;dbid=8&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=26781712&amp;cf=" onclick="newWindow(this.href);return false">PubMed</a></li>
<li><a href="/servlet/linkout?suffix=e_1_5_1_2_7_2&amp;dbid=16&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=10.1186%2Fs13059-016-0869-4&amp;cf=" onclick="newWindow(this.href);return false">Crossref</a></li>
<li><a class="google-scholar" href="http://scholar.google.com/scholar_lookup?hl=en&volume=17&publication_year=2016&pages=9&journal=Genome+Biol.&author=Gussow+A.B.+%2C+Petrovski+S.+%2C+Wang+Q.+%2C+Allen+A.S.+%2C++and+Goldstein+D.B.+&title=The+intolerance+to+functional+genetic+variation+of+protein+domains+predicts+the+localization+of+pathogenic+mutations+within+genes&">Google Scholar</a></li></ul></div></div><div class="ref"><ul title="list of authors" class="rlist inline-bullet-list ref__authors ref__inline-block"><li class="ref__authors__name">Lek M. </li><li class="ref__authors__name">Karczewski K.J. </li><li class="ref__authors__name">Minikel E.V. </li><li class="ref__authors__name">Samocha K.E. </li><li class="ref__authors__name">Banks E. </li><li class="ref__authors__name">Fennell T. </li><li class="ref__authors__name">OâDonnell-Luria A.H. </li><li class="ref__authors__name">Ware J.S. </li><li class="ref__authors__name">Hill A.J. </li><li class="ref__authors__name">Cummings B.B. </li><li>et al.</li><li>Exome Aggregation Consortium</li></ul><div class="ref__title">Analysis of protein-coding genetic variation in 60,706 humans.</div><span class="ref__series"><i>Nature.</i> <span class="ref__seriesDate">2016; </span><span class="ref__seriesVolume">536</span></span>: <span class="ref__seriesPages">285-291</span><div class="ref__controls"><a href="#back-bib8" title="View in Article" class="ref__controls__viewInArticle scroll-into-link">View
                        in Article
                        <i class="icon-section_arrow_u"></i></a><ul class="rlist inline-bullet-list"><li><a href="/servlet/linkout?suffix=e_1_5_1_2_8_2&amp;dbid=8&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=27535533&amp;cf=" onclick="newWindow(this.href);return false">PubMed</a></li>
<li><a href="/servlet/linkout?suffix=e_1_5_1_2_8_2&amp;dbid=16&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=10.1038%2Fnature19057&amp;cf=" onclick="newWindow(this.href);return false">Crossref</a></li>
<li><a class="google-scholar" href="http://scholar.google.com/scholar_lookup?hl=en&volume=536&publication_year=2016&pages=285-291&journal=Nature&author=Exome+Aggregation+Consortium&title=Analysis+of+protein-coding+genetic+variation+in+60%2C706+humans&">Google Scholar</a></li></ul></div></div><div class="ref"><ul title="list of authors" class="rlist inline-bullet-list ref__authors ref__inline-block"><li class="ref__authors__name">Bartha I. </li><li class="ref__authors__name">di Iulio J. </li><li class="ref__authors__name">Venter J.C. </li><li class="ref__authors__name">Telenti A. </li></ul><div class="ref__title">Human gene essentiality.</div><span class="ref__series"><i>Nat. Rev. Genet.</i> <span class="ref__seriesDate">2018; </span><span class="ref__seriesVolume">19</span></span>: <span class="ref__seriesPages">51-62</span><div class="ref__controls"><a href="#back-bib9" title="View in Article" class="ref__controls__viewInArticle scroll-into-link">View
                        in Article
                        <i class="icon-section_arrow_u"></i></a><ul class="rlist inline-bullet-list"><li><a href="/servlet/linkout?suffix=e_1_5_1_2_9_2&amp;dbid=137438953472&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=2-s2.0-85039852688&amp;cf=" onclick="newWindow(this.href);return false">Scopus (4)</a></li>
<li><a href="/servlet/linkout?suffix=e_1_5_1_2_9_2&amp;dbid=8&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=29082913&amp;cf=" onclick="newWindow(this.href);return false">PubMed</a></li>
<li><a href="/servlet/linkout?suffix=e_1_5_1_2_9_2&amp;dbid=16&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=10.1038%2Fnrg.2017.75&amp;cf=" onclick="newWindow(this.href);return false">Crossref</a></li>
<li><a class="google-scholar" href="http://scholar.google.com/scholar_lookup?hl=en&volume=19&publication_year=2018&pages=51-62&journal=Nat.+Rev.+Genet.&author=Bartha+I.+%2C+di+Iulio+J.+%2C+Venter+J.C.+%2C++and+Telenti+A.+&title=Human+gene+essentiality&">Google Scholar</a></li></ul></div></div><div class="ref"><ul title="list of authors" class="rlist inline-bullet-list ref__authors ref__inline-block"><li class="ref__authors__name">Papaspiliopoulos O. </li><li class="ref__authors__name">Roberts G.O. </li></ul><div class="ref__title">Non-centered parameterisations for hierarchical models and data augmentation.</div><span class="ref__series"><i>Bayesian Stat.</i> <span class="ref__seriesDate">2003; </span><span class="ref__seriesVolume">7</span></span>: <span class="ref__seriesPages">307-326</span><div class="ref__controls"><a href="#back-bib10" title="View in Article" class="ref__controls__viewInArticle scroll-into-link">View
                        in Article
                        <i class="icon-section_arrow_u"></i></a><ul class="rlist inline-bullet-list"><li><a class="google-scholar" href="http://scholar.google.com/scholar_lookup?hl=en&volume=7&publication_year=2003&pages=307-326&journal=Bayesian+Stat.&author=Papaspiliopoulos+O.++and+Roberts+G.O.+&title=Non-centered+parameterisations+for+hierarchical+models+and+data+augmentation&">Google Scholar</a></li></ul></div></div><div class="ref"><ul title="list of authors" class="rlist inline-bullet-list ref__authors ref__inline-block"><li class="ref__authors__name">Papaspiliopoulos O. </li><li class="ref__authors__name">Roberts G. </li></ul><div class="ref__title">Stability of the Gibbs sampler for Bayesian hierarchical models.</div><span class="ref__series"><i>Ann. Stat.</i> <span class="ref__seriesDate">2008; </span><span class="ref__seriesVolume">36</span></span>: <span class="ref__seriesPages">95-117</span><div class="ref__controls"><a href="#back-bib11" title="View in Article" class="ref__controls__viewInArticle scroll-into-link">View
                        in Article
                        <i class="icon-section_arrow_u"></i></a><ul class="rlist inline-bullet-list"><li><a href="/servlet/linkout?suffix=e_1_5_1_2_11_2&amp;dbid=137438953472&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=2-s2.0-50649109863&amp;cf=" onclick="newWindow(this.href);return false">Scopus (24)</a></li>
<li><a href="/servlet/linkout?suffix=e_1_5_1_2_11_2&amp;dbid=16&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=10.1214%2F009053607000000749&amp;cf=" onclick="newWindow(this.href);return false">Crossref</a></li>
<li><a class="google-scholar" href="http://scholar.google.com/scholar_lookup?hl=en&volume=36&publication_year=2008&pages=95-117&journal=Ann.+Stat.&author=Papaspiliopoulos+O.++and+Roberts+G.+&title=Stability+of+the+Gibbs+sampler+for+Bayesian+hierarchical+models&">Google Scholar</a></li></ul></div></div><div class="ref"><ul title="list of authors" class="rlist inline-bullet-list ref__authors ref__inline-block"><li class="ref__authors__name">Betancourt M. </li></ul><div class="ref__title">A general metric for Riemannian Manifold Hamiltonian Monte Carlo.</div><span class="ref__series"><i>Nature.</i> <span class="ref__seriesDate">2013; </span><span class="ref__seriesVolume">9</span></span>: <span class="ref__seriesPages">327-334</span><div class="ref__controls"><a href="#back-bib12" title="View in Article" class="ref__controls__viewInArticle scroll-into-link">View
                        in Article
                        <i class="icon-section_arrow_u"></i></a><ul class="rlist inline-bullet-list"><li><a class="google-scholar" href="http://scholar.google.com/scholar_lookup?hl=en&volume=9&publication_year=2013&pages=327-334&journal=Nature&author=Betancourt+M.+&title=A+general+metric+for+Riemannian+Manifold+Hamiltonian+Monte+Carlo&">Google Scholar</a></li></ul></div></div><div class="ref"><ul title="list of authors" class="rlist inline-bullet-list ref__authors ref__inline-block"><li class="ref__authors__name">Yu Y. </li><li class="ref__authors__name">Meng X.-L. </li></ul><div class="ref__title">To center or not to center: that is not the questionâAn ancillarityâsufficiency interweaving strategy (ASIS) for boosting MCMC efficiency.</div><span class="ref__series"><i>J.Â Comput. Graph. Stat.</i> <span class="ref__seriesDate">2011; </span><span class="ref__seriesVolume">20</span></span>: <span class="ref__seriesPages">531-570</span><div class="ref__controls"><a href="#back-bib13" title="View in Article" class="ref__controls__viewInArticle scroll-into-link">View
                        in Article
                        <i class="icon-section_arrow_u"></i></a><ul class="rlist inline-bullet-list"><li><a href="/servlet/linkout?suffix=e_1_5_1_2_13_2&amp;dbid=137438953472&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=2-s2.0-80053346023&amp;cf=" onclick="newWindow(this.href);return false">Scopus (55)</a></li>
<li><a href="/servlet/linkout?suffix=e_1_5_1_2_13_2&amp;dbid=16&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=10.1198%2Fjcgs.2011.203main&amp;cf=" onclick="newWindow(this.href);return false">Crossref</a></li>
<li><a class="google-scholar" href="http://scholar.google.com/scholar_lookup?hl=en&volume=20&publication_year=2011&pages=531-570&journal=J.%C2%A0Comput.+Graph.+Stat.&author=Yu+Y.++and+Meng+X.-L.+&title=To+center+or+not+to+center%3A+that+is+not+the+question%E2%80%94An+ancillarity%E2%80%93sufficiency+interweaving+strategy+%28ASIS%29+for+boosting+MCMC+efficiency&">Google Scholar</a></li></ul></div></div><div class="ref"><ul title="list of authors" class="rlist inline-bullet-list ref__authors ref__inline-block"><li class="ref__authors__name">Duan L.L. </li><li class="ref__authors__name">Johndrow J.E. </li><li class="ref__authors__name">Dunson D.B. </li></ul><div class="ref__title">Scaling up data augmentation MCMC via calibration.</div><span class="ref__series"><i>arXiv.</i> <span class="ref__seriesDate">2017; </span><span class="ref__seriesVolume"></span></span><div class="citation-host"><a href="https://arxiv.org/abs/1703.03123" target="_blank">https://arxiv.org/abs/1703.03123</a></div><div class="ref__controls"><a href="#back-bib14" title="View in Article" class="ref__controls__viewInArticle scroll-into-link">View
                        in Article
                        <i class="icon-section_arrow_u"></i></a><ul class="rlist inline-bullet-list"><li><a class="google-scholar" href="http://scholar.google.com/scholar_lookup?hl=en&publication_year=2017&journal=arXiv&author=Duan+L.L.+%2C+Johndrow+J.E.+%2C++and+Dunson+D.B.+&title=Scaling+up+data+augmentation+MCMC+via+calibration&">Google Scholar</a></li></ul></div></div><div class="ref"><ul title="list of authors" class="rlist inline-bullet-list ref__authors ref__inline-block"><li class="ref__authors__name">Liu J.S. </li></ul><div class="ref__title">The collapsed Gibbs sampler with applications to a gene regulation problem.</div><span class="ref__series"><i>J.Â Am. Stat. Assoc.</i> <span class="ref__seriesDate">1994; </span><span class="ref__seriesVolume">89</span></span>: <span class="ref__seriesPages">958-966</span><div class="ref__controls"><a href="#back-bib15" title="View in Article" class="ref__controls__viewInArticle scroll-into-link">View
                        in Article
                        <i class="icon-section_arrow_u"></i></a><ul class="rlist inline-bullet-list"><li><a href="/servlet/linkout?suffix=e_1_5_1_2_15_2&amp;dbid=16&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=10.1080%2F01621459.1994.10476829&amp;cf=" onclick="newWindow(this.href);return false">Crossref</a></li>
<li><a class="google-scholar" href="http://scholar.google.com/scholar_lookup?hl=en&volume=89&publication_year=1994&pages=958-966&journal=J.%C2%A0Am.+Stat.+Assoc.&author=Liu+J.S.+&title=The+collapsed+Gibbs+sampler+with+applications+to+a+gene+regulation+problem&">Google Scholar</a></li></ul></div></div><div class="ref"><ul title="list of authors" class="rlist inline-bullet-list ref__authors ref__inline-block"><li class="ref__authors__name">Duane S. </li><li class="ref__authors__name">Kennedy A.D. </li><li class="ref__authors__name">Pendleton B.J. </li><li class="ref__authors__name">Roweth D. </li></ul><div class="ref__title">Hybrid Monte Carlo.</div><span class="ref__series"><i>Phys. Lett. B.</i> <span class="ref__seriesDate">1987; </span><span class="ref__seriesVolume">195</span></span>: <span class="ref__seriesPages">216-222</span><div class="ref__controls"><a href="#back-bib16" title="View in Article" class="ref__controls__viewInArticle scroll-into-link">View
                        in Article
                        <i class="icon-section_arrow_u"></i></a><ul class="rlist inline-bullet-list"><li><a href="/servlet/linkout?suffix=e_1_5_1_2_16_2&amp;dbid=137438953472&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=2-s2.0-4243137056&amp;cf=" onclick="newWindow(this.href);return false">Scopus (0)</a></li>
<li><a href="/servlet/linkout?suffix=e_1_5_1_2_16_2&amp;dbid=16&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=10.1016%2F0370-2693%2887%2991197-X&amp;cf=" onclick="newWindow(this.href);return false">Crossref</a></li>
<li><a class="google-scholar" href="http://scholar.google.com/scholar_lookup?hl=en&volume=195&publication_year=1987&pages=216-222&journal=Phys.+Lett.+B&author=Duane+S.+%2C+Kennedy+A.D.+%2C+Pendleton+B.J.+%2C++and+Roweth+D.+&title=Hybrid+Monte+Carlo&">Google Scholar</a></li></ul></div></div><div class="ref"><ul title="list of authors" class="rlist inline-bullet-list ref__authors ref__inline-block"><li class="ref__authors__name">Madsen B.E. </li><li class="ref__authors__name">Browning S.R. </li></ul><div class="ref__title">A groupwise association test for rare mutations using a weighted sum statistic.</div><span class="ref__series"><i>PLoS Genet.</i> <span class="ref__seriesDate">2009; </span><span class="ref__seriesVolume">5</span></span>: <span class="ref__seriesPages">e1000384</span><div class="ref__controls"><a href="#back-bib17" title="View in Article" class="ref__controls__viewInArticle scroll-into-link">View
                        in Article
                        <i class="icon-section_arrow_u"></i></a><ul class="rlist inline-bullet-list"><li><a href="/servlet/linkout?suffix=e_1_5_1_2_17_2&amp;dbid=137438953472&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=2-s2.0-61449168010&amp;cf=" onclick="newWindow(this.href);return false">Scopus (648)</a></li>
<li><a href="/servlet/linkout?suffix=e_1_5_1_2_17_2&amp;dbid=8&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=19214210&amp;cf=" onclick="newWindow(this.href);return false">PubMed</a></li>
<li><a href="/servlet/linkout?suffix=e_1_5_1_2_17_2&amp;dbid=16&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=10.1371%2Fjournal.pgen.1000384&amp;cf=" onclick="newWindow(this.href);return false">Crossref</a></li>
<li><a class="google-scholar" href="http://scholar.google.com/scholar_lookup?hl=en&volume=5&publication_year=2009&pages=e1000384&journal=PLoS+Genet.&author=Madsen+B.E.++and+Browning+S.R.+&title=A+groupwise+association+test+for+rare+mutations+using+a+weighted+sum+statistic&">Google Scholar</a></li></ul></div></div><div class="ref"><ul title="list of authors" class="rlist inline-bullet-list ref__authors ref__inline-block"><li class="ref__authors__name">Lek M. </li><li class="ref__authors__name">Karczewski K.J. </li><li class="ref__authors__name">Minikel E.V. </li><li class="ref__authors__name">Samocha K.E. </li><li class="ref__authors__name">Banks E. </li><li class="ref__authors__name">Fennell T. </li><li class="ref__authors__name">OâDonnell-Luria A.H. </li><li class="ref__authors__name">Ware J.S. </li><li class="ref__authors__name">Hill A.J. </li><li class="ref__authors__name">Cummings B.B. </li><li>et al.</li><li>Exome Aggregation Consortium</li></ul><div class="ref__title">Analysis of protein-coding genetic variation in 60,706 humans.</div><span class="ref__series"><i>Nature.</i> <span class="ref__seriesDate">2016; </span><span class="ref__seriesVolume">536</span></span>: <span class="ref__seriesPages">285-291</span><div class="ref__controls"><a href="#back-bib18" title="View in Article" class="ref__controls__viewInArticle scroll-into-link">View
                        in Article
                        <i class="icon-section_arrow_u"></i></a><ul class="rlist inline-bullet-list"><li><a href="/servlet/linkout?suffix=e_1_5_1_2_18_2&amp;dbid=8&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=27535533&amp;cf=" onclick="newWindow(this.href);return false">PubMed</a></li>
<li><a href="/servlet/linkout?suffix=e_1_5_1_2_18_2&amp;dbid=16&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=10.1038%2Fnature19057&amp;cf=" onclick="newWindow(this.href);return false">Crossref</a></li>
<li><a class="google-scholar" href="http://scholar.google.com/scholar_lookup?hl=en&volume=536&publication_year=2016&pages=285-291&journal=Nature&author=Exome+Aggregation+Consortium&title=Analysis+of+protein-coding+genetic+variation+in+60%2C706+humans&">Google Scholar</a></li></ul></div></div><div class="ref"><ul title="list of authors" class="rlist inline-bullet-list ref__authors ref__inline-block"><li class="ref__authors__name">Traynelis J. </li><li class="ref__authors__name">Silk M. </li><li class="ref__authors__name">Wang Q. </li><li class="ref__authors__name">Berkovic S.F. </li><li class="ref__authors__name">Liu L. </li><li class="ref__authors__name">Ascher D.B. </li><li class="ref__authors__name">Balding D.J. </li><li class="ref__authors__name">Petrovski S. </li></ul><div class="ref__title">Optimizing genomic medicine in epilepsy through a gene-customized approach to missense variant interpretation.</div><span class="ref__series"><i>Genome Res.</i> <span class="ref__seriesDate">2017; </span><span class="ref__seriesVolume">27</span></span>: <span class="ref__seriesPages">1715-1729</span><div class="ref__controls"><a href="#back-bib19" title="View in Article" class="ref__controls__viewInArticle scroll-into-link">View
                        in Article
                        <i class="icon-section_arrow_u"></i></a><ul class="rlist inline-bullet-list"><li><a href="/servlet/linkout?suffix=e_1_5_1_2_19_2&amp;dbid=137438953472&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=2-s2.0-85030645609&amp;cf=" onclick="newWindow(this.href);return false">Scopus (9)</a></li>
<li><a href="/servlet/linkout?suffix=e_1_5_1_2_19_2&amp;dbid=8&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=28864458&amp;cf=" onclick="newWindow(this.href);return false">PubMed</a></li>
<li><a href="/servlet/linkout?suffix=e_1_5_1_2_19_2&amp;dbid=16&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=10.1101%2Fgr.226589.117&amp;cf=" onclick="newWindow(this.href);return false">Crossref</a></li>
<li><a class="google-scholar" href="http://scholar.google.com/scholar_lookup?hl=en&volume=27&publication_year=2017&pages=1715-1729&journal=Genome+Res.&author=Traynelis+J.+%2C+Silk+M.+%2C+Wang+Q.+%2C+Berkovic+S.F.+%2C+Liu+L.+%2C+Ascher+D.B.+%2C+Balding+D.J.+%2C++and+Petrovski+S.+&title=Optimizing+genomic+medicine+in+epilepsy+through+a+gene-customized+approach+to+missense+variant+interpretation&">Google Scholar</a></li></ul></div></div><div class="ref"><ul title="list of authors" class="rlist inline-bullet-list ref__authors ref__inline-block"><li class="ref__authors__name">Sivley R.M. </li><li class="ref__authors__name">Dou X. </li><li class="ref__authors__name">Meiler J. </li><li class="ref__authors__name">Bush W.S. </li><li class="ref__authors__name">Capra J.A. </li><li class="ref__authors__name">Blackwell T.S. </li><li class="ref__authors__name">Phillips J.A. </li><li class="ref__authors__name">Meiler J. </li><li class="ref__authors__name">Bush W.S. </li><li class="ref__authors__name">Capra J.A. </li></ul><div class="ref__title">Comprehensive analysis of constraint on the spatial distribution of missense variants in human protein structures.</div><span class="ref__series"><i>Am. J. Hum. Genet.</i> <span class="ref__seriesDate">2018; </span><span class="ref__seriesVolume">102</span></span>: <span class="ref__seriesPages">415-426</span><div class="ref__controls"><a href="#back-bib20" title="View in Article" class="ref__controls__viewInArticle scroll-into-link">View
                        in Article
                        <i class="icon-section_arrow_u"></i></a><ul class="rlist inline-bullet-list"><li><a href="/servlet/linkout?suffix=e_1_5_1_2_20_2&amp;dbid=137438953472&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=2-s2.0-85042045525&amp;cf=" onclick="newWindow(this.href);return false">Scopus (0)</a></li>
<li><a href="/servlet/linkout?suffix=e_1_5_1_2_20_2&amp;dbid=8&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=29455857&amp;cf=" onclick="newWindow(this.href);return false">PubMed</a></li>
<li><a href="/servlet/linkout?suffix=e_1_5_1_2_20_2&amp;dbid=4&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=10.1016%2Fj.ajhg.2018.01.017&amp;cf=abstract&amp;site=cell-site" onclick="newWindow(this.href);return false">Abstract</a></li>
<li><a href="/servlet/linkout?suffix=e_1_5_1_2_20_2&amp;dbid=4&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=10.1016%2Fj.ajhg.2018.01.017&amp;cf=fulltext&amp;site=cell-site" onclick="newWindow(this.href);return false">Full Text</a></li>
<li><a href="/servlet/linkout?suffix=e_1_5_1_2_20_2&amp;dbid=4&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=10.1016%2Fj.ajhg.2018.01.017&amp;cf=pdf&amp;site=cell-site" onclick="newWindow(this.href);return false">Full Text PDF</a></li>
<li><a class="google-scholar" href="http://scholar.google.com/scholar_lookup?hl=en&volume=102&publication_year=2018&pages=415-426&journal=Am.+J.+Hum.+Genet.&author=Sivley+R.M.+%2C+Dou+X.+%2C+Meiler+J.+%2C+Bush+W.S.+%2C+Capra+J.A.+%2C+Blackwell+T.S.+%2C+Phillips+J.A.+%2C+Meiler+J.+%2C+Bush+W.S.+%2C++and+Capra+J.A.+&title=Comprehensive+analysis+of+constraint+on+the+spatial+distribution+of+missense+variants+in+human+protein+structures&">Google Scholar</a></li></ul></div></div><div class="ref"><ul title="list of authors" class="rlist inline-bullet-list ref__authors ref__inline-block"><li class="ref__authors__name">Marchler-Bauer A. </li><li class="ref__authors__name">Zheng C. </li><li class="ref__authors__name">Chitsaz F. </li><li class="ref__authors__name">Derbyshire M.K. </li><li class="ref__authors__name">Geer L.Y. </li><li class="ref__authors__name">Geer R.C. </li><li class="ref__authors__name">Gonzales N.R. </li><li class="ref__authors__name">Gwadz M. </li><li class="ref__authors__name">Hurwitz D.I. </li><li class="ref__authors__name">Lanczycki C.J. </li><li>et al.</li></ul><div class="ref__title">CDD: conserved domains and protein three-dimensional structure.</div><span class="ref__series"><i>Nucleic Acids Res.</i> <span class="ref__seriesDate">2013; </span><span class="ref__seriesVolume">41</span></span>: <span class="ref__seriesPages">D348-D352</span><div class="ref__controls"><a href="#back-bib21" title="View in Article" class="ref__controls__viewInArticle scroll-into-link">View
                        in Article
                        <i class="icon-section_arrow_u"></i></a><ul class="rlist inline-bullet-list"><li><a href="/servlet/linkout?suffix=e_1_5_1_2_21_2&amp;dbid=137438953472&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=2-s2.0-84875445608&amp;cf=" onclick="newWindow(this.href);return false">Scopus (549)</a></li>
<li><a href="/servlet/linkout?suffix=e_1_5_1_2_21_2&amp;dbid=8&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=23197659&amp;cf=" onclick="newWindow(this.href);return false">PubMed</a></li>
<li><a href="/servlet/linkout?suffix=e_1_5_1_2_21_2&amp;dbid=16&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=10.1093%2Fnar%2Fgks1243&amp;cf=" onclick="newWindow(this.href);return false">Crossref</a></li>
<li><a class="google-scholar" href="http://scholar.google.com/scholar_lookup?hl=en&volume=41&publication_year=2013&pages=D348-D352&journal=Nucleic+Acids+Res.&author=Marchler-Bauer+A.+%2C+Zheng+C.+%2C+Chitsaz+F.+%2C+Derbyshire+M.K.+%2C+Geer+L.Y.+%2C+Geer+R.C.+%2C+Gonzales+N.R.+%2C+Gwadz+M.+%2C+Hurwitz+D.I.+%2C+Lanczycki+C.J.++et+al.&title=CDD%3A+conserved+domains+and+protein+three-dimensional+structure&">Google Scholar</a></li></ul></div></div><div class="ref"><ul title="list of authors" class="rlist inline-bullet-list ref__authors ref__inline-block"><li class="ref__authors__name">Pruitt K.D. </li><li class="ref__authors__name">Harrow J. </li><li class="ref__authors__name">Harte R.A. </li><li class="ref__authors__name">Wallin C. </li><li class="ref__authors__name">Diekhans M. </li><li class="ref__authors__name">Maglott D.R. </li><li class="ref__authors__name">Searle S. </li><li class="ref__authors__name">Farrell C.M. </li><li class="ref__authors__name">Loveland J.E. </li><li class="ref__authors__name">Ruef B.J. </li><li>et al.</li></ul><div class="ref__title">The consensus coding sequence (CCDS) project: Identifying a common protein-coding gene set for the human and mouse genomes.</div><span class="ref__series"><i>Genome Res.</i> <span class="ref__seriesDate">2009; </span><span class="ref__seriesVolume">19</span></span>: <span class="ref__seriesPages">1316-1323</span><div class="ref__controls"><a href="#back-bib22" title="View in Article" class="ref__controls__viewInArticle scroll-into-link">View
                        in Article
                        <i class="icon-section_arrow_u"></i></a><ul class="rlist inline-bullet-list"><li><a href="/servlet/linkout?suffix=e_1_5_1_2_22_2&amp;dbid=137438953472&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=2-s2.0-67650064594&amp;cf=" onclick="newWindow(this.href);return false">Scopus (326)</a></li>
<li><a href="/servlet/linkout?suffix=e_1_5_1_2_22_2&amp;dbid=8&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=19498102&amp;cf=" onclick="newWindow(this.href);return false">PubMed</a></li>
<li><a href="/servlet/linkout?suffix=e_1_5_1_2_22_2&amp;dbid=16&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=10.1101%2Fgr.080531.108&amp;cf=" onclick="newWindow(this.href);return false">Crossref</a></li>
<li><a class="google-scholar" href="http://scholar.google.com/scholar_lookup?hl=en&volume=19&publication_year=2009&pages=1316-1323&journal=Genome+Res.&author=Pruitt+K.D.+%2C+Harrow+J.+%2C+Harte+R.A.+%2C+Wallin+C.+%2C+Diekhans+M.+%2C+Maglott+D.R.+%2C+Searle+S.+%2C+Farrell+C.M.+%2C+Loveland+J.E.+%2C+Ruef+B.J.++et+al.&title=The+consensus+coding+sequence+%28CCDS%29+project%3A+Identifying+a+common+protein-coding+gene+set+for+the+human+and+mouse+genomes&">Google Scholar</a></li></ul></div></div><div class="ref"><ul title="list of authors" class="rlist inline-bullet-list ref__authors ref__inline-block"><li class="ref__authors__name">Landrum M.J. </li><li class="ref__authors__name">Lee J.M. </li><li class="ref__authors__name">Riley G.R. </li><li class="ref__authors__name">Jang W. </li><li class="ref__authors__name">Rubinstein W.S. </li><li class="ref__authors__name">Church D.M. </li><li class="ref__authors__name">Maglott D.R. </li></ul><div class="ref__title">ClinVar: public archive of relationships among sequence variation and human phenotype.</div><span class="ref__series"><i>Nucleic Acids Res.</i> <span class="ref__seriesDate">2014; </span><span class="ref__seriesVolume">42</span></span>: <span class="ref__seriesPages">D980-D985</span><div class="ref__controls"><a href="#back-bib23" title="View in Article" class="ref__controls__viewInArticle scroll-into-link">View
                        in Article
                        <i class="icon-section_arrow_u"></i></a><ul class="rlist inline-bullet-list"><li><a href="/servlet/linkout?suffix=e_1_5_1_2_23_2&amp;dbid=137438953472&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=2-s2.0-84891809093&amp;cf=" onclick="newWindow(this.href);return false">Scopus (707)</a></li>
<li><a href="/servlet/linkout?suffix=e_1_5_1_2_23_2&amp;dbid=8&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=24234437&amp;cf=" onclick="newWindow(this.href);return false">PubMed</a></li>
<li><a href="/servlet/linkout?suffix=e_1_5_1_2_23_2&amp;dbid=16&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=10.1093%2Fnar%2Fgkt1113&amp;cf=" onclick="newWindow(this.href);return false">Crossref</a></li>
<li><a class="google-scholar" href="http://scholar.google.com/scholar_lookup?hl=en&volume=42&publication_year=2014&pages=D980-D985&journal=Nucleic+Acids+Res.&author=Landrum+M.J.+%2C+Lee+J.M.+%2C+Riley+G.R.+%2C+Jang+W.+%2C+Rubinstein+W.S.+%2C+Church+D.M.+%2C++and+Maglott+D.R.+&title=ClinVar%3A+public+archive+of+relationships+among+sequence+variation+and+human+phenotype&">Google Scholar</a></li></ul></div></div><div class="ref"><ul title="list of authors" class="rlist inline-bullet-list ref__authors ref__inline-block"><li class="ref__authors__name">Dewey F.E. </li><li class="ref__authors__name">Murray M.F. </li><li class="ref__authors__name">Overton J.D. </li><li class="ref__authors__name">Habegger L. </li><li class="ref__authors__name">Leader J.B. </li><li class="ref__authors__name">Fetterolf S.N. </li><li class="ref__authors__name">OâDushlaine C. </li><li class="ref__authors__name">Van Hout C.V. </li><li class="ref__authors__name">Staples J. </li><li class="ref__authors__name">Gonzaga-Jauregui C. </li><li>et al.</li></ul><div class="ref__title">Distribution and clinical impact of functional variants in 50,726 whole-exome sequences from the DiscovEHR study.</div><span class="ref__series"><i>Science.</i> <span class="ref__seriesDate">2016; </span><span class="ref__seriesVolume">354</span></span>10.1126/science.aaf6814<div class="ref__controls"><a href="#back-bib24" title="View in Article" class="ref__controls__viewInArticle scroll-into-link">View
                        in Article
                        <i class="icon-section_arrow_u"></i></a><ul class="rlist inline-bullet-list"><li><a href="/servlet/linkout?suffix=e_1_5_1_2_24_2&amp;dbid=137438953472&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=2-s2.0-85007028930&amp;cf=" onclick="newWindow(this.href);return false">Scopus (87)</a></li>
<li><a href="/servlet/linkout?suffix=e_1_5_1_2_24_2&amp;dbid=16&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=10.1126%2Fscience.aaf6814&amp;cf=" onclick="newWindow(this.href);return false">Crossref</a></li>
<li><a class="google-scholar" href="http://scholar.google.com/scholar_lookup?hl=en&volume=354&publication_year=2016&journal=Science&author=Dewey+F.E.+%2C+Murray+M.F.+%2C+Overton+J.D.+%2C+Habegger+L.+%2C+Leader+J.B.+%2C+Fetterolf+S.N.+%2C+O%E2%80%99Dushlaine+C.+%2C+Van+Hout+C.V.+%2C+Staples+J.+%2C+Gonzaga-Jauregui+C.++et+al.&title=Distribution+and+clinical+impact+of+functional+variants+in+50%2C726+whole-exome+sequences+from+the+DiscovEHR+study&">Google Scholar</a></li></ul></div></div><div class="ref"><ul title="list of authors" class="rlist inline-bullet-list ref__authors ref__inline-block"><li class="ref__authors__name">Petrovski S. </li><li class="ref__authors__name">Gussow A.B. </li><li class="ref__authors__name">Wang Q. </li><li class="ref__authors__name">Halvorsen M. </li><li class="ref__authors__name">Han Y. </li><li class="ref__authors__name">Weir W.H. </li><li class="ref__authors__name">Allen A.S. </li><li class="ref__authors__name">Goldstein D.B. </li></ul><div class="ref__title">The intolerance of regulatory sequence to genetic variation predicts gene dosage sensitivity.</div><span class="ref__series"><i>PLoS Genet.</i> <span class="ref__seriesDate">2015; </span><span class="ref__seriesVolume">11</span></span>: <span class="ref__seriesPages">e1005492</span><div class="ref__controls"><a href="#back-bib25" title="View in Article" class="ref__controls__viewInArticle scroll-into-link">View
                        in Article
                        <i class="icon-section_arrow_u"></i></a><ul class="rlist inline-bullet-list"><li><a href="/servlet/linkout?suffix=e_1_5_1_2_25_2&amp;dbid=8&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=26332131&amp;cf=" onclick="newWindow(this.href);return false">PubMed</a></li>
<li><a href="/servlet/linkout?suffix=e_1_5_1_2_25_2&amp;dbid=16&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=10.1371%2Fjournal.pgen.1005492&amp;cf=" onclick="newWindow(this.href);return false">Crossref</a></li>
<li><a class="google-scholar" href="http://scholar.google.com/scholar_lookup?hl=en&volume=11&publication_year=2015&pages=e1005492&journal=PLoS+Genet.&author=Petrovski+S.+%2C+Gussow+A.B.+%2C+Wang+Q.+%2C+Halvorsen+M.+%2C+Han+Y.+%2C+Weir+W.H.+%2C+Allen+A.S.+%2C++and+Goldstein+D.B.+&title=The+intolerance+of+regulatory+sequence+to+genetic+variation+predicts+gene+dosage+sensitivity&">Google Scholar</a></li></ul></div></div><div class="ref"><ul title="list of authors" class="rlist inline-bullet-list ref__authors ref__inline-block"><li class="ref__authors__name">Firth H.V. </li><li class="ref__authors__name">Richards S.M. </li><li class="ref__authors__name">Bevan A.P. </li><li class="ref__authors__name">Clayton S. </li><li class="ref__authors__name">Corpas M. </li><li class="ref__authors__name">Rajan D. </li><li class="ref__authors__name">Van Vooren S. </li><li class="ref__authors__name">Moreau Y. </li><li class="ref__authors__name">Pettett R.M. </li><li class="ref__authors__name">Carter N.P. </li></ul><div class="ref__title">DECIPHER: Database of Chromosomal Imbalance and Phenotype in Humans Using Ensembl Resources.</div><span class="ref__series"><i>Am. J. Hum. Genet.</i> <span class="ref__seriesDate">2009; </span><span class="ref__seriesVolume">84</span></span>: <span class="ref__seriesPages">524-533</span><div class="ref__controls"><a href="#back-bib26" title="View in Article" class="ref__controls__viewInArticle scroll-into-link">View
                        in Article
                        <i class="icon-section_arrow_u"></i></a><ul class="rlist inline-bullet-list"><li><a href="/servlet/linkout?suffix=e_1_5_1_2_26_2&amp;dbid=137438953472&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=2-s2.0-64149099583&amp;cf=" onclick="newWindow(this.href);return false">Scopus (550)</a></li>
<li><a href="/servlet/linkout?suffix=e_1_5_1_2_26_2&amp;dbid=8&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=19344873&amp;cf=" onclick="newWindow(this.href);return false">PubMed</a></li>
<li><a href="/servlet/linkout?suffix=e_1_5_1_2_26_2&amp;dbid=4&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=10.1016%2Fj.ajhg.2009.03.010&amp;cf=abstract&amp;site=cell-site" onclick="newWindow(this.href);return false">Abstract</a></li>
<li><a href="/servlet/linkout?suffix=e_1_5_1_2_26_2&amp;dbid=4&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=10.1016%2Fj.ajhg.2009.03.010&amp;cf=fulltext&amp;site=cell-site" onclick="newWindow(this.href);return false">Full Text</a></li>
<li><a href="/servlet/linkout?suffix=e_1_5_1_2_26_2&amp;dbid=4&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=10.1016%2Fj.ajhg.2009.03.010&amp;cf=pdf&amp;site=cell-site" onclick="newWindow(this.href);return false">Full Text PDF</a></li>
<li><a class="google-scholar" href="http://scholar.google.com/scholar_lookup?hl=en&volume=84&publication_year=2009&pages=524-533&journal=Am.+J.+Hum.+Genet.&author=Firth+H.V.+%2C+Richards+S.M.+%2C+Bevan+A.P.+%2C+Clayton+S.+%2C+Corpas+M.+%2C+Rajan+D.+%2C+Van+Vooren+S.+%2C+Moreau+Y.+%2C+Pettett+R.M.+%2C++and+Carter+N.P.+&title=DECIPHER%3A+Database+of+Chromosomal+Imbalance+and+Phenotype+in+Humans+Using+Ensembl+Resources&">Google Scholar</a></li></ul></div></div><div class="ref"><ul title="list of authors" class="rlist inline-bullet-list ref__authors ref__inline-block"><li class="ref__authors__name">Zhu X. </li><li class="ref__authors__name">Padmanabhan R. </li><li class="ref__authors__name">Copeland B. </li><li class="ref__authors__name">Bridgers J. </li><li class="ref__authors__name">Ren Z. </li><li class="ref__authors__name">Kamalakaran S. </li><li class="ref__authors__name">Driscoll-collins A.O. </li><li class="ref__authors__name">Berkovic S.F. </li><li class="ref__authors__name">Scheffer I.E. </li><li class="ref__authors__name">Poduri A. </li><li>et al.</li></ul><div class="ref__title">A case-control collapsing analysis identifies epilepsy genes implicated in trio sequencing studies focused on de novo mutations.</div><span class="ref__series"><i>PLoS Genet.</i> <span class="ref__seriesDate">2017; </span><span class="ref__seriesVolume">13</span></span>: <span class="ref__seriesPages">e1007104</span><div class="ref__controls"><a href="#back-bib27" title="View in Article" class="ref__controls__viewInArticle scroll-into-link">View
                        in Article
                        <i class="icon-section_arrow_u"></i></a><ul class="rlist inline-bullet-list"><li><a href="/servlet/linkout?suffix=e_1_5_1_2_27_2&amp;dbid=137438953472&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=2-s2.0-85036612433&amp;cf=" onclick="newWindow(this.href);return false">Scopus (0)</a></li>
<li><a href="/servlet/linkout?suffix=e_1_5_1_2_27_2&amp;dbid=8&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=29186148&amp;cf=" onclick="newWindow(this.href);return false">PubMed</a></li>
<li><a href="/servlet/linkout?suffix=e_1_5_1_2_27_2&amp;dbid=16&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=10.1371%2Fjournal.pgen.1007104&amp;cf=" onclick="newWindow(this.href);return false">Crossref</a></li>
<li><a class="google-scholar" href="http://scholar.google.com/scholar_lookup?hl=en&volume=13&publication_year=2017&pages=e1007104&journal=PLoS+Genet.&author=Zhu+X.+%2C+Padmanabhan+R.+%2C+Copeland+B.+%2C+Bridgers+J.+%2C+Ren+Z.+%2C+Kamalakaran+S.+%2C+Driscoll-collins+A.O.+%2C+Berkovic+S.F.+%2C+Scheffer+I.E.+%2C+Poduri+A.++et+al.&title=A+case-control+collapsing+analysis+identifies+epilepsy+genes+implicated+in+trio+sequencing+studies+focused+on+de+novo+mutations&">Google Scholar</a></li></ul></div></div><div class="ref"><ul title="list of authors" class="rlist inline-bullet-list ref__authors ref__inline-block"><li class="ref__authors__name">Hoffman-Andrews L. </li></ul><div class="ref__title">The known unknown: the challenges of genetic variants of uncertain significance in clinical practice.</div><span class="ref__series"><i>J.Â Law Biosci.</i> <span class="ref__seriesDate">2018; </span><span class="ref__seriesVolume">4</span></span>: <span class="ref__seriesPages">648-657</span><div class="ref__controls"><a href="#back-bib28" title="View in Article" class="ref__controls__viewInArticle scroll-into-link">View
                        in Article
                        <i class="icon-section_arrow_u"></i></a><ul class="rlist inline-bullet-list"><li><a href="/servlet/linkout?suffix=e_1_5_1_2_28_2&amp;dbid=137438953472&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=2-s2.0-85056700984&amp;cf=" onclick="newWindow(this.href);return false">Scopus (2)</a></li>
<li><a href="/servlet/linkout?suffix=e_1_5_1_2_28_2&amp;dbid=8&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=29868193&amp;cf=" onclick="newWindow(this.href);return false">PubMed</a></li>
<li><a href="/servlet/linkout?suffix=e_1_5_1_2_28_2&amp;dbid=16&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=10.1093%2Fjlb%2Flsx038&amp;cf=" onclick="newWindow(this.href);return false">Crossref</a></li>
<li><a class="google-scholar" href="http://scholar.google.com/scholar_lookup?hl=en&volume=4&publication_year=2018&pages=648-657&journal=J.%C2%A0Law+Biosci.&author=Hoffman-Andrews+L.+&title=The+known+unknown%3A+the+challenges+of+genetic+variants+of+uncertain+significance+in+clinical+practice&">Google Scholar</a></li></ul></div></div><div class="ref"><p>Ray, T. (2016). Motherâs Negligence Suit Against Questâs Athena Could Broadly Impact Genetic Testing Labs. GenomeWeb, March 14, 2016. <a href="https://www.genomeweb.com/molecular-diagnostics/mothers-negligence-suit-against-quests-athena-could-broadly-impact-genetic#.XCy8t5NKjIU" target="_blank">https://www.genomeweb.com/molecular-diagnostics/mothers-negligence-suit-against-quests-athena-could-broadly-impact-genetic#.XCy8t5NKjIU</a>.</p><div class="ref__controls"><a href="#back-bib29" title="View in Article" class="ref__controls__viewInArticle scroll-into-link">View
                        in Article
                        <i class="icon-section_arrow_u"></i></a><ul class="rlist inline-bullet-list"><li><a class="google-scholar" href="http://scholar.google.com/scholar?hl=en&q=29">Google Scholar</a></li></ul></div></div><div class="ref"><ul title="list of authors" class="rlist inline-bullet-list ref__authors ref__inline-block"><li class="ref__authors__name">Samocha K.E. </li><li class="ref__authors__name">Kosmicki J.A. </li><li class="ref__authors__name">Karczewski K.J. </li><li class="ref__authors__name">OâDonnell-Luria A.H. </li><li class="ref__authors__name">Pierce-Hoffman E. </li><li class="ref__authors__name">MacArthur D.G. </li><li class="ref__authors__name">Neale B.M. </li><li class="ref__authors__name">Daly M.J. </li></ul><div class="ref__title">Regional missense constraint improves variant deleteriousness prediction.</div><span class="ref__series"><i>bioRxiv.</i> <span class="ref__seriesDate">2017; </span><span class="ref__seriesVolume"></span></span>10.1101/148353<div class="ref__controls"><a href="#back-bib30" title="View in Article" class="ref__controls__viewInArticle scroll-into-link">View
                        in Article
                        <i class="icon-section_arrow_u"></i></a><ul class="rlist inline-bullet-list"><li><a href="/servlet/linkout?suffix=e_1_5_1_2_30_2&amp;dbid=16&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=10.1101%2F148353&amp;cf=" onclick="newWindow(this.href);return false">Crossref</a></li>
<li><a class="google-scholar" href="http://scholar.google.com/scholar_lookup?hl=en&publication_year=2017&journal=bioRxiv&author=Samocha+K.E.+%2C+Kosmicki+J.A.+%2C+Karczewski+K.J.+%2C+O%E2%80%99Donnell-Luria+A.H.+%2C+Pierce-Hoffman+E.+%2C+MacArthur+D.G.+%2C+Neale+B.M.+%2C++and+Daly+M.J.+&title=Regional+missense+constraint+improves+variant+deleteriousness+prediction&">Google Scholar</a></li></ul></div></div><div class="ref"><ul title="list of authors" class="rlist inline-bullet-list ref__authors ref__inline-block"><li class="ref__authors__name">Adzhubei I.A. </li><li class="ref__authors__name">Schmidt S. </li><li class="ref__authors__name">Peshkin L. </li><li class="ref__authors__name">Ramensky V.E. </li><li class="ref__authors__name">Gerasimova A. </li><li class="ref__authors__name">Bork P. </li><li class="ref__authors__name">Kondrashov A.S. </li><li class="ref__authors__name">Sunyaev S.R. </li></ul><div class="ref__title">A method and server for predicting damaging missense mutations.</div><span class="ref__series"><i>Nat. Methods.</i> <span class="ref__seriesDate">2010; </span><span class="ref__seriesVolume">7</span></span>: <span class="ref__seriesPages">248-249</span><div class="ref__controls"><a href="#back-bib31" title="View in Article" class="ref__controls__viewInArticle scroll-into-link">View
                        in Article
                        <i class="icon-section_arrow_u"></i></a><ul class="rlist inline-bullet-list"><li><a href="/servlet/linkout?suffix=e_1_5_1_2_31_2&amp;dbid=137438953472&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=2-s2.0-77951640946&amp;cf=" onclick="newWindow(this.href);return false">Scopus (5843)</a></li>
<li><a href="/servlet/linkout?suffix=e_1_5_1_2_31_2&amp;dbid=8&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=20354512&amp;cf=" onclick="newWindow(this.href);return false">PubMed</a></li>
<li><a href="/servlet/linkout?suffix=e_1_5_1_2_31_2&amp;dbid=16&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=10.1038%2Fnmeth0410-248&amp;cf=" onclick="newWindow(this.href);return false">Crossref</a></li>
<li><a class="google-scholar" href="http://scholar.google.com/scholar_lookup?hl=en&volume=7&publication_year=2010&pages=248-249&journal=Nat.+Methods&author=Adzhubei+I.A.+%2C+Schmidt+S.+%2C+Peshkin+L.+%2C+Ramensky+V.E.+%2C+Gerasimova+A.+%2C+Bork+P.+%2C+Kondrashov+A.S.+%2C++and+Sunyaev+S.R.+&title=A+method+and+server+for+predicting+damaging+missense+mutations&">Google Scholar</a></li></ul></div></div><div class="ref"><ul title="list of authors" class="rlist inline-bullet-list ref__authors ref__inline-block"><li class="ref__authors__name">Kircher M. </li><li class="ref__authors__name">Witten D.M. </li><li class="ref__authors__name">Jain P. </li><li class="ref__authors__name">OâRoak B.J. </li><li class="ref__authors__name">Cooper G.M. </li><li class="ref__authors__name">Shendure J. </li></ul><div class="ref__title">A general framework for estimating the relative pathogenicity of human genetic variants.</div><span class="ref__series"><i>Nat. Genet.</i> <span class="ref__seriesDate">2014; </span><span class="ref__seriesVolume">46</span></span>: <span class="ref__seriesPages">310-315</span><div class="ref__controls"><a href="#back-bib32" title="View in Article" class="ref__controls__viewInArticle scroll-into-link">View
                        in Article
                        <i class="icon-section_arrow_u"></i></a><ul class="rlist inline-bullet-list"><li><a href="/servlet/linkout?suffix=e_1_5_1_2_32_2&amp;dbid=137438953472&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=2-s2.0-84895858942&amp;cf=" onclick="newWindow(this.href);return false">Scopus (1621)</a></li>
<li><a href="/servlet/linkout?suffix=e_1_5_1_2_32_2&amp;dbid=8&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=24487276&amp;cf=" onclick="newWindow(this.href);return false">PubMed</a></li>
<li><a href="/servlet/linkout?suffix=e_1_5_1_2_32_2&amp;dbid=16&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=10.1038%2Fng.2892&amp;cf=" onclick="newWindow(this.href);return false">Crossref</a></li>
<li><a class="google-scholar" href="http://scholar.google.com/scholar_lookup?hl=en&volume=46&publication_year=2014&pages=310-315&journal=Nat.+Genet.&author=Kircher+M.+%2C+Witten+D.M.+%2C+Jain+P.+%2C+O%E2%80%99Roak+B.J.+%2C+Cooper+G.M.+%2C++and+Shendure+J.+&title=A+general+framework+for+estimating+the+relative+pathogenicity+of+human+genetic+variants&">Google Scholar</a></li></ul></div></div><div class="ref"><ul title="list of authors" class="rlist inline-bullet-list ref__authors ref__inline-block"><li class="ref__authors__name">Ioannidis N.M. </li><li class="ref__authors__name">Rothstein J.H. </li><li class="ref__authors__name">Pejaver V. </li><li class="ref__authors__name">Middha S. </li><li class="ref__authors__name">McDonnell S.K. </li><li class="ref__authors__name">Baheti S. </li><li class="ref__authors__name">Musolf A. </li><li class="ref__authors__name">Li Q. </li><li class="ref__authors__name">Holzinger E. </li><li class="ref__authors__name">Karyadi D. </li><li>et al.</li></ul><div class="ref__title">REVEL: An Ensemble Method for Predicting the Pathogenicity of Rare Missense Variants.</div><span class="ref__series"><i>Am. J. Hum. Genet.</i> <span class="ref__seriesDate">2016; </span><span class="ref__seriesVolume">99</span></span>: <span class="ref__seriesPages">877-885</span><div class="ref__controls"><a href="#back-bib33" title="View in Article" class="ref__controls__viewInArticle scroll-into-link">View
                        in Article
                        <i class="icon-section_arrow_u"></i></a><ul class="rlist inline-bullet-list"><li><a href="/servlet/linkout?suffix=e_1_5_1_2_33_2&amp;dbid=8&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=27666373&amp;cf=" onclick="newWindow(this.href);return false">PubMed</a></li>
<li><a href="/servlet/linkout?suffix=e_1_5_1_2_33_2&amp;dbid=4&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=10.1016%2Fj.ajhg.2016.08.016&amp;cf=abstract&amp;site=cell-site" onclick="newWindow(this.href);return false">Abstract</a></li>
<li><a href="/servlet/linkout?suffix=e_1_5_1_2_33_2&amp;dbid=4&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=10.1016%2Fj.ajhg.2016.08.016&amp;cf=fulltext&amp;site=cell-site" onclick="newWindow(this.href);return false">Full Text</a></li>
<li><a href="/servlet/linkout?suffix=e_1_5_1_2_33_2&amp;dbid=4&amp;doi=10.1016/j.ajhg.2018.12.020&amp;key=10.1016%2Fj.ajhg.2016.08.016&amp;cf=pdf&amp;site=cell-site" onclick="newWindow(this.href);return false">Full Text PDF</a></li>
<li><a class="google-scholar" href="http://scholar.google.com/scholar_lookup?hl=en&volume=99&publication_year=2016&pages=877-885&journal=Am.+J.+Hum.+Genet.&author=Ioannidis+N.M.+%2C+Rothstein+J.H.+%2C+Pejaver+V.+%2C+Middha+S.+%2C+McDonnell+S.K.+%2C+Baheti+S.+%2C+Musolf+A.+%2C+Li+Q.+%2C+Holzinger+E.+%2C+Karyadi+D.++et+al.&title=REVEL%3A+An+Ensemble+Method+for+Predicting+the+Pathogenicity+of+Rare+Missense+Variants&">Google Scholar</a></li></ul></div></div></section><section class="article-info"><h2 class="top" id="articleInformation">Article Info</h2><section><h3 class="publicationHistory__label">Publication History</h3><div class="article-info__dates"><div class="article-info__date">Published: January 24, 2019</div><div class="article-info__date">Accepted:
            December 21,
            2018</div><div class="article-info__date">Received:
            September 20,
            2018</div></div></section><section><h3>IDENTIFICATION</h3><p>DOI: <a href="https://doi.org/10.1016/j.ajhg.2018.12.020">https://doi.org/10.1016/j.ajhg.2018.12.020</a></p></section><section><h3>Copyright</h3><div class="copyright">Â© 2019 American Society of Human Genetics.</div></section><section><h3>ScienceDirect</h3><a href="https://www.sciencedirect.com/science/article/pii/S0002929718305032">Access this article on ScienceDirect</a></section></section><div id="figure-viewer" class="showCaption showThumbstrip figure-viewer"><nav class="nav-bar figure-viewer__top"><a href="#" title="Close Figure Viewer" class="figure-viewer__close"><i aria-hidden="true" class="icon-tools_close"></i></a><div class="figure-viewer__title"></div></nav><div class="frames"><ul class="frames__list"></ul><a href="#" title="Previous figure" class="frameButton previousFrameButton frames__prev"><i class="icon-arrow_l"></i></a><a href="#" title="Next figure" class="frameButton nextFrameButton frames__next"><i class="icon-arrow_r"></i></a></div><div class="caption"><div class="figure-viewer__caption-body"></div></div><div class="toolbar figure-viewer__toolbar"><span class="figure-viewer__counter"></span><a href="#" class="button toggle-caption-button figure-viewer__caption__ctrl"><i aria-hidden="true" class="icon icon-info"></i><span class="label">Hide Caption</span></a><a href="#" class="button download-options-button"><i aria-hidden="true" class="icon icon-gizmo-folder-in download-size-icon"></i><span class="label">Download</span></a><a href="#figures" class="button see-figures figure-viewer__see-figure__ctrl"><i aria-hidden="true" class="icon icon-gizmo-image_figure viewer-size-icon"></i><span class="label">See figure in article</span></a></div><div class="thumbstrip"><a class="toggleButton thumbstrip__ctrl"><span>Toggle Thumbstrip</span></a><a href="#" title="Previous Thumbnails" class="thumbstripButton previousThumbstripButton thumbstrip__prev"><i aria-hidden="true" class="icon-arrow_l"></i></a><a href="#" title="Next Thumbnails" class="thumbstripButton nextThumbstripButton thumbstrip__next"><i aria-hidden="true" class="icon-arrow_r"></i></a><div class="thumbstrip__wrapper"><ul class="rlist figure-viewer__thumbslist"></ul></div></div><div class="download figure-viewer__download"><div class="pointer"></div><ul class="rlist downlad__wrapper"><li><a target="_blank" href="#" title="View Large Image in a new tab" class="viewLargeImage"><i class="icon icon-gizmo-image_figure figure-viewer__download__viewer-size-icon"></i><span class="figure-viewer__download__text">View Large Image</span></a></li><li><a target="_blank" href="#" title="View high resolution image in a new tab" class="downloadThisButton" download=""><i class="icon icon-gizmo-image_figure figure-viewer__download__viewer-size-icon"></i><span class="figure-viewer__download__text">Download Hi-res image </span></a></li><li class="hidden-xs hidden-sm"><a href="#" class="downloadAllButton"><i class="icon icon-gizmo-ppt figure-viewer__download__viewer-size-icon figure-viewer__download__viewer-size-icon--ppt"></i><span class="figure-viewer__download__text"> Download .PPT</span></a></li></ul></div></div><div class="figures-placeholder"></div><div class="tables-placeholder"></div>
    

        


    

        


    

            

            

            <div data-widget-id="393effff-5b44-4681-a9b4-4f029b3540b7" data-widget-name="UX3HTMLWidget"><style>
.linkback-ad .pb-ad {
    display: block;
   height: auto;
    margin-left: 0px;
    margin-right: auto;
}
</style></div>

            

        

    

        


    

        


    

            

            

            <div data-pb-widget-type="ux3-relatedArticles" data-widget-id="1ff3fe0d-c546-4157-8e21-384cabd8aafb" data-widget-name="UX3RelatedArticlesWidget" data-article-pii="S0002929718305032" class="ux3DeferredWidget"></div>

            

        
<div id="figure-viewer" class="showCaption showThumbstrip figure-viewer"><nav class="nav-bar figure-viewer__top"><a href="#" title="Close Figure Viewer" class="figure-viewer__close"><i aria-hidden="true" class="icon-tools_close"></i></a><div class="figure-viewer__title"></div></nav><div class="frames"><ul class="frames__list"></ul><a href="#" title="Previous Frame" class="frameButton previousFrameButton frames__prev"><!--span Previous Frame--><i class="icon-arrow_l"></i></a><a href="#" title="Next Frame" class="frameButton nextFrameButton frames__next"><!--span Next Frame--><i class="icon-arrow_r"></i></a></div><div class="caption"><div class="figure-viewer__caption-body"></div></div><div class="toolbar figure-viewer__toolbar"><span class="figure-viewer__counter"></span><a href="#" class="button toggle-caption-button figure-viewer__caption__ctrl"><i aria-hidden="true" class="icon icon-info"></i><span class="label">Hide Caption</span></a><a href="#" class="button download-options-button"><i aria-hidden="true" class="icon icon-gizmo-folder-in download-size-icon"></i><span class="label">Download</span></a><a href="#figures" class="button see-figures figure-viewer__see-figure__ctrl"><i aria-hidden="true" class="icon icon-gizmo-image_figure viewer-size-icon"></i><span class="label">See figure in Article</span></a></div><div class="thumbstrip"><a class="toggleButton thumbstrip__ctrl"><span>Toggle Thumbstrip</span></a><a href="#" title="Previous Thumbnails" class="thumbstripButton previousThumbstripButton thumbstrip__prev"><i aria-hidden="true" class="icon-arrow_l"></i></a><a href="#" title="Next Thumbnails" class="thumbstripButton nextThumbstripButton thumbstrip__next"><i aria-hidden="true" class="icon-arrow_r"></i></a><div class="thumbstrip__wrapper"><ul class="rlist figure-viewer__thumbslist"></ul></div></div><div class="download figure-viewer__download"><div class="pointer"></div><ul class="rlist downlad__wrapper"><li><a target="_blank" href="/specs/products/marlin/components/figure-viewer/images/facets-2015-0010-f10-standard.jpg" title="View high resolution image in a new tab" class="downloadThisButton"><i class="icon icon-gizmo-image_figure figure-viewer__download__viewer-size-icon"></i><span class="figure-viewer__download__text">Download Hi-res image</span></a></li><li class="hidden-xs hidden-sm"><a href="/action/downloadFigures?articleNumber=e00489" target="_blank" title="Doanload .PPT" class="downloadAllButton"><i class="icon icon-gizmo-ppt figure-viewer__download__viewer-size-icon figure-viewer__download__viewer-size-icon--ppt"></i><span class="figure-viewer__download__text"> Download .PPT </span></a></li></ul></div></div><div data-toggle="transplant" data-direction="from" data-transplant="self" data-target=".adplaceholder--slot2, .adplaceholder--slot3 " data-remove="false" class="adplaceholder__transplant hidden-md hidden-lg"></div></div><div class="col-md-3 col-lg-2"></div></div></div></div></article>

            

        

        </main>
    
</div></div>

            
                </div>
            

        

        </div>
    

    

        


    

            

            

            <div data-widget-id="9c9ff7a1-a035-49f6-a336-600c9419c518" data-widget-name="UX3HTMLWidget"><!-- Start of HubSpot Embed Code --> 

<script type="text/javascript" id="hs-script-loader" async defer src="//js.hs-scripts.com/519224.js"></script> 

<!-- End of HubSpot Embed Code --></div>

            

        

    

        


    

        


    

                
                
                
                
                
                
                    
                
                    
                
                    
                        
                    
                
                    
                
                    
                
                    
                
                    
                
                    
                
                    
                
                    
                
                    
                
                
                    
                
                    
                
                    
                        
                        
                        
                    
                
                    
                
                    
                
                    
                        
                        
                            
                        
                        
                            
                            
                            
                        
                    
                
                    
                
                    
                
                    
                
                    
                
                    
                
                
                
                <div class="widget sessionValidationWidget none  widget-none  widget-compact-all" id="0257aef7-6074-4898-a989-1c3f693225a2">
                
                

                <div class="wrapped " >
                    
                    <div class="widget-body body body-none  body-compact-all"><script>
    $.ajax({
        url: "https://www.cell.com/ssoSessionCheck",
        type: "GET",
        dataType: "jsonp",
        success: function (data) {
            if(data.RefreshSession) {
                location.reload();
            }
        },
        error : function (){
            console.warn("can't validate CSAS session.");
        }
    });
</script></div>
                </div>
                </div>

        










        <div data-widget-def="ux3-layout-widget" data-widget-id="d95b4a46-dc4d-4ca8-83f9-8b6724c649ec" class="ux3 ">
        
    

        


    

            

            

            <div data-widget-id="d95b4a46-dc4d-4ca8-83f9-8b6724c649ec" data-widget-name="ux3-layout-widget"><div class="footer clearfix">
    

        


    

                
                
                
                
                
                
                    
                
                    
                
                    
                
                    
                
                    
                
                    
                
                    
                
                    
                
                    
                
                    
                
                    
                
                
                    
                
                    
                
                    
                
                    
                
                    
                
                    
                        
                        
                        
                    
                
                    
                
                    
                
                    
                
                    
                
                    
                
                
                
                <div class="widget literatumAd none adplaceholder adplaceholder--header widget-horizontalAd" id="a07ee92b-3942-49f8-9186-a2a08fc3fc56">
                
                

                <div class="wrapped " >
                    
                    <div class="widget-body body body-horizontalAd "><div class="horizontalAdContainer">
    <div class="horizontalAdLabelContainer">
        <div class="horizontalAdLabel">
                <img src="/templates/jsp/_style2/_marlin/images/bg_adHoriz.gif" alt="Advertisement"/>
        </div>
    </div>
    <div class="horizontalAd">
        




        <div class="pb-ad">
            <!-- Beginning Async AdSlot 4 for Ad unit els.cellpress/cell  ### size: [[728,90]] -->
<div class="gpt-ad" data-adsize="cell-leaderboard" data-page="" data-site="cell" data-pos="4"></div>
<!-- End AdSlot 4 -->
        </div>
    

    </div>
</div></div>
                </div>
                </div>

        

    

        


    

            

            

            <div data-widget-id="23703bbc-af46-4f3f-a5bb-dd52891ff0ff" data-widget-name="UX3HTMLWidget"><div class="footer__top">
            <div class="footer__logo centered"><img alt="logo" src="/pb-assets/ux3/logos/cell/footer-logo-1532516631990.svg"></div>
            <div class="footer__social-links">
              <ul class="rlist rlist--inline centered">
                <li><a href="http://on.fb.me/1tgHakh" title="Facebook" target="_blank"><i aria-hidden="true" class="icon-Icon_Facebook"></i></a></li>
                <li><a href="https://www.linkedin.com/company/cell-press/" title="LinkedIn" target="_blank"><i aria-hidden="true" class="icon-linkedin"></i></a></li>
                <li><a href="http://bit.ly/1zq7Q5T" title="Twitter" target="_blank"><i aria-hidden="true" class="icon-Icon_Twitter"></i></a></li>
              </ul>
            </div>
</div>
          
          </div>

            

        

    

        


    

            

            
                <div class="footer__bottom text-onDark">
            

            <div data-widget-id="06227714-c621-4651-9879-b0ce5176cccc" data-widget-name="ux3-layout-widget"><div class="col-sm-4 col-md-5 clearfix">
    

        


    

            

            

            <div data-widget-id="07e1d276-b4dd-4870-b858-ed74e70d64ef" data-widget-name="UX3HTMLWidget"><div class="groups">
	<div class="group">
		<div class="group__title">Research Journals</div>
		<ul class="rlist clearfix sided">
		<li>
			<a alt="Cell" href="/cell/home">Cell</a>
		</li>
			<li>
				<a alt="Cancer Cell" href="/cancer-cell/home">Cancer Cell</a>
			</li>
			<li>
				<a alt="Cell Chemical Biology" href="/cell-chemical-biology">Cell Chemical Biology</a>
			</li>
			<li>
				<a alt="Cell Host and Microbe" href="/cell-host-microbe/home">Cell Host &amp; Microbe</a>
			</li>
			<li>
				<a alt="Cell Metabolism" href="/cell-metabolism/home">Cell Metabolism</a>
			</li>
			<li>
				<a alt="Cell Reports" href="/cell-reports/home">Cell Reports</a>
			</li>
			<li>
				<a alt="Cell Stem Cell" href="/cell-stem-cell/home">Cell Stem Cell</a>
			</li>
			<li>
				<a alt="Cell Systems" href="/cell-systems/home">Cell Systems</a>
			</li>
			<li>
				<a alt="Chem" href="/chem/home">Chem</a>
			</li>
			<li>
				<a alt="Current Biology" href="/current-biology/home">Current Biology</a>
			</li>
			<li>
				<a alt="Developmental Cell" href="/developmental-cell/home">Developmental Cell</a>
			</li>
			<li>
				<a alt="Immunity" href="/immunity/home">Immunity</a>
			</li>
			<li>
				<a alt="iScience" href="/iscience/home">iScience</a>
			</li>
			<li>
				<a alt="Joule" href="/joule/home">Joule</a>
			</li>
				<li>
				<a alt="Matter" href="/matter/home">Matter</a>
			</li>
			<li>
				<a alt="Molecular Cell" href="/molecular-cell/home">Molecular Cell</a>
			</li>
			<li>
				<a alt="Neuron" href="/neuron/home">Neuron</a>
			</li>
			<li>
				<a alt="One Earth" href="/one-earth/home">One Earth</a>
			</li>
			<li>
				<a alt="Structure" href="/structure/home">Structure</a>
			</li>
		</ul>
	</div>

	<div class="group">
		<div class="group__title">Trends Reviews Journals</div>
		<ul class="rlist clearfix sided">
			<li>
				<a alt="Trends in Biochemical Sciences" href="/trends/biochemical-sciences/home">Biochemical Sciences</a>
			</li>
			<li>
				<a alt="Trends in Biotechnology" href="/trends/biotechnology/home">Biotechnology</a>
			</li>
			<li>
				<a alt="Trends in Cancer" href="/trends/cancer/home">Cancer</a>
			</li>
			<li>
				<a alt="Trends in Cell Biology" href="/trends/cell-biology/home">Cell Biology</a>
			</li>
			<li>
				<a alt="Trends in Chemistry" href="/trends/chemistry/home">Chemistry</a>
			</li>
			<li>
				<a alt="Trends in Cognitive Sciences" href="/trends/cognitive-sciences/home">Cognitive Sciences</a>
			</li>
			<li>
				<a alt="Trends in Ecology and Evolution" href="/trends/ecology-evolution/home">Ecology &amp; Evolution</a>
			</li>
			<li>
				<a alt="Trends in Endocrinology and Metabolism" href="/trends/endocrinology-metabolism/home">Endocrinology &amp; Metabolism</a>
			</li>
			<li>
				<a alt="Trends in Genetics" href="/trends/genetics/home">Genetics</a>
			</li>
			<li>
				<a alt="Trends in Immunology" href="/trends/immunology/home">Immunology</a>
			</li>
			<li>
				<a alt="Trends in Microbiology" href="/trends/microbiology/home">Microbiology</a>
			</li>
			<li>
				<a alt="Trends in Molecular Medicine" href="/trends/molecular-medicine/home">Molecular Medicine</a>
			</li>
			<li>
				<a alt="Trends in Neurosciences" href="/trends/neurosciences/home">Neurosciences</a>
			</li>
			<li>
				<a alt="Trends in Parasitology" href="/trends/parasitology/home">Parasitology</a>
			</li>
			<li>
				<a alt="Trends in Pharmacological Sciences" href="/trends/pharmacological-sciences/home">Pharmacological Sciences</a>
			</li>
			<li>
				<a alt="Trends in Plant Science" href="/trends/plant-science/home">Plant Science</a>
			</li>
		</ul>
	</div>

<div class="group">
	<div class="group__title">Partner Journals</div>
	<ul class="rlist clearfix sided">
		<li>
			<a alt="American Journal of Human Genetics" href="/ajhg/home">AJHG</a>
		</li>
		<li>
			<a alt="Biophysical Journal" href="/biophysj/home">Biophysical Journal</a>
		</li>
		<li>
			<a alt="EBioMedicine" href="https://www.ebiomedicine.com/">EBioMedicine</a>
		</li>
		<li>
			<a alt="EBioMedicine" href="https://www.heliyon.com/">Heliyon</a>
		</li>		
		<li>
			<a alt="Molecular Plant" href="/molecular-plant/home">Molecular Plant</a>
		</li>
		<li>
			<a alt="Molecular Therapy Family" href="/molecular-therapy-family/home">Molecular Therapy Family</a>
		</li>
		<li>
			<a alt="Stem Cell Reports" href="/stem-cell-reports/home">Stem Cell Reports</a>
		</li>
	</ul>
</div>

</div>
</div>

            

        
</div><div class="col-sm-4 col-md-5 clearfix">
    

        


    

            

            

            <div data-widget-id="850476de-00cb-4c0d-943c-bf1881da24f9" data-widget-name="UX3HTMLWidget"><div class="groups">
	<div class="clearfix">
		<div class="group sided">
			<div class="group__title">Collections</div>
			<ul class="rlist clearfix">
				<li>
					<a alt="Best Ofs" href="/bestof">Best Ofs</a>
				</li>
				<li>
					<a alt="Cell Press Reviews" href="/reviews">Cell Press Reviews</a>
				</li>
				<li>
					<a alt="Cell Selections" href="/selections">Cell Selections</a>
				</li>
				<li>
					<a alt="Nucleus Collections" href="/nucleus">Nucleus Collections</a>
				</li>
				<li>
					<a alt="SnapShots" href="/snapshots">SnapShots</a>
				</li>
				<li>
					<a alt="Trends Limited Edition" href="/trends/limited-edition">Trends Limited Editions</a>
				</li>
			</ul>
		</div>

		<div class="group sided">
			<div class="group__title">Community</div>
			<ul class="rlist clearfix">
				<li>
					<a alt="Cell Career Network" target="_blank" href="http://careers.cell.com/">Cell Career Network</a>
				</li>
				<li>
					<a alt="China Center" href="/china">China Center</a>
				</li>
				<li>
					<a alt="Cell Symposia" href="/symposia">Cell Symposia</a>
				</li>
				<li>
					<a alt="CrossTalk" target="_blank" href="http://crosstalk.cell.com/">CrossTalk</a>
				</li>
				<li>
					<a alt="Lablinks" href="/lablinks">LabLinks</a>
				</li>
				<li>
					<a alt="Webinars" href="/webinars">Webinars</a>
				</li>
			</ul>
		</div>
	</div>


	<div class="clearfix">
		<div class="group sided">
			<div class="group__title">Evolving the Article</div>
			<ul class="rlist clearfix">
				<li>
					<a alt="Figure360" href="/figure360">Figure360</a>
				</li>
				<li>
					<a alt="Sneak Peek" href="/sneakpeek">Sneak Peek</a>
				</li>
				<li>
					<a alt="STAR Methods" href="/star-methods">STAR Methods</a>
				</li>
			</ul>
		</div>

		<div class="group sided">
			<div class="group__title">Science in Society</div>
			<ul class="rlist clearfix">
				<li>
					<a alt="Cell Picture Show" href="/pictureshow">Cell Picture Show</a>
				</li>
				<li>
					<a alt="Cell Press Podcast" href="/podcast">Cell Press Podcast</a>
				</li>
				<li>
					<a alt="Cell Press Videos" href="/video">Cell Press Videos</a>
				</li>
				<li>
					<a alt="Color and Comics" href="/comics">Coloring and Comics</a>
				</li>
				<li>
					<a alt="Research Arc" href="/research-arc">Research Arc</a>
				</li>
			</ul>
		</div>
	</div>

</div>
</div>

            

        
</div><div class="col-sm-4 col-md-2 clearfix">
    

        


    

            

            

            <div data-widget-id="beafac41-1cfc-440b-9b55-74aba4a5612c" data-widget-name="UX3HTMLWidget"><div class="groups">
	<div class="group">
		<div class="group__title">Connect</div>
		<ul class="rlist clearfix">
			<li>
				<a alt="About Cell Press" href="/about">About Cell Press</a>
			</li>
			<li>
				<a alt="Contact" href="/contact">Contact</a>
			</li>
			<li>
				<a alt="Help and Support" target="_blank" href="https://service.elsevier.com/app/overview/cell-press/">Help &amp; Support</a>
			</li>
			<li>
				<a alt="Newsroom" href="/newsroom">Newsroom</a>
			</li>
			<li>
				<a alt="Publication Alerts" href="/action/showPreferences?menuTab=Alerts">Publication Alerts</a>
			</li>
		</ul>
	</div>
	<div class="group">
		<div class="group__title">Access</div>
		<ul class="rlist clearfix">
			<li>
				<a alt="Subscribe" href="/subscribe">Subscribe</a>
			</li>
			<li>
				<a alt="Register" href="https://secure.jbs.elsevierhealth.com/action/registration?journal=&amp;redirectUri=https%3A%2F%2Fwww.cell.com%2F%3Fcode%3Dcell-site&amp;code=cell-site">Register</a>
			</li>
			<li>
				<a alt="Read It Now" href="/read-it-now">Read-It-Now</a>
			</li>
			<li>
				<a alt="Mobile" href="/mobile">Mobile</a>
			</li>
		</ul>
	</div>
	<div class="group">
		<div class="group__title">Information</div>
		<ul class="rlist clearfix">
			<li>
				<a alt="For Advertisters" target="_blank" href="http://cellpress.concevoirlabs.com/">For Advertisers</a>
			</li>
			<li>
				<a alt="For Recruiters" target="_blank" href="http://recruitmentads.concevoirlabs.com/">For Recruiters</a>
			</li>
			<li>
				<a alt="For Librarians" href="/librarians">For Librarians</a>
			</li>
			<li>
				<a alt="Cookies" href="/cookies">Cookies</a>
			</li>
			<li>
				<a alt="Terms and Conditions" target="_blank" href="https://www.elsevier.com/legal/elsevier-website-terms-and-conditions">Terms &amp; Conditions</a>
			</li>
			<li>
				<a alt="Privacy Policy" target="_blank" href="https://www.elsevier.com/legal/privacy-policy">Privacy Policy</a>
			</li>
			<li>
				<a alt="Accessibility" href="/accessibility">Accessibility</a>
			</li>
		</ul>
	</div>
</div>
</div>

            

        
</div></div>

            
                </div>
            

        

    

        


    

            

            

            <div data-widget-id="779ad11f-7bcd-4a2a-a2cd-8c67b649cbb2" data-widget-name="UX3HTMLWidget"><p class="align-center copy-right">
    <span>We use cookies to help provide and enhance our service and tailor content and ads. By continuing you agree to the <a href="/cookies" target="_blank">use of cookies</a>.</span><br />
            Copyright Â© 2019 Elsevier Inc. except certain content provided by third parties.
<br>
<br>
<a href="https://www.relx.com"><img src="//els-jbs-prod-cdn.literatumonline.com/pb/assets/raw/Health%20Advance/images/relx-group-1542288309497.png" alt="RELX" rel="nofollow" target="_new"></a>
</p></div>

            

        
</div></div>

            

        

        </div>
    

    

        


    

                
                
                
                
                
                
                    
                
                    
                
                    
                        
                    
                
                    
                
                    
                
                    
                
                    
                
                    
                
                    
                
                    
                
                    
                
                
                    
                
                    
                
                    
                        
                        
                        
                    
                
                    
                
                    
                
                    
                        
                        
                            
                        
                        
                            
                            
                            
                        
                    
                
                    
                
                    
                
                    
                
                    
                
                    
                
                
                
                <div class="widget csasSsoUrls none  widget-none  widget-compact-all" id="948bc5c4-8b0d-47be-b5ad-6570559802d1">
                
                

                <div class="wrapped " >
                    
                    <div class="widget-body body body-none  body-compact-all"><div id="sso-urls-div" class="hidden ssoUrls">
    
</div>

<!-- clear SSO URLs after all iFrames are rendered or after passing a time out-->
<script>
    setTimeout(clearSsoUrlsFun,10000);
    var iFramesList = document.getElementById("sso-urls-div").getElementsByTagName("iframe");
    var renderedIFramesCount = 0;
    var numberOfIFrames = iFramesList.length;
    var skipRefresh = false;
    for (var i = 0; i < iFramesList.length; i++) {
        var iFrame = iFramesList[i];
        iFrame.addEventListener("load", function() {
            renderedIFramesCount = renderedIFramesCount + 1;
            if (renderedIFramesCount >= numberOfIFrames)
            {
                clearSsoUrlsFun();
            }
        });
    }

    function clearSsoUrlsFun() {
        if (skipRefresh == false && numberOfIFrames > 0)
        {
            skipRefresh = true;
            $.ajax({
                url: "https://www.cell.com/action/clearSsoUrls",
                type: "GET",
                dataType: "jsonp",
                error : errorCallingAjx
            });
        }
    }

    function errorCallingAjx(){
        console.log("can't delete sso urls .")
    }
</script></div>
                </div>
                </div>

        
        </div>
    </div>













        
            <script type="text/javascript" src="/products/marlin/cell/releasedAssets/js/main.bundle.js"></script>
        
        
        


    




    




<script rel="preconnect" src="//assets.adobedtm.com/376c5346e33126fdb6b2dbac81e307cbacfd7935/satelliteLib-8e350dc4ceceb17a5d77c03f66d95608807a5585.js"></script>



<script type="text/javascript">
    var pageData =  {"journal":{"specialty":"Cell Press","issn":"0002-9297","name":"The American Journal of Human Genetics"},"page":{"loadTimestamp":"1549961969941","server":"cell","environment":"prod","businessUnit":"ELS:RAP:JBS","noTracking":"","name":"fulltext:html","section":"Cell","language":"en","type":"CP-CA","productName":"JB"},"visitor":{"departmentName":"ae:Journal Branded Solutions - Guests|ae:ScienceDirect Guests","accessType":"ae:ANON_GUEST|ae:ANON_GUEST","accountId":"ae:375335|ae:228598","accountName":"ae:Journal Branded Solutions - Guests|ae:ScienceDirect Guests","departmentId":"ae:388406|ae:291352","ipAddress":"131.159.28.186","loginStatus":"anonymous","userId":"ae:19319093|ae:12975512"},"content":[{"accessType":"article:free:standard","volumeNumber":"104","issueNumber":"2","viewState":"fulltext","format":"MIME-HTML","id":"pii:S0002929718305032","coverDate":"201902","type":"XOCS-JOURNAL:SCOPE-FULL","title":"Improved Pathogenic Variant Localization via a Hierarchical Model of Subregional Intolerance","status":"S300"}]};
</script>

<script type="text/javascript">pageDataTracker.trackPageLoad();</script>


</body>
</html>
